{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools\n",
    "import time\n",
    "import os\n",
    "import sys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import HTSeq\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def invert_strand(iv):\n",
    "    iv2 = iv.copy()\n",
    "    if iv2.strand == \"+\":\n",
    "        iv2.strand = \"-\"\n",
    "    elif iv2.strand == \"-\":\n",
    "        iv2.strand = \"+\"\n",
    "    else:\n",
    "        raise ValueError(\"Illegal strand\")\n",
    "    return iv2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# select gtf file: instantiate GFF_Reader Object\n",
    "gtf_file = HTSeq.GFF_Reader(\"/gcm-lfs1/pablo/data/RNAdeg/annotation/gff/schizosaccharomyces_pombe.chr.extended.gff3\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# CIGAR match characters (including alignment match, sequence match, and\n",
    "# sequence mismatch\n",
    "com = ('M', '=', 'X')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "## heterochromatic genes\n",
    "htc_genes = ('dg1', 'dh1', 'after_tlh', 'MAT2', 'MAT3', 'MAT1')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Counting gapped single-end reads:  **RNA-Seq data**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When sequencing RNA, many reads will pass over an **exon-exon junction** and hence **align to two (or more) disjunct intervals** on the genome, \n",
    "tyically with an intron in between. If the reads have been aligned with a **splice-aware alignment tool** (e.g. STAR), such gapped alignment is indicated in the `SAM` **file** by the `CIGAR` (Compact Idiosyncratic Gapped Alignment Report) **string**."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Obtain `GenomicArrayOfSets` for the `exon` **features** from the `GFF` **File** (for `RNA` **stranded**)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [],
   "source": [
    "# instantiate `GenomicArrayOfSets` for the `exons` features: (for `RNA` **stranded**)\n",
    "#exon_features = HTSeq.GenomicArrayOfSets(\"auto\", stranded=True)\n",
    "exon_features = HTSeq.GenomicArrayOfSets(\"auto\", stranded=False) ## Parastou used this instead!\n",
    "\n",
    "genes_dict = {}\n",
    "\n",
    "## loop over all features in gtf file\n",
    "for feature in gtf_file:\n",
    "    \n",
    "    ## store all exons in our `GenomicArrayOfSets`\n",
    "    if feature.type == \"exon\":\n",
    "        \n",
    "        ## identify each `exon` feature by parent transcript/gene\n",
    "        gene_id = feature.attr[\"Parent\"].split(':')[1][:-2] ## get rid of '.1'\n",
    "\n",
    "        exon_features[feature.iv] += gene_id\n",
    "\n",
    "        # Is this the first time we see this gene?\n",
    "        if gene_id not in genes_dict:\n",
    "            # If so, add to the 'genes_dict' an empty list \n",
    "            genes_dict[ gene_id ] = list()\n",
    "\n",
    "        # add the feature to the gene list\n",
    "        genes_dict[ gene_id ].append( feature )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6992"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(genes_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![img_overlap](https://htseq.readthedocs.io/en/release_0.11.1/_images/count_modes.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Parse `BAM` **File** to **count reads** falling into each `exon` **feature** "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "#sample_name = '1113_pA'\n",
    "sample_name = '1168_pA'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100000 alignment records processed. 5.148064374923706 s\n",
      "200000 alignment records processed. 10.611220836639404 s\n",
      "300000 alignment records processed. 15.98412275314331 s\n",
      "400000 alignment records processed. 21.070789098739624 s\n",
      "500000 alignment records processed. 26.10773253440857 s\n",
      "600000 alignment records processed. 31.32817554473877 s\n",
      "700000 alignment records processed. 36.56236934661865 s\n",
      "800000 alignment records processed. 41.88441467285156 s\n",
      "900000 alignment records processed. 47.05519652366638 s\n",
      "1000000 alignment records processed. 52.630645513534546 s\n",
      "1100000 alignment records processed. 57.98696780204773 s\n",
      "1200000 alignment records processed. 63.21794009208679 s\n",
      "1300000 alignment records processed. 68.13604068756104 s\n",
      "1400000 alignment records processed. 73.07154846191406 s\n",
      "1500000 alignment records processed. 78.13472437858582 s\n",
      "1600000 alignment records processed. 83.1444799900055 s\n",
      "1700000 alignment records processed. 88.29799723625183 s\n",
      "1800000 alignment records processed. 93.56368398666382 s\n",
      "1900000 alignment records processed. 99.01699757575989 s\n",
      "2000000 alignment records processed. 104.24005389213562 s\n",
      "2100000 alignment records processed. 109.46070337295532 s\n",
      "2200000 alignment records processed. 114.60179710388184 s\n",
      "2300000 alignment records processed. 119.96442484855652 s\n",
      "2400000 alignment records processed. 125.50934100151062 s\n",
      "2500000 alignment records processed. 130.97541165351868 s\n",
      "2600000 alignment records processed. 136.45956254005432 s\n",
      "2700000 alignment records processed. 141.98451375961304 s\n",
      "2800000 alignment records processed. 147.51049280166626 s\n",
      "2900000 alignment records processed. 153.17639780044556 s\n",
      "3000000 alignment records processed. 158.9577612876892 s\n",
      "3100000 alignment records processed. 164.68575143814087 s\n",
      "3200000 alignment records processed. 170.32992100715637 s\n",
      "3300000 alignment records processed. 175.8687813282013 s\n",
      "3400000 alignment records processed. 181.39940810203552 s\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Elapsed Time (Counting reads): 181.87013912200928\n"
     ]
    }
   ],
   "source": [
    "import collections\n",
    "counts = collections.Counter( )\n",
    "\n",
    "start_time = time.time()\n",
    "\n",
    "# select bam file: instantiate BAM_Reader Object (RNA-seq)\n",
    "#bam_file = HTSeq.BAM_Reader(os.path.join('/gcm-lfs1/pablo/data/RNAdeg/data/RNA/bams', sample_name, sample_name + '.Aligned.sortedByCoord.out.bam'))\n",
    "bam_file = HTSeq.BAM_Reader(os.path.join('/gcm-lfs1/pablo/data/RNAdeg/old_data/RNA/tagged_bams', sample_name, sample_name + '.Aligned.sortedByCoord.out.tagged.bam'))\n",
    "\n",
    "## -------\n",
    "## Options\n",
    "## -------\n",
    "\n",
    "minaqual = 10\n",
    "\n",
    "## How to deal with overlap of READ and FEATURE\n",
    "overlap_mode = \"union\"\n",
    "#overlap_mode = \"intersection-nonempty\"\n",
    "#overlap_mode = \"intersection-strict\"\n",
    "\n",
    "## How to deal with overlapping FEATURES and multimapped reads\n",
    "#multimapped_mode = 'all'\n",
    "multimapped_mode = 'fractional'\n",
    "#multimapped_mode = 'none'\n",
    "\n",
    "## How to deal with multimapped reads (secondary alignments!)\n",
    "#secondary_alignment_mode = 'ignore'\n",
    "secondary_alignment_mode = 'none'\n",
    "\n",
    "i = 0\n",
    "\n",
    "#for aln in itertools.islice(bam_file, 10000):\n",
    "for aln in bam_file:\n",
    "\n",
    "    if i > 0 and i % 100000 == 0:\n",
    "        sys.stderr.write(\"{} alignment records processed. {} s\\n\".format(i,  time.time() - start_time))\n",
    "        sys.stderr.flush()\n",
    "    i += 1          \n",
    "    \n",
    "    ## ----------------------\n",
    "    ## Inspect read alignment\n",
    "    ## ----------------------\n",
    "    \n",
    "    ## _mapped or _unmapped (our BAM files only contain _mapped)\n",
    "    counts[\"_total\"] += 1\n",
    "    \n",
    "    if not aln.aligned:\n",
    "        import pdb\n",
    "        pdb.set_trace()\n",
    "        counts[\"_unmapped\"] += 1\n",
    "        ## skips to next iteration\n",
    "        continue\n",
    "        \n",
    "    ## Multimapped reads are contained as separate entries in the BAM file.\n",
    "    try:\n",
    "        if aln.optional_field(\"NH\") > 1 and aln.optional_field(\"NH\") <= 15:\n",
    "            counts[\"_alignment_not_unique\"] += 1\n",
    "            \n",
    "            if multimapped_mode == 'none':\n",
    "                ## skips to next iteration\n",
    "                continue\n",
    "                \n",
    "            elif ((secondary_alignment_mode == 'ignore') and aln.not_primary_alignment):\n",
    "                counts[\"_not_primary_alignment\"] += 1\n",
    "                ## skips to next iteration\n",
    "                continue\n",
    "                \n",
    "        elif aln.optional_field(\"NH\") > 15:\n",
    "            counts[\"_ignore_alignment\"] += 1\n",
    "            \n",
    "            ## skips to next iteration\n",
    "            continue\n",
    "\n",
    "    except KeyError:\n",
    "        pass\n",
    "\n",
    "    #if aln.aQual < minaqual:\n",
    "    #    #import pdb\n",
    "    #    #pdb.set_trace()\n",
    "    #    counts[\"_too_low_aQual\"] +=  1\n",
    "    #    continue\n",
    "\n",
    "    ## -----------------------------\n",
    "    ## Read and Feature Overlap Mode\n",
    "    ## -----------------------------\n",
    "    \n",
    "    ## invert strand - due to sequencing is strand is reversed!\n",
    "    iv_seq = (invert_strand(co.ref_iv) for co in aln.cigar if (co.type in com and co.size > 0))\n",
    "    #iv_seq = (co.ref_iv for co in aln.cigar if (co.type in com and co.size > 0))\n",
    "    \n",
    "    ## loop over CIGAR operations (cig_op) for aligned read    \n",
    "\n",
    "    ## A. Union: the union of all the sets S(i). This mode is recommended for most use cases.\n",
    "    if overlap_mode == \"union\":\n",
    "        ## feature set\n",
    "        gene_ids = set()\n",
    "        \n",
    "        for iv in iv_seq:\n",
    "            #if iv.chrom not in exon_features.chrom_vectors:\n",
    "            #    raise UnknownChrom\n",
    "            \n",
    "            for iv2, fs in exon_features[ iv ].steps():\n",
    "                gene_ids = gene_ids.union(fs)\n",
    "\n",
    "    ## B. Intersection-strict: the intersection of all the sets S(i).\n",
    "    ## C. Intersection-nonempty: the intersection of all non-empty sets S(i).\n",
    "    elif overlap_mode in (\"intersection-strict\", \"intersection-nonempty\"):\n",
    "        ## feature set\n",
    "        gene_ids = None\n",
    "        \n",
    "        for iv in iv_seq:\n",
    "            #if iv.chrom not in exon_features.chrom_vectors:\n",
    "            #    raise UnknownChrom\n",
    "            \n",
    "            for iv2, fs in exon_features[ iv ].steps():\n",
    "                if ((len(fs) > 0) or (overlap_mode == \"intersection-strict\")):\n",
    "                    if gene_ids is None:\n",
    "                        gene_ids = fs.copy()\n",
    "                    else:\n",
    "                        gene_ids = gene_ids.intersection(fs)\n",
    "                    \n",
    "    ## Other: Ilegal!                   \n",
    "    else:\n",
    "        sys.exit(\"Illegal overlap mode.\")\n",
    "    \n",
    "    ## --------------\n",
    "    ## Count Features\n",
    "    ## --------------\n",
    "    \n",
    "    ## A. Mapped to unknown feature (it is empty)\n",
    "    if gene_ids is None or len(gene_ids) == 0:\n",
    "        #import pdb\n",
    "        #pdb.set_trace()\n",
    "        counts[\"_no_feature\"] += 1\n",
    "        \n",
    "    ## B. Mapped to a region with overlapping features (contains more than one element)\n",
    "    ## See next how to deal with this ambiguous read alignments! (multimapped_mode)\n",
    "    elif len(gene_ids) > 1:\n",
    "        counts[\"_ambiguous\"] += 1\n",
    "    \n",
    "    ## Deal with reads mapped to region with overlapping features!\n",
    "    if gene_ids is not None and len(gene_ids) > 0:\n",
    "        \n",
    "        ## - ignore reads mapped to region with overlapping features!\n",
    "        if multimapped_mode == 'none':\n",
    "            \n",
    "            ## C. Uniquely Mapped (contains exactly one element) \n",
    "            if len(gene_ids) == 1:\n",
    "                counts[list(gene_ids)[0]] += 1\n",
    "        \n",
    "        ## - count each multimapped feature once\n",
    "        elif multimapped_mode == 'all':\n",
    "            \n",
    "            ## D. Multi Mapped\n",
    "            for fsi in list(gene_ids):\n",
    "                counts[ fsi ] += 1\n",
    "                \n",
    "        ## - count each multimapped feature a fractio\n",
    "        elif multimapped_mode == 'fractional':\n",
    "            \n",
    "            ## D. Multi Mapped\n",
    "            for fsi in list(gene_ids):\n",
    "                counts[ fsi ] += 1 / aln.optional_field('NH')\n",
    "\n",
    "        ## Other: Ilegal!                   \n",
    "        else:\n",
    "            sys.exit(\"Illegal multimap mode.\")\n",
    "\n",
    "print('Elapsed Time (Counting reads):', time.time() - start_time)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [],
   "source": [
    "#{k: v for k, v in sorted(counts.items(), key=lambda item: item[1], reverse=True)}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Convert `counter` to **DataFrame**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene-id</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>_total</td>\n",
       "      <td>3.408617e+06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>_alignment_not_unique</td>\n",
       "      <td>1.574200e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SPAC212.11</td>\n",
       "      <td>3.736667e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SPAC212.10</td>\n",
       "      <td>5.750000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SPAC212.09c</td>\n",
       "      <td>8.191667e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SPNCRNA.70</td>\n",
       "      <td>3.333333e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SPAC212.08c</td>\n",
       "      <td>1.805000e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SPAC212.07c</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SPAC212.12</td>\n",
       "      <td>1.001667e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SPAC212.06c</td>\n",
       "      <td>2.633333e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SPAC212.05c</td>\n",
       "      <td>2.500000e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SPAC212.04c</td>\n",
       "      <td>1.377833e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SPAC212.03</td>\n",
       "      <td>1.333333e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SPAC212.02</td>\n",
       "      <td>1.333333e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SPAC212.01c</td>\n",
       "      <td>8.495000e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SPAC977.01</td>\n",
       "      <td>1.370833e+02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>SPAC977.18</td>\n",
       "      <td>5.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>_no_feature</td>\n",
       "      <td>1.957000e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>SPAC977.02</td>\n",
       "      <td>2.166667e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>_ambiguous</td>\n",
       "      <td>1.002946e+06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  gene-id         count\n",
       "0                  _total  3.408617e+06\n",
       "1   _alignment_not_unique  1.574200e+05\n",
       "2              SPAC212.11  3.736667e+01\n",
       "3              SPAC212.10  5.750000e+00\n",
       "4             SPAC212.09c  8.191667e+01\n",
       "5              SPNCRNA.70  3.333333e-01\n",
       "6             SPAC212.08c  1.805000e+02\n",
       "7             SPAC212.07c  1.000000e+00\n",
       "8              SPAC212.12  1.001667e+02\n",
       "9             SPAC212.06c  2.633333e+01\n",
       "10            SPAC212.05c  2.500000e-01\n",
       "11            SPAC212.04c  1.377833e+02\n",
       "12             SPAC212.03  1.333333e+00\n",
       "13             SPAC212.02  1.333333e+00\n",
       "14            SPAC212.01c  8.495000e+01\n",
       "15             SPAC977.01  1.370833e+02\n",
       "16             SPAC977.18  5.000000e+00\n",
       "17            _no_feature  1.957000e+03\n",
       "18             SPAC977.02  2.166667e+00\n",
       "19             _ambiguous  1.002946e+06"
      ]
     },
     "execution_count": 170,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_df = pd.DataFrame.from_dict(counts, orient='index').reset_index()\n",
    "counts_df = counts_df.rename(columns={'index':'gene-id', 0:'count'})\n",
    "counts_df.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(6690, 2)"
      ]
     },
     "execution_count": 171,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Summary of counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene-id</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>_total</td>\n",
       "      <td>3408617.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>_alignment_not_unique</td>\n",
       "      <td>157420.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>_no_feature</td>\n",
       "      <td>1957.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>_ambiguous</td>\n",
       "      <td>1002946.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  gene-id      count\n",
       "0                  _total  3408617.0\n",
       "1   _alignment_not_unique   157420.0\n",
       "17            _no_feature     1957.0\n",
       "19             _ambiguous  1002946.0"
      ]
     },
     "execution_count": 172,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_df[counts_df['gene-id'].str.startswith('_')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4411329.713095238"
      ]
     },
     "execution_count": 173,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## now contains multiple counting\n",
    "counts_df[~counts_df['gene-id'].str.startswith('_')]['count'].sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Show counts for genes of interest: **Heterochromatic Genes**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene-id</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2047</th>\n",
       "      <td>dh1</td>\n",
       "      <td>11.250000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2049</th>\n",
       "      <td>dg1</td>\n",
       "      <td>8.611111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4138</th>\n",
       "      <td>MAT1</td>\n",
       "      <td>209.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4151</th>\n",
       "      <td>MAT2</td>\n",
       "      <td>40.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4155</th>\n",
       "      <td>MAT3</td>\n",
       "      <td>188.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5461</th>\n",
       "      <td>after_tlh</td>\n",
       "      <td>108.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene-id       count\n",
       "2047        dh1   11.250000\n",
       "2049        dg1    8.611111\n",
       "4138       MAT1  209.000000\n",
       "4151       MAT2   40.000000\n",
       "4155       MAT3  188.500000\n",
       "5461  after_tlh  108.000000"
      ]
     },
     "execution_count": 174,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_df[counts_df['gene-id'].isin(htc_genes)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compare results with **Parastou's counting**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Results using **Parastou's counting script**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [],
   "source": [
    "#xp_rna = '/gcm-lfs1/pablo/data/RNAdeg/data/RNA/xp_data/pombe_gene_count_matrix.csv'\n",
    "xp_rna = '/gcm-lfs1/pablo/data/RNAdeg/old_data/RNA/xp_data/pombe_gene_count_matrix.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene-id</th>\n",
       "      <th>length</th>\n",
       "      <th>type</th>\n",
       "      <th>category</th>\n",
       "      <th>bio_type</th>\n",
       "      <th>1168_pA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SPAC212.11</td>\n",
       "      <td>5661.0</td>\n",
       "      <td>gene</td>\n",
       "      <td>repeat</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SPAC212.10</td>\n",
       "      <td>605.0</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>repeat</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SPAC212.09c</td>\n",
       "      <td>1655.0</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>repeat</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SPNCRNA.70</td>\n",
       "      <td>529.0</td>\n",
       "      <td>ncRNA_gene</td>\n",
       "      <td>repeat</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SPAC212.08c</td>\n",
       "      <td>1210.0</td>\n",
       "      <td>gene</td>\n",
       "      <td>repeat</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>476</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene-id  length        type category        bio_type  1168_pA\n",
       "0   SPAC212.11  5661.0        gene   repeat  protein_coding       80\n",
       "1   SPAC212.10   605.0  pseudogene   repeat      pseudogene       10\n",
       "2  SPAC212.09c  1655.0  pseudogene   repeat      pseudogene      219\n",
       "3   SPNCRNA.70   529.0  ncRNA_gene   repeat           ncRNA        1\n",
       "4  SPAC212.08c  1210.0        gene   repeat  protein_coding      476"
      ]
     },
     "execution_count": 176,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xp_rna_df = pd.read_csv(xp_rna, sep='\\t')\n",
    "xp_rna_df = xp_rna_df[['gene-id', 'length', 'type', 'category', 'bio_type', sample_name]].astype({sample_name: 'int64'})\n",
    "xp_rna_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3254080"
      ]
     },
     "execution_count": 177,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total = xp_rna_df[sample_name].sum()\n",
    "total"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **Merge** both counts DataFrames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene-id</th>\n",
       "      <th>count</th>\n",
       "      <th>length</th>\n",
       "      <th>type</th>\n",
       "      <th>category</th>\n",
       "      <th>bio_type</th>\n",
       "      <th>1168_pA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>_total</td>\n",
       "      <td>3.408617e+06</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>_alignment_not_unique</td>\n",
       "      <td>1.574200e+05</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SPAC212.11</td>\n",
       "      <td>3.736667e+01</td>\n",
       "      <td>5661.0</td>\n",
       "      <td>gene</td>\n",
       "      <td>repeat</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>80.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SPAC212.10</td>\n",
       "      <td>5.750000e+00</td>\n",
       "      <td>605.0</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>repeat</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SPAC212.09c</td>\n",
       "      <td>8.191667e+01</td>\n",
       "      <td>1655.0</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>repeat</td>\n",
       "      <td>pseudogene</td>\n",
       "      <td>219.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6944</th>\n",
       "      <td>SPMITTRNAARG.02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>72.0</td>\n",
       "      <td>tRNA_gene</td>\n",
       "      <td>gene</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6945</th>\n",
       "      <td>SPMITTRNAILE.02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>71.0</td>\n",
       "      <td>tRNA_gene</td>\n",
       "      <td>gene</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6946</th>\n",
       "      <td>SPMITTRNASER.02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>81.0</td>\n",
       "      <td>tRNA_gene</td>\n",
       "      <td>gene</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6947</th>\n",
       "      <td>SPMITTRNAALA.01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>72.0</td>\n",
       "      <td>tRNA_gene</td>\n",
       "      <td>gene</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6948</th>\n",
       "      <td>SPMITTRNAGLU.01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>71.0</td>\n",
       "      <td>tRNA_gene</td>\n",
       "      <td>gene</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6949 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    gene-id         count  length        type category  \\\n",
       "0                    _total  3.408617e+06     NaN         NaN      NaN   \n",
       "1     _alignment_not_unique  1.574200e+05     NaN         NaN      NaN   \n",
       "2                SPAC212.11  3.736667e+01  5661.0        gene   repeat   \n",
       "3                SPAC212.10  5.750000e+00   605.0  pseudogene   repeat   \n",
       "4               SPAC212.09c  8.191667e+01  1655.0  pseudogene   repeat   \n",
       "...                     ...           ...     ...         ...      ...   \n",
       "6944        SPMITTRNAARG.02           NaN    72.0   tRNA_gene     gene   \n",
       "6945        SPMITTRNAILE.02           NaN    71.0   tRNA_gene     gene   \n",
       "6946        SPMITTRNASER.02           NaN    81.0   tRNA_gene     gene   \n",
       "6947        SPMITTRNAALA.01           NaN    72.0   tRNA_gene     gene   \n",
       "6948        SPMITTRNAGLU.01           NaN    71.0   tRNA_gene     gene   \n",
       "\n",
       "            bio_type  1168_pA  \n",
       "0                NaN      NaN  \n",
       "1                NaN      NaN  \n",
       "2     protein_coding     80.0  \n",
       "3         pseudogene     10.0  \n",
       "4         pseudogene    219.0  \n",
       "...              ...      ...  \n",
       "6944            tRNA      0.0  \n",
       "6945            tRNA      0.0  \n",
       "6946            tRNA      0.0  \n",
       "6947            tRNA      0.0  \n",
       "6948            tRNA      0.0  \n",
       "\n",
       "[6949 rows x 7 columns]"
      ]
     },
     "execution_count": 178,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_xp_rna = pd.merge(counts_df, xp_rna_df, on='gene-id', how='outer')\n",
    "merged_xp_rna"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Bigger differences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_xp_rna['diff_count'] = abs(merged_xp_rna['count'] - merged_xp_rna[sample_name])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene-id</th>\n",
       "      <th>count</th>\n",
       "      <th>length</th>\n",
       "      <th>type</th>\n",
       "      <th>category</th>\n",
       "      <th>bio_type</th>\n",
       "      <th>1168_pA</th>\n",
       "      <th>diff_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>_total</td>\n",
       "      <td>3.408617e+06</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>_alignment_not_unique</td>\n",
       "      <td>1.574200e+05</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>_no_feature</td>\n",
       "      <td>1.957000e+03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>_ambiguous</td>\n",
       "      <td>1.002946e+06</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>SPRRNA.10</td>\n",
       "      <td>2.000000e+00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2631</th>\n",
       "      <td>SPNCRNA.1030</td>\n",
       "      <td>1.154000e+04</td>\n",
       "      <td>1320.0</td>\n",
       "      <td>ncRNA_gene</td>\n",
       "      <td>gene</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>5763.0</td>\n",
       "      <td>5777.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6220</th>\n",
       "      <td>SPCC13B11.01</td>\n",
       "      <td>4.330433e+04</td>\n",
       "      <td>3462.0</td>\n",
       "      <td>gene</td>\n",
       "      <td>gene</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>10827.0</td>\n",
       "      <td>32477.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6221</th>\n",
       "      <td>SPCC13B11.02c</td>\n",
       "      <td>4.330333e+04</td>\n",
       "      <td>3822.0</td>\n",
       "      <td>gene</td>\n",
       "      <td>gene</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>10825.0</td>\n",
       "      <td>32478.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6222</th>\n",
       "      <td>SPNCRNA.2001</td>\n",
       "      <td>4.330533e+04</td>\n",
       "      <td>3958.0</td>\n",
       "      <td>ncRNA_gene</td>\n",
       "      <td>gene</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>10826.0</td>\n",
       "      <td>32479.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6223</th>\n",
       "      <td>SPNCRNA.2002</td>\n",
       "      <td>4.330533e+04</td>\n",
       "      <td>3894.0</td>\n",
       "      <td>ncRNA_gene</td>\n",
       "      <td>gene</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>10825.0</td>\n",
       "      <td>32480.333333</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6949 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    gene-id         count  length        type category  \\\n",
       "0                    _total  3.408617e+06     NaN         NaN      NaN   \n",
       "1     _alignment_not_unique  1.574200e+05     NaN         NaN      NaN   \n",
       "17              _no_feature  1.957000e+03     NaN         NaN      NaN   \n",
       "19               _ambiguous  1.002946e+06     NaN         NaN      NaN   \n",
       "85                SPRRNA.10  2.000000e+00     NaN         NaN      NaN   \n",
       "...                     ...           ...     ...         ...      ...   \n",
       "2631           SPNCRNA.1030  1.154000e+04  1320.0  ncRNA_gene     gene   \n",
       "6220           SPCC13B11.01  4.330433e+04  3462.0        gene     gene   \n",
       "6221          SPCC13B11.02c  4.330333e+04  3822.0        gene     gene   \n",
       "6222           SPNCRNA.2001  4.330533e+04  3958.0  ncRNA_gene     gene   \n",
       "6223           SPNCRNA.2002  4.330533e+04  3894.0  ncRNA_gene     gene   \n",
       "\n",
       "            bio_type  1168_pA    diff_count  \n",
       "0                NaN      NaN           NaN  \n",
       "1                NaN      NaN           NaN  \n",
       "17               NaN      NaN           NaN  \n",
       "19               NaN      NaN           NaN  \n",
       "85               NaN      NaN           NaN  \n",
       "...              ...      ...           ...  \n",
       "2631           ncRNA   5763.0   5777.000000  \n",
       "6220  protein_coding  10827.0  32477.333333  \n",
       "6221  protein_coding  10825.0  32478.333333  \n",
       "6222           ncRNA  10826.0  32479.333333  \n",
       "6223           ncRNA  10825.0  32480.333333  \n",
       "\n",
       "[6949 rows x 8 columns]"
      ]
     },
     "execution_count": 180,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_xp_rna.sort_values('diff_count', na_position='first')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7f509f9a8630>"
      ]
     },
     "execution_count": 181,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYsAAAD4CAYAAAAdIcpQAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8li6FKAAAU7ElEQVR4nO3dbbBd1X3f8e/Pki0EsQCVQFWJRNDROBZMMaBQUWfSBJyg2I5FMqGVpy6qh0QdTFu76UwiJZ0meaEO7qSOQ1OIVdtB+InI+AEVlziyEifTGQb5YpMIASqKISBLQY5TAzGpCPK/L866zrE4uvsI7rn36N7vZ+bMXvt/9tp3LV2hH/vh7JOqQpKkqbxqtgcgSRp/hoUkqZNhIUnqZFhIkjoZFpKkTgtnewCjcs4559TKlStnexiSdEp54IEH/rKqvvf4+pwNi5UrVzIxMTHbw5CkU0qSPx9U9zSUJKmTYSFJ6mRYSJI6GRaSpE6GhSSpk2EhSepkWEiSOhkWkqROhoUkqdOc/QT33q89w8rNn5vtYQh44ua3zPYQJL1CHllIkjoZFpKkToaFJKmTYSFJ6mRYSJI6GRaSpE6GhSSpk2EhSepkWEiSOhkWkqROhoUkqZNhIUnqZFhIkjoZFpKkToaFJKmTYSFJ6jTSsEhyVpK7kjya5JEkVyZZmmRXksfa8uy+7bckOZBkf5Jr+uqXJ9nb3rslSUY5bknSdxv1kcVvAr9XVT8AXAI8AmwGdlfVKmB3WyfJamADcBGwDrg1yYK2n9uATcCq9lo34nFLkvqMLCySLAF+GPgQQFW9UFXfBNYD29tm24FrW3s9cGdVHa2qx4EDwBVJlgFLquq+qirgjr4+kqQZMMojiwuBrwO/k+QrST6Y5AzgvKo6DNCW57btlwNP9fU/2GrLW/v4+ksk2ZRkIsnEseefmd7ZSNI8NsqwWAhcBtxWVZcC36KdcjqBQdchaor6S4tV26pqTVWtWXD6mSc7XknSCYwyLA4CB6vq/rZ+F73weLqdWqItj/Rtf35f/xXAoVZfMaAuSZohIwuLqvoL4Kkkr2ulq4GHgZ3AxlbbCNzd2juBDUkWJbmA3oXsPe1U1XNJ1ra7oK7v6yNJmgELR7z/fwt8LMlrgK8C76QXUDuS3AA8CVwHUFX7kuygFygvAjdV1bG2nxuB24HFwL3tJUmaISMNi6p6EFgz4K2rT7D9VmDrgPoEcPH0jk6SNCw/wS1J6mRYSJI6GRaSpE6GhSSpk2EhSepkWEiSOhkWkqROhoUkqZNhIUnqZFhIkjoZFpKkToaFJKmTYSFJ6mRYSJI6GRaSpE6GhSSpk2EhSepkWEiSOhkWkqROhoUkqZNhIUnqZFhIkjqNNCySPJFkb5IHk0y02tIku5I81pZn922/JcmBJPuTXNNXv7zt50CSW5JklOOWJH23mTiy+NGqekNVrWnrm4HdVbUK2N3WSbIa2ABcBKwDbk2yoPW5DdgErGqvdTMwbklSMxunodYD21t7O3BtX/3OqjpaVY8DB4ArkiwDllTVfVVVwB19fSRJM2DUYVHA7yd5IMmmVjuvqg4DtOW5rb4ceKqv78FWW97ax9dfIsmmJBNJJo49/8w0TkOS5reFI97/G6vqUJJzgV1JHp1i20HXIWqK+kuLVduAbQCLlq0auI0k6eSN9Miiqg615RHgM8AVwNPt1BJteaRtfhA4v6/7CuBQq68YUJckzZCRhUWSM5K8drIN/DjwELAT2Ng22wjc3do7gQ1JFiW5gN6F7D3tVNVzSda2u6Cu7+sjSZoBozwNdR7wmXaX60Lg41X1e0m+BOxIcgPwJHAdQFXtS7IDeBh4Ebipqo61fd0I3A4sBu5tL0nSDBlZWFTVV4FLBtS/AVx9gj5bga0D6hPAxdM9RknScPwEtySpk2EhSepkWEiSOhkWkqROhoUkqZNhIUnqZFhIkjoZFpKkToaFJKmTYSFJ6mRYSJI6GRaSpE6GhSSpk2EhSepkWEiSOhkWkqROhoUkqdNQYZHEb6mTpHls2COL306yJ8m7kpw10hFJksbOUGFRVT8E/AvgfGAiyceT/NhIRyZJGhtDX7OoqseA/wj8IvBPgVuSPJrkp0c1OEnSeBj2msU/SvIbwCPAVcBPVtXrW/s3Rjg+SdIYGPbI4reALwOXVNVNVfVlgKo6RO9o44SSLEjylST3tPWlSXYleawtz+7bdkuSA0n2J7mmr355kr3tvVuS5GQnKkl6+YYNizcDH6+qvwFI8qokpwNU1Uc6+r6b3hHJpM3A7qpaBexu6yRZDWwALgLWAbcmWdD63AZsAla117ohxy1JmgbDhsUXgMV966e32pSSrADeAnywr7we2N7a24Fr++p3VtXRqnocOABckWQZsKSq7quqAu7o6yNJmgHDhsVpVfXXkyutffoQ/d4P/ALw7b7aeVV1uO3nMHBuqy8Hnurb7mCrLW/t4+uSpBkybFh8K8llkytJLgf+ZqoOSd4KHKmqB4b8GYOuQ9QU9UE/c1OSiSQTx55/ZsgfK0nqsnDI7d4DfDLJoba+DPjnHX3eCLwtyZuB04AlST4KPJ1kWVUdbqeYjrTtD9L7HMekFcChVl8xoP4SVbUN2AawaNmqgYEiSTp5w34o70vADwA3Au8CXt91xFBVW6pqRVWtpHfh+g+q6h3ATmBj22wjcHdr7wQ2JFmU5AJ6F7L3tFNVzyVZ2+6Cur6vjyRpBgx7ZAHwg8DK1ufSJFTVHS/jZ94M7EhyA/AkcB1AVe1LsgN4GHgRuKmqjrU+NwK307vIfm97SZJmyFBhkeQjwD8EHgQm/wGfvDOpU1V9Efhia38DuPoE220Ftg6oTwA+zFCSZsmwRxZrgNXt1lVJ0jwz7N1QDwF/f5QDkSSNr2GPLM4BHk6yBzg6Wayqt41kVJKksTJsWPzqKAchSRpvQ4VFVf1Rku8HVlXVF9pzoRZ09ZMkzQ3DPqL854C7gA+00nLgs6MalCRpvAx7gfsmep/Ifha+80VI507ZQ5I0ZwwbFker6oXJlSQLOcHzmSRJc8+wYfFHSX4JWNy+e/uTwP8c3bAkSeNk2LDYDHwd2Av8a+B/0fENeZKkuWPYu6G+DfyP9pIkzTPDPhvqcQZco6iqC6d9RJKksXMyz4aadBq9J8Uunf7hSJLG0bDfZ/GNvtfXqur9wFUjHpskaUwMexrqsr7VV9E70njtSEYkSRo7w56G+q997ReBJ4B/Nu2jkSSNpWHvhvrRUQ9EkjS+hj0N9fNTvV9V75ue4UiSxtHJ3A31g8DOtv6TwB8DT41iUJKk8XIyX350WVU9B5DkV4FPVtXPjmpgkqTxMezjPr4PeKFv/QVg5bSPRpI0loY9svgIsCfJZ+h9kvungDtGNipJ0lgZ9kN5W4F3Av8X+Cbwzqr6z1P1SXJakj1J/iTJviS/1upLk+xK8lhbnt3XZ0uSA0n2J7mmr355kr3tvVuS5OVMVpL08gx7GgrgdODZqvpN4GCSCzq2PwpcVVWXAG8A1iVZS+8JtrurahWwu62TZDWwAbgIWAfcmmTyq1tvAzYBq9pr3UmMW5L0Cg37taq/AvwisKWVXg18dKo+1fPXfdu/mt4prPXA9lbfDlzb2uuBO6vqaFU9DhwArkiyDFhSVfdVVdE7/TXZR5I0A4Y9svgp4G3AtwCq6hBDPO4jyYIkDwJHgF1VdT9wXlUdbvs5zN99PetyvvtW3IOttry1j69LkmbIsGHxQvu/+gJIcsYwnarqWFW9AVhB7yjh4ik2H3Qdoqaov3QHyaYkE0kmjj3/zDBDlCQNYdiw2JHkA8BZSX4O+AIn8UVIVfVN4Iv0rjU83U4t0ZZH2mYHgfP7uq0ADrX6igH1QT9nW1Wtqao1C04/c9jhSZI6DHs31K8DdwGfAl4H/Keq+m9T9UnyvUnOau3FwJuAR+l9Cnxj22wjcHdr7wQ2JFnULp6vAva0U1XPJVnb7oK6vq+PJGkGdH7Oot2R9PmqehOw6yT2vQzY3vq/CthRVfckuY/ekcoNwJP0vkiJqtqXZAfwML0n295UVcfavm4EbgcWA/e2lyRphnSGRVUdS/J8kjOraugLAVX1p8ClA+rfAK4+QZ+twNYB9QlgqusdkqQRGvYT3P8P2JtkF+2OKICq+ncjGZUkaawMGxafay9J0jw0ZVgk+b6qerKqtk+1nSRpbuu6G+qzk40knxrxWCRJY6orLPo/EHfhKAciSRpfXWFRJ2hLkuaRrgvclyR5lt4RxuLWpq1XVS0Z6egkSWNhyrCoqgVTvS9Jmh9O5vssJEnzlGEhSepkWEiSOhkWkqROhoUkqZNhIUnqZFhIkjoZFpKkToaFJKmTYSFJ6mRYSJI6GRaSpE6GhSSpk2EhSeo0srBIcn6SP0zySJJ9Sd7d6kuT7EryWFue3ddnS5IDSfYnuaavfnmSve29W5Jk0M+UJI3GKI8sXgT+Q1W9HlgL3JRkNbAZ2F1Vq4DdbZ323gbgImAdcGuSye/TuA3YBKxqr3UjHLck6TgjC4uqOlxVX27t54BHgOXAemB722w7cG1rrwfurKqjVfU4cAC4IskyYElV3VdVBdzR10eSNANm5JpFkpXApcD9wHlVdRh6gQKc2zZbDjzV1+1gqy1v7ePrkqQZMvKwSPI9wKeA91TVs1NtOqBWU9QH/axNSSaSTBx7/pmTH6wkaaCRhkWSV9MLio9V1adb+el2aom2PNLqB4Hz+7qvAA61+ooB9Zeoqm1Vtaaq1iw4/czpm4gkzXOjvBsqwIeAR6rqfX1v7QQ2tvZG4O6++oYki5JcQO9C9p52quq5JGvbPq/v6yNJmgELR7jvNwL/Etib5MFW+yXgZmBHkhuAJ4HrAKpqX5IdwMP07qS6qaqOtX43ArcDi4F720uSNENGFhZV9b8ZfL0B4OoT9NkKbB1QnwAunr7RSZJOhp/gliR1MiwkSZ0MC0lSJ8NCktTJsJAkdTIsJEmdDAtJUifDQpLUybCQJHUyLCRJnQwLSVInw0KS1MmwkCR1MiwkSZ0MC0lSJ8NCktTJsJAkdTIsJEmdDAtJUifDQpLUybCQJHUyLCRJnQwLSVKnkYVFkg8nOZLkob7a0iS7kjzWlmf3vbclyYEk+5Nc01e/PMne9t4tSTKqMUuSBhvlkcXtwLrjapuB3VW1Ctjd1kmyGtgAXNT63JpkQetzG7AJWNVex+9TkjRiIwuLqvpj4K+OK68Htrf2duDavvqdVXW0qh4HDgBXJFkGLKmq+6qqgDv6+kiSZshMX7M4r6oOA7Tlua2+HHiqb7uDrba8tY+vD5RkU5KJJBPHnn9mWgcuSfPZuFzgHnQdoqaoD1RV26pqTVWtWXD6mdM2OEma72Y6LJ5up5ZoyyOtfhA4v2+7FcChVl8xoC5JmkEzHRY7gY2tvRG4u6++IcmiJBfQu5C9p52qei7J2nYX1PV9fSRJM2ThqHac5BPAjwDnJDkI/ApwM7AjyQ3Ak8B1AFW1L8kO4GHgReCmqjrWdnUjvTurFgP3tpckaQaNLCyq6u0neOvqE2y/Fdg6oD4BXDyNQ5MknaRxucAtSRpjhoUkqZNhIUnqZFhIkjoZFpKkToaFJKmTYSFJ6mRYSJI6GRaSpE6GhSSpk2EhSepkWEiSOhkWkqROhoUkqZNhIUnqZFhIkjoZFpKkToaFJKmTYSFJ6mRYSJI6GRaSpE6GhSSp0ykTFknWJdmf5ECSzbM9HkmaT06JsEiyAPjvwE8Aq4G3J1k9u6OSpPlj4WwPYEhXAAeq6qsASe4E1gMPz+qoNJSVmz8320MQ8MTNb5ntIegUdqqExXLgqb71g8A/Pn6jJJuATW316J+/960PzcDYxsk5wF/O9iBm0HybL7yCOee90zySmePveWZ9/6DiqRIWGVCrlxSqtgHbAJJMVNWaUQ9snMy3Oc+3+YJzni/Gcc6nxDULekcS5/etrwAOzdJYJGneOVXC4kvAqiQXJHkNsAHYOctjkqR545Q4DVVVLyb5N8DngQXAh6tqX0e3baMf2diZb3Oeb/MF5zxfjN2cU/WSU/+SJH2XU+U0lCRpFhkWkqROcy4s5tJjQZKcn+QPkzySZF+Sd7f60iS7kjzWlmf39dnS5r4/yTV99cuT7G3v3ZJk0O3IYyHJgiRfSXJPW5/r8z0ryV1JHm2/6yvnwZz/ffs7/VCSTyQ5ba7NOcmHkxxJ8lBfbdrmmGRRkt9t9fuTrBzphKpqzrzoXfz+M+BC4DXAnwCrZ3tcr2A+y4DLWvu1wP+h97iT/wJsbvXNwHtbe3Wb8yLggvZnsaC9twe4kt5nVu4FfmK25zfFvH8e+DhwT1uf6/PdDvxsa78GOGsuz5neh2wfBxa39R3Av5prcwZ+GLgMeKivNm1zBN4F/HZrbwB+d6Tzme0/0Gn+5VwJfL5vfQuwZbbHNY3zuxv4MWA/sKzVlgH7B82X3t1jV7ZtHu2rvx34wGzP5wRzXAHsBq7qC4u5PN8l7R/OHFefy3OefCLDUnp3ZN4D/PhcnDOw8riwmLY5Tm7T2gvpfeI7o5rLXDsNNeixIMtnaSzTqh1iXgrcD5xXVYcB2vLcttmJ5r+8tY+vj6P3A78AfLuvNpfneyHwdeB32qm3DyY5gzk856r6GvDrwJPAYeCZqvp95vCc+0znHL/Tp6peBJ4B/t6oBj7XwmKox4KcapJ8D/Ap4D1V9exUmw6o1RT1sZLkrcCRqnpg2C4DaqfMfJuF9E5V3FZVlwLfond64kRO+Tm38/Tr6Z1u+QfAGUneMVWXAbVTas5DeDlznNH5z7WwmHOPBUnyanpB8bGq+nQrP51kWXt/GXCk1U80/4OtfXx93LwReFuSJ4A7gauSfJS5O1/ojfVgVd3f1u+iFx5zec5vAh6vqq9X1d8Cnwb+CXN7zpOmc47f6ZNkIXAm8FejGvhcC4s59ViQdtfDh4BHqup9fW/tBDa29kZ61zIm6xvaXRIXAKuAPe1w97kka9s+r+/rMzaqaktVraiqlfR+d39QVe9gjs4XoKr+Angqyeta6Wp6j96fs3Omd/ppbZLT21ivBh5hbs950nTOsX9fP0Pvv5fRHVnN9gWgEVxQejO9u4b+DPjl2R7PK5zLD9E7rPxT4MH2ejO985K7gcfacmlfn19uc99P350hwBrgofbebzHCC2HTNPcf4e8ucM/p+QJvACba7/mzwNnzYM6/BjzaxvsRencBzak5A5+gd03mb+kdBdwwnXMETgM+CRygd8fUhaOcj4/7kCR1mmunoSRJI2BYSJI6GRaSpE6GhSSpk2EhSepkWEiSOhkWkqRO/x/ncvHN3vnI3QAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "merged_xp_rna['diff_count'].plot.hist(xlim=(0, 11000))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7f509e9a12e8>"
      ]
     },
     "execution_count": 182,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZIAAAEICAYAAAB1f3LfAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8li6FKAAAgAElEQVR4nO3de3iU9Zn/8fc9kxBARDAgAgGhBesCRar5KS7VqrRK1YpdD2WtC9v6K1tXV9vuFqTdX0972Xroaf117a6rruLaKoW2sF2ptaC7davSYAMCSo1CJYCAaaREcUgy9/4xz+AkmUlmMocnyXxe15UrT+7nkO88trn5ns3dERER6a1I2AUQEZH+TYlERETyokQiIiJ5USIREZG8KJGIiEhelEhERCQvRU0kZnafme03sy0psTvM7EUz22xmPzGzESnnlplZg5ltN7MLU+Knm9nzwbk7zcyCeJWZPRLEnzWzScX8PCIi0pUVcx6JmZ0DtADL3X1GELsAWO/ubWZ2G4C7LzWzacAPgTOAccAvgZPdvd3MNgA3Ac8AjwJ3uvtaM/trYKa7f9rMFgAfdfeP9VSuUaNG+aRJkwr+eUVEBrKNGze+7u6jO8crivlL3f2/O9cS3P0XKT8+A1wRHM8HHnb3GLDDzBqAM8xsJzDc3Z8GMLPlwGXA2uCerwT3rwS+Z2bmPWTHSZMmUVdXl8cnExEpP2b2+3TxsPtIPkkiIQCMB3alnGsMYuOD487xDve4extwEKguYnlFRKST0BKJmX0RaAMeSobSXObdxLu7J93vW2xmdWZWd+DAgVyLKyIiGYSSSMxsEXAJ8PGUZqhGYELKZTXAniBekybe4R4zqwCOA/6Q7ne6+93uXuvutaNHd2niExGRXip5IjGzecBS4FJ3fyvl1BpgQTASazIwFdjg7nuBQ2Y2OxittRBYnXLPouD4ChKd+FqFUkSkhIra2W5mPwTOBUaZWSPwZWAZUAU8HozifcbdP+3uW81sBbCNRJPX9e7eHjzqOuB+YAiJPpVkv8q9wINBx/wfgAXF/DwiItJVUYf/9lW1tbWuUVvhaGqJ0dh8mJqRQ6geVhV2cUQkB2a20d1rO8eLWiMRSbW6fjdLV22mMhKhNR7n9stncums8T3fKCJ9WtjDf6VMNLXEWLpqM2+3xjkUa+Pt1jhLVm2mqSUWdtFEJE9KJFISjc2HqYx0/J9bZSRCY/PhkEokIoWiRCIlUTNyCK3xeIdYazxOzcghIZVIRApFiURKonpYFbdfPpPBlRGOrapgcGWE2y+fqQ53kQFAne1SMpfOGs+cKaM0aktkgFEikZKqHlalBCIywKhpS0RE8qJEIiIieVEiERGRvCiRiIhIXpRIREQkL0okIiJlYt2211i6chPrtr1W0Odq+K+ISBm44DtP8rt9bwLwSF0j7xlzDI999tyCPFs1EhGRAW7dtteOJpGk7fveLFjNRIlERGSA+8W2fTnFc6VEIiIywF0wbUxO8VwpkYiIDHBzp53Ie8Yc0yH2njHHMHfaiQV5vjrbRUTKwGOfPZd1217jF9v2ccG0MQVLIqBEIiJSNuZOO7GgCSRJTVsiIpIXJRIREcmLEomIiORFiURERPKiRCIiInkpaiIxs/vMbL+ZbUmJHW9mj5vZS8H3kSnnlplZg5ltN7MLU+Knm9nzwbk7zcyCeJWZPRLEnzWzScX8PCIi0lWxayT3A/M6xW4G1rn7VGBd8DNmNg1YAEwP7rnLzKLBPd8HFgNTg6/kM68Fmt19CvAd4LaifRIREUmrqInE3f8b+EOn8HzggeD4AeCylPjD7h5z9x1AA3CGmY0Fhrv70+7uwPJO9ySftRKYm6ytiIhIaYTRRzLG3fcCBN9PCOLjgV0p1zUGsfHBced4h3vcvQ04CFQXreQiItJFX+psT1eT8G7i3d3T9eFmi82szszqDhw40MsiiohIZ2Ekkn1BcxXB9/1BvBGYkHJdDbAniNekiXe4x8wqgOPo2pQGgLvf7e617l47evToAn0UEREJI5GsARYFx4uA1SnxBcFIrMkkOtU3BM1fh8xsdtD/sbDTPclnXQGsD/pRRESkRIq6aKOZ/RA4FxhlZo3Al4FbgRVmdi3wKnAlgLtvNbMVwDagDbje3duDR11HYgTYEGBt8AVwL/CgmTWQqIksKObnkcJpaonR2HyYmpFDqB5WFXZxRCQPVo7/gK+trfW6urqwi1G2VtfvZumqzVRGIrTG49x++UwunTW+5xtFJFRmttHdazvH+1Jnu5SBppYYS1dt5u3WOIdibbzdGmfJqs00tcTCLpqI9JISiZRUY/NhKiMd/2dXGYnQ2Hw4pBKJSL6USKSkakYOoTUe7xBrjcepGTkkpBKJSL6USKSkqodVcfvlMxlcGeHYqgoGV0a4/fKZ6nAX6ce01a6U3KWzxjNnyiiN2hIZIJRIJBTVw6qUQEQGCDVtScE0tcTYtOsNjcASKTOqkUhOMk0k1NwQkfKlRCJZe+iZ3/PV/9hKZTRCu/vRZJE6N+RtEiOylqzazJwpozokG81mFxmYlEgkKw8983u++NPERpdH2hMr1ySTRXJuSDKJwDtzQ5IJQzUWkYFLfSTSo6aWGF/92bYu8ajZ0RpGd3NDNJtdZGBTIpEeNTYfZlC069Yvre3xo81U3c0N0Wx2kYFNTVvSo5qRQ2iLd13c88sfmX40WXQ3N0Sz2UUGNtVIJCvXnzuFqooIx1RFGVQR4ZaPzuDjs0/qcE31sCpOnTCiS0e6ZrOLDGyqkUi3UjvJwfn0OVO4+syJOScBzWYXGbhUI5GMOneSx9qcf3qyodfPy1RjEZH+TYlEMlInuYhkQ4lEMlInuYhkQ4lEMlInuYhkQ53t0i11kotIT5RIJK3O62IpgYhIJkok0oXWxRKRXKiPRDrQulgikislEulAQ35FJFdKJNKBhvyKSK6USKQDDfkVkVyF1tluZp8F/i/gwPPAJ4ChwCPAJGAncJW7NwfXLwOuBdqBG939sSB+OnA/MAR4FLjJ3bsuVStZ05BfEclFKDUSMxsP3AjUuvsMIAosAG4G1rn7VGBd8DNmNi04Px2YB9xlZtHgcd8HFgNTg695JfwoA0pTS4xNu96gqSWmdbFEJGthDv+tAIaYWSuJmsgeYBlwbnD+AeBJYCkwH3jY3WPADjNrAM4ws53AcHd/GsDMlgOXAWtL9zEGBg35FZHeCqVG4u67gW8CrwJ7gYPu/gtgjLvvDa7ZC5wQ3DIe2JXyiMYgNj447hzvwswWm1mdmdUdOHCgkB+n39OQXxHJR1hNWyNJ1DImA+OAY8zsmu5uSRPzbuJdg+53u3utu9eOHj061yIPaBryKyL5CGvU1geBHe5+wN1bgR8DfwrsM7OxAMH3/cH1jcCElPtrSDSFNQbHneOSAw35FZF8hJVIXgVmm9lQMzNgLvACsAZYFFyzCFgdHK8BFphZlZlNJtGpviFo/jpkZrOD5yxMuUeypCG/IpKPUDrb3f1ZM1sJPAe0Ab8F7gaGASvM7FoSyebK4PqtZrYC2BZcf727twePu453hv+uRR3tvaIhvyLSW1aOUy5qa2u9rq4u7GKIiPQrZrbR3Ws7xzWzXURE8qJEIiIieVEiERGRvCiRiIhIXpRIREQkL0okclTqoo0iItnSnu1lqKkl1mW+iBZtFJHeUiIpM+kSxpwpo44u2vg2iaVSlqzazJwpozQxUUR6pKatMpJpld+tew5q0UYR6TUlkjLS2HyYikjHBZMTCcS0aKOI9JoSSRnZsvsgLbH2DrHWeJzp44Zr0UYR6TX1kZSJppYYX1q9pUv8/10yjephVVq0UUR6TYmkTNzzq1doT7M+54SRQ48eVw+rUgIRkZz1umnLzAab2ZWFLIwUR1NLjHuf2pHhbPmt/iwihZVTIjGzqJl92MyWA78HPlacYkkhNTYfpj3eNWFEIzB93HEhlEhEBpKsmrbM7BzgauBiYAMwB5js7m8VsWxSIK1t7WmbtW44b4qaskQkbz0mEjNrJLFb4feBz7v7ITPboSTSf2zZczBtPLV/RESkt7Jp2loFjCfRjPURMzsGNaz3G6vrd/P1R19Ie27WhBElLo2IDEQ9JhJ3vwmYBHwbOA/4HTDazK4ys2HFLZ7kIzmT/Uh713MLz5rIlDHHlr5QIjLgZNVH4omN3dcD682sEpgH/DlwFzCqeMWTfDQ2H6YyEjm6fhZAVTTCt646lUtOHRdiyURkIMlpHomZDQJOAXYCfwlEC18kKZSakUO6LH1iETjr3dUhlUhEBqKsh/+a2cXAy8CdwPeABuDc4hRLCqF6WJWWPhGRosulRvIt4Dx3bwAws3cD/wmsLUbBpDC09ImIFFsuiWR/MokEXgH2F7g8UiCdN69SAhGRYsklkWw1s0eBFSSG/14J/MbM/gzA3X9chPJJL6yu382SlZuJRoz2uHPHFdrtUESKJ5clUgYD+4APkOgbOQAcD3wEuCTXX2xmI8xspZm9aGYvmNlZZna8mT1uZi8F30emXL/MzBrMbLuZXZgSP93Mng/O3Wlmlv43loemlhh/96NNxNrivHWknVhbnL/90Sbtwy4iRZN1jcTdP9HdeTNb5u7fyOF3/yPwc3e/IhgNNhT4ArDO3W81s5uBm4GlZjYNWABMB8YBvzSzk929ncSM+8XAM8CjJIYml22/zdY9B2nttB5Ka7uzdc9Bzjn5hJBKJSIDWSE3tsp6JWAzGw6cA9wL4O5H3P0NYD7wQHDZA8BlwfF84GF3j7n7DhIjxs4ws7HAcHd/OpjrsjzlnrL09MtNGc6UdUVNRIqokIkkl79U7yLRNPZvZvZbM7snWHpljLvvBQi+J/8JPR7YlXJ/YxAbHxx3jnctnNliM6szs7oDBw7kUNT+o6klxn3/s7NLPGowfdzw0hdIRMpCIRNJLutvVQCnAd939/cBb5JoxsokXZLybuJdg+53u3utu9eOHj06h6L2H43NhxkU7fqf9DMfPLnPjNpq2HeIlXW7aNh3KOyiiEiBFHKHxFxqJI1Ao7s/G/y8kkQi2WdmY919b9BstT/l+gkp99cAe4J4TZp4WUo3k72qIsLVZ04MqUQdfemnz7P8mVeP/rzwrIl8bf57QyyRiBRCIWskP8r2Qnd/DdhlZu8JQnOBbcAaYFEQWwSsDo7XAAvMrMrMJgNTgQ1B89chM5sdjNZamHJP2dlxoIUPTB3FoKgdncl+xxV9YyZ7w75DHZIIwPKnX1XNRGQAyGY/kuOBG0j8S/9eEiOrzgJeAL7u7s0A7v71HH/33wAPBSO2XgE+QSKxrTCza0nsgXJl8OytZraCRLJpA64PRmwBXAfcDwwhMVqrLEdsXXPPMzzV8E5H+7RxQ7l30Rl9IokA1O96I2NcqxCL9G/ZNG39O/A8cDpwTXB8G/AhEn/A5/fmF7t7PVCb5tTcDNffAtySJl4HzOhNGQaKuh1NHZIIQP2uP7LjQEufSSSZ9j7Rnigi/V82iWScu18UNB01uvu5QfxXZlZfvKJJtv77pdczxmsn942VfqeMOZaFZ01k+dMd+0hUGxHp/7JJJJFghvmxwDAzm+TuO82sGhhU3OJJNjKNcjhnat/aKuZr89/LwtmTqN/1BrMmjFASERkgskkk3wBeDI4/CdwTrELyJ8BXi1QuyVJTS4x/+dUrXeKzJ4/sM7WRVFPGHKsEIjLA9JhI3P2HQUe3uXubma0GZgG7k5MHJTzpdkEcUmksu2haiKUSkXLS4/DfYFRV3N3bgtDZJBZunFXMgkl2jhkUJdbWcVN2x6gZOSSkEolIuclmHslvgBEAZvZ5EiOnhgCfM7NcFmmUAltdv5uL//9TJHtJqqKmXRBFpOSy6SOJJueKAB8Dznb3w2Z2K/AcsKxopZOMksvFp6702+7w8xverz4IESmpbGokfzSz5DyN10nsSwKJJFTImfGSg3TLxbfFnT0HD4dUIhEpV9nUSD5NYgb6JhJrX9WZ2X8BM4FcZ7NLgdzzqx1p47ublUhEpLSyGbW12cxOAy4ATgY2kVgs8XPBHiJSYg37DmWchPjlNds4pqpCW+uKSMlktfpvsK5V2a5j1dfc9eTLGc8daY+zZNVm5kwZpQ53ESmJvPo4zEyJpcSaWmKs2bS722sqIxEa1cQlIiWSzeq/p2U6heaSlFxj82GqKiK0HYlnvKY1Htc8EhEpmWyatn4D/Bfpl3TS0q0ltmX3Qd5Mk0Qqo8bgiiit8XhW80iaWmI0Nh+mZuQQNYGJSF6ySSQvAH/l7i91PmFmu9JcL0XS1BLjS6u3dIl/4aJTuPy0mqwTw+r63SxdtZnKSORo4lHnvIj0VjaJ5Ctk7kv5m8IVRXryr796hfY0O9KfcuJwqodVZVWzaGqJsXTVZt5ujR9dn0ud8yKSjx472919pbtvz3B6ZIHLIxk0tcS496n0c0cgTXbJILnIYyp1zotIPvKdma5l5Euksfkwg6Jd/3NVRIzp4447+nNTS4xNu96gqSWW9jk1I4fQGu/Yx6LOeRHJRzajtjZnOgWMKWxxJJNjBkVpbe/ayf7V+dOPNkll0/dRPayK2y+fyZJO16lZS0R6K5s+kjHAhUBzp7gBvy54iaSL1fW7WbJyM4lX7gyKJgbQffkj0/n4mScBufV9XDprPHOmjNKoLREpiGwSyc+AYe7eZX92M3uy4CWSDtKt8ht3+PlNZ3dY5TfdBlfJvo90iSLbznkRkZ5k09l+rbs/leHc1YUvkqTKdpVf9X2ISFi0DHyfl24eaNd4su9jcGWEY6sqtMGViJRMVos2SnimjxtORQTaUiobFZFEvDP1fYhIGFQj6Qdumnsyg6IwtDJKVYXx7atmZUwS1cOqOHXCCCURESmZUGskZhYF6oDd7n6JmR0PPAJMAnYCVyW3+TWzZcC1QDtwo7s/FsRPB+4nsY/8o8BN7p79DL0+LDlaKxoxzCJcd+67ufrMiUoSItKnhF0juYnEWl5JNwPr3H0qsC74GTObBiwApgPzgLuCJATwfWAxMDX4mleaohdXcrRWrC3OW0faibXFuXN9l+XORERCF1oiMbMa4GLgnpTwfOCB4PgB4LKU+MPuHnP3HUADcIaZjQWGu/vTQS1keco9/Vq60Vqt7c7WPQdDKpGISHph1ki+CywBUsesjnH3vQDB9xOC+HggdaXhxiA2PjjuHO/CzBabWZ2Z1R04cKAwn6CIblv7YoYzmUZxiYiEI5REYmaXAPvdfWO2t6SJeTfxrkH3u9291t1rR48eneWvDUfdjia27j3UJR4h/Wit7vS09hYk9oBfWbeLhn1df6eISE/C6myfA1xqZhcBg4HhZvbvwD4zG+vue4Nmq/3B9Y3AhJT7a4A9QbwmTbxf+9dfvZI2/qFpJ+TU0Z7N2ltf+unzLH/m1aM/LzxrIl+b/97eFVxEylIoNRJ3X+buNe4+iUQn+np3vwZYAywKLlsErA6O1wALzKzKzCaT6FTfEDR/HTKz2WZmwMKUe/qlppYY617Yn/bcp85+V07PSa69dSjWxtutcZas2tyhZrJu22sdkgjA8qdfVc1ERHLS1yYk3gqsMLNrgVeBKwHcfauZrQC2AW3A9e7eHtxzHe8M/10bfPVbf//TLbSlaZybNWE4tZOrs35OT2tv3fiDjazZ/Frae+t3vdFhHS8Rke6Enkjc/UngyeC4CZib4bpbgFvSxOuAGcUrYek07DvE2i1d/7hHgG9eMSunZ6Vbe+tIe5yDh1v5q+V1PLZtX8Z7Rw6tzOl3iUh5C3seiaT47i9/lzYeiRiXfO8p1tTvTns+XYd657W3KqNGezzOpx/sPokANL/V2vsPISJlJ/QaiSQ0tcT42fPpm5ra4k5b3NPuL9Jdh3py7a2tew7yqeV1xNqhLd51c6zOZk0YUZgPJSJlQTWSPuKOn2eaN/KOznurZ9OhXj2siuOGDGJQNJrukV0sPGui+kdEJCeqkfQBTS0xVtQ19nhd5/1FstnMqqklxsHDR3i7tS3jc89/zygueu84Zk0YoSQiIjlTIukDHnr2VTI1OA2KGlUV0bR7q/e0mVWy2as97rSmXGbAsg+fwvHHDFLyEJG8KZGErKklxp3rMi/G+OiNZ/Pmkfa0+4skO9SXdOojqR5W1aHZqzMHzj/lBCUQESkIJZKQbd1zkLZ4+lXvqyoivHmknVO76fzOtJnV1j0HiVjmdbk0V0RECkWJJGR3PdGQ8Zy7Z7XnevWwqi4juZas3EQs3czGgEZmiUihaNRWiBr2HeKZHc0Zz3/zylNz3sQq2aTVXRLRyCwRKSTVSEK0eHldxnPf+OiMLgssZiPdSK5Un7/gZK4/f2rOzxURyUQ1kpDU7Wjilaa3Mp6/YPqJWS0B31lrWztvHsk81PfC6SfmVE4RkZ6oRhKS+3+9I+O5Wy6bwVMNr/e4BHxn3S3ECGrSEpHiUCIJydOvNKWNTxgxmHkzTmTObet5uzV+tIkq3fIoSU0tMa5/aGPG/pbPXzCVC6ePVRIRkaJQIglBw75DNL2Zvvlp1oTjspqxnrS6fjeffbg+44RGgDHDhyiJiEjRqI8kBDf8IPMOw4v+dHKPM9aTmlpiLFm5qdskAhrqKyLFpURSYg37DvHivjfTnju2KkLt5OouS8AProx0WR4FEiO0PMNkxiT1i4hIsalpq8Tqd72R8dxZ7xp19DjTjPVUX/uPLRzJUB2ZMe5Yvvux9ymJiEjRKZGU2Lce357x3BmTj+/wc+cZ66m+84vtbHz1YNpzX/jwKSz+wLt7X0gRkRyoaauE1m17jb0HM88JOfG4npdDgUTfyPcyLK3yqbMnK4mISEmpRlJCP9q4q9vzw4dk/s/R1BKjsfkwxwyKcteTL9OeoWvkwmlj8imiiEjOlEhK6Odb92c8FzGYPu64tOeS+4p43IllyiDA7MkjqZ1cnXc5RURyoURSIpf845Pdnv+Hy2Z06A9JrYFk2lck1YLaGm694tRCFFVEJCdKJCXQ1BJjy970Q34BznvPKD5+5klHf07WQCojEWLtccy7H+JbGYXPzzulYOUVEcmFOttL4CfPdb8f+69fbjq6MGPqzoaHYm0caYt325wF8JWPzMh5uXkRkUJRIimBbzz6YvcXuNPYfBh4Zxn4bEQjxi0fncHHZ5/U88UiIkUSSiIxswlm9oSZvWBmW83spiB+vJk9bmYvBd9HptyzzMwazGy7mV2YEj/dzJ4Pzt1p1s3+siH46XO7aO/hmlg7HDx8hKaWWNrlUdKJGvzbX/6fDk1iIiJhCKtG0gb8rbv/CTAbuN7MpgE3A+vcfSqwLviZ4NwCYDowD7jLzKLBs74PLAamBl/zSvlBenLr2heyuu6vHtzIn966np9vfY3zTxnd4/UGTB83PM/SiYjkL5TOdnffC+wNjg+Z2QvAeGA+cG5w2QPAk8DSIP6wu8eAHWbWAJxhZjuB4e7+NICZLQcuA9aW7MN0o6klxmuHWrO69nAwKuuLP9mS1fVfna9+ERHpG0LvIzGzScD7gGeBMUGSSSabE4LLxgOps/kag9j44LhzvE946JmdBX9m1FC/iIj0KaEO/zWzYcAq4DPu/sduujfSnfBu4ul+12ISTWBMnDgx98L2wrd/mX4Zk96qjMDam87RQowi0qeEViMxs0oSSeQhd/9xEN5nZmOD82OB5FTwRmBCyu01wJ4gXpMm3oW73+3ute5eO3p0z30Q+fqXJ14q+DNvnHuykoiI9Dlhjdoy4F7gBXf/dsqpNcCi4HgRsDolvsDMqsxsMolO9Q1B89chM5sdPHNhyj2h+vYvf1fQ51VGjavPLE1NSkQkF2E1bc0B/gJ43szqg9gXgFuBFWZ2LfAqcCWAu281sxXANhIjvq539+So2uuA+4EhJDrZQ+9ob9h3iFhPY35z9JVLp6tzXUT6pLBGbT1F+v4NgLkZ7rkFuCVNvA6YUbjS5e/vf/J8QZ/3qbMna76IiPRZoY/aGoie2dlcsGedPvE4vnjxtII9T0Sk0JRICuzvHnmuYM/6izMnsOqv31+w54mIFIMSSYGt/O3egjznjJNG8g8fnZnzfU0tMTbteuPoIpAiIsWmZeQL6KfPdb8DYi6+/mfvzfme1OXnW+Nxbr98JpfO6jPzM0VkgFKNpIA+s2JzQZ7z4eljcp4v0nn5+bdb4yxZtVk1ExEpOiWSAlm37bWCPWvun+S+73q65ecrI5Gjy9OLiBSLEkmBfPrBjQV71qwJI3K+J93y863xODUjhxSqWCIiaSmRFEhr95sYZi0aMbbt/WPO91UPq+L2y2cyuDLCsVUVDK6McPvlMzWJUUSKTp3tBfDBb64v2LPa486SVZuZM2VUzkng0lnjmTNlFI3Nh6kZOURJRERKQomkABpez68fYlhVlJaUNVWSfRu9SQTVw6qUQESkpNS0lad8h/yef3I1bfGO7WLJvg3NCRGR/kA1kjzlO+T3vk/OZk39bpZ0mv/xVMPrmhMiIv2CEkmIdt56MdC1bwNgzm3rebs1ztskRmL1tt9ERKTYlEjyMOnm/+z1vRv//oMdfk7t29i06w0qI5GjSQTy6zcRESkm9ZGE4AsfPqXbhKA5ISLSnyiR9NKUXtZGTh49lMUfeHe312hOiIj0J2ra6qW2Xt53x1Xv6/Z8U0uMxubDzJkyiv9Zer7mhIhIn6dEkqOmlhjffbx3+7FXVUS6bZ5aXb+bJSs3EbUI7R7njitO1UgtEenzlEhykPhDv5lYW7zni9M4/5TMo66aWmL87Yp6Eo9OTE783Ip6jdQSkT5PfSRZamqJ8Xc/2tTrJAKwdst+lqysTzvJcOueP9L50W3xRFxEpC9TjSRLW/ccpLU9/5UZV9Tt5j83v0Zb3LnhvClcfebEoMaR6dnZ/c5k34r6U0Sk1JRIsrS7gPt6vHkk0XT1rcd/x/eeeIk7rjiVOVNGURm1DsmqMmpMH3dcj8/TzogiEiY1bWWpsfmtojw31pZY7RfgW1eeSlVFhKGDolRVRPjWlaf2WLvQzogiEjbVSLI0dFDxXlVy1npvloFP7oyoWfAiEhYlkiydcmJue6jnInXWeudl4Hvq+9AseBEJ24Bo2jKzeWa23cwazOzmYvyOa5cXbivdpN+IpG4AAAa5SURBVKGDot3OWl9dv5s5t63nmnueZc5t61lTv7vLNZoFLyJh6/c1EjOLAv8EfAhoBH5jZmvcfVu4JeteVYXxz9ecxvRxx6X9o5/a99HTCsDaGVFEwtTvEwlwBtDg7q8AmNnDwHygYIkkn1V+kyJARUWEqug7I6vOOfmEjNfn2vehnRFFJCwDIZGMB1K3KWwEzgypLBkNqozwsxvez5tH2rOqNajvQ0T6i4HQR2JpYl1m8ZnZYjOrM7O6AwcOlKBY76iqMG6/fCZTxhzLqRNGZFVzUN+HiPQXA6FG0ghMSPm5BtjT+SJ3vxu4G6C2tjb/KeoZRA0qohEGRSMcaY93mr2eG/V9iEh/MBASyW+AqWY2GdgNLACuLuQv2HnrxVn1k0QMvvOxWQX946++DxHp6/p9InH3NjO7AXgMiAL3ufvWQv+edMlk7imj+MjMcWzZc4gpo4/hQ9NPPPpHX3/8RaRcmHvRWnn6rNraWq+rqwu7GCIi/YqZbXT32s7xgdDZLiIiIVIiERGRvCiRiIhIXpRIREQkL0okIiKSl7IctWVmB4Df9/L2UcDrBSzOQKH3kpneTXp6L+n15fdykruP7hwsy0SSDzOrSzf8rdzpvWSmd5Oe3kt6/fG9qGlLRETyokQiIiJ5USLJ3d1hF6CP0nvJTO8mPb2X9Prde1EfiYiI5EU1EhERyYsSSQ7MbJ6ZbTezBjO7OezyFIOZ3Wdm+81sS0rseDN73MxeCr6PTDm3LHgf283swpT46Wb2fHDuTjOzIF5lZo8E8WfNbFIpP19vmdkEM3vCzF4ws61mdlMQL+t3Y2aDzWyDmW0K3stXg3hZv5ckM4ua2W/N7GfBzwPzvbi7vrL4IrFE/cvAu4BBwCZgWtjlKsLnPAc4DdiSErsduDk4vhm4LTieFryHKmBy8H6iwbkNwFkkdrBcC3w4iP818M/B8QLgkbA/c5bvZSxwWnB8LPC74POX9bsJPsOw4LgSeBaYXe7vJeX9fA74AfCz4OcB+V5Cf9H95Sv4D/lYys/LgGVhl6tIn3VSp0SyHRgbHI8Ftqd7ByT2hDkruObFlPifA/+Sek1wXEFi4pWF/Zl78Y5WAx/Su+nwToYCzwFn6r04JHZrXQecn5JIBuR7UdNW9sYDu1J+bgxi5WCMu+8FCL6fEMQzvZPxwXHneId73L0NOAhUF63kRRA0IbyPxL++y/7dBM039cB+4HF313tJ+C6wBIinxAbke1EiyZ6liZX7kLdM76S7d9Wv36OZDQNWAZ9x9z92d2ma2IB8N+7e7u6zSPwL/Awzm9HN5WXxXszsEmC/u2/M9pY0sX7zXpRIstcITEj5uQbYE1JZSm2fmY0FCL7vD+KZ3kljcNw53uEeM6sAjgP+ULSSF5CZVZJIIg+5+4+DsN5NwN3fAJ4E5qH3Mge41Mx2Ag8D55vZvzNA34sSSfZ+A0w1s8lmNohE59aakMtUKmuARcHxIhL9A8n4gmD0yGRgKrAhqLIfMrPZwQiThZ3uST7rCmC9B428fVnwOe4FXnD3b6ecKut3Y2ajzWxEcDwE+CDwImX+Xtx9mbvXuPskEn8r1rv7NQzU9xJ2h1R/+gIuIjFa52Xgi2GXp0if8YfAXqCVxL94riXR7roOeCn4fnzK9V8M3sd2gtEkQbwW2BKc+x7vTH4dDPwIaCAxGuVdYX/mLN/L+0k0G2wG6oOvi8r93QAzgd8G72UL8KUgXtbvpdM7Opd3OtsH5HvRzHYREcmLmrZERCQvSiQiIpIXJRIREcmLEomIiORFiURERPKiRCLST5nZZ8xsaNjlENHwX5F+Kpg1Xevur4ddFilvqpGIFJGZLTSzzcF+HQ+a2Ulmti6IrTOzicF195vZFSn3tQTfzzWzJ81spZm9aGYPWcKNwDjgCTN7IpxPJ5JQEXYBRAYqM5tOYrbyHHd/3cyOBx4Alrv7A2b2SeBO4LIeHvU+YDqJNZb+J3jenWb2OeA81UgkbKqRiBTP+cDK5B96d/8DiT0mfhCcf5DE0is92eDuje4eJ7E0y6QilFWk15RIRIrH6HlZ7+T5NoL/PwaL8w1KuSaWctyOWhKkj1EiESmedcBVZlYNif26gV+TWA0W4OPAU8HxTuD04Hg+iW1re3KIxLa/IqHSv2xEisTdt5rZLcB/mVk7iVVybwTuM7PPAweATwSX/yuw2sw2kEhAb2bxK+4G1prZXnc/r/CfQCQ7Gv4rIiJ5UdOWiIjkRYlERETyokQiIiJ5USIREZG8KJGIiEhelEhERCQvSiQiIpIXJRIREcnL/wICB/CnHe70dwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#merged_xp_rna.plot.scatter(x = 'count', y = sample_name, xlim=(0, 1000), ylim=(0, 1000))\n",
    "merged_xp_rna.plot.scatter(x = 'count', y = sample_name,)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Show counts for genes of interest: **Heterochromatic Genes**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene-id</th>\n",
       "      <th>count</th>\n",
       "      <th>length</th>\n",
       "      <th>type</th>\n",
       "      <th>category</th>\n",
       "      <th>bio_type</th>\n",
       "      <th>1168_pA</th>\n",
       "      <th>diff_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2047</th>\n",
       "      <td>dh1</td>\n",
       "      <td>11.250000</td>\n",
       "      <td>4000.0</td>\n",
       "      <td>gene</td>\n",
       "      <td>repeat</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>37.0</td>\n",
       "      <td>25.750000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2049</th>\n",
       "      <td>dg1</td>\n",
       "      <td>8.611111</td>\n",
       "      <td>3500.0</td>\n",
       "      <td>gene</td>\n",
       "      <td>repeat</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>30.0</td>\n",
       "      <td>21.388889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4138</th>\n",
       "      <td>MAT1</td>\n",
       "      <td>209.000000</td>\n",
       "      <td>1500.0</td>\n",
       "      <td>gene</td>\n",
       "      <td>repeat</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>211.0</td>\n",
       "      <td>2.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4151</th>\n",
       "      <td>MAT2</td>\n",
       "      <td>40.000000</td>\n",
       "      <td>4000.0</td>\n",
       "      <td>gene</td>\n",
       "      <td>repeat</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>35.0</td>\n",
       "      <td>5.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4155</th>\n",
       "      <td>MAT3</td>\n",
       "      <td>188.500000</td>\n",
       "      <td>4000.0</td>\n",
       "      <td>gene</td>\n",
       "      <td>repeat</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>182.0</td>\n",
       "      <td>6.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5461</th>\n",
       "      <td>after_tlh</td>\n",
       "      <td>108.000000</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>gene</td>\n",
       "      <td>repeat</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>111.0</td>\n",
       "      <td>3.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene-id       count  length  type category        bio_type  1168_pA  \\\n",
       "2047        dh1   11.250000  4000.0  gene   repeat  protein_coding     37.0   \n",
       "2049        dg1    8.611111  3500.0  gene   repeat  protein_coding     30.0   \n",
       "4138       MAT1  209.000000  1500.0  gene   repeat  protein_coding    211.0   \n",
       "4151       MAT2   40.000000  4000.0  gene   repeat  protein_coding     35.0   \n",
       "4155       MAT3  188.500000  4000.0  gene   repeat  protein_coding    182.0   \n",
       "5461  after_tlh  108.000000  5000.0  gene   repeat  protein_coding    111.0   \n",
       "\n",
       "      diff_count  \n",
       "2047   25.750000  \n",
       "2049   21.388889  \n",
       "4138    2.000000  \n",
       "4151    5.000000  \n",
       "4155    6.500000  \n",
       "5461    3.000000  "
      ]
     },
     "execution_count": 183,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_xp_rna[merged_xp_rna['gene-id'].isin(htc_genes)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tests"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Find **read alignments** mapping to a specific set of features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "# instantiate `GenomicArrayOfSets` for the `exons` features: (for `RNA` **stranded**)\n",
    "#test_features = HTSeq.GenomicArrayOfSets(\"auto\", stranded=True)\n",
    "test_features = HTSeq.GenomicArrayOfSets(\"auto\", stranded=False)\n",
    "\n",
    "genes_dict = {}\n",
    "\n",
    "## loop over all features in gtf file\n",
    "for feature in gtf_file:\n",
    "    \n",
    "    ## store all exons in our `GenomicArrayOfSets`\n",
    "    if feature.type == \"exon\":\n",
    "        \n",
    "        ## identify each `exon` feature by parent transcript/gene\n",
    "        gene_id = feature.attr[\"Parent\"].split(':')[1][:-2] ## get rid of '.1'\n",
    "        \n",
    "        if gene_id in htc_genes:\n",
    "\n",
    "            test_features[ feature.iv ] += gene_id\n",
    "\n",
    "            # Is this the first time we see this gene?\n",
    "            if gene_id not in genes_dict:\n",
    "                # If so, add to the 'genes_dict' an empty list \n",
    "                genes_dict[ gene_id ] = list()\n",
    "\n",
    "            # add the feature to the gene list\n",
    "            genes_dict[ gene_id ].append( feature )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'dg1': [<GenomicFeature: exon 'gene:dg1' at I: 3764499 -> 3760998 (strand '-')>],\n",
       " 'dh1': [<GenomicFeature: exon 'gene:dh1' at I: 3756999 -> 3752998 (strand '-')>],\n",
       " 'after_tlh': [<GenomicFeature: exon 'gene:after_tlh' at II: 4538999 -> 4533998 (strand '-')>],\n",
       " 'MAT2': [<GenomicFeature: exon 'gene:MAT2' at II: 2132999 -> 2128998 (strand '-')>],\n",
       " 'MAT3': [<GenomicFeature: exon 'gene:MAT3' at II: 2136999 -> 2132998 (strand '-')>],\n",
       " 'MAT1': [<GenomicFeature: exon 'gene:MAT1' at II: 2115499 -> 2113998 (strand '-')>]}"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "overlap_feature = genes_dict['MAT2'][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [],
   "source": [
    "#sample_name = '1113_pA'\n",
    "sample_name = '1168_pA'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100000 alignment records processed. 1.45521879196167 s\n",
      "200000 alignment records processed. 2.9313876628875732 s\n",
      "300000 alignment records processed. 4.411110877990723 s\n",
      "400000 alignment records processed. 5.869096279144287 s\n",
      "500000 alignment records processed. 7.261061906814575 s\n",
      "600000 alignment records processed. 8.715775966644287 s\n",
      "700000 alignment records processed. 10.164191246032715 s\n",
      "800000 alignment records processed. 11.5581636428833 s\n",
      "900000 alignment records processed. 13.015530586242676 s\n",
      "1000000 alignment records processed. 14.428688287734985 s\n",
      "1100000 alignment records processed. 15.822635650634766 s\n",
      "1200000 alignment records processed. 17.230992317199707 s\n",
      "1300000 alignment records processed. 18.653667211532593 s\n",
      "1400000 alignment records processed. 20.069377422332764 s\n",
      "1500000 alignment records processed. 21.5571928024292 s\n",
      "1600000 alignment records processed. 23.057801485061646 s\n",
      "1700000 alignment records processed. 24.540045738220215 s\n",
      "1800000 alignment records processed. 26.050745964050293 s\n",
      "1900000 alignment records processed. 27.51209855079651 s\n",
      "2000000 alignment records processed. 29.02902102470398 s\n",
      "2100000 alignment records processed. 30.548490047454834 s\n",
      "2200000 alignment records processed. 32.05630040168762 s\n",
      "2300000 alignment records processed. 33.54954385757446 s\n",
      "2400000 alignment records processed. 35.048887729644775 s\n",
      "2500000 alignment records processed. 36.497315406799316 s\n",
      "2600000 alignment records processed. 37.90671467781067 s\n",
      "2700000 alignment records processed. 39.29213500022888 s\n",
      "2800000 alignment records processed. 40.65205693244934 s\n",
      "2900000 alignment records processed. 42.0492639541626 s\n",
      "3000000 alignment records processed. 43.45267939567566 s\n",
      "3100000 alignment records processed. 44.879032135009766 s\n",
      "3200000 alignment records processed. 46.29083013534546 s\n",
      "3300000 alignment records processed. 47.697195053100586 s\n",
      "3400000 alignment records processed. 49.12601852416992 s\n",
      "3500000 alignment records processed. 50.54462218284607 s\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Elapsed Time (Counting reads): 50.6607940196991\n"
     ]
    }
   ],
   "source": [
    "start_time = time.time()\n",
    "\n",
    "# select bam file: instantiate BAM_Reader Object (RNA-seq)\n",
    "bam_file = HTSeq.BAM_Reader(os.path.join('/gcm-lfs1/pablo/data/RNAdeg/data/RNA/bams', sample_name, sample_name + '.Aligned.sortedByCoord.out.bam'))\n",
    "\n",
    "## -------\n",
    "## Options\n",
    "## ------- \n",
    "\n",
    "i = 0\n",
    "\n",
    "reads = {}\n",
    "\n",
    "#for aln in itertools.islice(bam_file, 1600000, 1700000):\n",
    "for aln in bam_file:\n",
    "    \n",
    "    #import pdb\n",
    "    #pdb.set_trace()\n",
    "\n",
    "    if i > 0 and i % 100000 == 0:\n",
    "        sys.stderr.write(\"{} alignment records processed. {} s\\n\".format(i,  time.time() - start_time))\n",
    "        sys.stderr.flush()\n",
    "    i += 1\n",
    "                \n",
    "    ## ----------------------\n",
    "    ## Inspect read alignment\n",
    "    ## ----------------------\n",
    "    \n",
    "    ## invert strand - due to sequencing is strand is reversed!\n",
    "    iv_seq = (invert_strand(co.ref_iv) for co in aln.cigar if (co.type in com and co.size > 0))\n",
    "    #iv_seq = (co.ref_iv for co in aln.cigar if (co.type in com and co.size > 0))\n",
    "    \n",
    "    gene_ids = set()\n",
    "    gene_ids_reduced = set()\n",
    "    \n",
    "    alignment_fs = {}\n",
    "        \n",
    "    jj = 0\n",
    "    ## loop over CIGAR operations (cig_op) for aligned read \n",
    "    for iv in iv_seq:\n",
    "        \n",
    "        jj += 1\n",
    "\n",
    "        if (iv.chrom == overlap_feature.iv.chrom) and (iv.end >= overlap_feature.iv.start)  and (iv.start <= overlap_feature.iv.end):\n",
    "            \n",
    "            #if jj >1:\n",
    "            #    import pdb\n",
    "            #    pdb.set_trace()\n",
    "            \n",
    "            fs = set()\n",
    "                         \n",
    "            for iv2, fs2 in exon_features[ iv ].steps():\n",
    "                gene_ids = gene_ids.union(fs2)\n",
    "                #print(iv2, fs2)\n",
    "                \n",
    "            for iv2, fs2 in test_features[ iv ].steps():\n",
    "                gene_ids_reduced = gene_ids_reduced.union(fs2)\n",
    "                fs = fs.union(fs2)\n",
    "                #print(iv2, fs2)\n",
    "            \n",
    "            alignment_fs[jj] = fs\n",
    "\n",
    "    if len(alignment_fs) > 0:\n",
    "        \n",
    "        alignment_fs['gene_ids'] = gene_ids\n",
    "        alignment_fs['gene_ids_reduced'] = gene_ids_reduced\n",
    "        \n",
    "        alignment_fs['NH']= aln.optional_field(\"NH\")\n",
    "        alignment_fs['alignment'] = aln\n",
    "\n",
    "        read_name = aln.read.name\n",
    "        \n",
    "        #if read_name not in reads.keys():\n",
    "        assert read_name not in reads.keys()\n",
    "        #reads[read_name] = {}   \n",
    "        reads[read_name] = alignment_fs\n",
    "\n",
    "    #else:\n",
    "    #    ky = '-'.join(sorted(list(gene_ids_reduced)))\n",
    "\n",
    "    #    if ky not in multi_aligned_reads.keys():\n",
    "    #        multi_aligned_reads[ky] = {}\n",
    "    #    multi_aligned_reads[ky][aln.read.name] = aln\n",
    "\n",
    "\n",
    "print('Elapsed Time (Counting reads):', time.time() - start_time)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "36"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(reads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['L183:365:CBF20ANXX:6:2209:9784:51525:1:N:0:0:CAGGCG', 'L183:365:CBF20ANXX:6:2209:11731:8029:1:N:0:0:CAGGCG', 'L183:365:CBF20ANXX:6:1210:3334:51173:1:Y:0:0:CAGGCG', 'L183:365:CBF20ANXX:6:1307:7211:69628:1:N:0:0:CAGGCG', 'L183:365:CBF20ANXX:6:2110:4363:14087:1:N:0:0:CAGGCG', 'L183:365:CBF20ANXX:6:2108:20607:14348:1:N:0:0:CAGGCG', 'L183:365:CBF20ANXX:6:2301:5203:62810:1:N:0:0:CAGGCG', 'L183:365:CBF20ANXX:6:2209:3431:99503:1:N:0:0:CAGGCG', 'L183:365:CBF20ANXX:6:1305:13966:12841:1:N:0:0:CAGGCG', 'L183:365:CBF20ANXX:6:1113:5035:21196:1:N:0:0:CAGGCG', 'L183:365:CBF20ANXX:6:1206:8002:38821:1:N:0:0:CAGGCG', 'L183:365:CBF20ANXX:6:1204:18755:41536:1:N:0:0:CAGGCG', 'L183:365:CBF20ANXX:6:1316:4551:6124:1:N:0:0:CAGGCG', 'L183:365:CBF20ANXX:6:1201:18976:23791:1:N:0:0:CAGGCG', 'L183:365:CBF20ANXX:6:2303:6998:82477:1:N:0:0:CAGGCG', 'L183:365:CBF20ANXX:6:1302:16959:84250:1:N:0:0:CAGGCG', 'L183:365:CBF20ANXX:6:1107:9334:70256:1:N:0:0:CAGGCG', 'L183:365:CBF20ANXX:6:2207:13785:62947:1:N:0:0:CAGGCG', 'L183:365:CBF20ANXX:6:1210:3103:30312:1:N:0:0:CAGGCG', 'L183:365:CBF20ANXX:6:2104:18462:99232:1:N:0:0:CAGGCG', 'L183:365:CBF20ANXX:6:1201:18781:56417:1:N:0:0:CAGGCG', 'L183:365:CBF20ANXX:6:1102:19765:81623:1:N:0:0:CAGGCG', 'L183:365:CBF20ANXX:6:1110:19280:36437:1:N:0:0:CAGGCG', 'L183:365:CBF20ANXX:6:1111:12241:94710:1:N:0:0:CAGGCG', 'L183:365:CBF20ANXX:6:2315:11572:31571:1:N:0:0:CAGGCG', 'L183:365:CBF20ANXX:6:2308:18909:12227:1:N:0:0:CAGGCG', 'L183:365:CBF20ANXX:6:2310:15406:29888:1:N:0:0:CAGGCG', 'L183:365:CBF20ANXX:6:2315:12903:95324:1:N:0:0:CAGGCG', 'L183:365:CBF20ANXX:6:2205:8888:5036:1:N:0:0:CAGGCG', 'L183:365:CBF20ANXX:6:2306:12980:27363:1:N:0:0:CAGGCG', 'L183:365:CBF20ANXX:6:2108:5839:51369:1:N:0:0:CAGGCG', 'L183:365:CBF20ANXX:6:1309:13642:18793:1:N:0:0:CAGGCG', 'L183:365:CBF20ANXX:6:2216:7032:15866:1:N:0:0:CAGGCG', 'L183:365:CBF20ANXX:6:2202:11872:10564:1:N:0:0:CAGGCG', 'L183:365:CBF20ANXX:6:2213:18606:100846:1:N:0:0:CAGGCG', 'L183:365:CBF20ANXX:6:2102:10490:25392:1:N:0:0:CAGGCG'])"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reads.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{1: {'MAT2'},\n",
       " 'gene_ids': {'MAT2', 'SPNCRNA.389'},\n",
       " 'gene_ids_reduced': {'MAT2'},\n",
       " 'NH': 2,\n",
       " 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:2209:11731:8029:1:N:0:0:CAGGCG' aligned to II:[2129911,2129961)/->}"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reads['L183:365:CBF20ANXX:6:2209:11731:8029:1:N:0:0:CAGGCG']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{2: {'MAT2'}, 'gene_ids': {'MAT2'}, 'gene_ids_reduced': {'MAT2'}, 'NH': 4, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:2209:9784:51525:1:N:0:0:CAGGCG' aligned to II:[1104943,2131897)/+>}\n",
      "{1: {'MAT2'}, 'gene_ids': {'MAT2', 'SPNCRNA.389'}, 'gene_ids_reduced': {'MAT2'}, 'NH': 2, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:2209:11731:8029:1:N:0:0:CAGGCG' aligned to II:[2129911,2129961)/->}\n",
      "{1: {'MAT2'}, 'gene_ids': {'MAT2', 'SPNCRNA.389'}, 'gene_ids_reduced': {'MAT2'}, 'NH': 2, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:1210:3334:51173:1:Y:0:0:CAGGCG' aligned to II:[2129925,2129975)/->}\n",
      "{1: {'MAT2'}, 'gene_ids': {'MAT2', 'SPNCRNA.389'}, 'gene_ids_reduced': {'MAT2'}, 'NH': 2, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:1307:7211:69628:1:N:0:0:CAGGCG' aligned to II:[2130452,2130502)/->}\n",
      "{1: {'MAT2'}, 'gene_ids': {'MAT2'}, 'gene_ids_reduced': {'MAT2'}, 'NH': 2, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:2110:4363:14087:1:N:0:0:CAGGCG' aligned to II:[2131243,2131293)/->}\n",
      "{1: {'MAT2'}, 'gene_ids': {'MAT2'}, 'gene_ids_reduced': {'MAT2'}, 'NH': 2, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:2108:20607:14348:1:N:0:0:CAGGCG' aligned to II:[2132467,2132517)/+>}\n",
      "{1: {'MAT2'}, 'gene_ids': {'MAT2'}, 'gene_ids_reduced': {'MAT2'}, 'NH': 2, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:2301:5203:62810:1:N:0:0:CAGGCG' aligned to II:[2132468,2132518)/+>}\n",
      "{1: {'MAT2'}, 'gene_ids': {'MAT2'}, 'gene_ids_reduced': {'MAT2'}, 'NH': 2, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:2209:3431:99503:1:N:0:0:CAGGCG' aligned to II:[2132511,2132561)/+>}\n",
      "{1: {'MAT2'}, 'gene_ids': {'MAT2', 'SPBC1711.01c'}, 'gene_ids_reduced': {'MAT2'}, 'NH': 2, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:1305:13966:12841:1:N:0:0:CAGGCG' aligned to II:[2132548,2132598)/+>}\n",
      "{1: {'MAT2'}, 'gene_ids': {'MAT2', 'SPBC1711.01c'}, 'gene_ids_reduced': {'MAT2'}, 'NH': 2, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:1113:5035:21196:1:N:0:0:CAGGCG' aligned to II:[2132549,2132598)/+>}\n",
      "{1: {'MAT2'}, 'gene_ids': {'MAT2', 'SPBC1711.01c'}, 'gene_ids_reduced': {'MAT2'}, 'NH': 2, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:1206:8002:38821:1:N:0:0:CAGGCG' aligned to II:[2132554,2132601)/+>}\n",
      "{1: {'MAT2'}, 'gene_ids': {'MAT2', 'SPBC1711.01c'}, 'gene_ids_reduced': {'MAT2'}, 'NH': 2, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:1204:18755:41536:1:N:0:0:CAGGCG' aligned to II:[2132554,2132604)/+>}\n",
      "{1: {'MAT2'}, 'gene_ids': {'MAT2', 'SPBC1711.01c'}, 'gene_ids_reduced': {'MAT2'}, 'NH': 2, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:1316:4551:6124:1:N:0:0:CAGGCG' aligned to II:[2132555,2132601)/+>}\n",
      "{1: {'MAT2'}, 'gene_ids': {'MAT2', 'SPBC1711.01c'}, 'gene_ids_reduced': {'MAT2'}, 'NH': 2, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:1201:18976:23791:1:N:0:0:CAGGCG' aligned to II:[2132555,2132601)/+>}\n",
      "{1: {'MAT2'}, 'gene_ids': {'MAT2', 'SPBC1711.01c'}, 'gene_ids_reduced': {'MAT2'}, 'NH': 2, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:2303:6998:82477:1:N:0:0:CAGGCG' aligned to II:[2132557,2132607)/+>}\n",
      "{1: {'MAT2'}, 'gene_ids': {'MAT2', 'SPBC1711.01c'}, 'gene_ids_reduced': {'MAT2'}, 'NH': 2, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:1302:16959:84250:1:N:0:0:CAGGCG' aligned to II:[2132558,2132606)/+>}\n",
      "{1: {'MAT2'}, 'gene_ids': {'MAT2', 'SPBC1711.01c'}, 'gene_ids_reduced': {'MAT2'}, 'NH': 2, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:1107:9334:70256:1:N:0:0:CAGGCG' aligned to II:[2132558,2132606)/+>}\n",
      "{1: {'MAT2'}, 'gene_ids': {'MAT2', 'SPBC1711.01c'}, 'gene_ids_reduced': {'MAT2'}, 'NH': 2, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:2207:13785:62947:1:N:0:0:CAGGCG' aligned to II:[2132569,2132606)/+>}\n",
      "{1: {'MAT2'}, 'gene_ids': {'MAT2', 'SPBC1711.01c'}, 'gene_ids_reduced': {'MAT2'}, 'NH': 2, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:1210:3103:30312:1:N:0:0:CAGGCG' aligned to II:[2132569,2132606)/+>}\n",
      "{1: {'MAT2'}, 'gene_ids': {'SPBC1711.02', 'MAT2'}, 'gene_ids_reduced': {'MAT2'}, 'NH': 2, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:2104:18462:99232:1:N:0:0:CAGGCG' aligned to II:[2132922,2132972)/->}\n",
      "{1: {'MAT2'}, 'gene_ids': {'SPBC1711.02', 'MAT2'}, 'gene_ids_reduced': {'MAT2'}, 'NH': 2, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:1201:18781:56417:1:N:0:0:CAGGCG' aligned to II:[2132940,2132990)/->}\n",
      "{1: {'MAT2'}, 'gene_ids': {'SPBC1711.02', 'MAT2'}, 'gene_ids_reduced': {'MAT2'}, 'NH': 2, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:1102:19765:81623:1:N:0:0:CAGGCG' aligned to II:[2132942,2132992)/->}\n",
      "{1: {'MAT3', 'MAT2'}, 'gene_ids': {'SPBC1711.02', 'MAT3', 'MAT2'}, 'gene_ids_reduced': {'MAT3', 'MAT2'}, 'NH': 2, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:1110:19280:36437:1:N:0:0:CAGGCG' aligned to II:[2132954,2133003)/->}\n",
      "{1: {'MAT3', 'MAT2'}, 'gene_ids': {'SPBC1711.02', 'MAT3', 'MAT2'}, 'gene_ids_reduced': {'MAT3', 'MAT2'}, 'NH': 2, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:1111:12241:94710:1:N:0:0:CAGGCG' aligned to II:[2132954,2133004)/->}\n",
      "{1: {'MAT3', 'MAT2'}, 'gene_ids': {'SPBC1711.02', 'MAT3', 'MAT2'}, 'gene_ids_reduced': {'MAT3', 'MAT2'}, 'NH': 2, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:2315:11572:31571:1:N:0:0:CAGGCG' aligned to II:[2132958,2133008)/->}\n",
      "{1: {'MAT3', 'MAT2'}, 'gene_ids': {'SPBC1711.02', 'MAT3', 'MAT2'}, 'gene_ids_reduced': {'MAT3', 'MAT2'}, 'NH': 2, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:2308:18909:12227:1:N:0:0:CAGGCG' aligned to II:[2132962,2133011)/->}\n",
      "{1: {'MAT3', 'MAT2'}, 'gene_ids': {'SPBC1711.02', 'MAT3', 'MAT2'}, 'gene_ids_reduced': {'MAT3', 'MAT2'}, 'NH': 2, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:2310:15406:29888:1:N:0:0:CAGGCG' aligned to II:[2132962,2133012)/->}\n",
      "{1: {'MAT3', 'MAT2'}, 'gene_ids': {'SPBC1711.02', 'MAT3', 'MAT2'}, 'gene_ids_reduced': {'MAT3', 'MAT2'}, 'NH': 2, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:2315:12903:95324:1:N:0:0:CAGGCG' aligned to II:[2132965,2133015)/->}\n",
      "{1: {'MAT3', 'MAT2'}, 'gene_ids': {'SPBC1711.02', 'MAT3', 'MAT2'}, 'gene_ids_reduced': {'MAT3', 'MAT2'}, 'NH': 2, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:2205:8888:5036:1:N:0:0:CAGGCG' aligned to II:[2132968,2133018)/->}\n",
      "{1: {'MAT3', 'MAT2'}, 'gene_ids': {'SPBC1711.02', 'MAT3', 'MAT2'}, 'gene_ids_reduced': {'MAT3', 'MAT2'}, 'NH': 2, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:2306:12980:27363:1:N:0:0:CAGGCG' aligned to II:[2132968,2133018)/->}\n",
      "{1: {'MAT3', 'MAT2'}, 'gene_ids': {'SPBC1711.02', 'MAT3', 'MAT2'}, 'gene_ids_reduced': {'MAT3', 'MAT2'}, 'NH': 2, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:2108:5839:51369:1:N:0:0:CAGGCG' aligned to II:[2132969,2133019)/->}\n",
      "{1: {'MAT3', 'MAT2'}, 'gene_ids': {'SPBC1711.02', 'MAT3', 'MAT2'}, 'gene_ids_reduced': {'MAT3', 'MAT2'}, 'NH': 2, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:1309:13642:18793:1:N:0:0:CAGGCG' aligned to II:[2132969,2133019)/->}\n",
      "{1: {'MAT3', 'MAT2'}, 'gene_ids': {'SPBC1711.02', 'MAT3', 'MAT2'}, 'gene_ids_reduced': {'MAT3', 'MAT2'}, 'NH': 2, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:2216:7032:15866:1:N:0:0:CAGGCG' aligned to II:[2132973,2133023)/->}\n",
      "{1: {'MAT3', 'MAT2'}, 'gene_ids': {'SPBC1711.02', 'MAT3', 'MAT2'}, 'gene_ids_reduced': {'MAT3', 'MAT2'}, 'NH': 2, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:2202:11872:10564:1:N:0:0:CAGGCG' aligned to II:[2132973,2133023)/->}\n",
      "{1: {'MAT3', 'MAT2'}, 'gene_ids': {'SPBC1711.02', 'MAT3', 'MAT2'}, 'gene_ids_reduced': {'MAT3', 'MAT2'}, 'NH': 2, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:2213:18606:100846:1:N:0:0:CAGGCG' aligned to II:[2132974,2133023)/->}\n",
      "{1: {'MAT3', 'MAT2'}, 'gene_ids': {'SPBC1711.02', 'MAT3', 'MAT2'}, 'gene_ids_reduced': {'MAT3', 'MAT2'}, 'NH': 2, 'alignment': <SAM_Alignment object: Read 'L183:365:CBF20ANXX:6:2102:10490:25392:1:N:0:0:CAGGCG' aligned to II:[2132987,2133037)/->}\n"
     ]
    }
   ],
   "source": [
    "for ii, vv in reads.items():\n",
    "    print(vv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "II:[2132954,2132999)/. {'SPBC1711.02', 'MAT2'}\n",
      "II:[2132999,2133000)/. {'SPBC1711.02', 'MAT3', 'MAT2'}\n",
      "II:[2133000,2133003)/. {'SPBC1711.02', 'MAT3'}\n",
      "{'SPBC1711.02', 'MAT3', 'MAT2'}\n"
     ]
    }
   ],
   "source": [
    "gene_ids = None\n",
    "\n",
    "for iv, fs in exon_features[reads['L183:365:CBF20ANXX:6:1110:19280:36437:1:N:0:0:CAGGCG']['alignment'].iv].steps():\n",
    "    print(iv, fs)\n",
    "\n",
    "    if ((len(fs) > 0) or (overlap_mode == \"intersection-strict\")):\n",
    "        if gene_ids is None:\n",
    "            gene_ids = fs.copy()\n",
    "        else:\n",
    "            gene_ids = gene_ids.union(fs)\n",
    "\n",
    "print(gene_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100000 alignment records processed. 268.18972301483154 s\n",
      "200000 alignment records processed. 269.1396098136902 s\n",
      "300000 alignment records processed. 270.1076776981354 s\n",
      "400000 alignment records processed. 271.0638234615326 s\n",
      "500000 alignment records processed. 272.05218982696533 s\n",
      "600000 alignment records processed. 273.02886414527893 s\n",
      "700000 alignment records processed. 274.006733417511 s\n",
      "800000 alignment records processed. 274.9754321575165 s\n",
      "900000 alignment records processed. 275.946084022522 s\n",
      "1000000 alignment records processed. 276.92605328559875 s\n",
      "1100000 alignment records processed. 277.9127058982849 s\n",
      "1200000 alignment records processed. 278.8938512802124 s\n",
      "1300000 alignment records processed. 279.88550305366516 s\n",
      "1400000 alignment records processed. 280.84897661209106 s\n",
      "1500000 alignment records processed. 281.8468029499054 s\n",
      "1600000 alignment records processed. 282.8613679409027 s\n",
      "1700000 alignment records processed. 283.8855996131897 s\n",
      "1800000 alignment records processed. 284.90057158470154 s\n",
      "1900000 alignment records processed. 285.8965382575989 s\n",
      "2000000 alignment records processed. 286.90704226493835 s\n",
      "2100000 alignment records processed. 287.9240925312042 s\n",
      "2200000 alignment records processed. 288.91779041290283 s\n",
      "2300000 alignment records processed. 289.9263114929199 s\n",
      "2400000 alignment records processed. 290.9412772655487 s\n",
      "2500000 alignment records processed. 291.9488093852997 s\n",
      "2600000 alignment records processed. 292.94491362571716 s\n",
      "2700000 alignment records processed. 293.9489686489105 s\n",
      "2800000 alignment records processed. 294.93218541145325 s\n",
      "2900000 alignment records processed. 295.93976068496704 s\n",
      "3000000 alignment records processed. 296.9364709854126 s\n",
      "3100000 alignment records processed. 297.9382109642029 s\n",
      "3200000 alignment records processed. 298.95654678344727 s\n",
      "3300000 alignment records processed. 299.96472215652466 s\n",
      "3400000 alignment records processed. 300.96124505996704 s\n",
      "3500000 alignment records processed. 301.9809923171997 s\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Elapsed Time (parsing reads): 302.0639624595642\n"
     ]
    }
   ],
   "source": [
    "nh = []\n",
    "\n",
    "start_time = time.time()\n",
    "i = 0\n",
    "#for aln in itertools.islice(bam_file, 10000):\n",
    "for aln in bam_file:\n",
    "\n",
    "    if i > 0 and i % 100000 == 0:\n",
    "        sys.stderr.write(\"{} alignment records processed. {} s\\n\".format(i,  time.time() - start_time))\n",
    "        sys.stderr.flush()\n",
    "    i += 1          \n",
    "        \n",
    "    ## Multimapped reads are contained as separate entries in the BAM file.\n",
    "    nh.append( aln.optional_field(\"NH\") )\n",
    "\n",
    "print('Elapsed Time (parsing reads):', time.time() - start_time)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([2.863813e+06, 8.456400e+04, 4.350300e+05, 5.664400e+04,\n",
       "        5.836000e+04, 4.026000e+03, 1.484000e+03, 1.384000e+03,\n",
       "        9.810000e+02, 1.490000e+03]),\n",
       " array([ 1. ,  1.9,  2.8,  3.7,  4.6,  5.5,  6.4,  7.3,  8.2,  9.1, 10. ]),\n",
       " <a list of 10 Patch objects>)"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZAAAAD8CAYAAABZ/vJZAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8li6FKAAAVkElEQVR4nO3cf6ye5X3f8fenOFCaDLCNQcyG2SvWFogUMo6Mt0hTBpXtlWmmEkiu1GBNllwhsiVTpAXyjzsIEpbWeEUqSDR4GJbFWCQRqAl1LaCKKjHDIbCCIchWYODggdkxhE4Krcl3fzzXUR4fDsfHl815cPx+SY+e+3zv6/re13mE+fj+8ThVhSRJx+o3Rr0ASdLJyQCRJHUxQCRJXQwQSVIXA0SS1MUAkSR1OWqAJPnNJE8m+V9J9iT5z62+IMmuJHvb+/yhOTcn2ZfkpSSrh+qXJ3mu7bsjSVr9jCQPtPruJEuH5qxvx9ibZP1QfVkbu7fNPf3EfCSSpNmYzRnIe8CVVfVZ4DJgTZKVwE3Ao1W1HHi0/UySS4B1wKXAGuDOJKe1XncBG4Hl7bWm1TcAh6rqYmALsLn1WgBsAq4AVgCbhoJqM7ClHf9Q6yFJmiNHDZAa+Nv24yfaq4C1wLZW3wZc07bXAtur6r2qehnYB6xIcgFwVlU9UYNvL943Zc5krweBq9rZyWpgV1VNVNUhYBeDAAtwZRs79fiSpDkwbzaD2hnE08DFwJ9W1e4k51fVAYCqOpDkvDZ8MfA/h6bvb7W/b9tT65NzXmu9Did5B1g4XJ8yZyHwdlUdnqbXhzr33HNr6dKls/mVJUnN008//VZVLZpan1WAVNX7wGVJzgG+n+QzMwzPdC1mqPfMmanXkYtJNjK4bMZFF13E+Pj4dMMkSR8iyf+ern5MT2FV1dvAXzG4d/FGuyxFe3+zDdsPXDg0bQnweqsvmaZ+xJwk84CzgYkZer0FnNPGTu01dc13V9VYVY0tWvSBAJUkdZrNU1iL2pkHSc4Efgf4CfAwMPlU1Hrgobb9MLCuPVm1jMHN8ifb5a53k6xs9zCunzJnste1wGPtPslOYFWS+e3m+SpgZ9v3eBs79fiSpDkwm0tYFwDb2n2Q3wB2VNWfJ3kC2JFkA/AqcB1AVe1JsgN4ATgM3NgugQHcANwLnAk80l4A9wD3J9nH4MxjXes1keRW4Kk27paqmmjbXwO2J/kG8EzrIUmaIzmV/jn3sbGx8h6IJB2bJE9X1djUut9ElyR1MUAkSV0MEElSFwNEktTFAJEkdZnVN9EFS2/6wUiO+8rtV4/kuJJ0NJ6BSJK6GCCSpC4GiCSpiwEiSepigEiSuhggkqQuBogkqYsBIknqYoBIkroYIJKkLgaIJKmLASJJ6mKASJK6GCCSpC4GiCSpiwEiSepigEiSuhggkqQuBogkqYsBIknqYoBIkrocNUCSXJjk8SQvJtmT5Mut/kdJfpbk2fb63aE5NyfZl+SlJKuH6pcnea7tuyNJWv2MJA+0+u4kS4fmrE+yt73WD9WXtbF729zTT8xHIkmajdmcgRwGvlpVnwZWAjcmuaTt21JVl7XXDwHavnXApcAa4M4kp7XxdwEbgeXttabVNwCHqupiYAuwufVaAGwCrgBWAJuSzG9zNrfjLwcOtR6SpDly1ACpqgNV9eO2/S7wIrB4hilrge1V9V5VvQzsA1YkuQA4q6qeqKoC7gOuGZqzrW0/CFzVzk5WA7uqaqKqDgG7gDVt35VtLG3uZC9J0hw4pnsg7dLS54DdrfSlJH+TZOvQmcFi4LWhaftbbXHbnlo/Yk5VHQbeARbO0Gsh8HYbO7WXJGkOzDpAknwK+C7wlar6OYPLUb8NXAYcAP54cug002uGes+cmXpNXffGJONJxg8ePDjdEElSh1kFSJJPMAiPb1fV9wCq6o2qer+qfgn8GYN7FDA4G7hwaPoS4PVWXzJN/Yg5SeYBZwMTM/R6CzinjZ3a6whVdXdVjVXV2KJFi2bz60qSZmE2T2EFuAd4saq+OVS/YGjY7wHPt+2HgXXtyaplDG6WP1lVB4B3k6xsPa8HHhqaM/mE1bXAY+0+yU5gVZL57RLZKmBn2/d4G0ubO9lLkjQH5h19CJ8Hvgg8l+TZVvs68PtJLmNw6egV4A8BqmpPkh3ACwye4Lqxqt5v824A7gXOBB5pLxgE1P1J9jE481jXek0kuRV4qo27paom2vbXgO1JvgE803pIkubIUQOkqv6a6e85/HCGObcBt01THwc+M039F8B1H9JrK7B1mvpP+dVlM0nSHPOb6JKkLgaIJKmLASJJ6mKASJK6GCCSpC4GiCSpiwEiSepigEiSuhggkqQuBogkqYsBIknqYoBIkroYIJKkLgaIJKmLASJJ6mKASJK6GCCSpC4GiCSpiwEiSepigEiSuhggkqQuBogkqYsBIknqYoBIkroYIJKkLgaIJKmLASJJ6mKASJK6HDVAklyY5PEkLybZk+TLrb4gya4ke9v7/KE5NyfZl+SlJKuH6pcnea7tuyNJWv2MJA+0+u4kS4fmrG/H2Jtk/VB9WRu7t809/cR8JJKk2ZjNGchh4KtV9WlgJXBjkkuAm4BHq2o58Gj7mbZvHXApsAa4M8lprdddwEZgeXutafUNwKGquhjYAmxuvRYAm4ArgBXApqGg2gxsacc/1HpIkubIUQOkqg5U1Y/b9rvAi8BiYC2wrQ3bBlzTttcC26vqvap6GdgHrEhyAXBWVT1RVQXcN2XOZK8Hgava2clqYFdVTVTVIWAXsKbtu7KNnXp8SdIcOKZ7IO3S0ueA3cD5VXUABiEDnNeGLQZeG5q2v9UWt+2p9SPmVNVh4B1g4Qy9FgJvt7FTe01d88Yk40nGDx48eCy/riRpBrMOkCSfAr4LfKWqfj7T0GlqNUO9Z85MvY4sVt1dVWNVNbZo0aLphkiSOswqQJJ8gkF4fLuqvtfKb7TLUrT3N1t9P3Dh0PQlwOutvmSa+hFzkswDzgYmZuj1FnBOGzu1lyRpDszmKawA9wAvVtU3h3Y9DEw+FbUeeGiovq49WbWMwc3yJ9tlrneTrGw9r58yZ7LXtcBj7T7JTmBVkvnt5vkqYGfb93gbO/X4kqQ5MO/oQ/g88EXguSTPttrXgduBHUk2AK8C1wFU1Z4kO4AXGDzBdWNVvd/m3QDcC5wJPNJeMAio+5PsY3Dmsa71mkhyK/BUG3dLVU207a8B25N8A3im9ZAkzZGjBkhV/TXT33MAuOpD5twG3DZNfRz4zDT1X9ACaJp9W4Gt09R/yuDRXknSCPhNdElSFwNEktTFAJEkdTFAJEldDBBJUhcDRJLUxQCRJHUxQCRJXQwQSVIXA0SS1MUAkSR1MUAkSV0MEElSFwNEktTFAJEkdTFAJEldDBBJUhcDRJLUxQCRJHUxQCRJXQwQSVIXA0SS1MUAkSR1MUAkSV0MEElSFwNEktTFAJEkdTlqgCTZmuTNJM8P1f4oyc+SPNtevzu07+Yk+5K8lGT1UP3yJM+1fXckSaufkeSBVt+dZOnQnPVJ9rbX+qH6sjZ2b5t7+vF/FJKkYzGbM5B7gTXT1LdU1WXt9UOAJJcA64BL25w7k5zWxt8FbASWt9dkzw3Aoaq6GNgCbG69FgCbgCuAFcCmJPPbnM3t+MuBQ62HJGkOHTVAqupHwMQs+60FtlfVe1X1MrAPWJHkAuCsqnqiqgq4D7hmaM62tv0gcFU7O1kN7Kqqiao6BOwC1rR9V7axtLmTvSRJc+R47oF8KcnftEtck2cGi4HXhsbsb7XFbXtq/Yg5VXUYeAdYOEOvhcDbbezUXh+QZGOS8STjBw8ePPbfUpI0rd4AuQv4beAy4ADwx62eacbWDPWeOTP1+uCOqruraqyqxhYtWvRhwyRJx6grQKrqjap6v6p+CfwZg3sUMDgbuHBo6BLg9VZfMk39iDlJ5gFnM7hk9mG93gLOaWOn9pIkzZGuAGn3NCb9HjD5hNbDwLr2ZNUyBjfLn6yqA8C7SVa2exjXAw8NzZl8wupa4LF2n2QnsCrJ/HaJbBWws+17vI2lzZ3sJUmaI/OONiDJd4AvAOcm2c/gyagvJLmMwaWjV4A/BKiqPUl2AC8Ah4Ebq+r91uoGBk90nQk80l4A9wD3J9nH4MxjXes1keRW4Kk27paqmryZ/zVge5JvAM+0HpKkOZTBX+hPDWNjYzU+Pt41d+lNPzjBq5mdV26/eiTHlaRJSZ6uqrGpdb+JLknqYoBIkroYIJKkLgaIJKmLASJJ6mKASJK6GCCSpC4GiCSpiwEiSepigEiSuhggkqQuBogkqYsBIknqYoBIkroYIJKkLgaIJKmLASJJ6mKASJK6GCCSpC4GiCSpiwEiSepigEiSuhggkqQuBogkqYsBIknqYoBIkroYIJKkLkcNkCRbk7yZ5Pmh2oIku5Lsbe/zh/bdnGRfkpeSrB6qX57kubbvjiRp9TOSPNDqu5MsHZqzvh1jb5L1Q/VlbezeNvf04/8oJEnHYjZnIPcCa6bUbgIerarlwKPtZ5JcAqwDLm1z7kxyWptzF7ARWN5ekz03AIeq6mJgC7C59VoAbAKuAFYAm4aCajOwpR3/UOshSZpDRw2QqvoRMDGlvBbY1ra3AdcM1bdX1XtV9TKwD1iR5ALgrKp6oqoKuG/KnMleDwJXtbOT1cCuqpqoqkPALmBN23dlGzv1+JKkOdJ7D+T8qjoA0N7Pa/XFwGtD4/a32uK2PbV+xJyqOgy8AyycoddC4O02dmqvD0iyMcl4kvGDBw8e468pSfowJ/omeqap1Qz1njkz9frgjqq7q2qsqsYWLVr0YcMkSceoN0DeaJelaO9vtvp+4MKhcUuA11t9yTT1I+YkmQeczeCS2Yf1egs4p42d2kuSNEd6A+RhYPKpqPXAQ0P1de3JqmUMbpY/2S5zvZtkZbuHcf2UOZO9rgUea/dJdgKrksxvN89XATvbvsfb2KnHlyTNkXlHG5DkO8AXgHOT7GfwZNTtwI4kG4BXgesAqmpPkh3AC8Bh4Maqer+1uoHBE11nAo+0F8A9wP1J9jE481jXek0kuRV4qo27paomb+Z/Ddie5BvAM62HJGkOZfAX+lPD2NhYjY+Pd81detMPTvBqZueV268eyXElaVKSp6tqbGrdb6JLkroYIJKkLgaIJKmLASJJ6mKASJK6GCCSpC4GiCSpiwEiSepigEiSuhggkqQuBogkqYsBIknqYoBIkroYIJKkLgaIJKmLASJJ6mKASJK6GCCSpC4GiCSpiwEiSepigEiSuhggkqQuBogkqYsBIknqYoBIkroYIJKkLgaIJKnLcQVIkleSPJfk2STjrbYgya4ke9v7/KHxNyfZl+SlJKuH6pe3PvuS3JEkrX5GkgdafXeSpUNz1rdj7E2y/nh+D0nSsTsRZyD/qqouq6qx9vNNwKNVtRx4tP1MkkuAdcClwBrgziSntTl3ARuB5e21ptU3AIeq6mJgC7C59VoAbAKuAFYAm4aDSpL00fsoLmGtBba17W3ANUP17VX1XlW9DOwDViS5ADirqp6oqgLumzJnsteDwFXt7GQ1sKuqJqrqELCLX4WOJGkOHG+AFPCXSZ5OsrHVzq+qAwDt/bxWXwy8NjR3f6stbttT60fMqarDwDvAwhl6SZLmyLzjnP/5qno9yXnAriQ/mWFspqnVDPXeOUcedBBsGwEuuuiiGZYnSToWx3UGUlWvt/c3ge8zuB/xRrssRXt/sw3fD1w4NH0J8HqrL5mmfsScJPOAs4GJGXpNt8a7q2qsqsYWLVrU94tKkj6gO0CSfDLJP5jcBlYBzwMPA5NPRa0HHmrbDwPr2pNVyxjcLH+yXeZ6N8nKdn/j+ilzJntdCzzW7pPsBFYlmd9unq9qNUnSHDmeS1jnA99vT9zOA/5HVf1FkqeAHUk2AK8C1wFU1Z4kO4AXgMPAjVX1fut1A3AvcCbwSHsB3APcn2QfgzOPda3XRJJbgafauFuqauI4fhdJ0jHqDpCq+inw2Wnq/xe46kPm3AbcNk19HPjMNPVf0AJomn1bga3Htmodi6U3/WAkx33l9qtHclxJx8ZvokuSuhggkqQuBogkqYsBIknqYoBIkroYIJKkLgaIJKmLASJJ6mKASJK6GCCSpC4GiCSpiwEiSepigEiSuhggkqQuBogkqYsBIknqYoBIkroYIJKkLgaIJKmLASJJ6mKASJK6GCCSpC4GiCSpiwEiSepigEiSuswb9QI0s6U3/WDUS5CkaXkGIknqclKfgSRZA/wJcBrwraq6fcRL0glwKp51vXL71aNegnTMTtozkCSnAX8K/GvgEuD3k1wy2lVJ0qnjpA0QYAWwr6p+WlV/B2wH1o54TZJ0yjiZA2Qx8NrQz/tbTZI0B07meyCZplYfGJRsBDa2H/82yUsf6ao+eucCb416ER8TvzafRTafkDa/Np/HCeLn8SvH+1n8o+mKJ3OA7AcuHPp5CfD61EFVdTdw91wt6qOWZLyqxka9jo8DP4sj+Xkcyc/jVz6qz+JkvoT1FLA8ybIkpwPrgIdHvCZJOmWctGcgVXU4yZeAnQwe491aVXtGvCxJOmWctAECUFU/BH446nXMsV+by3EngJ/Fkfw8juTn8SsfyWeRqg/cd5Yk6ahO5nsgkqQRMkBOAkkuTPJ4kheT7Eny5VGv6eMgyWlJnkny56Ney6glOSfJg0l+0v47+eejXtOoJPmP7c/J80m+k+Q3R72muZRka5I3kzw/VFuQZFeSve19/ok4lgFycjgMfLWqPg2sBG70n20B4MvAi6NexMfEnwB/UVX/FPgsp+jnkmQx8B+Asar6DIMHbNaNdlVz7l5gzZTaTcCjVbUceLT9fNwMkJNAVR2oqh+37XcZ/M/hlP7WfZIlwNXAt0a9llFLchbwL4F7AKrq76rq7dGuaqTmAWcmmQf8FtN8P+zXWVX9CJiYUl4LbGvb24BrTsSxDJCTTJKlwOeA3aNdycj9V+A/Ab8c9UI+Bv4xcBD4b+2S3reSfHLUixqFqvoZ8F+AV4EDwDtV9ZejXdXHwvlVdQAGfyEFzjsRTQ2Qk0iSTwHfBb5SVT8f9XpGJcm/Ad6sqqdHvZaPiXnAPwPuqqrPAf+PE3SJ4mTTru2vBZYB/xD4ZJI/GO2qfn0ZICeJJJ9gEB7frqrvjXo9I/Z54N8meYXBv8J8ZZL/PtoljdR+YH9VTZ6VPsggUE5FvwO8XFUHq+rvge8B/2LEa/o4eCPJBQDt/c0T0dQAOQkkCYPr2y9W1TdHvZ5Rq6qbq2pJVS1lcIP0sao6Zf+WWVX/B3gtyT9ppauAF0a4pFF6FViZ5Lfan5urOEUfKJjiYWB9214PPHQimp7U30Q/hXwe+CLwXJJnW+3r7Zv4EsC/B77d/l24nwL/bsTrGYmq2p3kQeDHDJ5efIZT7BvpSb4DfAE4N8l+YBNwO7AjyQYGIXvdCTmW30SXJPXwEpYkqYsBIknqYoBIkroYIJKkLgaIJKmLASJJ6mKASJK6GCCSpC7/H36qj6JJIy57AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.hist(nh)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Min: 1 Max: 10\n"
     ]
    }
   ],
   "source": [
    "print('Min:', min(nh), 'Max:', max(nh))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (heterochromatin)",
   "language": "python",
   "name": "heterochromatin"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
