{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import copy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import HTSeq\n",
    "import pysam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "plt.style.use('ggplot')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "project_dir = '/home/pmonteagudo/workspace/silencing_project'\n",
    "if project_dir not in sys.path: \n",
    "    sys.path.append(project_dir)\n",
    "from config_analysis import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Util import to_log2_tpm\n",
    "from Util import long_header"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "long_header.extend(['gene_length','intron_length'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- other **directories**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/gcm-lfs1/pablo/data/rna_silencing/results/xp_data/RNA'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/gcm-lfs1/pablo/data/rna_silencing/results/xp_data/ChIP'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chip_dir"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_gene_counts(gene_counts_file, index_col='gene_id', gdf='pombase', sample_ids=None):\n",
    "    \n",
    "    if gdf == 'pombase':\n",
    "        rename_cols = {}\n",
    "    \n",
    "    # import Gene Counts .csv file as DataFrame \n",
    "    counts_df = pd.read_csv(gene_counts_file, sep='\\t')\n",
    "\n",
    "    # rename columns\n",
    "    counts_df = counts_df.rename(columns=rename_cols)\n",
    "    \n",
    "    # contains info about columns being 'samples' or NOT\n",
    "    header_dict = { ii:('annotation' if ii in long_header else 'sample_col') for ii in counts_df.columns }\n",
    "    \n",
    "    if not isinstance(sample_ids, type(None)):\n",
    "        # select only a subset of samples\n",
    "        counts_df = counts_df[ [ii for ii in counts_df.columns if (ii in sample_ids or header_dict[ii] == 'annotation')] ]\n",
    "    \n",
    "    return counts_df.sort_values('gene_id')\n",
    "                                                                   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Normalization in Gene Counts: **ChIP-Seq data**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "params_dict = {\n",
    "    # sequencing type: ['rna', 'chip']\n",
    "    'seq_type': 'chip',\n",
    "    # gene expression matrix: ['gene_count', 'tpm'] \n",
    "    'expression': 'tpm',\n",
    "    #'expression': 'gene_count',\n",
    "    # batch with corresponding parameters\n",
    "    'batch': 'xp_data',\n",
    "    # normalization of counts\n",
    "    'norm': 'wo_nh-norm',\n",
    "    # h3k9me2\n",
    "    'h3k9me2': True,\n",
    "    #'h3k9me2': False, \n",
    "    # log transform data\n",
    "    'log_transform': True,\n",
    "    #'log_transform': False,\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **A**. Results <font color='blue'> **without** </font> **NH-normalization**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Import **gene expresion Matrix**: `*_pombe_tpm_matrix.csv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/gcm-lfs1/pablo/data/rna_silencing/results_wo_nh-norm/xp_data/ChIP'"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if params_dict['seq_type'] == 'rna':\n",
    "    # stranded=True, remove duplicate reads mapping to same gene, count only multimapped if repeat\n",
    "    xp_data_dir = os.path.join(data_results_dir + '_' + params_dict['norm'], params_dict['batch'], 'RNA')\n",
    "else:\n",
    "    # stranded=False, remove duplicate reads mapping to same gene, count only multimapped if repeat\n",
    "    xp_data_dir = os.path.join(data_results_dir + '_' + params_dict['norm'], params_dict['batch'], 'ChIP')\n",
    "xp_data_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/gcm-lfs1/pablo/data/rna_silencing/results_wo_nh-norm/xp_data/ChIP/h3k9me2_chip_pombe_tpm_matrix.csv'"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if (params_dict['seq_type'] == 'chip') and (params_dict['h3k9me2']):\n",
    "    counts_file = os.path.join(xp_data_dir, 'h3k9me2_' + params_dict['seq_type'] + '_pombe_' + params_dict['expression'] + '_matrix.csv')\n",
    "else:\n",
    "    counts_file = os.path.join(xp_data_dir, params_dict['seq_type'] + '_pombe_' + params_dict['expression'] + '_matrix.csv')\n",
    "counts_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>seqid</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>cds_length</th>\n",
       "      <th>utr_length</th>\n",
       "      <th>intron_length</th>\n",
       "      <th>...</th>\n",
       "      <th>302_H3K9me2_1</th>\n",
       "      <th>324_H3K9me2_1</th>\n",
       "      <th>491_H3K9me2_1</th>\n",
       "      <th>504_H3K9me2_1</th>\n",
       "      <th>510_H3K9me2_1</th>\n",
       "      <th>530_H3K9me2_1</th>\n",
       "      <th>638_H3K9me2_1</th>\n",
       "      <th>WT_H3K9me2_1</th>\n",
       "      <th>chip-fake-reads_simulated-data_1</th>\n",
       "      <th>chip-fake-reads_simulated-data_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FP565355_region_1..2120</td>\n",
       "      <td>mating_type_region</td>\n",
       "      <td>region</td>\n",
       "      <td>1</td>\n",
       "      <td>2120</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>408.094181</td>\n",
       "      <td>4502.883538</td>\n",
       "      <td>4993.117682</td>\n",
       "      <td>3063.305824</td>\n",
       "      <td>5711.894487</td>\n",
       "      <td>2515.868824</td>\n",
       "      <td>4165.457917</td>\n",
       "      <td>3124.282423</td>\n",
       "      <td>581.140715</td>\n",
       "      <td>581.140715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FP565355_region_15417..15473</td>\n",
       "      <td>mating_type_region</td>\n",
       "      <td>region</td>\n",
       "      <td>15417</td>\n",
       "      <td>15473</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1162.048668</td>\n",
       "      <td>1971.046770</td>\n",
       "      <td>1686.235695</td>\n",
       "      <td>7319.396234</td>\n",
       "      <td>7398.231799</td>\n",
       "      <td>2747.824806</td>\n",
       "      <td>8045.427287</td>\n",
       "      <td>3527.447203</td>\n",
       "      <td>147.177306</td>\n",
       "      <td>147.177306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FP565355_region_15474..15608</td>\n",
       "      <td>mating_type_region</td>\n",
       "      <td>region</td>\n",
       "      <td>15474</td>\n",
       "      <td>15608</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1391.168477</td>\n",
       "      <td>6279.042506</td>\n",
       "      <td>7482.761671</td>\n",
       "      <td>13815.154482</td>\n",
       "      <td>14955.159873</td>\n",
       "      <td>6049.474767</td>\n",
       "      <td>15653.804910</td>\n",
       "      <td>9781.204416</td>\n",
       "      <td>588.709225</td>\n",
       "      <td>588.709225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FP565355_region_15609..16735</td>\n",
       "      <td>mating_type_region</td>\n",
       "      <td>region</td>\n",
       "      <td>15609</td>\n",
       "      <td>16735</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>439.435683</td>\n",
       "      <td>2448.427202</td>\n",
       "      <td>5792.039436</td>\n",
       "      <td>2234.060733</td>\n",
       "      <td>5488.584094</td>\n",
       "      <td>2627.398594</td>\n",
       "      <td>6195.784590</td>\n",
       "      <td>3577.713987</td>\n",
       "      <td>441.531919</td>\n",
       "      <td>441.531919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FP565355_region_16736..16794</td>\n",
       "      <td>mating_type_region</td>\n",
       "      <td>region</td>\n",
       "      <td>16736</td>\n",
       "      <td>16794</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>242.514505</td>\n",
       "      <td>4121.908333</td>\n",
       "      <td>2797.902190</td>\n",
       "      <td>9192.532044</td>\n",
       "      <td>4486.233712</td>\n",
       "      <td>2686.762033</td>\n",
       "      <td>10219.535118</td>\n",
       "      <td>2041.416254</td>\n",
       "      <td>588.709225</td>\n",
       "      <td>588.709225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7013</th>\n",
       "      <td>dg1a</td>\n",
       "      <td>I</td>\n",
       "      <td>gene</td>\n",
       "      <td>3761000</td>\n",
       "      <td>3764500</td>\n",
       "      <td>+</td>\n",
       "      <td>dg1_plus</td>\n",
       "      <td>3501.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>58234.743863</td>\n",
       "      <td>96152.109937</td>\n",
       "      <td>48565.266936</td>\n",
       "      <td>112852.707168</td>\n",
       "      <td>112718.195628</td>\n",
       "      <td>48103.172703</td>\n",
       "      <td>4709.563541</td>\n",
       "      <td>89743.227845</td>\n",
       "      <td>1185.774983</td>\n",
       "      <td>1185.774983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7014</th>\n",
       "      <td>dg1b</td>\n",
       "      <td>I</td>\n",
       "      <td>gene</td>\n",
       "      <td>3761000</td>\n",
       "      <td>3764500</td>\n",
       "      <td>-</td>\n",
       "      <td>dg1_minus</td>\n",
       "      <td>3501.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>58234.743863</td>\n",
       "      <td>96152.109937</td>\n",
       "      <td>48565.266936</td>\n",
       "      <td>112852.707168</td>\n",
       "      <td>112718.195628</td>\n",
       "      <td>48103.172703</td>\n",
       "      <td>4709.563541</td>\n",
       "      <td>89743.227845</td>\n",
       "      <td>1185.774983</td>\n",
       "      <td>1185.774983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7015</th>\n",
       "      <td>dh1a</td>\n",
       "      <td>I</td>\n",
       "      <td>gene</td>\n",
       "      <td>3753000</td>\n",
       "      <td>3757000</td>\n",
       "      <td>+</td>\n",
       "      <td>dh1_plus</td>\n",
       "      <td>4001.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>59047.863616</td>\n",
       "      <td>64496.167096</td>\n",
       "      <td>42670.309302</td>\n",
       "      <td>72909.101415</td>\n",
       "      <td>92407.529966</td>\n",
       "      <td>38165.098682</td>\n",
       "      <td>5199.304858</td>\n",
       "      <td>67757.508482</td>\n",
       "      <td>912.894055</td>\n",
       "      <td>912.894055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7016</th>\n",
       "      <td>dh1b</td>\n",
       "      <td>I</td>\n",
       "      <td>gene</td>\n",
       "      <td>3753000</td>\n",
       "      <td>3757000</td>\n",
       "      <td>-</td>\n",
       "      <td>dh1_minus</td>\n",
       "      <td>4001.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>59047.863616</td>\n",
       "      <td>64496.167096</td>\n",
       "      <td>42670.309302</td>\n",
       "      <td>72909.101415</td>\n",
       "      <td>92407.529966</td>\n",
       "      <td>38165.098682</td>\n",
       "      <td>5199.304858</td>\n",
       "      <td>67757.508482</td>\n",
       "      <td>912.894055</td>\n",
       "      <td>912.894055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7017</th>\n",
       "      <td>spncrna.9001</td>\n",
       "      <td>II</td>\n",
       "      <td>gene</td>\n",
       "      <td>4442542</td>\n",
       "      <td>4445970</td>\n",
       "      <td>+</td>\n",
       "      <td>prt2</td>\n",
       "      <td>3429.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>79.164005</td>\n",
       "      <td>53.117797</td>\n",
       "      <td>77.013263</td>\n",
       "      <td>21.794728</td>\n",
       "      <td>19.749407</td>\n",
       "      <td>106.083019</td>\n",
       "      <td>98.119399</td>\n",
       "      <td>15.720966</td>\n",
       "      <td>142.953579</td>\n",
       "      <td>142.953579</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7018 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           gene_id               seqid    type    start  \\\n",
       "0          FP565355_region_1..2120  mating_type_region  region        1   \n",
       "1     FP565355_region_15417..15473  mating_type_region  region    15417   \n",
       "2     FP565355_region_15474..15608  mating_type_region  region    15474   \n",
       "3     FP565355_region_15609..16735  mating_type_region  region    15609   \n",
       "4     FP565355_region_16736..16794  mating_type_region  region    16736   \n",
       "...                            ...                 ...     ...      ...   \n",
       "7013                          dg1a                   I    gene  3761000   \n",
       "7014                          dg1b                   I    gene  3761000   \n",
       "7015                          dh1a                   I    gene  3753000   \n",
       "7016                          dh1b                   I    gene  3753000   \n",
       "7017                  spncrna.9001                  II    gene  4442542   \n",
       "\n",
       "          end strand  gene_name  cds_length  utr_length  intron_length  ...  \\\n",
       "0        2120      +        NaN         NaN         NaN            NaN  ...   \n",
       "1       15473      +        NaN         NaN         NaN            NaN  ...   \n",
       "2       15608      +        NaN         NaN         NaN            NaN  ...   \n",
       "3       16735      +        NaN         NaN         NaN            NaN  ...   \n",
       "4       16794      +        NaN         NaN         NaN            NaN  ...   \n",
       "...       ...    ...        ...         ...         ...            ...  ...   \n",
       "7013  3764500      +   dg1_plus      3501.0         NaN            NaN  ...   \n",
       "7014  3764500      -  dg1_minus      3501.0         NaN            NaN  ...   \n",
       "7015  3757000      +   dh1_plus      4001.0         NaN            NaN  ...   \n",
       "7016  3757000      -  dh1_minus      4001.0         NaN            NaN  ...   \n",
       "7017  4445970      +       prt2      3429.0         NaN            NaN  ...   \n",
       "\n",
       "      302_H3K9me2_1 324_H3K9me2_1 491_H3K9me2_1  504_H3K9me2_1  510_H3K9me2_1  \\\n",
       "0        408.094181   4502.883538   4993.117682    3063.305824    5711.894487   \n",
       "1       1162.048668   1971.046770   1686.235695    7319.396234    7398.231799   \n",
       "2       1391.168477   6279.042506   7482.761671   13815.154482   14955.159873   \n",
       "3        439.435683   2448.427202   5792.039436    2234.060733    5488.584094   \n",
       "4        242.514505   4121.908333   2797.902190    9192.532044    4486.233712   \n",
       "...             ...           ...           ...            ...            ...   \n",
       "7013   58234.743863  96152.109937  48565.266936  112852.707168  112718.195628   \n",
       "7014   58234.743863  96152.109937  48565.266936  112852.707168  112718.195628   \n",
       "7015   59047.863616  64496.167096  42670.309302   72909.101415   92407.529966   \n",
       "7016   59047.863616  64496.167096  42670.309302   72909.101415   92407.529966   \n",
       "7017      79.164005     53.117797     77.013263      21.794728      19.749407   \n",
       "\n",
       "      530_H3K9me2_1  638_H3K9me2_1  WT_H3K9me2_1  \\\n",
       "0       2515.868824    4165.457917   3124.282423   \n",
       "1       2747.824806    8045.427287   3527.447203   \n",
       "2       6049.474767   15653.804910   9781.204416   \n",
       "3       2627.398594    6195.784590   3577.713987   \n",
       "4       2686.762033   10219.535118   2041.416254   \n",
       "...             ...            ...           ...   \n",
       "7013   48103.172703    4709.563541  89743.227845   \n",
       "7014   48103.172703    4709.563541  89743.227845   \n",
       "7015   38165.098682    5199.304858  67757.508482   \n",
       "7016   38165.098682    5199.304858  67757.508482   \n",
       "7017     106.083019      98.119399     15.720966   \n",
       "\n",
       "      chip-fake-reads_simulated-data_1  chip-fake-reads_simulated-data_2  \n",
       "0                           581.140715                        581.140715  \n",
       "1                           147.177306                        147.177306  \n",
       "2                           588.709225                        588.709225  \n",
       "3                           441.531919                        441.531919  \n",
       "4                           588.709225                        588.709225  \n",
       "...                                ...                               ...  \n",
       "7013                       1185.774983                       1185.774983  \n",
       "7014                       1185.774983                       1185.774983  \n",
       "7015                        912.894055                        912.894055  \n",
       "7016                        912.894055                        912.894055  \n",
       "7017                        142.953579                        142.953579  \n",
       "\n",
       "[7018 rows x 24 columns]"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if params_dict['batch'] == 'gene_count':\n",
    "    # for raw gene counts\n",
    "    counts_df = load_gene_counts(counts_file, index_col='gene_id', gdf='pombase')\n",
    "else:\n",
    "    # for tpm\n",
    "    #counts_df = load_gene_counts(counts_file, index_col='gene_id', gdf='pombase', as_int=False)\n",
    "    counts_df = load_gene_counts(counts_file, index_col='gene_id', gdf='pombase')\n",
    "counts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7018, 24)"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **Log-transform** and filter for **Heterochromatic** and **Protein coding** (mRNA) genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>end</th>\n",
       "      <th>category</th>\n",
       "      <th>seqid</th>\n",
       "      <th>strand</th>\n",
       "      <th>utr_length</th>\n",
       "      <th>start</th>\n",
       "      <th>cds_length</th>\n",
       "      <th>gene_length</th>\n",
       "      <th>...</th>\n",
       "      <th>302_H3K9me2_1</th>\n",
       "      <th>324_H3K9me2_1</th>\n",
       "      <th>491_H3K9me2_1</th>\n",
       "      <th>504_H3K9me2_1</th>\n",
       "      <th>510_H3K9me2_1</th>\n",
       "      <th>530_H3K9me2_1</th>\n",
       "      <th>638_H3K9me2_1</th>\n",
       "      <th>WT_H3K9me2_1</th>\n",
       "      <th>chip-fake-reads_simulated-data_1</th>\n",
       "      <th>chip-fake-reads_simulated-data_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>FP565355_region_1..2120</td>\n",
       "      <td>2120</td>\n",
       "      <td>repeat</td>\n",
       "      <td>mating_type_region</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2120</td>\n",
       "      <td>...</td>\n",
       "      <td>8.676289</td>\n",
       "      <td>12.136954</td>\n",
       "      <td>12.286014</td>\n",
       "      <td>11.581345</td>\n",
       "      <td>12.480006</td>\n",
       "      <td>11.297414</td>\n",
       "      <td>12.024606</td>\n",
       "      <td>11.609771</td>\n",
       "      <td>9.185224</td>\n",
       "      <td>9.185224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>FP565355_region_15417..15473</td>\n",
       "      <td>15473</td>\n",
       "      <td>repeat</td>\n",
       "      <td>mating_type_region</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15417</td>\n",
       "      <td>NaN</td>\n",
       "      <td>57</td>\n",
       "      <td>...</td>\n",
       "      <td>10.183696</td>\n",
       "      <td>10.945478</td>\n",
       "      <td>10.720446</td>\n",
       "      <td>12.837706</td>\n",
       "      <td>12.853160</td>\n",
       "      <td>11.424599</td>\n",
       "      <td>12.974133</td>\n",
       "      <td>11.784818</td>\n",
       "      <td>7.211181</td>\n",
       "      <td>7.211181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>FP565355_region_15474..15608</td>\n",
       "      <td>15608</td>\n",
       "      <td>repeat</td>\n",
       "      <td>mating_type_region</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15474</td>\n",
       "      <td>NaN</td>\n",
       "      <td>135</td>\n",
       "      <td>...</td>\n",
       "      <td>10.443118</td>\n",
       "      <td>12.616559</td>\n",
       "      <td>12.869548</td>\n",
       "      <td>13.754068</td>\n",
       "      <td>13.868452</td>\n",
       "      <td>12.562833</td>\n",
       "      <td>13.934318</td>\n",
       "      <td>13.255944</td>\n",
       "      <td>9.203860</td>\n",
       "      <td>9.203860</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>FP565355_region_15609..16735</td>\n",
       "      <td>16735</td>\n",
       "      <td>repeat</td>\n",
       "      <td>mating_type_region</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15609</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1127</td>\n",
       "      <td>...</td>\n",
       "      <td>8.782788</td>\n",
       "      <td>11.258229</td>\n",
       "      <td>12.500105</td>\n",
       "      <td>11.126098</td>\n",
       "      <td>12.422481</td>\n",
       "      <td>11.359968</td>\n",
       "      <td>12.597304</td>\n",
       "      <td>11.805226</td>\n",
       "      <td>8.789638</td>\n",
       "      <td>8.789638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>FP565355_region_16736..16794</td>\n",
       "      <td>16794</td>\n",
       "      <td>repeat</td>\n",
       "      <td>mating_type_region</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16736</td>\n",
       "      <td>NaN</td>\n",
       "      <td>59</td>\n",
       "      <td>...</td>\n",
       "      <td>7.927864</td>\n",
       "      <td>12.009447</td>\n",
       "      <td>11.450645</td>\n",
       "      <td>13.166404</td>\n",
       "      <td>12.131611</td>\n",
       "      <td>11.392190</td>\n",
       "      <td>13.319183</td>\n",
       "      <td>10.996061</td>\n",
       "      <td>9.203860</td>\n",
       "      <td>9.203860</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name                       gene_id    end category               seqid  \\\n",
       "0       NaN       FP565355_region_1..2120   2120   repeat  mating_type_region   \n",
       "1       NaN  FP565355_region_15417..15473  15473   repeat  mating_type_region   \n",
       "2       NaN  FP565355_region_15474..15608  15608   repeat  mating_type_region   \n",
       "3       NaN  FP565355_region_15609..16735  16735   repeat  mating_type_region   \n",
       "4       NaN  FP565355_region_16736..16794  16794   repeat  mating_type_region   \n",
       "\n",
       "  strand  utr_length  start  cds_length  gene_length  ...  302_H3K9me2_1  \\\n",
       "0      +         NaN      1         NaN         2120  ...       8.676289   \n",
       "1      +         NaN  15417         NaN           57  ...      10.183696   \n",
       "2      +         NaN  15474         NaN          135  ...      10.443118   \n",
       "3      +         NaN  15609         NaN         1127  ...       8.782788   \n",
       "4      +         NaN  16736         NaN           59  ...       7.927864   \n",
       "\n",
       "  324_H3K9me2_1 491_H3K9me2_1  504_H3K9me2_1  510_H3K9me2_1  530_H3K9me2_1  \\\n",
       "0     12.136954     12.286014      11.581345      12.480006      11.297414   \n",
       "1     10.945478     10.720446      12.837706      12.853160      11.424599   \n",
       "2     12.616559     12.869548      13.754068      13.868452      12.562833   \n",
       "3     11.258229     12.500105      11.126098      12.422481      11.359968   \n",
       "4     12.009447     11.450645      13.166404      12.131611      11.392190   \n",
       "\n",
       "   638_H3K9me2_1  WT_H3K9me2_1  chip-fake-reads_simulated-data_1  \\\n",
       "0      12.024606     11.609771                          9.185224   \n",
       "1      12.974133     11.784818                          7.211181   \n",
       "2      13.934318     13.255944                          9.203860   \n",
       "3      12.597304     11.805226                          8.789638   \n",
       "4      13.319183     10.996061                          9.203860   \n",
       "\n",
       "   chip-fake-reads_simulated-data_2  \n",
       "0                          9.185224  \n",
       "1                          7.211181  \n",
       "2                          9.203860  \n",
       "3                          8.789638  \n",
       "4                          9.203860  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if (params_dict['log_transform']):\n",
    "    #counts_df = to_log2_tpm(counts_df) ## default shift = 1\n",
    "    counts_df = to_log2_tpm(counts_df, gene_id_col='gene_id', shift=1)\n",
    "\n",
    "    # Keep only Heterochromatic genes and  protein coding genes\n",
    "    #counts_df = counts_df[(counts_df['type'] == 'gene') | (counts_df['category'] == 'repeat')]\n",
    "    counts_df = counts_df[(counts_df['bio_type'] == 'mRNA') | (counts_df['category'] == 'repeat')]\n",
    "\n",
    "counts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5159, 24)"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Select `sample_id`s for samples present in `counts_df`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#sample_ids = [ii for ii in counts_df.columns if ii != 'gene_id']\n",
    "sample_ids = [ss for ss in counts_df.columns if ss not in long_header]\n",
    "len(sample_ids)\n",
    "#sample_ids"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Select only `gene_id` and `sample_id`'s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "counts_df = counts_df[['gene_id'] + sample_ids]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "#counts_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **B**. Results <font color='blue'> **with** </font> **NH-normalization**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Import **gene expresion Matrix**: `*_pombe_tpm_matrix.csv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/gcm-lfs1/pablo/data/rna_silencing/results/xp_data/ChIP'"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if params_dict['seq_type'] == 'rna':\n",
    "    # stranded=True, remove duplicate reads mapping to same gene, count only multimapped if repeat (divide by NH)\n",
    "    norm_xp_data_dir = rna_dir\n",
    "    #norm_xp_data_dir = os.path.join(data_results_dir, params_dict['batch'], 'RNA')\n",
    "else:\n",
    "    # stranded=False, remove duplicate reads mapping to same gene, count only multimapped if repeat (divide by NH)\n",
    "    norm_xp_data_dir = chip_dir\n",
    "    #norm_xp_data_dir = os.path.join(data_results_dir, params_dict['batch'], 'ChIP')\n",
    "norm_xp_data_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/gcm-lfs1/pablo/data/rna_silencing/results/xp_data/ChIP/h3k9me2_chip_pombe_tpm_matrix.csv'"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if (params_dict['seq_type'] == 'chip') and (params_dict['h3k9me2']):\n",
    "    norm_counts_file = os.path.join(norm_xp_data_dir,  'h3k9me2_' + params_dict['seq_type'] + '_pombe_' + params_dict['expression'] + '_matrix.csv')\n",
    "else:\n",
    "    norm_counts_file = os.path.join(norm_xp_data_dir, params_dict['seq_type'] + '_pombe_' + params_dict['expression'] + '_matrix.csv')\n",
    "norm_counts_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>seqid</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>cds_length</th>\n",
       "      <th>utr_length</th>\n",
       "      <th>intron_length</th>\n",
       "      <th>...</th>\n",
       "      <th>302_H3K9me2_1</th>\n",
       "      <th>324_H3K9me2_1</th>\n",
       "      <th>491_H3K9me2_1</th>\n",
       "      <th>504_H3K9me2_1</th>\n",
       "      <th>510_H3K9me2_1</th>\n",
       "      <th>530_H3K9me2_1</th>\n",
       "      <th>638_H3K9me2_1</th>\n",
       "      <th>WT_H3K9me2_1</th>\n",
       "      <th>chip-fake-reads_simulated-data_1</th>\n",
       "      <th>chip-fake-reads_simulated-data_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FP565355_region_1..2120</td>\n",
       "      <td>mating_type_region</td>\n",
       "      <td>region</td>\n",
       "      <td>1</td>\n",
       "      <td>2120</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>164.336295</td>\n",
       "      <td>2307.272915</td>\n",
       "      <td>1834.725481</td>\n",
       "      <td>2253.683503</td>\n",
       "      <td>4356.955689</td>\n",
       "      <td>940.184246</td>\n",
       "      <td>1336.748605</td>\n",
       "      <td>2025.518399</td>\n",
       "      <td>149.274199</td>\n",
       "      <td>149.274199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FP565355_region_15417..15473</td>\n",
       "      <td>mating_type_region</td>\n",
       "      <td>region</td>\n",
       "      <td>15417</td>\n",
       "      <td>15473</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1861.995895</td>\n",
       "      <td>4003.507846</td>\n",
       "      <td>2457.144726</td>\n",
       "      <td>21265.246117</td>\n",
       "      <td>22372.720331</td>\n",
       "      <td>4068.558264</td>\n",
       "      <td>10208.141304</td>\n",
       "      <td>9067.231621</td>\n",
       "      <td>149.292221</td>\n",
       "      <td>149.292221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FP565355_region_15474..15608</td>\n",
       "      <td>mating_type_region</td>\n",
       "      <td>region</td>\n",
       "      <td>15474</td>\n",
       "      <td>15608</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>557.280874</td>\n",
       "      <td>3188.432197</td>\n",
       "      <td>2725.927999</td>\n",
       "      <td>10034.388453</td>\n",
       "      <td>11306.337056</td>\n",
       "      <td>2239.284006</td>\n",
       "      <td>4965.437098</td>\n",
       "      <td>6285.596982</td>\n",
       "      <td>149.292221</td>\n",
       "      <td>149.292221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FP565355_region_15609..16735</td>\n",
       "      <td>mating_type_region</td>\n",
       "      <td>region</td>\n",
       "      <td>15609</td>\n",
       "      <td>16735</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>234.708310</td>\n",
       "      <td>1657.714345</td>\n",
       "      <td>2813.343195</td>\n",
       "      <td>2163.559662</td>\n",
       "      <td>5532.608345</td>\n",
       "      <td>1296.749856</td>\n",
       "      <td>2620.430311</td>\n",
       "      <td>3065.480457</td>\n",
       "      <td>149.292221</td>\n",
       "      <td>149.292221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FP565355_region_16736..16794</td>\n",
       "      <td>mating_type_region</td>\n",
       "      <td>region</td>\n",
       "      <td>16736</td>\n",
       "      <td>16794</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>97.147612</td>\n",
       "      <td>2093.061997</td>\n",
       "      <td>1019.260035</td>\n",
       "      <td>6676.829964</td>\n",
       "      <td>3391.663539</td>\n",
       "      <td>994.536464</td>\n",
       "      <td>3241.669300</td>\n",
       "      <td>1311.854788</td>\n",
       "      <td>149.292221</td>\n",
       "      <td>149.292221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7013</th>\n",
       "      <td>dg1a</td>\n",
       "      <td>I</td>\n",
       "      <td>gene</td>\n",
       "      <td>3761000</td>\n",
       "      <td>3764500</td>\n",
       "      <td>+</td>\n",
       "      <td>dg1_plus</td>\n",
       "      <td>3501.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>11139.692310</td>\n",
       "      <td>21691.546349</td>\n",
       "      <td>8659.794733</td>\n",
       "      <td>36676.451155</td>\n",
       "      <td>42290.611930</td>\n",
       "      <td>8818.917732</td>\n",
       "      <td>682.669596</td>\n",
       "      <td>27952.456058</td>\n",
       "      <td>149.355127</td>\n",
       "      <td>149.355127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7014</th>\n",
       "      <td>dg1b</td>\n",
       "      <td>I</td>\n",
       "      <td>gene</td>\n",
       "      <td>3761000</td>\n",
       "      <td>3764500</td>\n",
       "      <td>-</td>\n",
       "      <td>dg1_minus</td>\n",
       "      <td>3501.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>11139.692310</td>\n",
       "      <td>21691.546349</td>\n",
       "      <td>8659.794733</td>\n",
       "      <td>36676.451155</td>\n",
       "      <td>42290.611930</td>\n",
       "      <td>8818.917732</td>\n",
       "      <td>682.669596</td>\n",
       "      <td>27952.456058</td>\n",
       "      <td>149.355127</td>\n",
       "      <td>149.355127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7015</th>\n",
       "      <td>dh1a</td>\n",
       "      <td>I</td>\n",
       "      <td>gene</td>\n",
       "      <td>3753000</td>\n",
       "      <td>3757000</td>\n",
       "      <td>+</td>\n",
       "      <td>dh1_plus</td>\n",
       "      <td>4001.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>14610.461803</td>\n",
       "      <td>17351.236631</td>\n",
       "      <td>9517.936436</td>\n",
       "      <td>27721.865201</td>\n",
       "      <td>40728.792993</td>\n",
       "      <td>8639.608773</td>\n",
       "      <td>923.452207</td>\n",
       "      <td>24701.222884</td>\n",
       "      <td>149.292221</td>\n",
       "      <td>149.292221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7016</th>\n",
       "      <td>dh1b</td>\n",
       "      <td>I</td>\n",
       "      <td>gene</td>\n",
       "      <td>3753000</td>\n",
       "      <td>3757000</td>\n",
       "      <td>-</td>\n",
       "      <td>dh1_minus</td>\n",
       "      <td>4001.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>14610.461803</td>\n",
       "      <td>17351.236631</td>\n",
       "      <td>9517.936436</td>\n",
       "      <td>27721.865201</td>\n",
       "      <td>40728.792993</td>\n",
       "      <td>8639.608773</td>\n",
       "      <td>923.452207</td>\n",
       "      <td>24701.222884</td>\n",
       "      <td>149.292221</td>\n",
       "      <td>149.292221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7017</th>\n",
       "      <td>spncrna.9001</td>\n",
       "      <td>II</td>\n",
       "      <td>gene</td>\n",
       "      <td>4442542</td>\n",
       "      <td>4445970</td>\n",
       "      <td>+</td>\n",
       "      <td>prt2</td>\n",
       "      <td>3429.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>126.847572</td>\n",
       "      <td>107.890651</td>\n",
       "      <td>112.221995</td>\n",
       "      <td>63.320832</td>\n",
       "      <td>59.723454</td>\n",
       "      <td>157.071493</td>\n",
       "      <td>124.495151</td>\n",
       "      <td>40.410424</td>\n",
       "      <td>145.007799</td>\n",
       "      <td>145.007799</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7018 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           gene_id               seqid    type    start  \\\n",
       "0          FP565355_region_1..2120  mating_type_region  region        1   \n",
       "1     FP565355_region_15417..15473  mating_type_region  region    15417   \n",
       "2     FP565355_region_15474..15608  mating_type_region  region    15474   \n",
       "3     FP565355_region_15609..16735  mating_type_region  region    15609   \n",
       "4     FP565355_region_16736..16794  mating_type_region  region    16736   \n",
       "...                            ...                 ...     ...      ...   \n",
       "7013                          dg1a                   I    gene  3761000   \n",
       "7014                          dg1b                   I    gene  3761000   \n",
       "7015                          dh1a                   I    gene  3753000   \n",
       "7016                          dh1b                   I    gene  3753000   \n",
       "7017                  spncrna.9001                  II    gene  4442542   \n",
       "\n",
       "          end strand  gene_name  cds_length  utr_length  intron_length  ...  \\\n",
       "0        2120      +        NaN         NaN         NaN            NaN  ...   \n",
       "1       15473      +        NaN         NaN         NaN            NaN  ...   \n",
       "2       15608      +        NaN         NaN         NaN            NaN  ...   \n",
       "3       16735      +        NaN         NaN         NaN            NaN  ...   \n",
       "4       16794      +        NaN         NaN         NaN            NaN  ...   \n",
       "...       ...    ...        ...         ...         ...            ...  ...   \n",
       "7013  3764500      +   dg1_plus      3501.0         NaN            NaN  ...   \n",
       "7014  3764500      -  dg1_minus      3501.0         NaN            NaN  ...   \n",
       "7015  3757000      +   dh1_plus      4001.0         NaN            NaN  ...   \n",
       "7016  3757000      -  dh1_minus      4001.0         NaN            NaN  ...   \n",
       "7017  4445970      +       prt2      3429.0         NaN            NaN  ...   \n",
       "\n",
       "      302_H3K9me2_1 324_H3K9me2_1 491_H3K9me2_1  504_H3K9me2_1  510_H3K9me2_1  \\\n",
       "0        164.336295   2307.272915   1834.725481    2253.683503    4356.955689   \n",
       "1       1861.995895   4003.507846   2457.144726   21265.246117   22372.720331   \n",
       "2        557.280874   3188.432197   2725.927999   10034.388453   11306.337056   \n",
       "3        234.708310   1657.714345   2813.343195    2163.559662    5532.608345   \n",
       "4         97.147612   2093.061997   1019.260035    6676.829964    3391.663539   \n",
       "...             ...           ...           ...            ...            ...   \n",
       "7013   11139.692310  21691.546349   8659.794733   36676.451155   42290.611930   \n",
       "7014   11139.692310  21691.546349   8659.794733   36676.451155   42290.611930   \n",
       "7015   14610.461803  17351.236631   9517.936436   27721.865201   40728.792993   \n",
       "7016   14610.461803  17351.236631   9517.936436   27721.865201   40728.792993   \n",
       "7017     126.847572    107.890651    112.221995      63.320832      59.723454   \n",
       "\n",
       "      530_H3K9me2_1  638_H3K9me2_1  WT_H3K9me2_1  \\\n",
       "0        940.184246    1336.748605   2025.518399   \n",
       "1       4068.558264   10208.141304   9067.231621   \n",
       "2       2239.284006    4965.437098   6285.596982   \n",
       "3       1296.749856    2620.430311   3065.480457   \n",
       "4        994.536464    3241.669300   1311.854788   \n",
       "...             ...            ...           ...   \n",
       "7013    8818.917732     682.669596  27952.456058   \n",
       "7014    8818.917732     682.669596  27952.456058   \n",
       "7015    8639.608773     923.452207  24701.222884   \n",
       "7016    8639.608773     923.452207  24701.222884   \n",
       "7017     157.071493     124.495151     40.410424   \n",
       "\n",
       "      chip-fake-reads_simulated-data_1  chip-fake-reads_simulated-data_2  \n",
       "0                           149.274199                        149.274199  \n",
       "1                           149.292221                        149.292221  \n",
       "2                           149.292221                        149.292221  \n",
       "3                           149.292221                        149.292221  \n",
       "4                           149.292221                        149.292221  \n",
       "...                                ...                               ...  \n",
       "7013                        149.355127                        149.355127  \n",
       "7014                        149.355127                        149.355127  \n",
       "7015                        149.292221                        149.292221  \n",
       "7016                        149.292221                        149.292221  \n",
       "7017                        145.007799                        145.007799  \n",
       "\n",
       "[7018 rows x 24 columns]"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if params_dict['batch'] == 'gene_count':\n",
    "    # for raw gene counts\n",
    "    norm_counts_df = load_gene_counts(norm_counts_file, index_col='gene_id', gdf='pombase', sample_ids=sample_ids)\n",
    "else:\n",
    "    # for tpm\n",
    "    #norm_counts_df = load_gene_counts(norm_counts_file, index_col='gene_id', gdf='pombase', sample_ids=sample_ids, as_int=False)\n",
    "    norm_counts_df = load_gene_counts(norm_counts_file, index_col='gene_id', gdf='pombase', sample_ids=sample_ids) \n",
    "\n",
    "norm_counts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7018, 24)"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "norm_counts_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **Log-transform** and filter for **Heterochromatic** and **Protein coding** (mRNA) genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>end</th>\n",
       "      <th>category</th>\n",
       "      <th>seqid</th>\n",
       "      <th>strand</th>\n",
       "      <th>utr_length</th>\n",
       "      <th>start</th>\n",
       "      <th>cds_length</th>\n",
       "      <th>gene_length</th>\n",
       "      <th>...</th>\n",
       "      <th>302_H3K9me2_1</th>\n",
       "      <th>324_H3K9me2_1</th>\n",
       "      <th>491_H3K9me2_1</th>\n",
       "      <th>504_H3K9me2_1</th>\n",
       "      <th>510_H3K9me2_1</th>\n",
       "      <th>530_H3K9me2_1</th>\n",
       "      <th>638_H3K9me2_1</th>\n",
       "      <th>WT_H3K9me2_1</th>\n",
       "      <th>chip-fake-reads_simulated-data_1</th>\n",
       "      <th>chip-fake-reads_simulated-data_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>FP565355_region_1..2120</td>\n",
       "      <td>2120</td>\n",
       "      <td>repeat</td>\n",
       "      <td>mating_type_region</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2120</td>\n",
       "      <td>...</td>\n",
       "      <td>7.369260</td>\n",
       "      <td>11.172598</td>\n",
       "      <td>10.842135</td>\n",
       "      <td>11.138709</td>\n",
       "      <td>12.089436</td>\n",
       "      <td>9.878333</td>\n",
       "      <td>10.385591</td>\n",
       "      <td>10.984788</td>\n",
       "      <td>7.231454</td>\n",
       "      <td>7.231454</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>FP565355_region_15417..15473</td>\n",
       "      <td>15473</td>\n",
       "      <td>repeat</td>\n",
       "      <td>mating_type_region</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15417</td>\n",
       "      <td>NaN</td>\n",
       "      <td>57</td>\n",
       "      <td>...</td>\n",
       "      <td>10.863409</td>\n",
       "      <td>11.967409</td>\n",
       "      <td>11.263354</td>\n",
       "      <td>14.376278</td>\n",
       "      <td>14.449518</td>\n",
       "      <td>11.990656</td>\n",
       "      <td>13.317574</td>\n",
       "      <td>13.146606</td>\n",
       "      <td>7.231627</td>\n",
       "      <td>7.231627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>FP565355_region_15474..15608</td>\n",
       "      <td>15608</td>\n",
       "      <td>repeat</td>\n",
       "      <td>mating_type_region</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15474</td>\n",
       "      <td>NaN</td>\n",
       "      <td>135</td>\n",
       "      <td>...</td>\n",
       "      <td>9.124847</td>\n",
       "      <td>11.639084</td>\n",
       "      <td>11.413061</td>\n",
       "      <td>13.292809</td>\n",
       "      <td>13.464972</td>\n",
       "      <td>11.129466</td>\n",
       "      <td>12.277996</td>\n",
       "      <td>12.618064</td>\n",
       "      <td>7.231627</td>\n",
       "      <td>7.231627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>FP565355_region_15609..16735</td>\n",
       "      <td>16735</td>\n",
       "      <td>repeat</td>\n",
       "      <td>mating_type_region</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15609</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1127</td>\n",
       "      <td>...</td>\n",
       "      <td>7.880859</td>\n",
       "      <td>10.695850</td>\n",
       "      <td>11.458583</td>\n",
       "      <td>11.079858</td>\n",
       "      <td>12.434005</td>\n",
       "      <td>10.341797</td>\n",
       "      <td>11.356138</td>\n",
       "      <td>11.582368</td>\n",
       "      <td>7.231627</td>\n",
       "      <td>7.231627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>FP565355_region_16736..16794</td>\n",
       "      <td>16794</td>\n",
       "      <td>repeat</td>\n",
       "      <td>mating_type_region</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>16736</td>\n",
       "      <td>NaN</td>\n",
       "      <td>59</td>\n",
       "      <td>...</td>\n",
       "      <td>6.616881</td>\n",
       "      <td>11.032088</td>\n",
       "      <td>9.994721</td>\n",
       "      <td>12.705164</td>\n",
       "      <td>11.728203</td>\n",
       "      <td>9.959330</td>\n",
       "      <td>11.662966</td>\n",
       "      <td>10.358492</td>\n",
       "      <td>7.231627</td>\n",
       "      <td>7.231627</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name                       gene_id    end category               seqid  \\\n",
       "0       NaN       FP565355_region_1..2120   2120   repeat  mating_type_region   \n",
       "1       NaN  FP565355_region_15417..15473  15473   repeat  mating_type_region   \n",
       "2       NaN  FP565355_region_15474..15608  15608   repeat  mating_type_region   \n",
       "3       NaN  FP565355_region_15609..16735  16735   repeat  mating_type_region   \n",
       "4       NaN  FP565355_region_16736..16794  16794   repeat  mating_type_region   \n",
       "\n",
       "  strand  utr_length  start  cds_length  gene_length  ...  302_H3K9me2_1  \\\n",
       "0      +         NaN      1         NaN         2120  ...       7.369260   \n",
       "1      +         NaN  15417         NaN           57  ...      10.863409   \n",
       "2      +         NaN  15474         NaN          135  ...       9.124847   \n",
       "3      +         NaN  15609         NaN         1127  ...       7.880859   \n",
       "4      +         NaN  16736         NaN           59  ...       6.616881   \n",
       "\n",
       "  324_H3K9me2_1 491_H3K9me2_1  504_H3K9me2_1  510_H3K9me2_1  530_H3K9me2_1  \\\n",
       "0     11.172598     10.842135      11.138709      12.089436       9.878333   \n",
       "1     11.967409     11.263354      14.376278      14.449518      11.990656   \n",
       "2     11.639084     11.413061      13.292809      13.464972      11.129466   \n",
       "3     10.695850     11.458583      11.079858      12.434005      10.341797   \n",
       "4     11.032088      9.994721      12.705164      11.728203       9.959330   \n",
       "\n",
       "   638_H3K9me2_1  WT_H3K9me2_1  chip-fake-reads_simulated-data_1  \\\n",
       "0      10.385591     10.984788                          7.231454   \n",
       "1      13.317574     13.146606                          7.231627   \n",
       "2      12.277996     12.618064                          7.231627   \n",
       "3      11.356138     11.582368                          7.231627   \n",
       "4      11.662966     10.358492                          7.231627   \n",
       "\n",
       "   chip-fake-reads_simulated-data_2  \n",
       "0                          7.231454  \n",
       "1                          7.231627  \n",
       "2                          7.231627  \n",
       "3                          7.231627  \n",
       "4                          7.231627  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if (params_dict['log_transform']):\n",
    "    #norm_counts_df = to_log2_tpm(norm_counts_df) ## default shift = 1\n",
    "    norm_counts_df = to_log2_tpm(norm_counts_df, gene_id_col='gene_id', shift=1)\n",
    "\n",
    "    # Keep only Heterochromatic genes and  protein coding genes\n",
    "    #norm_counts_df = norm_counts_df[(norm_counts_df['type'] == 'gene') | (counts_df['category'] == 'repeat')]\n",
    "    norm_counts_df = norm_counts_df[(norm_counts_df['bio_type'] == 'mRNA') | (norm_counts_df['category'] == 'repeat')]\n",
    "\n",
    "norm_counts_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5159, 24)"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "norm_counts_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Select only `gene_id` and `sample_id`'s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "norm_counts_df = norm_counts_df[['gene_id'] + sample_ids]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "#norm_counts_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### **C**. Merge Results: <font color='blue'> **without** </font>  and <font color='blue'> **with** </font>  **NH-normalization**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Check for **missing genes**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "missing_genes = set(norm_counts_df['gene_id']).symmetric_difference(counts_df['gene_id'])\n",
    "len(missing_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **Merge** both counts DataFrames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "#sample_ids = list(set(sample_ids).symmetric_difference(norm_counts_df.columns))\n",
    "#sample_ids = list(set(sample_ids).intersection(norm_counts_df.columns))\n",
    "#sample_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "# use concat with `set_index` with 'gene_id' in both DataFrames, then `swaplevel` with `sort_index` for expected MultiIndex in columns:\n",
    "merged_xp = (pd.concat([counts_df.set_index('gene_id'), \n",
    "                norm_counts_df.set_index('gene_id')], \n",
    "                axis=1, \n",
    "                keys=['wo_norm', 'norm'], \n",
    "                sort=True)\n",
    "        .swaplevel(0, 1, axis=1)\n",
    "        .sort_index(axis=1, ascending=[True, False])\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"2\" halign=\"left\">301_H3K9me2_1</th>\n",
       "      <th colspan=\"2\" halign=\"left\">302_H3K9me2_1</th>\n",
       "      <th colspan=\"2\" halign=\"left\">324_H3K9me2_1</th>\n",
       "      <th colspan=\"2\" halign=\"left\">491_H3K9me2_1</th>\n",
       "      <th colspan=\"2\" halign=\"left\">504_H3K9me2_1</th>\n",
       "      <th>...</th>\n",
       "      <th colspan=\"2\" halign=\"left\">530_H3K9me2_1</th>\n",
       "      <th colspan=\"2\" halign=\"left\">638_H3K9me2_1</th>\n",
       "      <th colspan=\"2\" halign=\"left\">WT_H3K9me2_1</th>\n",
       "      <th colspan=\"2\" halign=\"left\">chip-fake-reads_simulated-data_1</th>\n",
       "      <th colspan=\"2\" halign=\"left\">chip-fake-reads_simulated-data_2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>wo_norm</th>\n",
       "      <th>norm</th>\n",
       "      <th>wo_norm</th>\n",
       "      <th>norm</th>\n",
       "      <th>wo_norm</th>\n",
       "      <th>norm</th>\n",
       "      <th>wo_norm</th>\n",
       "      <th>norm</th>\n",
       "      <th>wo_norm</th>\n",
       "      <th>norm</th>\n",
       "      <th>...</th>\n",
       "      <th>wo_norm</th>\n",
       "      <th>norm</th>\n",
       "      <th>wo_norm</th>\n",
       "      <th>norm</th>\n",
       "      <th>wo_norm</th>\n",
       "      <th>norm</th>\n",
       "      <th>wo_norm</th>\n",
       "      <th>norm</th>\n",
       "      <th>wo_norm</th>\n",
       "      <th>norm</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>FP565355_region_1..2120</th>\n",
       "      <td>8.144838</td>\n",
       "      <td>7.770820</td>\n",
       "      <td>8.676289</td>\n",
       "      <td>7.369260</td>\n",
       "      <td>12.136954</td>\n",
       "      <td>11.172598</td>\n",
       "      <td>12.286014</td>\n",
       "      <td>10.842135</td>\n",
       "      <td>11.581345</td>\n",
       "      <td>11.138709</td>\n",
       "      <td>...</td>\n",
       "      <td>11.297414</td>\n",
       "      <td>9.878333</td>\n",
       "      <td>12.024606</td>\n",
       "      <td>10.385591</td>\n",
       "      <td>11.609771</td>\n",
       "      <td>10.984788</td>\n",
       "      <td>9.185224</td>\n",
       "      <td>7.231454</td>\n",
       "      <td>9.185224</td>\n",
       "      <td>7.231454</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FP565355_region_15417..15473</th>\n",
       "      <td>10.089078</td>\n",
       "      <td>11.701693</td>\n",
       "      <td>10.183696</td>\n",
       "      <td>10.863409</td>\n",
       "      <td>10.945478</td>\n",
       "      <td>11.967409</td>\n",
       "      <td>10.720446</td>\n",
       "      <td>11.263354</td>\n",
       "      <td>12.837706</td>\n",
       "      <td>14.376278</td>\n",
       "      <td>...</td>\n",
       "      <td>11.424599</td>\n",
       "      <td>11.990656</td>\n",
       "      <td>12.974133</td>\n",
       "      <td>13.317574</td>\n",
       "      <td>11.784818</td>\n",
       "      <td>13.146606</td>\n",
       "      <td>7.211181</td>\n",
       "      <td>7.231627</td>\n",
       "      <td>7.211181</td>\n",
       "      <td>7.231627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FP565355_region_15474..15608</th>\n",
       "      <td>10.488893</td>\n",
       "      <td>10.102709</td>\n",
       "      <td>10.443118</td>\n",
       "      <td>9.124847</td>\n",
       "      <td>12.616559</td>\n",
       "      <td>11.639084</td>\n",
       "      <td>12.869548</td>\n",
       "      <td>11.413061</td>\n",
       "      <td>13.754068</td>\n",
       "      <td>13.292809</td>\n",
       "      <td>...</td>\n",
       "      <td>12.562833</td>\n",
       "      <td>11.129466</td>\n",
       "      <td>13.934318</td>\n",
       "      <td>12.277996</td>\n",
       "      <td>13.255944</td>\n",
       "      <td>12.618064</td>\n",
       "      <td>9.203860</td>\n",
       "      <td>7.231627</td>\n",
       "      <td>9.203860</td>\n",
       "      <td>7.231627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FP565355_region_15609..16735</th>\n",
       "      <td>9.567602</td>\n",
       "      <td>9.596110</td>\n",
       "      <td>8.782788</td>\n",
       "      <td>7.880859</td>\n",
       "      <td>11.258229</td>\n",
       "      <td>10.695850</td>\n",
       "      <td>12.500105</td>\n",
       "      <td>11.458583</td>\n",
       "      <td>11.126098</td>\n",
       "      <td>11.079858</td>\n",
       "      <td>...</td>\n",
       "      <td>11.359968</td>\n",
       "      <td>10.341797</td>\n",
       "      <td>12.597304</td>\n",
       "      <td>11.356138</td>\n",
       "      <td>11.805226</td>\n",
       "      <td>11.582368</td>\n",
       "      <td>8.789638</td>\n",
       "      <td>7.231627</td>\n",
       "      <td>8.789638</td>\n",
       "      <td>7.231627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FP565355_region_16736..16794</th>\n",
       "      <td>6.936350</td>\n",
       "      <td>6.553472</td>\n",
       "      <td>7.927864</td>\n",
       "      <td>6.616881</td>\n",
       "      <td>12.009447</td>\n",
       "      <td>11.032088</td>\n",
       "      <td>11.450645</td>\n",
       "      <td>9.994721</td>\n",
       "      <td>13.166404</td>\n",
       "      <td>12.705164</td>\n",
       "      <td>...</td>\n",
       "      <td>11.392190</td>\n",
       "      <td>9.959330</td>\n",
       "      <td>13.319183</td>\n",
       "      <td>11.662966</td>\n",
       "      <td>10.996061</td>\n",
       "      <td>10.358492</td>\n",
       "      <td>9.203860</td>\n",
       "      <td>7.231627</td>\n",
       "      <td>9.203860</td>\n",
       "      <td>7.231627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPRPTCENA.4</th>\n",
       "      <td>17.160078</td>\n",
       "      <td>15.559783</td>\n",
       "      <td>16.086106</td>\n",
       "      <td>13.657590</td>\n",
       "      <td>16.878727</td>\n",
       "      <td>14.689923</td>\n",
       "      <td>15.791995</td>\n",
       "      <td>13.211908</td>\n",
       "      <td>17.014930</td>\n",
       "      <td>15.336742</td>\n",
       "      <td>...</td>\n",
       "      <td>15.743017</td>\n",
       "      <td>13.200185</td>\n",
       "      <td>12.326876</td>\n",
       "      <td>9.496698</td>\n",
       "      <td>16.458359</td>\n",
       "      <td>14.670575</td>\n",
       "      <td>10.299144</td>\n",
       "      <td>7.232119</td>\n",
       "      <td>10.299144</td>\n",
       "      <td>7.232119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dg1a</th>\n",
       "      <td>17.000530</td>\n",
       "      <td>15.471866</td>\n",
       "      <td>15.829617</td>\n",
       "      <td>13.443551</td>\n",
       "      <td>16.553046</td>\n",
       "      <td>14.404912</td>\n",
       "      <td>15.567667</td>\n",
       "      <td>13.080284</td>\n",
       "      <td>16.784094</td>\n",
       "      <td>15.162606</td>\n",
       "      <td>...</td>\n",
       "      <td>15.553874</td>\n",
       "      <td>13.106549</td>\n",
       "      <td>12.201684</td>\n",
       "      <td>9.417155</td>\n",
       "      <td>16.453532</td>\n",
       "      <td>14.770739</td>\n",
       "      <td>10.212831</td>\n",
       "      <td>7.232230</td>\n",
       "      <td>10.212831</td>\n",
       "      <td>7.232230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dg1b</th>\n",
       "      <td>17.000530</td>\n",
       "      <td>15.471866</td>\n",
       "      <td>15.829617</td>\n",
       "      <td>13.443551</td>\n",
       "      <td>16.553046</td>\n",
       "      <td>14.404912</td>\n",
       "      <td>15.567667</td>\n",
       "      <td>13.080284</td>\n",
       "      <td>16.784094</td>\n",
       "      <td>15.162606</td>\n",
       "      <td>...</td>\n",
       "      <td>15.553874</td>\n",
       "      <td>13.106549</td>\n",
       "      <td>12.201684</td>\n",
       "      <td>9.417155</td>\n",
       "      <td>16.453532</td>\n",
       "      <td>14.770739</td>\n",
       "      <td>10.212831</td>\n",
       "      <td>7.232230</td>\n",
       "      <td>10.212831</td>\n",
       "      <td>7.232230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dh1a</th>\n",
       "      <td>16.630883</td>\n",
       "      <td>15.345470</td>\n",
       "      <td>15.849622</td>\n",
       "      <td>13.834813</td>\n",
       "      <td>15.976948</td>\n",
       "      <td>14.082834</td>\n",
       "      <td>15.380979</td>\n",
       "      <td>13.216585</td>\n",
       "      <td>16.153831</td>\n",
       "      <td>14.758789</td>\n",
       "      <td>...</td>\n",
       "      <td>15.220004</td>\n",
       "      <td>13.076917</td>\n",
       "      <td>12.344380</td>\n",
       "      <td>9.852455</td>\n",
       "      <td>16.048114</td>\n",
       "      <td>14.592353</td>\n",
       "      <td>9.835883</td>\n",
       "      <td>7.231627</td>\n",
       "      <td>9.835883</td>\n",
       "      <td>7.231627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dh1b</th>\n",
       "      <td>16.630883</td>\n",
       "      <td>15.345470</td>\n",
       "      <td>15.849622</td>\n",
       "      <td>13.834813</td>\n",
       "      <td>15.976948</td>\n",
       "      <td>14.082834</td>\n",
       "      <td>15.380979</td>\n",
       "      <td>13.216585</td>\n",
       "      <td>16.153831</td>\n",
       "      <td>14.758789</td>\n",
       "      <td>...</td>\n",
       "      <td>15.220004</td>\n",
       "      <td>13.076917</td>\n",
       "      <td>12.344380</td>\n",
       "      <td>9.852455</td>\n",
       "      <td>16.048114</td>\n",
       "      <td>14.592353</td>\n",
       "      <td>9.835883</td>\n",
       "      <td>7.231627</td>\n",
       "      <td>9.835883</td>\n",
       "      <td>7.231627</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5159 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             301_H3K9me2_1            302_H3K9me2_1  \\\n",
       "                                   wo_norm       norm       wo_norm   \n",
       "gene_id                                                               \n",
       "FP565355_region_1..2120           8.144838   7.770820      8.676289   \n",
       "FP565355_region_15417..15473     10.089078  11.701693     10.183696   \n",
       "FP565355_region_15474..15608     10.488893  10.102709     10.443118   \n",
       "FP565355_region_15609..16735      9.567602   9.596110      8.782788   \n",
       "FP565355_region_16736..16794      6.936350   6.553472      7.927864   \n",
       "...                                    ...        ...           ...   \n",
       "SPRPTCENA.4                      17.160078  15.559783     16.086106   \n",
       "dg1a                             17.000530  15.471866     15.829617   \n",
       "dg1b                             17.000530  15.471866     15.829617   \n",
       "dh1a                             16.630883  15.345470     15.849622   \n",
       "dh1b                             16.630883  15.345470     15.849622   \n",
       "\n",
       "                                        324_H3K9me2_1             \\\n",
       "                                   norm       wo_norm       norm   \n",
       "gene_id                                                            \n",
       "FP565355_region_1..2120        7.369260     12.136954  11.172598   \n",
       "FP565355_region_15417..15473  10.863409     10.945478  11.967409   \n",
       "FP565355_region_15474..15608   9.124847     12.616559  11.639084   \n",
       "FP565355_region_15609..16735   7.880859     11.258229  10.695850   \n",
       "FP565355_region_16736..16794   6.616881     12.009447  11.032088   \n",
       "...                                 ...           ...        ...   \n",
       "SPRPTCENA.4                   13.657590     16.878727  14.689923   \n",
       "dg1a                          13.443551     16.553046  14.404912   \n",
       "dg1b                          13.443551     16.553046  14.404912   \n",
       "dh1a                          13.834813     15.976948  14.082834   \n",
       "dh1b                          13.834813     15.976948  14.082834   \n",
       "\n",
       "                             491_H3K9me2_1            504_H3K9me2_1  \\\n",
       "                                   wo_norm       norm       wo_norm   \n",
       "gene_id                                                               \n",
       "FP565355_region_1..2120          12.286014  10.842135     11.581345   \n",
       "FP565355_region_15417..15473     10.720446  11.263354     12.837706   \n",
       "FP565355_region_15474..15608     12.869548  11.413061     13.754068   \n",
       "FP565355_region_15609..16735     12.500105  11.458583     11.126098   \n",
       "FP565355_region_16736..16794     11.450645   9.994721     13.166404   \n",
       "...                                    ...        ...           ...   \n",
       "SPRPTCENA.4                      15.791995  13.211908     17.014930   \n",
       "dg1a                             15.567667  13.080284     16.784094   \n",
       "dg1b                             15.567667  13.080284     16.784094   \n",
       "dh1a                             15.380979  13.216585     16.153831   \n",
       "dh1b                             15.380979  13.216585     16.153831   \n",
       "\n",
       "                                         ... 530_H3K9me2_1             \\\n",
       "                                   norm  ...       wo_norm       norm   \n",
       "gene_id                                  ...                            \n",
       "FP565355_region_1..2120       11.138709  ...     11.297414   9.878333   \n",
       "FP565355_region_15417..15473  14.376278  ...     11.424599  11.990656   \n",
       "FP565355_region_15474..15608  13.292809  ...     12.562833  11.129466   \n",
       "FP565355_region_15609..16735  11.079858  ...     11.359968  10.341797   \n",
       "FP565355_region_16736..16794  12.705164  ...     11.392190   9.959330   \n",
       "...                                 ...  ...           ...        ...   \n",
       "SPRPTCENA.4                   15.336742  ...     15.743017  13.200185   \n",
       "dg1a                          15.162606  ...     15.553874  13.106549   \n",
       "dg1b                          15.162606  ...     15.553874  13.106549   \n",
       "dh1a                          14.758789  ...     15.220004  13.076917   \n",
       "dh1b                          14.758789  ...     15.220004  13.076917   \n",
       "\n",
       "                             638_H3K9me2_1            WT_H3K9me2_1             \\\n",
       "                                   wo_norm       norm      wo_norm       norm   \n",
       "gene_id                                                                         \n",
       "FP565355_region_1..2120          12.024606  10.385591    11.609771  10.984788   \n",
       "FP565355_region_15417..15473     12.974133  13.317574    11.784818  13.146606   \n",
       "FP565355_region_15474..15608     13.934318  12.277996    13.255944  12.618064   \n",
       "FP565355_region_15609..16735     12.597304  11.356138    11.805226  11.582368   \n",
       "FP565355_region_16736..16794     13.319183  11.662966    10.996061  10.358492   \n",
       "...                                    ...        ...          ...        ...   \n",
       "SPRPTCENA.4                      12.326876   9.496698    16.458359  14.670575   \n",
       "dg1a                             12.201684   9.417155    16.453532  14.770739   \n",
       "dg1b                             12.201684   9.417155    16.453532  14.770739   \n",
       "dh1a                             12.344380   9.852455    16.048114  14.592353   \n",
       "dh1b                             12.344380   9.852455    16.048114  14.592353   \n",
       "\n",
       "                             chip-fake-reads_simulated-data_1            \\\n",
       "                                                      wo_norm      norm   \n",
       "gene_id                                                                   \n",
       "FP565355_region_1..2120                              9.185224  7.231454   \n",
       "FP565355_region_15417..15473                         7.211181  7.231627   \n",
       "FP565355_region_15474..15608                         9.203860  7.231627   \n",
       "FP565355_region_15609..16735                         8.789638  7.231627   \n",
       "FP565355_region_16736..16794                         9.203860  7.231627   \n",
       "...                                                       ...       ...   \n",
       "SPRPTCENA.4                                         10.299144  7.232119   \n",
       "dg1a                                                10.212831  7.232230   \n",
       "dg1b                                                10.212831  7.232230   \n",
       "dh1a                                                 9.835883  7.231627   \n",
       "dh1b                                                 9.835883  7.231627   \n",
       "\n",
       "                             chip-fake-reads_simulated-data_2            \n",
       "                                                      wo_norm      norm  \n",
       "gene_id                                                                  \n",
       "FP565355_region_1..2120                              9.185224  7.231454  \n",
       "FP565355_region_15417..15473                         7.211181  7.231627  \n",
       "FP565355_region_15474..15608                         9.203860  7.231627  \n",
       "FP565355_region_15609..16735                         8.789638  7.231627  \n",
       "FP565355_region_16736..16794                         9.203860  7.231627  \n",
       "...                                                       ...       ...  \n",
       "SPRPTCENA.4                                         10.299144  7.232119  \n",
       "dg1a                                                10.212831  7.232230  \n",
       "dg1b                                                10.212831  7.232230  \n",
       "dh1a                                                 9.835883  7.231627  \n",
       "dh1b                                                 9.835883  7.231627  \n",
       "\n",
       "[5159 rows x 22 columns]"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_xp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "# https://stackoverflow.com/questions/40225683/how-to-simply-add-a-column-level-to-a-pandas-dataframe\n",
    "#merged_xp.columns = pd.MultiIndex.from_tuples(map(lambda x: ( x, '_'.join(x.split('_')[:-1]) ), merged_xp.columns))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "#merged_xp = pd.merge(counts_df, xp_parastou_df, on='gene_id', how='outer')\n",
    "#merged_xp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5159, 22)"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_xp.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Visualization:  <font color='blue'> **without** </font>  vs <font color='blue'> **with** </font>  **NH-normalization**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **Visualization**: `scatter_plot`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "import collections"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "def map_color(ratio, n_clusters):\n",
    "    \n",
    "    clust = np.linspace(0, 1, n_clusters)\n",
    "    # init vars\n",
    "    lower_bound = 0\n",
    "    color = 0\n",
    "    \n",
    "    for ii, upper_bound in enumerate(np.linspace(0, math.pi/2, n_clusters)):\n",
    "        \n",
    "        if (ratio > math.tan(lower_bound)) and (ratio <= math.tan(upper_bound)):\n",
    "            color = clust[ii]\n",
    "\n",
    "        ii =+ 1\n",
    "        lower_bound = upper_bound\n",
    "    \n",
    "    return color"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "def correlation_scatter_plot(merged_xp, x, y, sample_ids=None, n_clusters=6, n_max=50000, out_dir=None):\n",
    "\n",
    "    deviate_genes = collections.Counter( )\n",
    "    \n",
    "    if isinstance(sample_ids, type(None)):\n",
    "        sample_ids = merged_xp.columns.get_level_values(0).unique()\n",
    "                      \n",
    "    for sample_id in sample_ids:\n",
    "\n",
    "        print(' sample_id: {}'.format(sample_id))\n",
    "\n",
    "        # remove nan's\n",
    "        df = merged_xp[sample_id].dropna()\n",
    "\n",
    "        #df['x'].plot.kde()\n",
    "        #df['y'].plot.kde()\n",
    "\n",
    "        # filter extreme counts\n",
    "        df = df[ (df[x] < n_max) & (df[y] < n_max) ]\n",
    "\n",
    "        ## Compute ratios: \n",
    "        df['ratio'] =  df[y] / df[x]\n",
    "        \n",
    "        # add color column\n",
    "        #df['color'] = df['ratio'].apply(lambda x: map_color(x)\n",
    "        df['color'] = df['ratio'].apply(lambda x: map_color(x, n_clusters))\n",
    "\n",
    "        #df['diff_count'] = abs(df[y] - df[x])\n",
    "        #df.sort_values('diff_count', na_position='first')\n",
    "\n",
    "        # print distribution ammong clusters\n",
    "        df_by_type = df['color'].value_counts().sort_index()\n",
    "        #print(df_by_type, '\\n')\n",
    "        \n",
    "        #for gg in df[df['color'] >= 0.8].index.tolist():\n",
    "        for gg in df[df['color'] <= 0.4].index.tolist():\n",
    "            deviate_genes[gg] += 1\n",
    "        \n",
    "        ## ------------\n",
    "        ## Scatter Plot\n",
    "        ## ------------\n",
    "        \n",
    "        plot_lim = max(max(df[x]), max(df[y]))\n",
    "\n",
    "        plt.figure()\n",
    "        #merged_xp.plot.scatter(x = 'count', y = sample_name, xlim=(0, 1000), ylim=(0, 1000))\n",
    "        df.plot.scatter(x=x, y=y, c='color', \n",
    "                        colormap='viridis',\n",
    "                        xlim=(0, plot_lim),\n",
    "                        ylim=(0, plot_lim))\n",
    "        plt.title(' sample_id: {} (genes={})'.format(sample_id, df_by_type.sum()))\n",
    "        \n",
    "        ## y = x - perfect correlation\n",
    "        xx = np.linspace(0, plot_lim, 1000)\n",
    "        plt.plot(xx, xx, '--k') # dashdot black\n",
    "        \n",
    "        if not isinstance(out_dir, type(None)):\n",
    "            plt.savefig(os.path.join(out_dir, sample_id + '_' + x + '_vs_' + y + '.pdf' ))\n",
    "            \n",
    "        #plt.close() \n",
    "\n",
    "    return deviate_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Scatter plots: **all Genes**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "if params_dict['expression'] == 'gene_count':\n",
    "    fig_dir = os.path.join(plots_dir, 'gxp_plots')\n",
    "else:\n",
    "    fig_dir = os.path.join(plots_dir, 'tpm_plots')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/gcm-lfs1/pablo/data/rna_silencing/results/Ratios/Plots_16-06-21/tpm_plots'"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if not os.path.isdir(fig_dir):    \n",
    "    !mkdir -p $fig_dir\n",
    "fig_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " sample_id: 301_H3K9me2_1\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 432x288 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAY8AAAD+CAYAAADCk4k3AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8li6FKAAAgAElEQVR4nOydd3wURfvAv5NOCD10UIqIoiIqFmygiIIKvIqO8CIiXZHXggIiFkQpKhb8AUoREEVgRBQUBSsqiICiKN3QQwkldBJSbn9/7AYvd5dc29xdkvny2Q/Z3Zlnnr3b22fneZ6ZEYZhoNFoNBqNP0SFWwGNRqPRFD+08dBoNBqN32jjodFoNBq/0cZDo9FoNH6jjYdGo9Fo/EYbD41Go9H4jTYegBBiqRBiqo3yhgshUryUaSWEMIQQdexqV1OyEUJ0EkL8JYTQv9swIIRoIYTYJYRIDLcukYC+CYuGscA1RSFYCDFWCLFJCHFSCHFMCPGLEOIOD+UeFEJsFkKcscp3dTl/oxBigRBip2XEnvVTjwKNnxBihhDiW6f924QQK4QQh4QQmUKIrUKIl4UQcS765rjIqWLV2ySEqGcdu0QIsciSdUII8VneuaLGavsDIcQO6zq2CyHeEkJU9EPGXUKIr4QQ+63P734f68UArwEvGIbhCPQaihPWfWR42GKcyni9j4UQPYQQPwghDlr3zO8efg81hRCzhBDrhRA5zvdvHoZhrADWAQOL4nqLG9p4FAGGYZw0DONQEYlfDzwCXApcDfwELBBCXJFXQAjxH+A94F2r3BRgphCinZOcJGADMBjYX0S65nEcGAe0AhoDTwJ9gVcKqmAZhF8AA7jOMIwdQogawA9AOnAjcD0QDXwrhChTdOqf5XLgJNAbaAI8BLQHZvshIwlYBTzsZ9t3AQnAQj/rFXd+Bmo6b4ZhOL9k+HIft8b83G4HLgPmAB8IIe5zKhOPeV+9AbgZDiemAo8IIWL9v5QShmEYIdmAZ4BtwBngILAEKGOdqw/MB/YCp4G/gW4u9ZdiPhBfBg4AR4GRmAbweSDNkjvSpd4Oq9xUzIfYIcyHVpSL7Kku9f4HbAIygX+AYUCMj9c6HEjxIC/Vur4lwAOYD8Y6Nny2R4DHnPZ/AT5yKfMxsLSA+juAZ/1ss1VB+gMzgG+91H8T+MNp/0Egx/q7GbAPWJB3j1jHewMngFinY5UtPXo4HTOsz3sucArYBdwDVABmWTK2AZ1cdKpu6X7QKrMcuNHLdXQCHED5AL43A7jfx7KfAZM9HH/c5b7q5vq9AFcAX2MavoOYv7VzXe9XoKN1z5/CNNINXdryJqcO8AnmbyzD+owHBXFfe72PAr2Pgc+BT/xtF9OAnwHaBnpdJWULSc9DCHE38DTwGNAIaAN85VQkCfgOaAtcAkwGpgshbnIRdQ8Qi/nGORDTIH1h1b8BeAp4xuUNG8wHyV7gSuAJYADmj64gfYdbsoYCF1p69wNe8P2q88nriPmwfAPzwagwXRCu5QyrbV/lxgghumFe/8/WsTjM61zsUnwxcI0QIjqQa7ATIcQFQDvMB5TruVuwelPA3YZhZDidTgCyAec3z0zMh/eNLqKGAV9i9ry+AGZivnF+g/n2uQizN1bFareMpU85S7fLrPrfCCEuLORyKmA+KE97u+4gaYnZYzmL9bsai3kvXYrZA3rFpUwT4EdgBdAcuBnIxbyuBKeiNTF7Q12Ba4GKwDQ/5UzE/Dxuwfzd9MI0bHky3rXcrYVt+dxJwFWWi2+7EOITIcRFfnxmhVEB08j5hWEYmcBawPXZVPoIhYXCfGBvwemN0Yc6C4ApTvtLgT9dyqwH/nY5thYY67S/A/jZpcwoINVF9lTr70TMB0FblzoPAEd91H04Tj0PYBkwy6XMWNzfEDcBA3yQfyfm218uZq/jTqdztSy5t7rUucM6XtWDvB0E3vM4ZenivGXj4c0N80Fyxqo3CYh2OvegdfwM8F4BbV4AZAEvYRqSJOAdq94Sp3IG8JbTflXr2P85HatkHbvTqf1UXHqXwPfOslzO1QB2O99vfn6GPvU8MB/kBtDO5fhy4AOXY2Oc7yvMt+g5LmXirXv8P073a47zvQF0xjTKCX7IWQsML+Q6qgHnednKuejQAbgY0yB9bbV3cQHyfbqPgfut++jyAs7P8HT/Op2fD3wcyHdekrazgaciRgGPAjuFEF9j9jI+MwzjBICVvfA8pv+4JhCHeWO6vpmuddnfj7ufcz/mTerMCpf95cBQIUR5wzCOu5y7CCgDfCKEcJ41MhpIEEJUNQzjYIFX6pkmuPvFl2H6/s9iGMYFPsr7AbMHUwm4F/MN+hbDMNb4UNfumTBvw/07eAXzzc6VGzCN82VWmf3k783lYr403CuEeN8wjJ+cKxuGkRf4fxOz1+kAPgR+t+o6s9ap3kEhRC7wl9OxI0KILP69V67ENAZHhRDOcuIxexb5EEJUw3yY/YXZQy1K8uI5mS7HmwAfuRxzvdevBM4TQpx0OZ6A6QXIY6/Lfb0HEJifzy4f5bwFTLJ6/kuBRc7foWEYBzBdzj5hGMYcp911QoifMOMbj2LGzPzG8gJMAXr5+HvxRCZQPsC6JYaQGA/DMPZYroqbMLu7zwGvCCGuNgxjN2a3uyPmwzTP5/o67g+gbFfRBRzz5o4ThZzLq3svZm/JlXQvsgvCtoe2YRinMH3UAKuFEJdiBgw7Y3bFczAfhM5Ux3yrP2KXHhY7DMNIdT4ghDiBB+NhGMZ268/11sP8QyHEq9b15NEFM7a1WAjxH8MwvnaR8THwsfXwPmMYxjEhRBruD03X+8LTMed7JQrYiBmYdiWfS0qYGWbfYH4H9xiG4aktOzmEqWtlD+e83VdRwAeYPRJXDjv9nVWAXOfPp1A5hmFMF0IsxnQ/3wR8JYT41DCM+8F0W2G+9RdGP8MwZnk6YRhGlhDiN6CeFxkeEUJ0xuxV9DEM44NAZFhUxozJlWpC1fPAMIwzmH73xUKI5zAD3P8B/g/TXz3LMIy5AMLMYz/fKmMHrmmzLTDftFx7HWC6wjKBBoZhfGlT+xuA6zB9wnlcZ5NsMH/Y8XD2B7Yas0cw06lMW+BXwzBc39DDRZS15ctaMQwjVwjRA/OBvVAIIQ3DcMswst5i82Ik1TBdCcHwG6Zr8niebE8IIRpiZuOsATqHwHBgGEa2EGIdZq/4E6dTGzDvZef7yvVe/w1oCmw1LJ9LgPgkxzCMfcB0zJjll8BsIUR/67f2PKa7tjAK/M1b8bqmuL8oeEUI0QfzWdM97zkTBJdgBtxLNSExHkKIXpgPilWYWVKtMQOTG6wim4GOQohPMH3mAzF993YZj2ZWIPojzGDfY5h+XjcMwzgphBgFjLLcF99gfk6XAJcZhjEkgPZfx3xbXoUZhL0eMysmH0KITcB4wzDGexIihKgO9Ldk7Mf0hXfB/Dw7ORV9FZhntbcYM95xN6ZbME9WEqaPGUw3YQ0hRDPgpGEYhQ5w9BchRF6PcgvmG21zS8eFhmEcdS1vPZz6Wy6ST4QQ9zu9WDwCrMS8j67HdGF9aBiGW/DdT2ZhxuYWCSGGWbpWx+wpbzQM4zMraPwtpqvqUaCKk4vroC+GWQhRGTjH6dA51ueebhjGrkKqfokZNHfmdWCu9T1/hRnofsA6l/eAH4X5u/tQCDEOM0uqHuaL2zjDMLZ509lXOUKI8ZaemzHdWXdjxoROgH9uK+v+HIFpLPdgviAMAhrg1Hvx5T4WQjyB6d14BPjRSvkGyDIMI91JVjPrz8pAUt6+YRh/OpVphOlad074KZ2EIrCCeRP9gukyOY050KaX0/m6mGmGpzC7gy9iui6WOpVZins67bfADJdjizEfJnn7OzBTdadjpuqmY95I0V5k9wL+xOyFHMF8YD3s4/UOxz1V9zHMH0GGpXd33APmBoUHHCthxgT2YroZ9mMaN7e0QcwA8Bar3GZcArP8G/B23Zb6eI2tXPV3OjcDp4AjZqbdBuv7PWF9/88AiS765niQ9QKmG66HtT8N88GVZV3f087fpdPn6Hq9OcCDLscygd5O+1UwA/B7LPl7gE8xXxryvldPn5kB1PPxc3uwgPozvNRrgOl2q+ty/Amn+2oJZizAAKo4lbnEum+OWOVSMDMaKxdyv17vel0+yJlgfScZmK6sRcBFAT4zymD+lvc7fRcLcQly+3IfYz4DvN7rBX23LmVexCk5ozRvwvpASixCiB2YhuHlcOui0QSDEOI94IRhGIWlmT+POeanSug0Kx1YvZwUzOyyX8OtT7jRI8w1muLDUGC/FRNECBErhHhaCNFUCNFQCNEb07UzJaxallzqY6YCl3rDAeieRwDybqBwf2c7wzB+tqOtcOIhJdOZUYZhjAqZMsUEIcQzmO44jxiGkWRzezGYAyCvwIwhbsdMknjNyD+FhyZCkFK2xZyqJxqYqpQa43K+Amb6+TmYsdaxSqnphdWVUlbGnE2hHqaLTiql7M6qdKPEGw+7sUYi1y6kyB4j/6joYokQ4rxCTqcbToFGjYkVDPeUTguAYXMigqZ4IaWMxowJtcEckLoa6KKU2uBU5hmgglJqiJSyKma8sgbmOCaPdaWUrwLpSqkxUsqngUpKqUASe/wiZKm6JQXLMJT4h4B+0PmPZVC1UdUUxFVAilJqG4CUcg7m+LYNTmUMoJyUUmDOoJCOmexxdSF1O2ImDgC8j5kApI2HB3RXSaPR+ENhg4K9YuSmGiLa+7I7p0+fPvzggw/ucDo0WSk12Wm/Nmbqch6pmEbBmfGYWWV7MV2R9ymlHFLKwupWV0rtA1BK7ZNSus6wUSQUR+PB3r17w62CR5KTkzl0qKhmYg8OrVtgaN0CI1y65eTk0LVrV5YtW0avXr2YOjX4Nd5EdB0c+8/3Wi6xxpYqSqnkwkR5OOb6Mnwb5hCBm4GGwDdSyp99rBtSdLaVRqMpMcTExHD11VfzxhtvMGLECNvkOnz45wOpmGPa8qiD2cNwpgcwXyllKKVSMJMgLvBSN01KWRPA+t/n+cOCoVj2PDQajcaZL7/8kqpVq3LllVcycKD9C/1l+zCrjw8P09VAIyllfcxBj52B/7qU2YU5Y8TPUsrqmIunbcOcUaGgugsxBx2Psf5f4F2V4NE9D41GU2xxOByMHTuWPn368M477xRdOzb0PJRSOZhrCS3BnIRTKaXWSykfklI+ZBV7CbhWSvk35uzjQ5RShwqqa9UZA7SRUv6DmY3lafJK2ymOqbqGjnn4j9YtMLRugREK3U6ePMmjjz7KkiVLkFIyevRoEhIS8pWpVasWBBkwB4zje8/xWqh8rV12tFVs0G4rjUZT7Dh8+DD33nsvKSkpjBgxgp49e+KyDoutOHSSpxvaeGg0mmJHpUqVaNq0KS+++CI33HBDkbeXq42HG9p4aDSaYoFhGLz//vu0adOG2rVr89Zbb4Wsbd3zcEcHzDUaTcSTkZHBY489xrBhw/jgg2AWAQyMbMPwupU2dM9Do9FENPv27aN37978+eefPPXUUzz22GMh10G7rdzRxkOj0UQsmzZtokuXLpw6dYpp06Zx2223hUWPXG073AiJ8ZBSTgPuBA4opS52Ov4/zNzlHGCRUmpwKPTRaDTFgzp16tCsWTOefvppGjduHDY9fBo/XsoIVcxjBtDW+YCU8ibM2SCbKqUuAsaGSBeNRhPBZGdnM3HiRDIyMkhKSmL69OlhNRwAuQivW2kjJMZDKfUT7lNVPwyMUUqdscqEZD4WjUYTuaSnp9OlSxdGjhzJ4sWLw63OWbIN4XUrbYQz5nE+cIOUciSQCTyllFrtqaCUsi/QF0ApRXJyYRNXho+YmBitWwBo3QKjpOn2119/cc8997B//36mTZtG165di0g7/ymNPQtvhNN4xACVgGuAKwElpWyglHILTVlz4ufNi2+U5ikZAkXrFhhat8DwV7fvv/+evn37UqFCBebPn0+zZs1suTZrepKgcZTCnoU3wjnOI5V/px5ehRmTiszXKI1GU6Q0aNCAa6+9li+//JJmzZqFWx03dMzDnXAaj88wFzxBSnk+EAdE5muURqOxnRMnTjBp0iQMw6BevXrMnDmT6tWrBy3XMAzmvLyAER3esEFLk1yivG6ljVCl6s7GXGM3WUqZCrwATAOmSSnXAVlAd08uK41GU/LYtm0bPXv2ZNu2bbRo0YKmTZsGLXPPln289t93OHn0FNlncmxdZ0+7rdwJifFQSnUp4NT9oWhfo9FEDkuXLqV///5ERUUxZ86cgAxHVmYWe/9Jo0bD6iQkxpF5Oovht79OVmZ2EWgMWUZ0kcgtzugR5hqNJmTMmDGD5557jsaNGzN9+nTq1q3rvZILP6uVzHh6LrnZuUTFRCOfaU/t82sUmeEAcJRCt5Q39Cei0WhCxnnnnUf79u1ZuHBhQIYDYOaz88jJysUwIDc7l3ljvqByrcqIKBfXko2eJh0wd0cbD41GU6Ts2bOHuXPnAnD99dczceJEEhMTA5aXm51/PfHcHAc1G1al8TUNzy4IFRMXzSPv9ghcadc2jSivW2lDu600Gk2RsXr1anr37k1WVhZt2rShcuXKhZZf9cUf/PjRCkRUFHcNbEvDy+u5lUksl8CJ9FNn9xPKxhEVFcVQ9T9S/thB2raDXNLyAsonl7PtOhw29SyklG2BcUA0MFUpNcbl/CAgb3RkDHAhUNXa5joVbQA8r5R6S0o5HOgDHLTOPaOU+tIWhQtBGw+NRlMkzJo1i2HDhlGnTh2mT5/u1XCs+2kTM4fN48ThkwDsTdnP02oA1c7JP/zrpa+H8GL71zl9LIOEpHiGzf93ivbzLqvHeZfVs/1asozgH5VSymhgAtAGc5zbainlQqXUhrwySqnXgNes8u2BJ5RS6ZjTOzVzkrMH+NRJ/JtKqZDOD6iNh0ajsZ0nn3yS8ePH06pVKyZMmEDFihW91lk+b/VZwwFwOPUIaxb/Rdu+N+crV6lGBd5aPcJ2nQvDpoD5VUCKUmobgJRyDubksBsKKN8FmO3heGtgq1Jqpx1KBYo2HhqNxnYaNGjAww8/zNChQ4mO9i3NtWL18vn2Y+JiqFKn8N5KqMi1Z5xHbWC3034qcLWnglLKRMyZyAd4ON0Zd6MyQEr5APAb8KRS6kjw6haONh4ajcYW1q1bx8GDB7npppt45JFH/J6b6q6B7di6Zic71qUSExNFk+sb07zdpUWkrX/4OoJcSvmb0+5ka16+PDxZoIKGMrYHllsuK2f5cUAHYKjT4XeAlyxZLwGvAz19UjgItPHQaDRBs2DBAgYOHMg555zDjTfe6Hd9R66Dj178lDOZWdRoUJWOj7fl8lsvPps9FW4cPmZTKaWaF3I6FXDOT64D7C2grKfeBUA7YI1SKs2pzbN/SymnAF/4pGyQlL78Mo1GYxu5ubmMHj2a/v37c8kll6CU8tlN5cyckQv5cc6v7Fi7m51/p/Lx6M/JysgqAo0Dw6a5rVYDjaSU9a0eRGdgoWshKWUFoCWwwIMMtziIlLKm0+5dwDofLysodM9Do9EExJkzZ+jTpw/fffcdXbt25eWXXyYuLi4gWbvWpeLI+Xex1yP7jpK24xDnNKltl7pBkW3D9CRKqRwp5QBgCWaq7jSl1Hop5UPW+XetoncBXyulTjnXt+IgbYB+LqJflVI2w3Rb7fBwvkjQxkOj0QREXFwcderUYfTo0TzwwANBySpXuWy+/cSKZahYvUJQMu3ErkGA1viLL12OveuyPwNz6W7XuqeBKh6Od7NFOT/RxkOj0fjF999/T61atbjgggsYNWqULTK7j5ak7z3KwdR04hJiub1/a8pXSbJFth3YNUiwJKGNh0aj8QnDMJg4cSKjR4+mXbt2TJkyxTbZSZXK8uyCxzl55BQJZROIjY+sR1NpnH7EG5H1DWk0mogkIyODJ598kgULFtChQwfeeMO+hZbyEEJQrnLk9DacKY2LPXlDGw+NRlMoBw8e5P7772f9+vUMHTqURx55JGJSaEOFXgzKnVCtJDgNuBM4oJS62OXcU5hzuVRVSullaDWaCKNChQpUr16dQYMGccstt4RbnbCQbcPcViWNUPXFZmAOtc+HlLIuZurZrhDpodFofOTjjz8mPT2duLg4Zs6cWWoNB+j1PDwREuOhlPoJc1ZIV94EBmPrasMajSYYsrKyGDJkCI8//jjTpk0LtzoRgcOI8rqVNsLWF5NSdgD2KKXWSim9le0L9AVQSpGcnFxo+XARExOjdQsArVtgFIVuaWlp3H///SxfvpzBgwczfPjwgEaMR/LnFgilsWfhjbAYD2uk5DDgVl/KW5OL5U0wZvg74VqoSE5O9nsyuFChdQuM4qxb5slM3uo1hfS9R2nd/QZu692qUHmbNm2iW7dupKenM3HiRDp27MiRI4FNzhopn1utWrVskVMaexbeCNcn0hCoD6yVUu7AnCBsjZSyRpj00WhKFDk5OQy85kU2Lk8hbfshZr/4GbNf+qzQOsnJydSpU4fPPvuMjh07hkjT4kG2Ee11K22EpeehlPobqJa3bxmQ5jrbSqOxh3U/bubU0dNn9w3D4Oe5K+ny3H/ylcvNzeWjjz6ic+fOJCcnM3/+/FKXhusLepCgOyH5RKSUs4EVQGMpZaqUslco2tVoSiuxcd7fC48ePUr37t15+umn+eqrrwC04SgAhyG8bqWNkPQ8lFJdvJyvFwo9NJrSwoXXNaJijQoc3X8MgKgoQftH25w9/88//9CjRw92797NmDFj6NChQ7hULRboEebu6JEvGk0JJCoqitd/fZ4Pnv2E/VsP0LbfTVx2izk+d+nSpfTr14+EhAQ+/vhjrrrqqjBrG/mUxp6FN7Tx0GhKKDExMfQYc5/b8eTkZJo0acL48eOpXTsy1suIdBy65+GG/kQ0mlLA6dOnUUoBcPHFFzN//nxtOPwg2xHldStt6J6HRlPC2b17Nz179mTTpk00bdqUCy64QAfG/USP83BHGw+NpgTzyy+/0LdvX3Jzc5k5cyYXXHBBuFUqlugR5u5oc6rRlFA+/PBDOnfuTJUqVfjiiy+46aabwq1SsUWn6rqjex4aTQmlQoUKtG7dmrfffpty5cqFW51ijV1uKyllW2AcEA1MVUqNcTk/COhq7cYAF2IuV5FuDaY+AeQCOUqp5ladysBcoB6wA5BKqcDmlfED3fPQaEoQBw4cYMmSJQC0b9+eadOmacNhAw6E180bUspoYALQDmgCdJFSNnEuo5R6TSnVTCnVDBgK/KiUcp6R/CbrfHOnY08D3ymlGgHfWftFju55aDQlhD///JNevXqRkZHBr7/+Svny5XVg3CayHbbMXXUVkKKU2gYgpZwDdAQ2FFC+CzDbB7kdgVbW3+8DS4EhwSjqC9p4aDQlgHnz5jF48GCqVq3Kxx9/TPny5cOtUonC15iGlPI3p93J1ozgedQGdjvtpwJXFyAnEXMBvQFOhw3gaymlAUxykl1dKbUPQCm1T0pZjRCgjYdGE8HkZOey4tPfOHX0FNf8pzkVq+U3CoZh8PLLL/Puu+/SokULJk2aRJUqVcKkbcnFF7cUgIs7yRVPQgpaCK89sNzFZXWdUmqvZRy+kVJushbaCws65qHRRCiOXAevd3uXaYNmM3vEAkZ1epu0HQfzlRFCEB8fT48ePZg9e7Y2HEWETdlWqUBdp/06wN4CynbGxWWllNpr/X8A+BTTDQaQJqWsCWD9f8DnCwsC3fPQaCKUDcu3sOnXFBy55stp2vaDfPLql/Sf2J3NmzeTkpLCeeedx6BBg3Rso4ixKdtqNdBISlkf2INpIP7rWkhKWQFoCdzvdKwsEKWUOmH9fSswwjq9EOgOjLH+X2CHst7QPQ+NJkLJPpODI8eR75gjJ5clS5bQvn17BgwYgMPh0IYjBOQYUV43byilcjBjGEuAjeYhtV5K+ZCU8iGnoncBXyulTjkdqw4sk1KuBVYBi5RSi61zY4A2Usp/gDbWfpEjDKMgl1vEYuzdW1BPL7xEytKbntC6BUY4dcvKyGJkp7fZ8ZcZY61YowLlbxJM+2gql156KfPnzychISEsunkjUr5TaxnaYK2rcdfyR7wW+vS6CXa0VWzQPQ+NJkKJKxPHkDmPcGvvllx992UcarCNaR9N5e677+aTTz6hTp064Vax1KBHmLujYx4aTQSTWL4MXYffjcPh4I9+y3n++efp27evdlWFmNJoHLwREuMhpZwG3AkcUEpdbB17DTMdLQvYCvRQSh0NhT4aTXFh+fLlnHvuudSpU4fJkydroxEmtPFwJ1RuqxmYA16c+Qa4WCnVFNiCORRfo4lYvv9gOS/fNY6Rd49j9aI/i7QtwzB477336NKlC6NHjwb0+uLhxI7pSUoaoVrD/CcpZT2XY1877f4K3BMKXTSaQPjzu/V88uoiTh4xE2DSdhyiWr2qnHuR/QsqnTlzhqFDhzJ37lxuu+02XnnlFdvb0PhHTilc7MkbkRLz6Ik5K6RHpJR9gb4ASimSk5NDpZdfxMTEaN0CIJJ1O3Mqi3censn65ZvOGg6AYweOs3lZCle0vNTW9tLS0ujSpQsrV65k2LBhPPvss0RFeX5wRfLnFsm6BYJ2W7kTduMhpRwG5ACzCipjzeGSN4+LEQkpgJ6IlPRET2jd/GfvP/t5+T/jOHXstNu52IRYKtauYLveGRkZZGdnM3nyZO644w7S09MLLBupnxtEjm5Wqm7QaOPhTliNh5SyO2YgvbVSqtgNONGUbN4d8IGb4RBRgvLJSTRt1YQr77Cv1/HVV1/RsmVLkpKSWLhwoY5vRBiGNh5uhM2RZy2KMgTooJRyf7XTaMJIVmY2J46cdDteo0E1Rn03lN5v/NeWB3xOTg4vvPACvXv3ZurUqYAOjEciOmDuTqhSdWdjzjefLKVMBV7AzK6Kx5wdEuBXpdRDBQrRaIqYMxlZTHtqNmk7D1G2YiLxZeLynY8rE0ePVyRJlcp6rP96t3f55/ftREdH0ePVzjRvV3jPJD09nYcffphly5bRq1cv+vfvb9u1aOxFu63cCVW2VRcPh98LRdsaja9MfWIWq774NwW39vk1aHhZPY4dOk755HIMmPQgVUzuUO4AACAASURBVGpV9lj3//q+x18/bDy7P/Hh93nlx2FUPdfzLLdbtmzhwQcfZN++fbzxxhvcd9999l6MxlZydbaVG2EPmGs0kcL+7fmnOz99IoNxv4zkZMYJr66kzSu35dvPzcll1aI/uKP/LR7LJyQkkJiYyLx587jiiiuCU1xT5OiYhzvanGo0FmXK5Z9kMCExnoSy8T7FIMokxbsdq31+jXz7DoeDBQsWYBgG55xzDl9//bU2HMUEPbeVO9p4aEolDoeDQ6npnEj/Nyje45X7qHthLSpUK0f1BlW579mOPgevB895hPhEM0YiogT1L61Ls1suPnv+5MmT9OnTh/79+/P11+b42ILGb2giD8PwvpU2tNtKU+rIOJnJ690msW9rGjGx0bS4qzmdn+1IzYbVGbFkECcOn6RsxbLExEb7LLNq3SqM/2skm1akULF6Bc5p8u/I8x07dtCzZ09SUlIYPnw4t956a1FclqYIKY3ZVN7QxkNT6pjz0mf8s/rfGMWPH63guk5XUvfCWkRFRVGhavlCahdMXEIcTW9qku/YsmXL6NevHwAffvghN954Y+CKa8KGDpi7oz8RTanj+KH84zdOH8/g0O6CR3J7IuNkJqu/XMu6nzbhcDgKLBcVFUXt2rVZtGiRNhzFGO22ckf3PDQlHkeuAzX6c3auS6VC1fKcf1UD1i/bwplTZwCodm4VGl5+rs/yjh8+ySv3jSd10z6iY6O5+IbGPD6jz9kYRmZmJj/99BO33nor1157LYsXL9bxjWKOzrZyRxsPTYln5rCP+WnOr+Ra64Ef3pNO+wG38PfSTUTHRHHv0A6UTy7nszw1ciGpm/YBkJudy7qfN/HX9xtpdstF7Nu3jz59+rB27VqWLl1Kw4YNteEoAdhlPKyZNcYB0cBUpdQYl/ODgK7WbgxwIVAVKAvMBGoADmCyUmqcVWc40AfIyzV/Rin1pS0KF4I2HpoSz/a/dp01HAAHdh2mdfcbaP+/wALXZzKy8u3nZjs4fTyD3377jT59+nDq1CmmTJlCw4YNg9JbEznYkYorpYwGJgBtgFRgtZRyoVJqQ14ZpdRrwGtW+fbAE0qpdCllPPCkUmqNlLIc8LuU8hunum8qpcYGraQf6FciTYknNj42335cfMzZtNpAaP3AdVSo9m9QvVaj6mw9tpF7772XMmXKsHDhQtq2dV37TFOcsSnmcRWQopTappTKAuYAHQsp3wWYDaCU2qeUWmP9fQLYCNi/mIwf6J6HpkRhGAZH044jogQVrQd85+f+w+THPiR93xHKVUqiXb+biY7xPQ3XlQtaNOLh8d34bsYyYhJiue+ZDixY/ClXX30177zzDpUqVbLrcjQRgsPHbCsp5W9Ou5Ot5STyqA3sdtpPBa4uQE4i5uqrAzycqwdcBqx0OjxASvkA8BtmD+WITwoHgTYemhJD5qkzDLnxZY4fOgFCcMHV5zF4Tn/Ou7wewxc9Sdr2g1SqWfGsUQmGC689n+oXJLN582Yq1ahA9+7d6datG9HRgRslTeTiazKVUqp5Iac9+b4KEt0eWK6UypcGKKVMAj4BHldKHbcOvwO8ZMl6CXgdc4G9IkUbD02J4ZXOEzialvd7MtiwfAs/zPqFm++/jsTyZah/6Tm2tbVhwwZ69uzJsaPHubfRg0Q5omnUvD7/HX6XnlK9BGJTwDwVqOu0XwfYW0DZzlguqzyklLGYhmOWUmp+3nGlVJpTmSnAF3Yo6w0d89CUGA7vcR+rsX3tLtvb+eKLL+jQoQNnzpzhmrI3s/OPPWxfu4vvP1jGgnFLbG9PEwEYPmzeWQ00klLWl1LGYRqIha6FpJQVgJbAAqdjAnMm8o1KqTdcytd02r0LWOfrZQWD7nloijVq1Of8+d064uJjSUhK4NiBE2fPiShBi7sL8yIUzIFdh5j6xEdknjxDueQk+r7VlfLJ5XjttdcYN24cl19+OYMfeoapD805WycnK5dta3YGfU2ayMOOnodSKkdKOQBYgpmqO00ptV5K+ZB1/l2r6F3A10qpU07VrwO6AX9LKfPWDchLyX1VStkM04TtAPoFrawPCKP4DY009u4tqKcXXiJl3WZPlDTdDMNgzH0T2PTLP2ePJVUqS3zZOI4dOEFUtKD1AzfQ+bnCklkKZvgdY9m+9t/YZmx8DJffdgk7y20kKyuLqvvqk/r3Xk6fyMRw/PsbavnfFvR8tXNAbfpLSftOiwJrDfNgn/xGw7kjvRbaet8wO9oqNoRqJcFpmGuVH1BKXWwdqwzMBephWksZigwBTclg4/J/2LJya75jJ4+c4pYHr6fDY7cFlU2Vk5XDsYP/9mBOGydxZOby+1d/0eHx21j73Xo2r3FqW0DFauWp2bA6XYffFXC7mghGjzB3I1QxjxmYaWfOPA18p5RqBHxn7Ws0hbL2+w280X0Si6f8gCM3/5xSUdFRnHtJ3aAMB0BMXAyJ5csAcNjYzyq+YwO/kZ2Vw94t+0ndvD9/BQP6jevG02oA8Ynu63poij96bit3QmI8lFI/Aa7RzI7A+9bf7wP/CYUumuLLgnFLePPByaz9bgNrv9tAVLTT26CAi29szOW3XmJLWw+OkaRX2sMfLCOBRC7hGhKSEmhyQ2Oio11+NoICl5vVlBDsCZiXKMIZMK+ulNoH5uhJKWW1ggpKKfsCfa2yJCcnh0hF/4iJidG6BYA33XJzHcwZ/Smfvbk4X3zBcBg0vqYRcXEx3NDpau7o18YWfTIzM5k89x3WpK/guitu4ILcy4kRsTRv24xOA+6kctVKjO0+8d8KBnw14QcGTn3IlvZ9pTh/p8UNPTGiO8Ui28oapZk3UtOIhECcJyIlSOiJ4qzbpEc/YOXCNThy8rupDKDzs+1p0MycEdfX6zMMg8//7xs2/ZpCYrkyPDDynnwTI2ZlZbF9+3YGDRrESy+9xOHDh8+eO3ToEI2va0DF6uWdxpTAljXbQv75FufvNFRYAfPgKYU9C2+E03ikSSlrWr2OmsCBMOqiiUAO7DzEjx+t4K+lG/NNbJhHUqWy1Gta10PNgjEMg9e6vsOGZVvO9mJ2bdjDc589zubtm2jQoAGVK1dmzpw5xMTEeBzwFxMX7RbbiI0vFu9hmgAxHLrn4Uo4BwkuBLpbf3fHaUCMRvPJa4t4uuUovpjwLSfTT+U7Fx0TTeOrG/D6r8P9nu589ojP2PDz5nzur7TtB+nWqjedOnXi5ZdfBky3S0FERUVx5yO3UKV2JWLjY6hWLznglGBNcUH4sJUuQpWqOxtoBSRLKVOBF4AxgJJS9gJ2AfeGQhdNZONwOHjm5jHsS0nzeL5C1XJ0GnwHLbu0CEj+pl9T8mXGOAwHW1hLavpWzq3cgOeff94nOTd2voZmt1zE4T1HqF6/6tnsLE0JRbut3AiJ8VBKdSngVOtQtK8pHhiGwXNtXyvQcLR76GZa/bcFNRoUmFvhFedMqSzjDH+xgqMc4lzO584L76JixYo+yyqfXM6vRaQ0xRhtPNzQjlpNRJC+7yhv955K6gbPswck163MPYPvICYu8Ft2xae/cfzwSaKio8wxIsI0IBdxJfWSzufSmy4KWLamhFOCsq2sRam6Y06weCZQOT7/EqWUMZiLk1wGJDmfU0r1DVQBjQZgwkPT800H4kyZcgmM/HZIUIbj4O7DzBm5kKP7j3HYSKNKTHXayJZcf/oazpzM4uKWjbm1Z8uA5WtKNiVpEKBSKldK+YZSalowcvz5NX4IXAJ8BXj2K2g0fvLJa4tYt3QTuzbu8Xi+Wr1kXvlpWEDrgP/57XrWfr+B8644l9iEWI7sO8pW1rODTTTKacqJw5fw2LTewV6CpjRQ8rKtPpdStldKfR6oAH+MR1ugrrUEokYTFLk5uTx90ygObPc8FiCxfBnuf6kT13W6MiD5n4//hi8nfsfp4xks+3glTW5uxPq4VezP2k0t6nNu9HnUbFQ9mEvQlCJECep5WCQA86SUKzBXNzx7hUqpB3wR4I/x2ABUBrTx0ATFhmVbeKXzBI/nEiuUoWrdKjw84QFqNgz84b7q8z84fTwDgCOn05m8+G1OOU5wVdUbaZB4AfWbnkOnQbcHLF9Tyih5xmMdQa774Y/xuB+YKqX8Ghe3lVJqZjBKaEoP65dv4dUCDIcQ0H/CA1zSqomtbTrIJdfIYfbs2Vx3/XU4HI6A3GCaUkwJCpgDKKVeDFaGP8bjQeAGoBKQ4XTcALTx0Hjl6ZtGse+fgsNlF1zbyDbD0eKuK9i4Yz1lT1WkUkIyt5S9mwVDf2BZrT945N0eJFUqa0s7mlJCyet5IKW8CXOBqdrAHuBDpdT3vtb3x3g8BlymlNron4qa0k5ubi69GzyJI7fgX+AN911N79f/a0t7GRkZLPp7PitOfkvv2/tz7M8sjuw7yqn006RtP8iUgR/xxPQ+trSlKSW4z45TrJFS9gZGAVOBlcA5wEdSyueUUlN8keGP8UjDHAmu0fiMYRg8fsXzhRqOFh2a22I4HA4HU5//kIkfj+PgyTQe7f8YD/V7iOdvfS1fuaNpx4JuS1PKKGFuK2Aw0EYptTbvgJRyLvAJYLvxeBOYJaUcg8skhkqpbX7I0ZQSNq/cytt93+Pk4VMez0fHRDP6x6FcdMWFtszA+lLf15jx1SRyyeVSWrBx8h5eXTqRuMS4fOUqVq8QdFua0oVd2VZSyrbAOMw1zKcqpca4nB8EdLV2Y4ALgapKqfSC6ga4KmsVzCQoZzZjJkX5hD9RwwlAB+AXIMVp+6ewSprSybheUxjV6e0CDQfA+L9HUv3cqra1+dt3fxBDLFdxM1VFLYxcSN24j/iycdS7pC7V61elyfXn0+cNe9xjmlKEDYtBWSO7JwDtgCZAFyllviCfUuo1pVQzpVQzYCjwo2U4CqsbyKqsy4A3pJSJlm5lgdcwn+8+4XPPQyml01M0PjHy7nFsWVVwZzQ2IZZJm18hOjq45WLBXHtjzZo1ZO2EytnVuYZbiRL5b1XDAS8ueSrotjSaILkKSMnz1Egp52CuqOraA8ijCzDbh7odMSeeBXNV1qXAEC+6PATMAY5JKdMxexy/WG36hE/Gw7J6W4AmwcyFoin5/Pnd+gINR2xCDN1e6kTLLtfa0tahQ4fo168fv//+O92u6AfgZjgAqtSqZEt7mtKLr24rKeVvTruTrYXs8qiNOSAvj1Tg6gLkJGIOzB7gQ12fV2XNwyrfUkpZF6gJ7FVKpXqr54xPxsOaCyUXKANo46FxY9/WNP6vzzTS9x/1XEDAyG+epnp9e9xU69ato0ePHqSnp/PWW2+x/bMD7GJ/vjKVa1fi3Ca16Tvuflva1JRifJyeRCnVvJDTnoQUZJbaA8uVUukB1PWIlNKT92iPtZ09r5TyKbfMn4D5W8BcKeUoTKvnPJxdB8xLMVv/2MnIu97yuNpfHq+vfIHkWj7H4gplwYIFDBw4kMqVK/Ppp5/StGlT9l2Sxv5tB9m/7QDxZeK4qv1l9NaxDY1d2BMwTwWcl76sA3ieRho686/LyltdX1dlzaHwKxHWeZ/8yf4Yj/HW/21cjvvcmKbk8XKncfyz0v3dISpaEJ8YT42G1Rj26aPExsba1ubWrVtp2rQpU6ZMITk5GYCaDavz3ILH2fxrChWrV6Dh5fVsa0+jsSnbajXQSEpZH/NtvzPg9oYjpawAtMSc1cOXunmrso6h8FVZ69twDWcJe8BcSvkE0BvTCP0N9FBKZRZFWxr7yMrKou95g/Mt5+pM1XOqMObHwGbD9cSxY8fYtWsXl1xyCY8//jgDBgwgLi5/Cm5SpbJc0e5SW9rTaPJhg/FQSuVIKQcASzBfuKcppdZLKR+yzr9rFb0L+FopdcpbXeu0T6uyKqV2uh6zXFXVgTRf3VV5CMPPieqllOdgBm9SlVKeF2DwXVZtzJSxJkqpDCmlAr5USs0opJqxd29BPb3wkpycbMt4haLATt12bkh1G3jnTHRsNMO/GMg5F9WxRbeUlBR69OjB6dOnWbZsGWXKhG7J19LyndpNpOhWq1YtCH6BcaPh2De8Ftr61EA72goJUsrymN6kzpidiGzM7KtHlVI+jaL1ZzGompbwFsBhoIqU8legs1IqmKd5DFBGSpkNJFKwD1ATAcx//UsWvLmkwPPxibGMWDKEGjYFxr/99tuzvYzJkyeH1HBoNHmUwCnZ3wbKAhcDO4FzgZHW8e6+CPAn5vEOsBa4XSl1yhpUMgp4F3PwoN8opfZIKcdidrUyMLtqX7uWk1L2Bfpadc76uSONmJiYEq1bt/r9Sd9X8EuJiILZeycTnxBXYBlfdTMMg7Fjx/Lcc89x6aWX8t6k6dSsVoPKVSohROhe7kr6d1pURLJuAVHyFoNqCzRQSp229rdIKXsAW30V4I/xuB6oqZTKBrAMyGCsNK9AkFJWwhzgUh84CnwspbxfKfWhczkrVzovX9qIhO6wJyKlq+6JYHXrXuexQs9Hx0UzbdsbnDh5nBMng9fN4XDwyy+/0KFDB5I21ODpa0cB5piNV35+lpjY0ORolOTvtCiJFN0st1XQlMCeRyZQFbPXkUcyfgzF8CeaeQRzWLwzjTEf+oFyC7BdKXXQMkrzAXtGkGls4fC+dK+Go1KtCkzb5t0n7Aupqans3r2bqKgoxo8fz421biUt5TCGw8BwGBxKTeedR963pS2NxmdsmJ4kwpgKfCOlfEhK2c4K2i/Bx0kRwb+ex6vAt1LK9zCtVT3MNT6e80OGK7uAa6zRlBlAa+C3wqtoQsVHIz5lyeSlhZZp/2gb7hl8py3trVixgr59+3L++eczb9484uPjSfl9h1u51E06LKYJLSWw5zES02vUFaiFGWt+VSn1nq8CfO55WHO8S8yuzZ2Yc6F0dhl+7xdKqZXAPGANZppuFP+6pzRhpHfjJ70ajhHfDLHFcBiGwfvvv0/nzp2pVKkSr7766tm4Rou7mrvlr1x5R7Og29Ro/KLk9TzGAZuVUrcopZoopW4BNkop3/JVgD/ZVnFAI8yUrnQgHughpezh64LpnlBKvQC8EGh9jf14c1MBvLNlNImJiUG3lZWVRf/+/Zk2bRqtW7dm/PjxlC9f/uz51g9czz+/bef3xX+BYdDk+sbcM8Seno5G4yuihC0GhTkBoutsob8DnwGP+yLAH7fV+8ClwOfgMomQpkRw4MABBl0+stAy8UlxTN5U8BgPf8nOzmb16tX873//Y9CgQR5n2n3o7W62tafRaADPM4NE44c3yh/j0Raor5QKJkCuiVCmPDmTZXN/L7RMpZoVeGv1CFvaW7duHfXr16ds2bL8/PPPnDpV8LofGk3YKX5uKW/8DLwkpRyslHJYI82HW8d9wp9sq12YripNCaN7nce8Go7L211sm+H45JNP6NixI6NGmem3euCfJtIRhvetmPEYZrbrPinlKsyAeRvgf74K8KfnMRNYIKUch7me+VmUUt/7IUcTQfgS3/jf1F40b9s06LZyc3MZNWoU7777Li1atGDgwIFBy9RoQkLxMw6FopRKlVJejrnIVF3MtUJW+TO/lT/GI29RklEuxw2ggR9yNBGCL4ZjUsorJCQkBN3W0aNH6d+/Pz/++CMPPvggw4cPt3WmXY2mSClhxgPOrtvxq7X5jT+z6to6na8mfDwrR7D7l8Ney72fOs62Nk+ePMnmzZt59dVX6dq1q21yNZpQUAKzrYLGn56HpgTgS28D7DMca9asoVmzZtSpUyfkM+JqNHZRDGMaRU6RrNGhiUx8MRyirD2GwzAMxo0bR4cOHZg1axagA+OaYkzJGyQYNLrnUUq4I8G7q+jWgTfSdWCnoNs6deoUTzzxBIsWLeLuu+/mnnvuCVqmRhNWSqFx8IY2HiWcHTt28ML1b3otZ5ebateuXfTs2ZPNmzfz3HPP0a9fv5BOoa7RFAXabeWONh4lmFDHNwD27t3LwYMHGf3cq6x6dwOPTHiGJtc2YsCknra1odGEHG083NDGo4QSSsNhGAZ//vknl112Gddccw1fLVzCcze9Sk5WLgCrF61lYv8Z9J/4YNBtaTThwK5sKyllW8xJCaOBqUqpMR7KtALeAmKBQ0qpllLKxsBcp2INgOeVUm9JKYcDfYCD1rlnlFJf2qNxwWjjUQIJpeE4c+YMzzzzDHPmzOHTTz/lqquuYuPSf84ajjw2LN8SdFsaTdiwoechpYwGJmCO5E4FVkspFyqlNjiVqQhMBNoqpXZJKasBKKU2A82c5OwBPnUS/6ZSamzwWvqONh4ljFAajrS0NPr06cPvv//O448/TvPmzQGoUqeyW9nYeD0gUFN8sSnmcRWQopTaBiClnIO5kuoGpzL/BeYrpXYBKKUOeJDTGtiqlNrp4VzI0MajhDB48GDSPvK+gqRd8Y0//viD3r17c+zYMSZNmsSdd/47TXqz1hdR98Ja7N60FwyIiY9h4Mx+trSr0YQFH42HlNJ5MbvJLusd1cacBiSPVOBqFxHnA7FSyqVAOWCcUmqmS5nOwGyXYwOklA9gLqb3pFLqiG8aB442HiWAcATG161bR2xsLAsXLqRJE9fVieHlb4awYfkW0vcdpVnri0iqVNa2tjWakOOj8VBKNS/ktKe0Q1fJMcAVmL2LMsAKKeWvSqktcHZdpQ7AUKc67wAvWbJeAl4HijxDRRuPYo6vhmNR5iwOHToUVFs5OTls2rSJiy++mG7dunH33XdTtmzBRqHJdecH1Z5GEynY5LZKxZyEMI86mLPZupY5pJQ6BZySUv6EuY5SXtCwHbBGKXV2clrnv6WUU4AvbNHWC2E3HlaAaCpwMabl7KmUWhFerYoHoexxHDlyhIcffpjff/+dn3/+mRo1ahRqODSakoRNxmM10EhKWR8z4N0ZM8bhzAJgvJQyBojDdGs5D9TqgovLSkpZUym1z9q9C1hni7ZeiITpScYBi5VSF2Ba2I1h1qdYEErDsXnzZu644w5WrlzJSy+9RI0aNYKWqdEUK2yYnkQplYM5O/kSzOecUkqtl1I+JKV8yCqzEVgM/AWswkznXQcgpUzEzNSa7yL6VSnl31LKv4CbgCeCulYfEYYRvtEvUsrywFqggVLKV0WMvXtde3qRQXJyctCuIV/wxXC4Go1AdVu8eDGPPvooZcuWZcqUKWczquwkVJ9bIGjdAiNSdKtVqxZ4jjX4g3Hpo95naVj79hN2tFVsCLfbqgHmwJbpUspLMRdgf8zy951FStkX6AuglCI5OTnkivpCTExMkerW8aKu5Gz1Xm5R5iy3Y4HqtmrVKi688EKUUtSuXdvv+r5Q1J9bMGjdAiOSdQsEPT2JO+HueTTHXIjkOqXUSmuVwuNKqecKqVYqex5Lxt4KVGT96rKsXV6+wHIFuan80e3UqVOkpaXRoEEDsrOzycnJKdIZcSPlLdUTWrfAiBTd7Op5NBvgvefx53jd8wglqUCqUmqltT8PeDqM+kQkFWnFA4NAiAMcPRxN18ubkJPtHq6yI76xc+dOevbsSWZmJj/88ANxcXF6xT9NqUcvBuVOWAPmSqn9wG5r3hYwc5s3FFKl1JFMK+IF5E1MW7FKLm9+/g9xCbmcjdIJewzHzz//zO23387+/fsZPXo0cXFxQcvUaEoCwvC+lTYiIdvqf8AsK1OgGe5rpJdaqtKKaCfDkcf5TTN5eMReQPB+6jje3x2c4TAMgylTptC1a1eqV6/OokWLuPHGG4OSqdGUKPRiUG6E222FUupPwP4UnmLNLqrxAMKD4QDIzYWfPi9r24jx3NxclixZQps2bRg3bhxJSUm2yNVoSgyl0Dh4I+zGQ+OKorqYCHg2HIYBOTnw+OwPgm5p//79xMbGUqVKFWbMmEFiYiJRUZHQGdVoIovS6Jbyhn5SRBBleaBAw2EY5nbcgCOxS4Nua82aNdx+++088YQ5nigpKUkbDo2mAITD8LqVNvTTIkKoRCuSxC6Prqq8bOo040YyWBp0W3PnzqVTp07Ex8czdOhQ7xU0mtKOjnm4od1WEUByAYFx+LfHcYCFQMHjO3whJyeHESNG8N5773H99dfzzjvvULmy+9obGo0mP9pt5Y7ueYSZgjKqwDQaDgMOsJRgDQfAiRMn+Prrr+nduzezZs3ShkOj8RXd83BD9zzCRjbVaFNgRpVhQI4Bh21wU6WkpFC+fHkqVarEkiVLqFChQtAyNZrShO55uKN7HmFhB9VF4YYj04i3xXAsWrSIdu3a8eKLLwJow6HRBILuebihjUeIieYHqosHgYIzqtKN2znGkqDacTgcjB07lr59+9K4cWMeeeSRoORpNKUZ4fC+lTa02yqExPITlcWLXgLjo4Brg2rn5MmTPProoyxZsgQpJaNHj6ZWrVoRMVGdRlMc0W4rd7TxCBFlmEp58aEXw7EYSAi6rT179rBixQpGjBhBz549EZ4a1Wg0vhPG2ccjFW08QkBlOhErDheaUXXQhvjGli1bOP/882ncuDErVqygYsWKQcvUaDS65+EJHfMoYpK5s1DDkWuIoA2HYRhMnjyZ1q1bM3++uUKlNhwajY3ogLkbuudRhFSjVaEZVceN+8mgd1BtZGZmMmTIEObNm8ftt9/ObbfdFpQ8jUbjjl0BcSllW2AcEI25PvkYD2VaAW8BscAhpVRL6/gO4ASQC+QopZpbxysDc4F6wA5AKqWO2KNxweieR5Hg8Go4Dhv/F7Th2LdvH506dWLevHk89dRTTJo0ibJlywYlU6PRuGNHtpWUMhqYALQDmgBdpJRNXMpUBCYCHZRSFwH3uoi5SSnVLM9wWDwNfKeUagR8R4gW1NM9D9vJporoiqDgOaoOGBMxaOJW01/+/vtvUlJSmDZtmu5xaDRFiT0B86uAFKXUNgAp5RygI/kXwPsvMF8ptQtAKXXAB7kdgVbW3+8DS4EhdihcGNp42EnONqqLDkDBYzgO8BkQXDwiJSWF8847j1tvvZUVgaQmwQAAEFlJREFUK1boaUY0miLG14C5lPI3p93JSqnJTvu1gd1O+6nA1S4izgdipZRLgXLAOKXUTOucAXwtpTSASU6yqyul9gEopfZJKav5pm1wRITxsLpzvwF7lFJ3hlufQIhiN9HHu3lJxf0BCDxtNjs7mxdffJEPPviAzz//nKZNm2rDodGEAh+Nh4s7yRVPP35XyTHAFZhLcpcBVkgpf1VKbQGuU0rttYzDN1LKTUqpn3zTzH4iJebxGLAx3EoEzhmSRV/Pd4YBOUZVa3LDwA1Heno6Xbp0Yfr06fTq1YsmTYJ3e2k0Gt+waQ3zVKCu034dYK+HMouVUqeUUoeAn4BLAZRSe63/DwCfYrrBANKklDUBrP99cXUFTdh7HlLKOsAdwEhgYJjVCQCDOH5HkOF+xoATRldO0yeoFtavX0+vXr04cOAA48aN45577glKnkaj8Q+bFntaDTSSUtYH9gCdMWMcziwAxkspY4A4TLfWm1LKskCUUuqE9fetwAirzkKgOzDG+n+BHcp6I+zGAzMlbTCmf88jUsq+QF8ApRTJyckhUs0LWauIPj0CHMfdThlAbsxNJFYYSmKQzaxcuRKHw8H3339P8+aBLfceExMTOZ+bC1q3wNC6hRAbbIdSKkdKOQBYgpmqO00ptV5K+ZB1/l2l1EYp5WLgL8CBmc67TkrZAPhUSgnmc/sjpdRiS/QYQEkpewG7cM/QKhKEEcZh91LKO4HblVL9rdzmp3yIeRh797r29EJPWd4kSSzIF+MwMB1ThhFNujGMbG4OWL7D4WDnzp3Ur18fwzA4cuRIUPGN5OTkiJ3bSusWGFo379SqVQuC8RebGC3bv+a10I+fD7KjrWJDuGMe1wEdrMEvc4CbpZQfhlcl75RhlpvhMKnFUcdgDhjzgzIcJ06coGfPnrRv357Dhw8jhNCBcY0mnDgM71spI6xuK6XUUGAonB1V+ZRS6v5w6uSNBL6ivHjPc1ZVbEMyz9welPxt27bRs2dPtm3bxosvvqiNhkYTCZQ+2+CVSIh5FBti2Eo58R7Cw3BShxGPo+wrcCYrYPlLly6lf//+REVFMWfOHK69Nrip2TUajT3oiRHdiRjjoZRaCjZMLVtEJDKHsmI20eJYvuNmyCiGA8Y8kqPKA4H7eZVS1KpVi+nTp1O3bl3vFTQaTUiwKduqRBExxiOyMUgUX3g0HDnU47AxBXMOM//JyMjg6NGj1KxZk7FjxwKQmBhsfpZGo7EVbTvc0MbDC4ko4sUKojiY77jDKMNp4y5O0gsz685/9uzZQ+/evcnOzmbx4sXaaGg0EYrQi0G5oY1HIZThE5LENKJEJmD2NIQAhxFHhnEbJ82hJwGxatUq+vTpQ2ZmJuPHjycmRn8VGk3EUgrXKPeGfmIVQGX6Eys2uGVVnTEu5YzRgtPcF7DsWbNmMWzYMOrUqcO8efNo1KhRkNpqNJqiRPc83NHGwwOV6Ums2OY+My7lOG4MJJdzA5adlZXF+++/z3XXXceECRP0in8aTXFA2w43tPFwIYkJng2HIThttA3YcBw+fJj4+HiSkpKYPXs2FStWJDo6sFiJRqMJLTrbyp1wjzCPKOL5nkTxuce1ODKMVpzkkYDkrlu3jnbt2jFkiLk+S5UqVbTh0GiKE3nrKhS2lTK08bBI4m3+v737j42yvgM4/n7aK9AKSCtQW2SDoSyr8RcqbHGgbigSVjtI+Phzak1aiWsyNhaFLfIjxlVIRjUCbhVNhCyyD+hWndDGGX9HRU2YyBAwZpMWYz1QIG1t7d13f9zRtb3Sux7Xu17v80qe9J7nnu/3Pk9Ing/f7/f5fp9x3h+6BsdPcQ4CbgInWBVXvXV1dZSVlREMBrnnnnsSEaoxJskS8Rra4caSB6EWx1ne83heZ4/jznl0uBL86IDrDAQCVFdXc++993LRRRexa9cuLr744kSFbIxJJmt5RLDkAeR69RGJI+h8nHBL+YpNxLNQpt/v55lnnuG2225DVZkwYUKCojXGJJ2LYcswNmAO9J7k5xy0ux/RRtmAa2pqaqKoqIjCwkJeeuklJk6ciNfXKorGmLThBTOwXyoKa3kAJ10Fna4YAOd8dHAZx7te0hW7+vp65s6dy8aNGwEoLCy0xGHMcBCMYcswGdryaGcUrwGOduYQ4Hscc48xwr1LkHPoYCYD6apyzrFp0yaqq6spKSlh4cKFgxa5MSb5bJJgpIxLHh5t5Hu/IYf9eB58657lmHuEIOfwDQN/F0dbWxvLli2jrq6OxYsXU11dTW5u7iBEboxJGUseETIseQTIZykjvANdR3K8g+S5bbRwd1w1HjhwgPr6elasWMGqVas4evRoooI1xgwVCUoeInID8CihgdbNqvpwH+dcAzxCaKluv6peLSKTgS3AuYQ6yWpV9dHw+auBCuhavfV3qrozIQH3I6PGPM7hTnK6JY5TsmgdcF1NTU0AXHrppbz11ltUVVXZ+IYxw1UCxjxEJBvYCMwHSoBbRKSk1znjgE3Ajap6IbA4/FUnsExVfwD8EPhlr7I1qnppeBv0xAEpbnn0l00Tzcen+LymiNnjna6IFhYNqK4tW7awcuVKamtruf766ykqKkpgpMaYoSZBT1vNBD5R1U8BRGQbUAb8u9s5twLPqepnAKraHP77OfB5+PNJEdkPTOpVNqlS3fKIlk0TxuMEvR/Gdi6Lr9xaghTHVEdHRwf3338/K1asYPbs2cyaNWsQIjXGDDmJmSQ4CTjcbb8xfKy76UC+iLwqIh+IyB29KxGRKcBlwLvdDleJyIci8pSI5A/o2uKU0pZHMrNpJ9MJUIiPL4DQ7PE2dx0BvhNT+S+//JLKykp2795NVVUV9913n61PZUymiHHMQ0Te77Zbq6q13fb76tfuXbEPuBz4KZALvC0i76jqwXD9o4FngaWqeiJc5nHgwXBdDwJ/hDgHcQdgyAyYnyabnvquEkJvXlJVxo8fH1ul7huyWlbhBQ7jssYSzN1AsHUDnjtGMGceI/LuIsaaaGhoYO/evWzZsoWbbur7XR4+ny/22JLMYouPxRafoRxbXGLstVLVK/r5uhGY3G3/POBIH+f4VbUFaBGR14FLgIMikkMocfxFVZ/r9ptfnPosIk8A/4gt2jMzJJLHabJpl3D2PpXBnd/vj6nes7015HqvAOAFoLPDj9/9KfTlt0Br9HqOHDlCcXEx8+bN47XXXmPSpEmc7vfHjx9/2u9SzWKLj8UWn6ESW3FxbF3S0SRonsd7wAUiMhVoAm4mNMbRXR2wQUR8wAhgFlAjIh7wJLBfVdd3LyAiReFeHICFwEeJCDaaVI95cLpsmgg+GnvsZ3MUL8YnqwKBAA899BBz5szh4MGDAEya1Lt70hiTERIw5qGqnUAV0ADsDx3SfSKyRESWhM/ZD9QDHwK7CT3O+xFwFfAL4Ccisie8nZqYtk5E9orIh8C1wK8TfPV98lwKJ7+Es+nTwDFVXRpjMXfkSO+WXt/yvV8x0vtX136nm4zfPU20nPn1119TVVXFK6+8wh133MGaNWsYMWJE1N8bKv/b6ovFFh+LLT5DJbZwy+NMn6F386ffH/WkXQfXJuK30kaqu61OZdO9IrInfCwBE1xCCfGEW8Y4VpPFVzjO4oS7h2iJ49ChQ5SXl9PY2MjatWu5/fbbzywUY0z6sxnmEVL9tNWbJDhTj+bPjPLeAKDdzeKoe4IsvibIWEITNvu3bds2Tp48iaoyc+bMRIZmjElXljwipLrlkVAjeYs8r44sLzSukYWfDvd92rm+33LOOZqbmyksLGT58uVUVFRw7rnnJiNkY0w6sHeYR0j5gHki5bC3K3EAZHnfMJK9/ZZpbW1lyZIllJaWcvz4cXJycixxGGN6csHoW4YZVsmjgxkE3eiu/YDLo53TP3Z9+PBhysrKePHFFykvL2fs2LHJCNMYk24CwehbhhlW3VYdzKTF3cIo/glAu/sx7Vzd57lvv/02lZWVdHZ2snXrVq699tpkhmqMSSc25hFhGCSPdsayHp93mCBjOOHuo8Xd1m8J5xyPPfYYBQUFPPXUU0ybNi1JsRpj0pIljwhpnzzO9tYxipe7VsvN4vccc4/T10Nc7e3ttLa2kp+fz4YNG8jJyWHMmDHJDdgYk34seURI8+Th8HGoxzLr2fjxaMExuseZzc3NVFRUkJ2dzY4dOygoKEhyrMaYtJWYJdmHlTQeMA8yznsAX48VjsExCkfP18Du2bOH+fPns2/fPsrLy8nKSuPLNsYkX2KWZB9W0vYuOpI3GMk7eN7//9EC7ixOugpCb3gM2bFjB4sWLcLn81FXV0dpaWkKojXGpDV72ipC2nZb+TiI53X2ONbBrB5PV7W1tbF+/XpmzJhBbW2tdVUZY+LiMnAeRzRpmTxy+Tt53vM9jgVcPm3uZ0BoYcO8vDxyc3PZvn07EydOJCcn+tIkxhjTJ5thHiEtu63yvL+R7Z3s2g+6XI6739LBDA4cOMCCBQtYvXo1EFpG3RKHMeaM2JhHhLRMHh49u6uCTKCDq2hoaKC0tJSWlhYWLlyYouiMMcNOMBh9yzBpmTy+pQTnQj1uzvn4JnAhNTU13H333Zx//vns3LmTK6+8MsVRGmOGDWt5REjLMY/jbjmdFJPDp3zrpvNx0xwef/w6Fi1axLp168jNzY1eiTHGxMgFAqkOYchJy+QB2bRQzlH/UQoKCpg82aOhoYEpU6bgeRnzIi9jTLLYgHmEtOy2AnjzzTeZM2cOW7duBWDq1KmWOIwxg8OWZI+Q8paHiNwAPEpoZt9mVX04Wpknn3ySNWvWMG3aNGbPnj3oMRpjMptLUMsjlvudiFwDPELo1ad+Vb26v7IiUgD8FZgC/AcQVf0qIQH3I6UtDxHJBjYC84ES4BYRKYlWbuXKlcydO5cXXniBqVOnDnaYxphMl4CWRyz3OxEZB2wCblTVC4HFMZRdDrysqhcAL4f3B12qu61mAp+o6qeq2gFsA8qiFVq6dCmbN29m9OjR0U41xpgz5gKBqFsMYrnf3Qo8p6qfAahqcwxly4Cnw5+fBn4e94UOQKq7rSZBj5UNG4FZvU8SkUqgEkBVqampSU50cSguLk51CKdlscXHYovPUI5tgP77UnD7d6Od1NraevSuu+56v9uhWlWt7bYfy/1uOpAjIq8CY4BHVXVLlLKFqvo5gKp+LiITY7imM5bqlkdfI9wRnYuqWquqV6jqFSLyQbjckNssNovNYhuSsZ2pKbH8Tl5e3vhT96nwVturnr5i6X2/8wGXAwuAecADIjI9xrJJlerk0QhM7rZ/HnAkRbEYY8xgiuV+1wjUq2qLqvqB14FLopT9QkSKAMJ/m0mCVHdbvQdcICJTgSbgZkJ9fsYYM9zEcr+rAzaIiA8YQahrqgb4uJ+yzwN3Ag+H/9YN8nUAKW55qGonUAU0APtDh3RflGK9m4JDicUWH4stPhZbGjnd/U5ElojIkvA5+4F64ENgN6FHcj+Kcq98GLhORA4B14X3B53nMnBNFmOMMWcm1WMexhhj0pAlD2OMMQNmycMYY8yAWfIwxhgzYJY8jDHGDJglD2OMMQNmycMYY8yA/Q8ehy4X26Vg+wAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "deviate_genes = correlation_scatter_plot(merged_xp, x='wo_norm', y ='norm', n_max=50000, sample_ids=sample_ids[0:1], out_dir=fig_dir)\n",
    "#deviate_genes = correlation_scatter_plot(merged_xp, x='wo_norm', y ='norm', n_max=50000, out_dir=fig_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "#df[df['color'] > 0 ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({'FP565355_region_16736..16794': 1, 'FP565355_region_4498..4556': 1})"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(deviate_genes)\n",
    "deviate_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Genes conntaining big deviations across all samples (or many samples):\n",
    "    - After using equivalent `'count_mode' = 'union'` to Parastou (instead of intersectiom-empty)\n",
    "    - Only **10 genes** contain big differences, most of which can be explained by differences between `ensembl` and `Pombase` **annotation**.\n",
    "    - Unexplained differences: (ChIP-seq)\n",
    "        - SPNCRNA.390 (This it's in the region where we MAT locus htc_genes are defined)\n",
    "        - SPNCRNA.1306"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "weird_genes = ['SPNCRNA.390', 'SPNCRNA.1306'] # still not explained by differences in annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#big_deviate_genes = {k:v for k,v in deviate_genes.items() if v >= 40}\n",
    "big_deviate_genes = {k:v for k,v in deviate_genes.items() if v >= 20}\n",
    "#big_deviate_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#missing_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#big_deviate_genes_df = merged_xp_rna[merged_xp_rna.index.isin(big_deviate_genes) & ~merged_xp_rna.index.isin(missing_genes)]\n",
    "big_deviate_genes_df = merged_xp_rna[merged_xp_rna.index.isin(weird_genes)]\n",
    "\n",
    "## many deviates are due to 0 counts, we require ~ 1 count per sample\n",
    "#big_deviate_genes_df[big_deviate_genes_df.sum(1) > 100] \n",
    "#big_deviate_genes_df[big_deviate_genes_df.xs('htseq', level=1, axis=1).sum(1) > 52]\n",
    "#big_deviate_genes_df[big_deviate_genes_df.xs('parastou', level=1, axis=1).sum(1) > 52]\n",
    "#big_deviate_genes_df[(big_deviate_genes_df.xs('htseq', level=1, axis=1).sum(1) > 100) & (big_deviate_genes_df.xs('parastou', level=1, axis=1).sum(1) > 100)]\n",
    "big_deviate_genes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "big_deviate_genes_df.xs('htseq', level=1, axis=1).sum(1)\n",
    "big_deviate_genes_df.xs('parastou', level=1, axis=1).sum(1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Investigate **heterochromatic genes**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import viz_strands ## get deg1, deg2 and non_degraded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## centromeric genes: `deg1`\n",
    "old_deg1 = ['dh1', 'dg1']\n",
    "deg1 = viz_strands.deg1\n",
    "\n",
    "## subtelomeric genes: `deg2`\n",
    "old_deg2 = ['SPAC212.11', 'SPAC212.10']\n",
    "deg2 = viz_strands.deg2\n",
    "\n",
    "# Mating type region (MTR) gene counts\n",
    "deg3 = ['MAT2', 'MAT3', 'MAT1']\n",
    "\n",
    "## rest of Heterochromatic genes, including mat: `deg3`\n",
    "# non_degraded = ['SPAC212.09c', 'SPNCRNA.70', 'SPAC212.08c', 'SPAC212.07c', 'SPAC212.12', 'SPAC212.06c',\n",
    "#                 'SPAC212.04c', 'SPAC212.03', 'SPAC212.02', 'SPAC212.01c', 'SPAC977.01', 'SPAC977.18',\n",
    "#                 'SPAC977.02', 'SPAC977.03', 'SPAC977.04', 'SPAC212.05c', 'MAT2', 'MAT3', 'MAT1']\n",
    "non_degraded = viz_strands.non_degraded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "old_all_htc_genes = old_deg1 + old_deg2 + deg3 + non_degraded\n",
    "old_htc_genes = old_deg1 + old_deg2 + deg3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_htc_genes = deg1 + deg2 + non_degraded\n",
    "htc_genes = deg1 + deg2 + deg3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Show counts for genes of interest: **manual entries Heterochromatic Genes (UNSTRANDED)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#merged_xp_rna[(merged_xp_rna.index.isin(old_htc_genes)) | (merged_xp_rna.index.isin(new_htc_genes))]\n",
    "merged_xp_rna[merged_xp_rna.index.isin(old_htc_genes)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Show counts for genes of interest: **manual entries Heterochromatic Genes (STRANDED)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_xp_rna[merged_xp_rna.index.isin(htc_genes)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Show counts for genes of interest: **all Heterochromatic Genes**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#[xx for xx in merged_xp_rna.columns if '80' in xx[0]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#htc_df = merged_xp_rna[merged_xp_rna.index.isin(het_genes)]\n",
    "#htc_df = merged_xp_rna[merged_xp_rna.index.isin(old_all_htc_genes)]\n",
    "#htc_df = merged_xp_rna[merged_xp_rna.index.isin(all_htc_genes)]\n",
    "htc_df = merged_xp_rna[merged_xp_rna.index.isin(non_degraded)].xs([xx for xx in merged_xp_rna.columns if '80' in xx[0]], axis=1)\n",
    "htc_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "htc_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Scatter plots: **all Genes**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if params_dict['expression'] == 'gene_count':\n",
    "    plot_dir = os.path.join(xp_data_dir, 'gxp_htc_plots')\n",
    "else:\n",
    "    plot_dir = os.path.join(xp_data_dir, 'tpm_htc_plots')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.isdir(plot_dir):    \n",
    "    !mkdir -p $plot_dir\n",
    "plot_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "htc_deviate_genes = correlation_scatter_plot(htc_df, n_max=50000, out_dir=plot_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (heterochromatin)",
   "language": "python",
   "name": "heterochromatin"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
