{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---------------------------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "project_dir = '/home/pmonteagudo/workspace/silencing_project'\n",
    "if project_dir not in sys.path: \n",
    "    sys.path.append(project_dir)\n",
    "from config_analysis import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.options.mode.chained_assignment = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import Util"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Result **directories**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#annot_dir = '/data/parastou/RNAdeg/annotation/'\n",
    "#annot_dir = os.path.join(project_data_dir, 'annotation/gff')\n",
    "annot_dir = os.path.join(project_data_dir, 'annotation/gff_v2')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---------------------------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- reference `GTF/GFF` **file**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "## - Parastous:\n",
    "#in_gff = os.path.join(annot_dir, 'schizosaccharomyces_pombe.chr.extended.gff3')\n",
    "\n",
    "## - sPomBase:\n",
    "in_gff = os.path.join(annot_dir, 'Schizosaccharomyces_pombe_all_chromosomes.extended.gff3')\n",
    "#in_gff = os.path.join(annot_dir, 'Schizosaccharomyces_pombe_all_chromosomes.extended_with_mat_locus.gff3') ## include MAT1, MAT2, MAT3\n",
    "\n",
    "## - Ensembl:\n",
    "#in_gff = os.path.join(annot_dir, 'Schizosaccharomyces_pombe.ASM294v2.45.gff3')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Heterochromatic Features (**repeats**) `GTF/GFF` **file**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "in_repeats = os.path.join(annot_dir, 'repeats_subtel.gff3')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---------------------------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Explore `GFF` "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Load `GFF` **File** as **DataFrame**: `schizosaccharomyces_pombe.chr.extended.gff3`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "gff_header = ['seqid', 'source', 'type', 'start', 'end', 'score', 'strand', 'phase', 'attributes']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "gff = pd.read_csv(in_gff, sep='\\t', comment='#', names=gff_header)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Parse `attributes` colummn\n",
    "attributes_df = gff['attributes'].apply(lambda x: Util.parse_attribute_col(x))\n",
    "#attributes_df = attributes_df.apply(pd.Series) # slower\n",
    "\n",
    "attributes_df = pd.DataFrame(attributes_df.values.tolist(), index=gff.index)\n",
    "gff = pd.concat([gff.drop('attributes', axis=1), attributes_df], axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Attribute Columns:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'Name', 'Parent'], dtype='object')"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attribute_columns = gff.columns.difference(gff_header)\n",
    "attribute_columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ID', 'Name', 'Parent'], dtype='object')"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attribute_columns = gff.columns.difference(gff_header)\n",
    "attribute_columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>ID</th>\n",
       "      <th>Name</th>\n",
       "      <th>Parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>1798347</td>\n",
       "      <td>1799015</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.01</td>\n",
       "      <td>mrx11</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>1798347</td>\n",
       "      <td>1799015</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.01.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPAC1002.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1798347</td>\n",
       "      <td>1798830</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.01.1:exon:1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPAC1002.01.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>intron</td>\n",
       "      <td>1798831</td>\n",
       "      <td>1798959</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.01.1:intron:1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPAC1002.01.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1798960</td>\n",
       "      <td>1799015</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.01.1:exon:2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPAC1002.01.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqid   source    type    start      end score strand phase  \\\n",
       "0     I  PomBase    gene  1798347  1799015     .      +     .   \n",
       "1     I  PomBase    mRNA  1798347  1799015     .      +     .   \n",
       "2     I  PomBase     CDS  1798347  1798830     .      +     .   \n",
       "3     I  PomBase  intron  1798831  1798959     .      +     .   \n",
       "4     I  PomBase     CDS  1798960  1799015     .      +     .   \n",
       "\n",
       "                       ID   Name         Parent  \n",
       "0             SPAC1002.01  mrx11            NaN  \n",
       "1           SPAC1002.01.1    NaN    SPAC1002.01  \n",
       "2    SPAC1002.01.1:exon:1    NaN  SPAC1002.01.1  \n",
       "3  SPAC1002.01.1:intron:1    NaN  SPAC1002.01.1  \n",
       "4    SPAC1002.01.1:exon:2    NaN  SPAC1002.01.1  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(41755, 11)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "trna_df = gff[gff['type'] == 'tRNA']\n",
    "trna_df['feature_length'] = trna_df['end'] - trna_df['start'] + 1\n",
    "trna_df.sort_values('feature_length', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "15084"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# tRNA features are extremely small\n",
    "trna_df['feature_length'].sum() # total length 15kb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>ID</th>\n",
       "      <th>Name</th>\n",
       "      <th>Parent</th>\n",
       "      <th>feature_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>40975</th>\n",
       "      <td>III</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>2443713</td>\n",
       "      <td>2447197</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.49.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.49</td>\n",
       "      <td>3485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40972</th>\n",
       "      <td>III</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>16927</td>\n",
       "      <td>20411</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.48.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.48</td>\n",
       "      <td>3485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40969</th>\n",
       "      <td>III</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>6057</td>\n",
       "      <td>9541</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.47.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.47</td>\n",
       "      <td>3485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40849</th>\n",
       "      <td>mitochondrial</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>1</td>\n",
       "      <td>2822</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.01.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.01</td>\n",
       "      <td>2822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40966</th>\n",
       "      <td>III</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>2440994</td>\n",
       "      <td>2442835</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.46.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.46</td>\n",
       "      <td>1842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40960</th>\n",
       "      <td>III</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>21289</td>\n",
       "      <td>23130</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.44.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.44</td>\n",
       "      <td>1842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40957</th>\n",
       "      <td>III</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>10419</td>\n",
       "      <td>12260</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.43.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.43</td>\n",
       "      <td>1842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40852</th>\n",
       "      <td>mitochondrial</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>3131</td>\n",
       "      <td>4552</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.02.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.02</td>\n",
       "      <td>1422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40954</th>\n",
       "      <td>III</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>1</td>\n",
       "      <td>1390</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.42.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.42</td>\n",
       "      <td>1390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40963</th>\n",
       "      <td>III</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>2451883</td>\n",
       "      <td>2452883</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.45.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.45</td>\n",
       "      <td>1001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40924</th>\n",
       "      <td>II</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>1406385</td>\n",
       "      <td>1406824</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.32.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.32</td>\n",
       "      <td>440</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40921</th>\n",
       "      <td>II</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>783253</td>\n",
       "      <td>783498</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.31.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.31</td>\n",
       "      <td>246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40915</th>\n",
       "      <td>II</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>520254</td>\n",
       "      <td>520494</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.29.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.29</td>\n",
       "      <td>241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40855</th>\n",
       "      <td>III</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>1369974</td>\n",
       "      <td>1370202</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.03.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.03</td>\n",
       "      <td>229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40939</th>\n",
       "      <td>II</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>4161556</td>\n",
       "      <td>4161775</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.37.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.37</td>\n",
       "      <td>220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40858</th>\n",
       "      <td>III</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>1474219</td>\n",
       "      <td>1474427</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.04.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.04</td>\n",
       "      <td>209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40867</th>\n",
       "      <td>III</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>2308122</td>\n",
       "      <td>2308319</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.07.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.07</td>\n",
       "      <td>198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40882</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>3165354</td>\n",
       "      <td>3165546</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.14.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.14</td>\n",
       "      <td>193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40879</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>2976790</td>\n",
       "      <td>2976979</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.13.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.13</td>\n",
       "      <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40903</th>\n",
       "      <td>III</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>523289</td>\n",
       "      <td>523475</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.24.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.24</td>\n",
       "      <td>187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40906</th>\n",
       "      <td>III</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>926108</td>\n",
       "      <td>926294</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.26.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.26</td>\n",
       "      <td>187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40918</th>\n",
       "      <td>II</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>569121</td>\n",
       "      <td>569305</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.30.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.30</td>\n",
       "      <td>185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40909</th>\n",
       "      <td>II</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>1199882</td>\n",
       "      <td>1200066</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.27.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.27</td>\n",
       "      <td>185</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40897</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>4184341</td>\n",
       "      <td>4184518</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.19.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.19</td>\n",
       "      <td>178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40876</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>912230</td>\n",
       "      <td>912396</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.12.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.12</td>\n",
       "      <td>167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40927</th>\n",
       "      <td>II</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>1504332</td>\n",
       "      <td>1504498</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.33.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.33</td>\n",
       "      <td>167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40978</th>\n",
       "      <td>III</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>9842</td>\n",
       "      <td>10006</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.50.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.50</td>\n",
       "      <td>165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40981</th>\n",
       "      <td>III</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>20712</td>\n",
       "      <td>20876</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.51.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.51</td>\n",
       "      <td>165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40984</th>\n",
       "      <td>III</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>2443248</td>\n",
       "      <td>2443412</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.52.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.52</td>\n",
       "      <td>165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40885</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>3547889</td>\n",
       "      <td>3548049</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.15.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.15</td>\n",
       "      <td>161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40948</th>\n",
       "      <td>II</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>2442700</td>\n",
       "      <td>2442852</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.40.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.40</td>\n",
       "      <td>153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40933</th>\n",
       "      <td>II</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>2322705</td>\n",
       "      <td>2322851</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.35.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.35</td>\n",
       "      <td>147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40942</th>\n",
       "      <td>II</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>4210360</td>\n",
       "      <td>4210500</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.38.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.38</td>\n",
       "      <td>141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40888</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>3634370</td>\n",
       "      <td>3634504</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.16.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.16</td>\n",
       "      <td>135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40861</th>\n",
       "      <td>III</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>1690945</td>\n",
       "      <td>1691075</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.05.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.05</td>\n",
       "      <td>131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40864</th>\n",
       "      <td>III</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>1702123</td>\n",
       "      <td>1702252</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.06.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.06</td>\n",
       "      <td>130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40936</th>\n",
       "      <td>II</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>3938578</td>\n",
       "      <td>3938707</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.36.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.36</td>\n",
       "      <td>130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40894</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>3891662</td>\n",
       "      <td>3891784</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.18.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.18</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40873</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>456527</td>\n",
       "      <td>456645</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.11.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.11</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40870</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>149658</td>\n",
       "      <td>149776</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.10.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.10</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40945</th>\n",
       "      <td>II</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>2032410</td>\n",
       "      <td>2032528</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.39.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.39</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40930</th>\n",
       "      <td>II</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>2261371</td>\n",
       "      <td>2261489</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.34.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.34</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40912</th>\n",
       "      <td>II</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>464527</td>\n",
       "      <td>464645</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.28.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.28</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40900</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>4240850</td>\n",
       "      <td>4240968</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.20.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.20</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40891</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>3830380</td>\n",
       "      <td>3830498</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.17.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.17</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40987</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>1563479</td>\n",
       "      <td>1563593</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.53.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.53</td>\n",
       "      <td>115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40993</th>\n",
       "      <td>III</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>853314</td>\n",
       "      <td>853428</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.55.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.55</td>\n",
       "      <td>115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40990</th>\n",
       "      <td>II</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>3096968</td>\n",
       "      <td>3097078</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.54.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.54</td>\n",
       "      <td>111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40951</th>\n",
       "      <td>II</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>3171566</td>\n",
       "      <td>3171645</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRRNA.41.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPRRNA.41</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               seqid   source  type    start      end score strand phase  \\\n",
       "40975            III  PomBase  rRNA  2443713  2447197     .      +     .   \n",
       "40972            III  PomBase  rRNA    16927    20411     .      -     .   \n",
       "40969            III  PomBase  rRNA     6057     9541     .      -     .   \n",
       "40849  mitochondrial  PomBase  rRNA        1     2822     .      +     .   \n",
       "40966            III  PomBase  rRNA  2440994  2442835     .      +     .   \n",
       "40960            III  PomBase  rRNA    21289    23130     .      -     .   \n",
       "40957            III  PomBase  rRNA    10419    12260     .      -     .   \n",
       "40852  mitochondrial  PomBase  rRNA     3131     4552     .      +     .   \n",
       "40954            III  PomBase  rRNA        1     1390     .      -     .   \n",
       "40963            III  PomBase  rRNA  2451883  2452883     .      +     .   \n",
       "40924             II  PomBase  rRNA  1406385  1406824     .      +     .   \n",
       "40921             II  PomBase  rRNA   783253   783498     .      +     .   \n",
       "40915             II  PomBase  rRNA   520254   520494     .      +     .   \n",
       "40855            III  PomBase  rRNA  1369974  1370202     .      -     .   \n",
       "40939             II  PomBase  rRNA  4161556  4161775     .      -     .   \n",
       "40858            III  PomBase  rRNA  1474219  1474427     .      -     .   \n",
       "40867            III  PomBase  rRNA  2308122  2308319     .      -     .   \n",
       "40882              I  PomBase  rRNA  3165354  3165546     .      +     .   \n",
       "40879              I  PomBase  rRNA  2976790  2976979     .      +     .   \n",
       "40903            III  PomBase  rRNA   523289   523475     .      +     .   \n",
       "40906            III  PomBase  rRNA   926108   926294     .      +     .   \n",
       "40918             II  PomBase  rRNA   569121   569305     .      +     .   \n",
       "40909             II  PomBase  rRNA  1199882  1200066     .      +     .   \n",
       "40897              I  PomBase  rRNA  4184341  4184518     .      -     .   \n",
       "40876              I  PomBase  rRNA   912230   912396     .      -     .   \n",
       "40927             II  PomBase  rRNA  1504332  1504498     .      +     .   \n",
       "40978            III  PomBase  rRNA     9842    10006     .      -     .   \n",
       "40981            III  PomBase  rRNA    20712    20876     .      -     .   \n",
       "40984            III  PomBase  rRNA  2443248  2443412     .      +     .   \n",
       "40885              I  PomBase  rRNA  3547889  3548049     .      -     .   \n",
       "40948             II  PomBase  rRNA  2442700  2442852     .      -     .   \n",
       "40933             II  PomBase  rRNA  2322705  2322851     .      +     .   \n",
       "40942             II  PomBase  rRNA  4210360  4210500     .      -     .   \n",
       "40888              I  PomBase  rRNA  3634370  3634504     .      +     .   \n",
       "40861            III  PomBase  rRNA  1690945  1691075     .      +     .   \n",
       "40864            III  PomBase  rRNA  1702123  1702252     .      +     .   \n",
       "40936             II  PomBase  rRNA  3938578  3938707     .      +     .   \n",
       "40894              I  PomBase  rRNA  3891662  3891784     .      -     .   \n",
       "40873              I  PomBase  rRNA   456527   456645     .      +     .   \n",
       "40870              I  PomBase  rRNA   149658   149776     .      +     .   \n",
       "40945             II  PomBase  rRNA  2032410  2032528     .      +     .   \n",
       "40930             II  PomBase  rRNA  2261371  2261489     .      -     .   \n",
       "40912             II  PomBase  rRNA   464527   464645     .      +     .   \n",
       "40900              I  PomBase  rRNA  4240850  4240968     .      -     .   \n",
       "40891              I  PomBase  rRNA  3830380  3830498     .      +     .   \n",
       "40987              I  PomBase  rRNA  1563479  1563593     .      +     .   \n",
       "40993            III  PomBase  rRNA   853314   853428     .      -     .   \n",
       "40990             II  PomBase  rRNA  3096968  3097078     .      -     .   \n",
       "40951             II  PomBase  rRNA  3171566  3171645     .      -     .   \n",
       "\n",
       "                ID Name     Parent  feature_length  \n",
       "40975  SPRRNA.49.1  NaN  SPRRNA.49            3485  \n",
       "40972  SPRRNA.48.1  NaN  SPRRNA.48            3485  \n",
       "40969  SPRRNA.47.1  NaN  SPRRNA.47            3485  \n",
       "40849  SPRRNA.01.1  NaN  SPRRNA.01            2822  \n",
       "40966  SPRRNA.46.1  NaN  SPRRNA.46            1842  \n",
       "40960  SPRRNA.44.1  NaN  SPRRNA.44            1842  \n",
       "40957  SPRRNA.43.1  NaN  SPRRNA.43            1842  \n",
       "40852  SPRRNA.02.1  NaN  SPRRNA.02            1422  \n",
       "40954  SPRRNA.42.1  NaN  SPRRNA.42            1390  \n",
       "40963  SPRRNA.45.1  NaN  SPRRNA.45            1001  \n",
       "40924  SPRRNA.32.1  NaN  SPRRNA.32             440  \n",
       "40921  SPRRNA.31.1  NaN  SPRRNA.31             246  \n",
       "40915  SPRRNA.29.1  NaN  SPRRNA.29             241  \n",
       "40855  SPRRNA.03.1  NaN  SPRRNA.03             229  \n",
       "40939  SPRRNA.37.1  NaN  SPRRNA.37             220  \n",
       "40858  SPRRNA.04.1  NaN  SPRRNA.04             209  \n",
       "40867  SPRRNA.07.1  NaN  SPRRNA.07             198  \n",
       "40882  SPRRNA.14.1  NaN  SPRRNA.14             193  \n",
       "40879  SPRRNA.13.1  NaN  SPRRNA.13             190  \n",
       "40903  SPRRNA.24.1  NaN  SPRRNA.24             187  \n",
       "40906  SPRRNA.26.1  NaN  SPRRNA.26             187  \n",
       "40918  SPRRNA.30.1  NaN  SPRRNA.30             185  \n",
       "40909  SPRRNA.27.1  NaN  SPRRNA.27             185  \n",
       "40897  SPRRNA.19.1  NaN  SPRRNA.19             178  \n",
       "40876  SPRRNA.12.1  NaN  SPRRNA.12             167  \n",
       "40927  SPRRNA.33.1  NaN  SPRRNA.33             167  \n",
       "40978  SPRRNA.50.1  NaN  SPRRNA.50             165  \n",
       "40981  SPRRNA.51.1  NaN  SPRRNA.51             165  \n",
       "40984  SPRRNA.52.1  NaN  SPRRNA.52             165  \n",
       "40885  SPRRNA.15.1  NaN  SPRRNA.15             161  \n",
       "40948  SPRRNA.40.1  NaN  SPRRNA.40             153  \n",
       "40933  SPRRNA.35.1  NaN  SPRRNA.35             147  \n",
       "40942  SPRRNA.38.1  NaN  SPRRNA.38             141  \n",
       "40888  SPRRNA.16.1  NaN  SPRRNA.16             135  \n",
       "40861  SPRRNA.05.1  NaN  SPRRNA.05             131  \n",
       "40864  SPRRNA.06.1  NaN  SPRRNA.06             130  \n",
       "40936  SPRRNA.36.1  NaN  SPRRNA.36             130  \n",
       "40894  SPRRNA.18.1  NaN  SPRRNA.18             123  \n",
       "40873  SPRRNA.11.1  NaN  SPRRNA.11             119  \n",
       "40870  SPRRNA.10.1  NaN  SPRRNA.10             119  \n",
       "40945  SPRRNA.39.1  NaN  SPRRNA.39             119  \n",
       "40930  SPRRNA.34.1  NaN  SPRRNA.34             119  \n",
       "40912  SPRRNA.28.1  NaN  SPRRNA.28             119  \n",
       "40900  SPRRNA.20.1  NaN  SPRRNA.20             119  \n",
       "40891  SPRRNA.17.1  NaN  SPRRNA.17             119  \n",
       "40987  SPRRNA.53.1  NaN  SPRRNA.53             115  \n",
       "40993  SPRRNA.55.1  NaN  SPRRNA.55             115  \n",
       "40990  SPRRNA.54.1  NaN  SPRRNA.54             111  \n",
       "40951  SPRRNA.41.1  NaN  SPRRNA.41              80  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rrna_df = gff[gff['type'] == 'rRNA']\n",
    "rrna_df['feature_length'] = rrna_df['end'] - rrna_df['start'] + 1\n",
    "rrna_df.sort_values('feature_length', ascending=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A **non-coding RNA** (`ncRNA`) is an RNA molecule that **is not `translated` into a protein**. The **DNA sequence** from which a functional non-coding RNA is `transcribed` is often called **an RNA gene**. \n",
    "\n",
    "Abundant and functionally important types of non-coding RNAs include:\n",
    "\n",
    "* **transfer RNAs** (`tRNAs`)\n",
    "* **ribosomal RNAs** (`rRNAs`)\n",
    "\n",
    "as well as small RNAs such as: \n",
    "\n",
    "* **Small nucleolar RNAs** (`snoRNAs`) \n",
    "* **Small nuclear RNA** (`snRNA`)\n",
    "* microRNAs (`miRNA`)\n",
    "* Small interfering RNA (`siRNA`)\n",
    "* Piwi-interacting RNA (`piRNA`) \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For all these **non-coding RNA** `gene` features (except for the actual `ncRNA_gene` and `tRNA_gene`), the corresponding `transcript` features are **NOT** found under `transcript`. Instead, they can be found in their own corresponding category:\n",
    "- snRNA_gene -> snRNA\n",
    "- rRNA_gene -> rRNA\n",
    "- snoRNA_gene -> snoRNA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Additionally, for **pseudogenes** (`56`) both `gene` and `transcript` features can be found under the `pseudogene` category:\n",
    "- `28` of the `56` **pseudogenes** behave as `transcripts` (e.g. contain a `Parent`, no `gene_id`, contain a `transcript_id`)\n",
    "- the other `28` **pseudogenes** behave as `genes` (e.g. contain no `Parent`, a `gene_id`, no `transcript_id`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>type</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>polyA_signal_sequence</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mating_type_region</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNP</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TR_box</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>regional_centromere</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>regional_centromere_central_core</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gap</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_group</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>snRNA</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>regional_centromere_inner_repeat_region</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dg_repeat</th>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LTR_retrotransposon</th>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dh_repeat</th>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>origin_of_replication</th>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>polyA_site</th>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>nuclear_mt_pseudogene</th>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogenic_transcript</th>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>region</th>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rRNA</th>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>repeat_region</th>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>snoRNA</th>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>low_complexity_region</th>\n",
       "      <td>59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>promoter</th>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tRNA</th>\n",
       "      <td>196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>long_terminal_repeat</th>\n",
       "      <td>239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ncRNA</th>\n",
       "      <td>1528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>three_prime_UTR</th>\n",
       "      <td>4791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>five_prime_UTR</th>\n",
       "      <td>4863</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mRNA</th>\n",
       "      <td>5140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>intron</th>\n",
       "      <td>5372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <td>7005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CDS</th>\n",
       "      <td>12162</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            ID\n",
       "type                                          \n",
       "polyA_signal_sequence                        1\n",
       "mating_type_region                           1\n",
       "SNP                                          2\n",
       "TR_box                                       2\n",
       "regional_centromere                          3\n",
       "regional_centromere_central_core             4\n",
       "gap                                          4\n",
       "gene_group                                   5\n",
       "snRNA                                        6\n",
       "regional_centromere_inner_repeat_region      6\n",
       "dg_repeat                                   11\n",
       "LTR_retrotransposon                         14\n",
       "dh_repeat                                   16\n",
       "origin_of_replication                       16\n",
       "polyA_site                                  17\n",
       "nuclear_mt_pseudogene                       17\n",
       "pseudogenic_transcript                      29\n",
       "region                                      29\n",
       "rRNA                                        49\n",
       "repeat_region                               49\n",
       "snoRNA                                      57\n",
       "low_complexity_region                       59\n",
       "promoter                                    62\n",
       "tRNA                                       196\n",
       "long_terminal_repeat                       239\n",
       "ncRNA                                     1528\n",
       "three_prime_UTR                           4791\n",
       "five_prime_UTR                            4863\n",
       "mRNA                                      5140\n",
       "intron                                    5372\n",
       "gene                                      7005\n",
       "CDS                                      12162"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gff_by_type = gff[['type', 'ID']].groupby(['type']).count()\n",
    "gff_by_type.sort_values('ID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "41755"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(gff_by_type.ID)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Interestingly, in this `gff` there are annotated **features** that might be of importance:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "!['spombe_chr'](https://cshperspectives.cshlp.org/content/7/7/a018770/F2.large.jpg)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "!['spombe_chr'](https://ars.els-cdn.com/content/image/1-s2.0-S0014579315004652-gr1.jpg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "features = ['mating_type_region', 'regional_centromere', 'regional_centromere_central_core',\n",
    "            'regional_centromere_inner_repeat_region', 'dg_repeat', 'dh_repeat', 'repeat_region', 'long_terminal_repeat']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "#features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>ID</th>\n",
       "      <th>Name</th>\n",
       "      <th>Parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>41362</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>dg_repeat</td>\n",
       "      <td>3759165</td>\n",
       "      <td>3763441</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRPTCENA.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41668</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>dg_repeat</td>\n",
       "      <td>3777115</td>\n",
       "      <td>3778839</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRPTCENA.8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41430</th>\n",
       "      <td>II</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>dg_repeat</td>\n",
       "      <td>1604786</td>\n",
       "      <td>1609649</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRPTCENB.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41383</th>\n",
       "      <td>II</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>dg_repeat</td>\n",
       "      <td>1638230</td>\n",
       "      <td>1644747</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRPTCENB.17</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41685</th>\n",
       "      <td>III</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>dg_repeat</td>\n",
       "      <td>1076115</td>\n",
       "      <td>1078499</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRPTCENC.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41202</th>\n",
       "      <td>III</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>dg_repeat</td>\n",
       "      <td>1082876</td>\n",
       "      <td>1083348</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRPTCENC.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41517</th>\n",
       "      <td>III</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>dg_repeat</td>\n",
       "      <td>1084748</td>\n",
       "      <td>1087132</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRPTCENC.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41629</th>\n",
       "      <td>III</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>dg_repeat</td>\n",
       "      <td>1111813</td>\n",
       "      <td>1114198</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRPTCENC.18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41444</th>\n",
       "      <td>III</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>dg_repeat</td>\n",
       "      <td>1118573</td>\n",
       "      <td>1120957</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRPTCENC.21</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41583</th>\n",
       "      <td>III</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>dg_repeat</td>\n",
       "      <td>1125333</td>\n",
       "      <td>1125582</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRPTCENC.24</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41330</th>\n",
       "      <td>III</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>dg_repeat</td>\n",
       "      <td>1132093</td>\n",
       "      <td>1134478</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPRPTCENC.27</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqid   source       type    start      end score strand phase  \\\n",
       "41362     I  PomBase  dg_repeat  3759165  3763441     .      +     .   \n",
       "41668     I  PomBase  dg_repeat  3777115  3778839     .      +     .   \n",
       "41430    II  PomBase  dg_repeat  1604786  1609649     .      -     .   \n",
       "41383    II  PomBase  dg_repeat  1638230  1644747     .      +     .   \n",
       "41685   III  PomBase  dg_repeat  1076115  1078499     .      +     .   \n",
       "41202   III  PomBase  dg_repeat  1082876  1083348     .      +     .   \n",
       "41517   III  PomBase  dg_repeat  1084748  1087132     .      +     .   \n",
       "41629   III  PomBase  dg_repeat  1111813  1114198     .      +     .   \n",
       "41444   III  PomBase  dg_repeat  1118573  1120957     .      +     .   \n",
       "41583   III  PomBase  dg_repeat  1125333  1125582     .      +     .   \n",
       "41330   III  PomBase  dg_repeat  1132093  1134478     .      +     .   \n",
       "\n",
       "                 ID Name Parent  \n",
       "41362   SPRPTCENA.4  NaN    NaN  \n",
       "41668   SPRPTCENA.8  NaN    NaN  \n",
       "41430   SPRPTCENB.1  NaN    NaN  \n",
       "41383  SPRPTCENB.17  NaN    NaN  \n",
       "41685   SPRPTCENC.2  NaN    NaN  \n",
       "41202   SPRPTCENC.5  NaN    NaN  \n",
       "41517   SPRPTCENC.7  NaN    NaN  \n",
       "41629  SPRPTCENC.18  NaN    NaN  \n",
       "41444  SPRPTCENC.21  NaN    NaN  \n",
       "41583  SPRPTCENC.24  NaN    NaN  \n",
       "41330  SPRPTCENC.27  NaN    NaN  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#feature_df = gff[gff['type'].isin(features)]\n",
    "#feature_df = gff[gff['type'].isin(['dh_repeat'])].sort_values(['seqid','start', 'end'])\n",
    "feature_df = gff[gff['type'].isin(['dg_repeat'])].sort_values(['seqid','start', 'end'])\n",
    "\n",
    "feature_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---------------------------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add `length` and `Category` Information"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Because in the new `gtf` we are using the `CDS` **features**, as oposed to the old `gtf` where had `exon` **features**, the lengths of the `gene`'s differ between the two.\n",
    "\n",
    "To recover the same lengths, it should be enough to add the corresponding **three_prime_UTR** and **five_prime_UTR** lengths, to the lengths obtained from `CDS` **features**\n",
    "\n",
    "**Remember:** `exon` = `CDS` + `UTR` "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **`CDS`s**: `12162`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This `gff` contains no **exon** features instead we start by looking at the **CDS** features"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The **coding region of a gene**, also known as the `CDS` (from **coding sequence**), is the portion of a gene's DNA or RNA that codes for protein.\n",
    "\n",
    "Although this term is also sometimes used interchangeably with **exon**, it is not the exact same thing: the **exon** is composed of the **coding region** (`CDS`) as well as the **3' and 5' untranslated regions** of the RNA, and so therefore, an exon would be partially made up of coding regions."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Get the `cds_df` filtering by **CDS** the `type` **Column**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "cds_df = gff[gff['type'].str.contains('CDS')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Filter `columns` that contain **all NaN's**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "cols = cds_df.columns[~cds_df.isna().all()]\n",
    "cds_df = cds_df[cols]\n",
    "#cds_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Add `length` **Column** to the `exon_df`: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "## compute exon lengths: end - start + 1\n",
    "cds_df['cds_length'] = cds_df['end'] - cds_df['start'] + 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Get corresponding `transcript_id`s from the `Parent` **Column**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "## used the `Parent` to get the `transcript_id` attribute\n",
    "#cds_df['transcript_id'] = cds_df['Parent'].apply(lambda x: x.split(':')[1])\n",
    "cds_df['transcript_id'] = cds_df['Parent']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "## check every `transcript` has an associated `transcript_id`\n",
    "assert not any(cds_df['transcript_id'].isnull())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>ID</th>\n",
       "      <th>Parent</th>\n",
       "      <th>cds_length</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1798347</td>\n",
       "      <td>1798830</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.01.1:exon:1</td>\n",
       "      <td>SPAC1002.01.1</td>\n",
       "      <td>484</td>\n",
       "      <td>SPAC1002.01.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1798960</td>\n",
       "      <td>1799015</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.01.1:exon:2</td>\n",
       "      <td>SPAC1002.01.1</td>\n",
       "      <td>56</td>\n",
       "      <td>SPAC1002.01.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1799128</td>\n",
       "      <td>1799817</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.02.1:exon:1</td>\n",
       "      <td>SPAC1002.02.1</td>\n",
       "      <td>690</td>\n",
       "      <td>SPAC1002.02.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1800212</td>\n",
       "      <td>1802983</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.03c.1:exon:1</td>\n",
       "      <td>SPAC1002.03c.1</td>\n",
       "      <td>2772</td>\n",
       "      <td>SPAC1002.03c.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1803773</td>\n",
       "      <td>1804372</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.04c.1:exon:1</td>\n",
       "      <td>SPAC1002.04c.1</td>\n",
       "      <td>600</td>\n",
       "      <td>SPAC1002.04c.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqid   source type    start      end score strand phase  \\\n",
       "2      I  PomBase  CDS  1798347  1798830     .      +     .   \n",
       "4      I  PomBase  CDS  1798960  1799015     .      +     .   \n",
       "8      I  PomBase  CDS  1799128  1799817     .      +     .   \n",
       "13     I  PomBase  CDS  1800212  1802983     .      -     .   \n",
       "18     I  PomBase  CDS  1803773  1804372     .      -     .   \n",
       "\n",
       "                       ID          Parent  cds_length   transcript_id  \n",
       "2    SPAC1002.01.1:exon:1   SPAC1002.01.1         484   SPAC1002.01.1  \n",
       "4    SPAC1002.01.1:exon:2   SPAC1002.01.1          56   SPAC1002.01.1  \n",
       "8    SPAC1002.02.1:exon:1   SPAC1002.02.1         690   SPAC1002.02.1  \n",
       "13  SPAC1002.03c.1:exon:1  SPAC1002.03c.1        2772  SPAC1002.03c.1  \n",
       "18  SPAC1002.04c.1:exon:1  SPAC1002.04c.1         600  SPAC1002.04c.1  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cds_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(12162, 12)"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cds_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Get `transcript_cds_lengths_df`: **summarize** `cds_df` by `transcript_id` and **sum** over individual `cds_length`s"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each `transcript` is associated to one or more `exons`, in this case `CDS`. This defines properly the **length** of an `transcript` as: the sum of exon (`CDS`) length's associated to that transcript."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "transcript_cds_lengths_df = cds_df[['transcript_id', 'cds_length']].groupby('transcript_id').sum().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>cds_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SPAC1002.01.1</td>\n",
       "      <td>540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SPAC1002.02.1</td>\n",
       "      <td>690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SPAC1002.03c.1</td>\n",
       "      <td>2772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SPAC1002.04c.1</td>\n",
       "      <td>600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SPAC1002.05c.1</td>\n",
       "      <td>2148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7000</th>\n",
       "      <td>dg1a.1</td>\n",
       "      <td>3501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7001</th>\n",
       "      <td>dg1b.1</td>\n",
       "      <td>3501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7002</th>\n",
       "      <td>dh1a.1</td>\n",
       "      <td>4001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7003</th>\n",
       "      <td>dh1b.1</td>\n",
       "      <td>4001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7004</th>\n",
       "      <td>spncrna.9001.1</td>\n",
       "      <td>3429</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7005 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       transcript_id  cds_length\n",
       "0      SPAC1002.01.1         540\n",
       "1      SPAC1002.02.1         690\n",
       "2     SPAC1002.03c.1        2772\n",
       "3     SPAC1002.04c.1         600\n",
       "4     SPAC1002.05c.1        2148\n",
       "...              ...         ...\n",
       "7000          dg1a.1        3501\n",
       "7001          dg1b.1        3501\n",
       "7002          dh1a.1        4001\n",
       "7003          dh1b.1        4001\n",
       "7004  spncrna.9001.1        3429\n",
       "\n",
       "[7005 rows x 2 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcript_cds_lengths_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7005, 2)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcript_cds_lengths_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- List containing all `transcript_ids` present in the `cds_df`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7005"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcript_ids = transcript_cds_lengths_df['transcript_id'].unique()\n",
    "len(transcript_ids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **`UTR`'s**: `9654`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In molecular genetics, an **untranslated region** (or `UTR`) refers to either of two sections, one on each side of a **coding sequence** (`CDS`) on a strand of **mRNA**.\n",
    "* If it is found on the **5' side, it is called the 5' UTR (or leader sequence)**.\n",
    "* If it is found on the **3' side, it is called the 3' UTR (or trailer sequence)**.\n",
    "\n",
    "**mRNA** is RNA that carries information from DNA to the ribosome, the site of protein synthesis (translation) within a cell. The mRNA is initially transcribed from the corresponding DNA sequence and then translated into protein. However, several regions of the mRNA are usually not translated into protein, **including the 5' and 3' UTRs**.\n",
    "\n",
    "Although they are called untranslated regions, and do not form the protein-coding region of the gene, uORFs located within the 5' UTR can be translated into peptides"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Get the `utr_df` filtering by **three_prime_UTR** and **five_prime_UTR** the `type` **Column**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "utr_features = ['three_prime_UTR', 'five_prime_UTR']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>ID</th>\n",
       "      <th>Name</th>\n",
       "      <th>Parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>1799061</td>\n",
       "      <td>1799127</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.02.1:five_prime_UTR:1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPAC1002.02.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>1799818</td>\n",
       "      <td>1800053</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.02.1:three_prime_UTR:1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPAC1002.02.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>1802984</td>\n",
       "      <td>1803141</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.03c.1:five_prime_UTR:1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPAC1002.03c.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>1799915</td>\n",
       "      <td>1800211</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.03c.1:three_prime_UTR:1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPAC1002.03c.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>1804373</td>\n",
       "      <td>1804491</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.04c.1:five_prime_UTR:1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPAC1002.04c.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35952</th>\n",
       "      <td>III</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>708064</td>\n",
       "      <td>708196</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPCPJ732.03.1:five_prime_UTR:1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPCPJ732.03.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35954</th>\n",
       "      <td>III</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>708330</td>\n",
       "      <td>708339</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPCPJ732.03.1:five_prime_UTR:2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPCPJ732.03.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35960</th>\n",
       "      <td>III</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>708891</td>\n",
       "      <td>709145</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPCPJ732.03.1:three_prime_UTR:1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPCPJ732.03.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36243</th>\n",
       "      <td>mating_type_region</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>3711</td>\n",
       "      <td>3753</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPMTR.01.1:five_prime_UTR:1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPMTR.01.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36247</th>\n",
       "      <td>mating_type_region</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>3861</td>\n",
       "      <td>3888</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPMTR.02.1:five_prime_UTR:1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPMTR.02.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9654 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    seqid   source             type    start      end score  \\\n",
       "7                       I  PomBase   five_prime_UTR  1799061  1799127     .   \n",
       "9                       I  PomBase  three_prime_UTR  1799818  1800053     .   \n",
       "12                      I  PomBase   five_prime_UTR  1802984  1803141     .   \n",
       "14                      I  PomBase  three_prime_UTR  1799915  1800211     .   \n",
       "17                      I  PomBase   five_prime_UTR  1804373  1804491     .   \n",
       "...                   ...      ...              ...      ...      ...   ...   \n",
       "35952                 III  PomBase   five_prime_UTR   708064   708196     .   \n",
       "35954                 III  PomBase   five_prime_UTR   708330   708339     .   \n",
       "35960                 III  PomBase  three_prime_UTR   708891   709145     .   \n",
       "36243  mating_type_region  PomBase   five_prime_UTR     3711     3753     .   \n",
       "36247  mating_type_region  PomBase   five_prime_UTR     3861     3888     .   \n",
       "\n",
       "      strand phase                                ID Name          Parent  \n",
       "7          +     .    SPAC1002.02.1:five_prime_UTR:1  NaN   SPAC1002.02.1  \n",
       "9          +     .   SPAC1002.02.1:three_prime_UTR:1  NaN   SPAC1002.02.1  \n",
       "12         -     .   SPAC1002.03c.1:five_prime_UTR:1  NaN  SPAC1002.03c.1  \n",
       "14         -     .  SPAC1002.03c.1:three_prime_UTR:1  NaN  SPAC1002.03c.1  \n",
       "17         -     .   SPAC1002.04c.1:five_prime_UTR:1  NaN  SPAC1002.04c.1  \n",
       "...      ...   ...                               ...  ...             ...  \n",
       "35952      +     .    SPCPJ732.03.1:five_prime_UTR:1  NaN   SPCPJ732.03.1  \n",
       "35954      +     .    SPCPJ732.03.1:five_prime_UTR:2  NaN   SPCPJ732.03.1  \n",
       "35960      +     .   SPCPJ732.03.1:three_prime_UTR:1  NaN   SPCPJ732.03.1  \n",
       "36243      -     .       SPMTR.01.1:five_prime_UTR:1  NaN      SPMTR.01.1  \n",
       "36247      +     .       SPMTR.02.1:five_prime_UTR:1  NaN      SPMTR.02.1  \n",
       "\n",
       "[9654 rows x 11 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "utr_df = gff[gff['type'].isin(utr_features)]\n",
    "utr_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Filter `columns` that contain **all NaN's**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "cols = utr_df.columns[~utr_df.isna().all()]\n",
    "utr_df = utr_df[cols]\n",
    "#utr_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Add `length` **Column** to the `utr_df`: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "## compute exon lengths: end - start + 1\n",
    "utr_df['utr_length'] = utr_df['end'] - utr_df['start'] + 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Get corresponding `transcript_id`s from the `Parent` **Column**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "## used the `Parent` to get the `transcript_id` attribute\n",
    "#utr_df['transcript_id'] = utr_df['Parent'].apply(lambda x: x.split(':')[1])\n",
    "utr_df['transcript_id'] = utr_df['Parent']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "## check every `transcript` has an associated `transcript_id`\n",
    "assert not any(utr_df['transcript_id'].isnull())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>ID</th>\n",
       "      <th>Parent</th>\n",
       "      <th>utr_length</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>1799061</td>\n",
       "      <td>1799127</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.02.1:five_prime_UTR:1</td>\n",
       "      <td>SPAC1002.02.1</td>\n",
       "      <td>67</td>\n",
       "      <td>SPAC1002.02.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>1799818</td>\n",
       "      <td>1800053</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.02.1:three_prime_UTR:1</td>\n",
       "      <td>SPAC1002.02.1</td>\n",
       "      <td>236</td>\n",
       "      <td>SPAC1002.02.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>1802984</td>\n",
       "      <td>1803141</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.03c.1:five_prime_UTR:1</td>\n",
       "      <td>SPAC1002.03c.1</td>\n",
       "      <td>158</td>\n",
       "      <td>SPAC1002.03c.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>three_prime_UTR</td>\n",
       "      <td>1799915</td>\n",
       "      <td>1800211</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.03c.1:three_prime_UTR:1</td>\n",
       "      <td>SPAC1002.03c.1</td>\n",
       "      <td>297</td>\n",
       "      <td>SPAC1002.03c.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>five_prime_UTR</td>\n",
       "      <td>1804373</td>\n",
       "      <td>1804491</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.04c.1:five_prime_UTR:1</td>\n",
       "      <td>SPAC1002.04c.1</td>\n",
       "      <td>119</td>\n",
       "      <td>SPAC1002.04c.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqid   source             type    start      end score strand phase  \\\n",
       "7      I  PomBase   five_prime_UTR  1799061  1799127     .      +     .   \n",
       "9      I  PomBase  three_prime_UTR  1799818  1800053     .      +     .   \n",
       "12     I  PomBase   five_prime_UTR  1802984  1803141     .      -     .   \n",
       "14     I  PomBase  three_prime_UTR  1799915  1800211     .      -     .   \n",
       "17     I  PomBase   five_prime_UTR  1804373  1804491     .      -     .   \n",
       "\n",
       "                                  ID          Parent  utr_length  \\\n",
       "7     SPAC1002.02.1:five_prime_UTR:1   SPAC1002.02.1          67   \n",
       "9    SPAC1002.02.1:three_prime_UTR:1   SPAC1002.02.1         236   \n",
       "12   SPAC1002.03c.1:five_prime_UTR:1  SPAC1002.03c.1         158   \n",
       "14  SPAC1002.03c.1:three_prime_UTR:1  SPAC1002.03c.1         297   \n",
       "17   SPAC1002.04c.1:five_prime_UTR:1  SPAC1002.04c.1         119   \n",
       "\n",
       "     transcript_id  \n",
       "7    SPAC1002.02.1  \n",
       "9    SPAC1002.02.1  \n",
       "12  SPAC1002.03c.1  \n",
       "14  SPAC1002.03c.1  \n",
       "17  SPAC1002.04c.1  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "utr_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(9654, 12)"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "utr_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Get `transcript_utr_lengths_df`: **summarize** `utr_df` by `transcript_id` and **sum** over individual `utr_length`s"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each `transcript` is associated to two or less `utr`, we can properly the define the **length** of a `transcript_utrs` as: the sum of individual `utr` length's associated to that transcript."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "transcript_utr_lengths_df = utr_df[['transcript_id', 'utr_length']].groupby('transcript_id').sum().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>utr_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SPAC1002.02.1</td>\n",
       "      <td>303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SPAC1002.03c.1</td>\n",
       "      <td>455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SPAC1002.04c.1</td>\n",
       "      <td>268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SPAC1002.05c.1</td>\n",
       "      <td>102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SPAC1002.07c.1</td>\n",
       "      <td>438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4852</th>\n",
       "      <td>SPCPJ732.01.1</td>\n",
       "      <td>536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4853</th>\n",
       "      <td>SPCPJ732.02c.1</td>\n",
       "      <td>712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4854</th>\n",
       "      <td>SPCPJ732.03.1</td>\n",
       "      <td>398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4855</th>\n",
       "      <td>SPMTR.01.1</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4856</th>\n",
       "      <td>SPMTR.02.1</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4857 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       transcript_id  utr_length\n",
       "0      SPAC1002.02.1         303\n",
       "1     SPAC1002.03c.1         455\n",
       "2     SPAC1002.04c.1         268\n",
       "3     SPAC1002.05c.1         102\n",
       "4     SPAC1002.07c.1         438\n",
       "...              ...         ...\n",
       "4852   SPCPJ732.01.1         536\n",
       "4853  SPCPJ732.02c.1         712\n",
       "4854   SPCPJ732.03.1         398\n",
       "4855      SPMTR.01.1          43\n",
       "4856      SPMTR.02.1          28\n",
       "\n",
       "[4857 rows x 2 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcript_utr_lengths_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4857, 2)"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcript_utr_lengths_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- List containing all `transcript_ids` present in the `utr_df`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4857"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "utr_transcript_ids = transcript_utr_lengths_df['transcript_id'].tolist()\n",
    "len(utr_transcript_ids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **`Transcript`s**: `7005`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Get `transcript_df` by filtering the **gff** for `transcript_id`s present in the `cds_df`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7005, 11)"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#transcript_df = gff[gff['transcript_id'].isin(transcript_ids)]\n",
    "transcript_df = gff[gff['ID'].isin(transcript_ids)]\n",
    "transcript_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Filter `columns` that contain **all NaN's**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>ID</th>\n",
       "      <th>Parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>1798347</td>\n",
       "      <td>1799015</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.01.1</td>\n",
       "      <td>SPAC1002.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>1799061</td>\n",
       "      <td>1800053</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.02.1</td>\n",
       "      <td>SPAC1002.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>1799915</td>\n",
       "      <td>1803141</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.03c.1</td>\n",
       "      <td>SPAC1002.03c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>1803624</td>\n",
       "      <td>1804491</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.04c.1</td>\n",
       "      <td>SPAC1002.04c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>1804548</td>\n",
       "      <td>1806797</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.05c.1</td>\n",
       "      <td>SPAC1002.05c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41741</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>pseudogenic_transcript</td>\n",
       "      <td>5726</td>\n",
       "      <td>6331</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC212.10b.1</td>\n",
       "      <td>SPAC212.10b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41744</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>3761000</td>\n",
       "      <td>3764500</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>dg1b.1</td>\n",
       "      <td>dg1b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41747</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>3761000</td>\n",
       "      <td>3764500</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>dg1a.1</td>\n",
       "      <td>dg1a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41750</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>3753000</td>\n",
       "      <td>3757000</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>dh1b.1</td>\n",
       "      <td>dh1b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41753</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>3753000</td>\n",
       "      <td>3757000</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>dh1a.1</td>\n",
       "      <td>dh1a</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7005 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqid   source                    type    start      end score strand  \\\n",
       "1         I  PomBase                    mRNA  1798347  1799015     .      +   \n",
       "6         I  PomBase                    mRNA  1799061  1800053     .      +   \n",
       "11        I  PomBase                    mRNA  1799915  1803141     .      -   \n",
       "16        I  PomBase                    mRNA  1803624  1804491     .      -   \n",
       "21        I  PomBase                    mRNA  1804548  1806797     .      -   \n",
       "...     ...      ...                     ...      ...      ...   ...    ...   \n",
       "41741     I  PomBase  pseudogenic_transcript     5726     6331     .      +   \n",
       "41744     I  PomBase                    mRNA  3761000  3764500     .      -   \n",
       "41747     I  PomBase                    mRNA  3761000  3764500     .      +   \n",
       "41750     I  PomBase                    mRNA  3753000  3757000     .      -   \n",
       "41753     I  PomBase                    mRNA  3753000  3757000     .      +   \n",
       "\n",
       "      phase              ID        Parent  \n",
       "1         .   SPAC1002.01.1   SPAC1002.01  \n",
       "6         .   SPAC1002.02.1   SPAC1002.02  \n",
       "11        .  SPAC1002.03c.1  SPAC1002.03c  \n",
       "16        .  SPAC1002.04c.1  SPAC1002.04c  \n",
       "21        .  SPAC1002.05c.1  SPAC1002.05c  \n",
       "...     ...             ...           ...  \n",
       "41741     .   SPAC212.10b.1   SPAC212.10b  \n",
       "41744     .          dg1b.1          dg1b  \n",
       "41747     .          dg1a.1          dg1a  \n",
       "41750     .          dh1b.1          dh1b  \n",
       "41753     .          dh1a.1          dh1a  \n",
       "\n",
       "[7005 rows x 10 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cols = transcript_df.columns[~transcript_df.isna().all()]\n",
    "transcript_df = transcript_df[cols]\n",
    "transcript_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Get corresponding `transcript_id`s from the `ID` **Column**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "## use the `ID` to get the `transcript_id` attribute\n",
    "transcript_df['transcript_id'] = transcript_df['ID']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Add `cds_length` **Column** to the `transcript_df`: **merge** `transcript_cds_lengths_df` to `transcript_df`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "transcript_df = transcript_df.merge(transcript_cds_lengths_df, on='transcript_id')\n",
    "#transcript_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Add `utr_length` **Column** to the `transcript_df`: **merge** `transcript_utr_lengths_df` to `transcript_df`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "transcript_df = transcript_df.merge(transcript_utr_lengths_df, on='transcript_id', how='outer')\n",
    "#transcript_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Get corresponding `gene_id` from the `Parent` **Column**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "## used the `Parent` to get the `gene_id` attribute\n",
    "#transcript_df['gene_id'] = transcript_df['Parent'].apply(lambda x: x.split(':')[1])\n",
    "transcript_df['gene_id'] = transcript_df['Parent']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "## check every `transcript` has an associated `gene-id`\n",
    "assert not any(transcript_df['gene_id'].isnull())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>ID</th>\n",
       "      <th>Parent</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>cds_length</th>\n",
       "      <th>utr_length</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>1798347</td>\n",
       "      <td>1799015</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.01.1</td>\n",
       "      <td>SPAC1002.01</td>\n",
       "      <td>SPAC1002.01.1</td>\n",
       "      <td>540</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SPAC1002.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>1799061</td>\n",
       "      <td>1800053</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.02.1</td>\n",
       "      <td>SPAC1002.02</td>\n",
       "      <td>SPAC1002.02.1</td>\n",
       "      <td>690</td>\n",
       "      <td>303.0</td>\n",
       "      <td>SPAC1002.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>1799915</td>\n",
       "      <td>1803141</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.03c.1</td>\n",
       "      <td>SPAC1002.03c</td>\n",
       "      <td>SPAC1002.03c.1</td>\n",
       "      <td>2772</td>\n",
       "      <td>455.0</td>\n",
       "      <td>SPAC1002.03c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>1803624</td>\n",
       "      <td>1804491</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.04c.1</td>\n",
       "      <td>SPAC1002.04c</td>\n",
       "      <td>SPAC1002.04c.1</td>\n",
       "      <td>600</td>\n",
       "      <td>268.0</td>\n",
       "      <td>SPAC1002.04c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>1804548</td>\n",
       "      <td>1806797</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.05c.1</td>\n",
       "      <td>SPAC1002.05c</td>\n",
       "      <td>SPAC1002.05c.1</td>\n",
       "      <td>2148</td>\n",
       "      <td>102.0</td>\n",
       "      <td>SPAC1002.05c</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqid   source  type    start      end score strand phase              ID  \\\n",
       "0     I  PomBase  mRNA  1798347  1799015     .      +     .   SPAC1002.01.1   \n",
       "1     I  PomBase  mRNA  1799061  1800053     .      +     .   SPAC1002.02.1   \n",
       "2     I  PomBase  mRNA  1799915  1803141     .      -     .  SPAC1002.03c.1   \n",
       "3     I  PomBase  mRNA  1803624  1804491     .      -     .  SPAC1002.04c.1   \n",
       "4     I  PomBase  mRNA  1804548  1806797     .      -     .  SPAC1002.05c.1   \n",
       "\n",
       "         Parent   transcript_id  cds_length  utr_length       gene_id  \n",
       "0   SPAC1002.01   SPAC1002.01.1         540         NaN   SPAC1002.01  \n",
       "1   SPAC1002.02   SPAC1002.02.1         690       303.0   SPAC1002.02  \n",
       "2  SPAC1002.03c  SPAC1002.03c.1        2772       455.0  SPAC1002.03c  \n",
       "3  SPAC1002.04c  SPAC1002.04c.1         600       268.0  SPAC1002.04c  \n",
       "4  SPAC1002.05c  SPAC1002.05c.1        2148       102.0  SPAC1002.05c  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcript_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7005, 14)"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcript_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Have a look at distribution of **feature** `types`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>type</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>snRNA</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogenic_transcript</th>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rRNA</th>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>snoRNA</th>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tRNA</th>\n",
       "      <td>196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ncRNA</th>\n",
       "      <td>1528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mRNA</th>\n",
       "      <td>5140</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          ID\n",
       "type                        \n",
       "snRNA                      6\n",
       "pseudogenic_transcript    29\n",
       "rRNA                      49\n",
       "snoRNA                    57\n",
       "tRNA                     196\n",
       "ncRNA                   1528\n",
       "mRNA                    5140"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcripts_df_by_type = transcript_df[['type', 'ID']].groupby(['type']).count()\n",
    "transcripts_df_by_type.sort_values('ID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7005"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(transcripts_df_by_type.ID)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- List containing all `gene_ids`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7005"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_ids = transcript_df['gene_id'].unique().tolist()\n",
    "len(gene_ids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## **`Gene`s**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Get `gene_df` by filtering the **gff** for `gene_id`s present in the `transcript_df`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7005, 11)"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#gene_df = gff[gff['gene_id'].isin(gene_ids)]\n",
    "gene_df = gff[gff['ID'].isin(gene_ids)]\n",
    "gene_df.shape\n",
    "#gene_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Filter `columns` that contain **all NaN's**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>ID</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>1798347</td>\n",
       "      <td>1799015</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.01</td>\n",
       "      <td>mrx11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>1799061</td>\n",
       "      <td>1800053</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.02</td>\n",
       "      <td>pom34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>1799915</td>\n",
       "      <td>1803141</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.03c</td>\n",
       "      <td>gls2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>1803624</td>\n",
       "      <td>1804491</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.04c</td>\n",
       "      <td>taf11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>1804548</td>\n",
       "      <td>1806797</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.05c</td>\n",
       "      <td>jmj2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41740</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>5726</td>\n",
       "      <td>6331</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC212.10b</td>\n",
       "      <td>SPAC212.10_plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41743</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>3761000</td>\n",
       "      <td>3764500</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>dg1b</td>\n",
       "      <td>dg1_minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41746</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>3761000</td>\n",
       "      <td>3764500</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>dg1a</td>\n",
       "      <td>dg1_plus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41749</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>3753000</td>\n",
       "      <td>3757000</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>dh1b</td>\n",
       "      <td>dh1_minus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41752</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>3753000</td>\n",
       "      <td>3757000</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>dh1a</td>\n",
       "      <td>dh1_plus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7005 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqid   source  type    start      end score strand phase            ID  \\\n",
       "0         I  PomBase  gene  1798347  1799015     .      +     .   SPAC1002.01   \n",
       "5         I  PomBase  gene  1799061  1800053     .      +     .   SPAC1002.02   \n",
       "10        I  PomBase  gene  1799915  1803141     .      -     .  SPAC1002.03c   \n",
       "15        I  PomBase  gene  1803624  1804491     .      -     .  SPAC1002.04c   \n",
       "20        I  PomBase  gene  1804548  1806797     .      -     .  SPAC1002.05c   \n",
       "...     ...      ...   ...      ...      ...   ...    ...   ...           ...   \n",
       "41740     I  PomBase  gene     5726     6331     .      +     .   SPAC212.10b   \n",
       "41743     I  PomBase  gene  3761000  3764500     .      -     .          dg1b   \n",
       "41746     I  PomBase  gene  3761000  3764500     .      +     .          dg1a   \n",
       "41749     I  PomBase  gene  3753000  3757000     .      -     .          dh1b   \n",
       "41752     I  PomBase  gene  3753000  3757000     .      +     .          dh1a   \n",
       "\n",
       "                  Name  \n",
       "0                mrx11  \n",
       "5                pom34  \n",
       "10                gls2  \n",
       "15               taf11  \n",
       "20                jmj2  \n",
       "...                ...  \n",
       "41740  SPAC212.10_plus  \n",
       "41743        dg1_minus  \n",
       "41746         dg1_plus  \n",
       "41749        dh1_minus  \n",
       "41752         dh1_plus  \n",
       "\n",
       "[7005 rows x 10 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cols = gene_df.columns[~gene_df.isna().all()]\n",
    "gene_df = gene_df[cols]\n",
    "gene_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Because not all genes contain a `Name` category, we can't use this as an identifier but this are not real NaNs, we fix this to resolve future issues:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_df['Name'] = gene_df['Name'].fillna('not_defined')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Get corresponding `transcript_id`s from the `ID` **Column**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "## use the `ID` to get the `gene_id` attribute\n",
    "gene_df['gene_id'] = gene_df['ID']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each `gene` is associated to one and only one `transcript`. This defines properly the **length** of a `gene` as: the length of the associated transcript (e.g. the sum of exon length's of the associated transcript)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Add `cds_length` and `utr_length` **Column**s to the `gene_df`: **merge** `transcripts_df` to `gene_df`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>ID</th>\n",
       "      <th>Name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>cds_length</th>\n",
       "      <th>utr_length</th>\n",
       "      <th>bio_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>1798347</td>\n",
       "      <td>1799015</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.01</td>\n",
       "      <td>mrx11</td>\n",
       "      <td>SPAC1002.01</td>\n",
       "      <td>540</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>1799061</td>\n",
       "      <td>1800053</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.02</td>\n",
       "      <td>pom34</td>\n",
       "      <td>SPAC1002.02</td>\n",
       "      <td>690</td>\n",
       "      <td>303.0</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>1799915</td>\n",
       "      <td>1803141</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.03c</td>\n",
       "      <td>gls2</td>\n",
       "      <td>SPAC1002.03c</td>\n",
       "      <td>2772</td>\n",
       "      <td>455.0</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>1803624</td>\n",
       "      <td>1804491</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.04c</td>\n",
       "      <td>taf11</td>\n",
       "      <td>SPAC1002.04c</td>\n",
       "      <td>600</td>\n",
       "      <td>268.0</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>1804548</td>\n",
       "      <td>1806797</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.05c</td>\n",
       "      <td>jmj2</td>\n",
       "      <td>SPAC1002.05c</td>\n",
       "      <td>2148</td>\n",
       "      <td>102.0</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7000</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>5726</td>\n",
       "      <td>6331</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC212.10b</td>\n",
       "      <td>SPAC212.10_plus</td>\n",
       "      <td>SPAC212.10b</td>\n",
       "      <td>606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pseudogenic_transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7001</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>3761000</td>\n",
       "      <td>3764500</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>dg1b</td>\n",
       "      <td>dg1_minus</td>\n",
       "      <td>dg1b</td>\n",
       "      <td>3501</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7002</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>3761000</td>\n",
       "      <td>3764500</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>dg1a</td>\n",
       "      <td>dg1_plus</td>\n",
       "      <td>dg1a</td>\n",
       "      <td>3501</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7003</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>3753000</td>\n",
       "      <td>3757000</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>dh1b</td>\n",
       "      <td>dh1_minus</td>\n",
       "      <td>dh1b</td>\n",
       "      <td>4001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7004</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>3753000</td>\n",
       "      <td>3757000</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>dh1a</td>\n",
       "      <td>dh1_plus</td>\n",
       "      <td>dh1a</td>\n",
       "      <td>4001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7005 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     seqid   source  type    start      end score strand phase            ID  \\\n",
       "0        I  PomBase  gene  1798347  1799015     .      +     .   SPAC1002.01   \n",
       "1        I  PomBase  gene  1799061  1800053     .      +     .   SPAC1002.02   \n",
       "2        I  PomBase  gene  1799915  1803141     .      -     .  SPAC1002.03c   \n",
       "3        I  PomBase  gene  1803624  1804491     .      -     .  SPAC1002.04c   \n",
       "4        I  PomBase  gene  1804548  1806797     .      -     .  SPAC1002.05c   \n",
       "...    ...      ...   ...      ...      ...   ...    ...   ...           ...   \n",
       "7000     I  PomBase  gene     5726     6331     .      +     .   SPAC212.10b   \n",
       "7001     I  PomBase  gene  3761000  3764500     .      -     .          dg1b   \n",
       "7002     I  PomBase  gene  3761000  3764500     .      +     .          dg1a   \n",
       "7003     I  PomBase  gene  3753000  3757000     .      -     .          dh1b   \n",
       "7004     I  PomBase  gene  3753000  3757000     .      +     .          dh1a   \n",
       "\n",
       "                 Name       gene_id  cds_length  utr_length  \\\n",
       "0               mrx11   SPAC1002.01         540         NaN   \n",
       "1               pom34   SPAC1002.02         690       303.0   \n",
       "2                gls2  SPAC1002.03c        2772       455.0   \n",
       "3               taf11  SPAC1002.04c         600       268.0   \n",
       "4                jmj2  SPAC1002.05c        2148       102.0   \n",
       "...               ...           ...         ...         ...   \n",
       "7000  SPAC212.10_plus   SPAC212.10b         606         NaN   \n",
       "7001        dg1_minus          dg1b        3501         NaN   \n",
       "7002         dg1_plus          dg1a        3501         NaN   \n",
       "7003        dh1_minus          dh1b        4001         NaN   \n",
       "7004         dh1_plus          dh1a        4001         NaN   \n",
       "\n",
       "                    bio_type  \n",
       "0                       mRNA  \n",
       "1                       mRNA  \n",
       "2                       mRNA  \n",
       "3                       mRNA  \n",
       "4                       mRNA  \n",
       "...                      ...  \n",
       "7000  pseudogenic_transcript  \n",
       "7001                    mRNA  \n",
       "7002                    mRNA  \n",
       "7003                    mRNA  \n",
       "7004                    mRNA  \n",
       "\n",
       "[7005 rows x 14 columns]"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#gene_df = gene_df.drop('type', axis=1).merge(transcript_df[['gene_id', 'cds_length', 'utr_length', 'type']], on='gene_id')\n",
    "gene_df = gene_df.merge(transcript_df[['gene_id', 'cds_length', 'utr_length', 'type']].rename(columns={'type':'bio_type'}), on='gene_id')\n",
    "gene_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7005, 14)"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Add `gene_length` **Column** to the `gene_df`: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>ID</th>\n",
       "      <th>Name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>cds_length</th>\n",
       "      <th>utr_length</th>\n",
       "      <th>bio_type</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>1798347</td>\n",
       "      <td>1799015</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.01</td>\n",
       "      <td>mrx11</td>\n",
       "      <td>SPAC1002.01</td>\n",
       "      <td>540</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>540.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>1799061</td>\n",
       "      <td>1800053</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.02</td>\n",
       "      <td>pom34</td>\n",
       "      <td>SPAC1002.02</td>\n",
       "      <td>690</td>\n",
       "      <td>303.0</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>993.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>1799915</td>\n",
       "      <td>1803141</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.03c</td>\n",
       "      <td>gls2</td>\n",
       "      <td>SPAC1002.03c</td>\n",
       "      <td>2772</td>\n",
       "      <td>455.0</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>3227.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>1803624</td>\n",
       "      <td>1804491</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.04c</td>\n",
       "      <td>taf11</td>\n",
       "      <td>SPAC1002.04c</td>\n",
       "      <td>600</td>\n",
       "      <td>268.0</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>868.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>1804548</td>\n",
       "      <td>1806797</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.05c</td>\n",
       "      <td>jmj2</td>\n",
       "      <td>SPAC1002.05c</td>\n",
       "      <td>2148</td>\n",
       "      <td>102.0</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>2250.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7000</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>5726</td>\n",
       "      <td>6331</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC212.10b</td>\n",
       "      <td>SPAC212.10_plus</td>\n",
       "      <td>SPAC212.10b</td>\n",
       "      <td>606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pseudogenic_transcript</td>\n",
       "      <td>606.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7001</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>3761000</td>\n",
       "      <td>3764500</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>dg1b</td>\n",
       "      <td>dg1_minus</td>\n",
       "      <td>dg1b</td>\n",
       "      <td>3501</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>3501.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7002</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>3761000</td>\n",
       "      <td>3764500</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>dg1a</td>\n",
       "      <td>dg1_plus</td>\n",
       "      <td>dg1a</td>\n",
       "      <td>3501</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>3501.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7003</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>3753000</td>\n",
       "      <td>3757000</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>dh1b</td>\n",
       "      <td>dh1_minus</td>\n",
       "      <td>dh1b</td>\n",
       "      <td>4001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>4001.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7004</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>3753000</td>\n",
       "      <td>3757000</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>dh1a</td>\n",
       "      <td>dh1_plus</td>\n",
       "      <td>dh1a</td>\n",
       "      <td>4001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>4001.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7005 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     seqid   source  type    start      end score strand phase            ID  \\\n",
       "0        I  PomBase  gene  1798347  1799015     .      +     .   SPAC1002.01   \n",
       "1        I  PomBase  gene  1799061  1800053     .      +     .   SPAC1002.02   \n",
       "2        I  PomBase  gene  1799915  1803141     .      -     .  SPAC1002.03c   \n",
       "3        I  PomBase  gene  1803624  1804491     .      -     .  SPAC1002.04c   \n",
       "4        I  PomBase  gene  1804548  1806797     .      -     .  SPAC1002.05c   \n",
       "...    ...      ...   ...      ...      ...   ...    ...   ...           ...   \n",
       "7000     I  PomBase  gene     5726     6331     .      +     .   SPAC212.10b   \n",
       "7001     I  PomBase  gene  3761000  3764500     .      -     .          dg1b   \n",
       "7002     I  PomBase  gene  3761000  3764500     .      +     .          dg1a   \n",
       "7003     I  PomBase  gene  3753000  3757000     .      -     .          dh1b   \n",
       "7004     I  PomBase  gene  3753000  3757000     .      +     .          dh1a   \n",
       "\n",
       "                 Name       gene_id  cds_length  utr_length  \\\n",
       "0               mrx11   SPAC1002.01         540         NaN   \n",
       "1               pom34   SPAC1002.02         690       303.0   \n",
       "2                gls2  SPAC1002.03c        2772       455.0   \n",
       "3               taf11  SPAC1002.04c         600       268.0   \n",
       "4                jmj2  SPAC1002.05c        2148       102.0   \n",
       "...               ...           ...         ...         ...   \n",
       "7000  SPAC212.10_plus   SPAC212.10b         606         NaN   \n",
       "7001        dg1_minus          dg1b        3501         NaN   \n",
       "7002         dg1_plus          dg1a        3501         NaN   \n",
       "7003        dh1_minus          dh1b        4001         NaN   \n",
       "7004         dh1_plus          dh1a        4001         NaN   \n",
       "\n",
       "                    bio_type  gene_length  \n",
       "0                       mRNA        540.0  \n",
       "1                       mRNA        993.0  \n",
       "2                       mRNA       3227.0  \n",
       "3                       mRNA        868.0  \n",
       "4                       mRNA       2250.0  \n",
       "...                      ...          ...  \n",
       "7000  pseudogenic_transcript        606.0  \n",
       "7001                    mRNA       3501.0  \n",
       "7002                    mRNA       3501.0  \n",
       "7003                    mRNA       4001.0  \n",
       "7004                    mRNA       4001.0  \n",
       "\n",
       "[7005 rows x 15 columns]"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_df['gene_length'] = gene_df[[\"cds_length\", \"utr_length\"]].sum(axis=1)\n",
    "gene_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Load **repeats** `GTF/GFF` **file** as `gdf`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "repeat_df = pd.read_csv(in_repeats, sep='\\t', comment='#', names=gff_header)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Parse `attributes` colummn\n",
    "attributes_df = repeat_df['attributes'].apply(lambda x: Util.parse_attribute_col(x))\n",
    "attributes_df = pd.DataFrame(attributes_df.values.tolist(), index=repeat_df.index)\n",
    "repeat_df = pd.concat([repeat_df.drop('attributes', axis=1), attributes_df], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>type</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CDS</th>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>five_prime_UTR</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>intron</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mRNA</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ncRNA</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pseudogenic_transcript</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>three_prime_UTR</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        ID\n",
       "type                      \n",
       "CDS                     32\n",
       "five_prime_UTR           7\n",
       "gene                    27\n",
       "intron                   5\n",
       "mRNA                    21\n",
       "ncRNA                    1\n",
       "pseudogenic_transcript   5\n",
       "three_prime_UTR          7"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "repeat_df[['ID', 'type']].groupby('type').count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Get `gene` **features** by selecting those entries without a `Parent`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>ID</th>\n",
       "      <th>Name</th>\n",
       "      <th>Parent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>1</td>\n",
       "      <td>5662</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC212.11</td>\n",
       "      <td>tlh1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>5726</td>\n",
       "      <td>6331</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC212.10</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>7619</td>\n",
       "      <td>9274</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC212.09c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>11027</td>\n",
       "      <td>11556</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPNCRNA.70</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>11784</td>\n",
       "      <td>12994</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC212.08c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>13665</td>\n",
       "      <td>14555</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC212.07c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>15855</td>\n",
       "      <td>16226</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC212.12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>18042</td>\n",
       "      <td>18974</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC212.06c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>20824</td>\n",
       "      <td>21015</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC212.05c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>21381</td>\n",
       "      <td>23050</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC212.04c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>23589</td>\n",
       "      <td>24054</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC212.03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>27353</td>\n",
       "      <td>27763</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC212.02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>28738</td>\n",
       "      <td>29657</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC212.01c</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>29764</td>\n",
       "      <td>31069</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC977.01</td>\n",
       "      <td>ftm1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>31140</td>\n",
       "      <td>32345</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC977.18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>32034</td>\n",
       "      <td>33012</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC977.02</td>\n",
       "      <td>ftm2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>33835</td>\n",
       "      <td>34272</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC977.03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>34298</td>\n",
       "      <td>34978</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC977.04</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>1</td>\n",
       "      <td>5662</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC212.11b</td>\n",
       "      <td>tlh1_plus</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>5726</td>\n",
       "      <td>6331</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC212.10b</td>\n",
       "      <td>SPAC212.10_plus</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>3761000</td>\n",
       "      <td>3764500</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>dg1b</td>\n",
       "      <td>dg1_minus</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>3761000</td>\n",
       "      <td>3764500</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>dg1a</td>\n",
       "      <td>dg1_plus</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>3753000</td>\n",
       "      <td>3757000</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>dh1b</td>\n",
       "      <td>dh1_minus</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>3753000</td>\n",
       "      <td>3757000</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>dh1a</td>\n",
       "      <td>dh1_plus</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>II</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>2129000</td>\n",
       "      <td>2133000</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>MAT2</td>\n",
       "      <td>MAT2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>II</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>2133000</td>\n",
       "      <td>2137000</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>MAT3</td>\n",
       "      <td>MAT3</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>II</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>2114000</td>\n",
       "      <td>2115500</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>MAT1</td>\n",
       "      <td>MAT1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    seqid   source  type    start      end score strand phase           ID  \\\n",
       "0       I  PomBase  gene        1     5662     .      -     .   SPAC212.11   \n",
       "3       I  PomBase  gene     5726     6331     .      -     .   SPAC212.10   \n",
       "6       I  PomBase  gene     7619     9274     .      +     .  SPAC212.09c   \n",
       "9       I  PomBase  gene    11027    11556     .      -     .   SPNCRNA.70   \n",
       "12      I  PomBase  gene    11784    12994     .      +     .  SPAC212.08c   \n",
       "16      I  PomBase  gene    13665    14555     .      +     .  SPAC212.07c   \n",
       "19      I  PomBase  gene    15855    16226     .      +     .   SPAC212.12   \n",
       "22      I  PomBase  gene    18042    18974     .      +     .  SPAC212.06c   \n",
       "29      I  PomBase  gene    20824    21015     .      +     .  SPAC212.05c   \n",
       "32      I  PomBase  gene    21381    23050     .      +     .  SPAC212.04c   \n",
       "39      I  PomBase  gene    23589    24054     .      -     .   SPAC212.03   \n",
       "43      I  PomBase  gene    27353    27763     .      -     .   SPAC212.02   \n",
       "46      I  PomBase  gene    28738    29657     .      +     .  SPAC212.01c   \n",
       "52      I  PomBase  gene    29764    31069     .      +     .   SPAC977.01   \n",
       "56      I  PomBase  gene    31140    32345     .      -     .   SPAC977.18   \n",
       "65      I  PomBase  gene    32034    33012     .      +     .   SPAC977.02   \n",
       "70      I  PomBase  gene    33835    34272     .      +     .   SPAC977.03   \n",
       "73      I  PomBase  gene    34298    34978     .      +     .   SPAC977.04   \n",
       "78      I  PomBase  gene        1     5662     .      +     .  SPAC212.11b   \n",
       "81      I  PomBase  gene     5726     6331     .      +     .  SPAC212.10b   \n",
       "84      I  PomBase  gene  3761000  3764500     .      -     .         dg1b   \n",
       "87      I  PomBase  gene  3761000  3764500     .      +     .         dg1a   \n",
       "90      I  PomBase  gene  3753000  3757000     .      -     .         dh1b   \n",
       "93      I  PomBase  gene  3753000  3757000     .      +     .         dh1a   \n",
       "96     II  PomBase  gene  2129000  2133000     .      -     .         MAT2   \n",
       "99     II  PomBase  gene  2133000  2137000     .      -     .         MAT3   \n",
       "102    II  PomBase  gene  2114000  2115500     .      -     .         MAT1   \n",
       "\n",
       "                Name Parent  \n",
       "0               tlh1    NaN  \n",
       "3                NaN    NaN  \n",
       "6                NaN    NaN  \n",
       "9                NaN    NaN  \n",
       "12               NaN    NaN  \n",
       "16               NaN    NaN  \n",
       "19               NaN    NaN  \n",
       "22               NaN    NaN  \n",
       "29               NaN    NaN  \n",
       "32               NaN    NaN  \n",
       "39               NaN    NaN  \n",
       "43               NaN    NaN  \n",
       "46               NaN    NaN  \n",
       "52              ftm1    NaN  \n",
       "56               NaN    NaN  \n",
       "65              ftm2    NaN  \n",
       "70               NaN    NaN  \n",
       "73               NaN    NaN  \n",
       "78         tlh1_plus    NaN  \n",
       "81   SPAC212.10_plus    NaN  \n",
       "84         dg1_minus    NaN  \n",
       "87          dg1_plus    NaN  \n",
       "90         dh1_minus    NaN  \n",
       "93          dh1_plus    NaN  \n",
       "96              MAT2    NaN  \n",
       "99              MAT3    NaN  \n",
       "102             MAT1    NaN  "
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "repeat_df = repeat_df[repeat_df['Parent'].isnull()]\n",
    "repeat_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(27, 11)"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "repeat_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Get list of **repeat** `gene_id`'s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "repeat_names = list(repeat_df['ID'])\n",
    "#repeat_names = list(repeat_df['Name'])\n",
    "#repeat_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "27"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(repeat_names))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Add `category` column to `gene_df`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_df['category'] = gene_df[['gene_id', 'Name']].apply(lambda row: Util.get_category(row['gene_id'], row['Name'], repeat_names), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>ID</th>\n",
       "      <th>Name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>cds_length</th>\n",
       "      <th>utr_length</th>\n",
       "      <th>bio_type</th>\n",
       "      <th>gene_length</th>\n",
       "      <th>category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>1798347</td>\n",
       "      <td>1799015</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.01</td>\n",
       "      <td>mrx11</td>\n",
       "      <td>SPAC1002.01</td>\n",
       "      <td>540</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>540.0</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>1799061</td>\n",
       "      <td>1800053</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.02</td>\n",
       "      <td>pom34</td>\n",
       "      <td>SPAC1002.02</td>\n",
       "      <td>690</td>\n",
       "      <td>303.0</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>993.0</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>1799915</td>\n",
       "      <td>1803141</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.03c</td>\n",
       "      <td>gls2</td>\n",
       "      <td>SPAC1002.03c</td>\n",
       "      <td>2772</td>\n",
       "      <td>455.0</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>3227.0</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>1803624</td>\n",
       "      <td>1804491</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.04c</td>\n",
       "      <td>taf11</td>\n",
       "      <td>SPAC1002.04c</td>\n",
       "      <td>600</td>\n",
       "      <td>268.0</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>868.0</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>1804548</td>\n",
       "      <td>1806797</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC1002.05c</td>\n",
       "      <td>jmj2</td>\n",
       "      <td>SPAC1002.05c</td>\n",
       "      <td>2148</td>\n",
       "      <td>102.0</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>2250.0</td>\n",
       "      <td>gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7000</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>5726</td>\n",
       "      <td>6331</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>SPAC212.10b</td>\n",
       "      <td>SPAC212.10_plus</td>\n",
       "      <td>SPAC212.10b</td>\n",
       "      <td>606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pseudogenic_transcript</td>\n",
       "      <td>606.0</td>\n",
       "      <td>repeat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7001</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>3761000</td>\n",
       "      <td>3764500</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>dg1b</td>\n",
       "      <td>dg1_minus</td>\n",
       "      <td>dg1b</td>\n",
       "      <td>3501</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>3501.0</td>\n",
       "      <td>repeat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7002</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>3761000</td>\n",
       "      <td>3764500</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>dg1a</td>\n",
       "      <td>dg1_plus</td>\n",
       "      <td>dg1a</td>\n",
       "      <td>3501</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>3501.0</td>\n",
       "      <td>repeat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7003</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>3753000</td>\n",
       "      <td>3757000</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>dh1b</td>\n",
       "      <td>dh1_minus</td>\n",
       "      <td>dh1b</td>\n",
       "      <td>4001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>4001.0</td>\n",
       "      <td>repeat</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7004</th>\n",
       "      <td>I</td>\n",
       "      <td>PomBase</td>\n",
       "      <td>gene</td>\n",
       "      <td>3753000</td>\n",
       "      <td>3757000</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>dh1a</td>\n",
       "      <td>dh1_plus</td>\n",
       "      <td>dh1a</td>\n",
       "      <td>4001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mRNA</td>\n",
       "      <td>4001.0</td>\n",
       "      <td>repeat</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7005 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     seqid   source  type    start      end score strand phase            ID  \\\n",
       "0        I  PomBase  gene  1798347  1799015     .      +     .   SPAC1002.01   \n",
       "1        I  PomBase  gene  1799061  1800053     .      +     .   SPAC1002.02   \n",
       "2        I  PomBase  gene  1799915  1803141     .      -     .  SPAC1002.03c   \n",
       "3        I  PomBase  gene  1803624  1804491     .      -     .  SPAC1002.04c   \n",
       "4        I  PomBase  gene  1804548  1806797     .      -     .  SPAC1002.05c   \n",
       "...    ...      ...   ...      ...      ...   ...    ...   ...           ...   \n",
       "7000     I  PomBase  gene     5726     6331     .      +     .   SPAC212.10b   \n",
       "7001     I  PomBase  gene  3761000  3764500     .      -     .          dg1b   \n",
       "7002     I  PomBase  gene  3761000  3764500     .      +     .          dg1a   \n",
       "7003     I  PomBase  gene  3753000  3757000     .      -     .          dh1b   \n",
       "7004     I  PomBase  gene  3753000  3757000     .      +     .          dh1a   \n",
       "\n",
       "                 Name       gene_id  cds_length  utr_length  \\\n",
       "0               mrx11   SPAC1002.01         540         NaN   \n",
       "1               pom34   SPAC1002.02         690       303.0   \n",
       "2                gls2  SPAC1002.03c        2772       455.0   \n",
       "3               taf11  SPAC1002.04c         600       268.0   \n",
       "4                jmj2  SPAC1002.05c        2148       102.0   \n",
       "...               ...           ...         ...         ...   \n",
       "7000  SPAC212.10_plus   SPAC212.10b         606         NaN   \n",
       "7001        dg1_minus          dg1b        3501         NaN   \n",
       "7002         dg1_plus          dg1a        3501         NaN   \n",
       "7003        dh1_minus          dh1b        4001         NaN   \n",
       "7004         dh1_plus          dh1a        4001         NaN   \n",
       "\n",
       "                    bio_type  gene_length category  \n",
       "0                       mRNA        540.0     gene  \n",
       "1                       mRNA        993.0     gene  \n",
       "2                       mRNA       3227.0     gene  \n",
       "3                       mRNA        868.0     gene  \n",
       "4                       mRNA       2250.0     gene  \n",
       "...                      ...          ...      ...  \n",
       "7000  pseudogenic_transcript        606.0   repeat  \n",
       "7001                    mRNA       3501.0   repeat  \n",
       "7002                    mRNA       3501.0   repeat  \n",
       "7003                    mRNA       4001.0   repeat  \n",
       "7004                    mRNA       4001.0   repeat  \n",
       "\n",
       "[7005 rows x 16 columns]"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7005, 16)"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "#set(gene_df[gene_df['category'] == 'repeat']['ID']).symmetric_difference(repeat_names)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Have a look at distribution of **feature** `category`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>category</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>repeat</th>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ribosomal</th>\n",
       "      <td>177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <td>6804</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             ID\n",
       "category       \n",
       "repeat       24\n",
       "ribosomal   177\n",
       "gene       6804"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_df_by_type = gene_df[['category', 'ID']].groupby(['category']).count()\n",
    "gene_df_by_type.sort_values('ID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7005"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(gene_df_by_type.ID)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Have a look at distribution of **feature** `types`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>type</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <td>7005</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        ID\n",
       "type      \n",
       "gene  7005"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_df_by_type = gene_df[['type', 'ID']].groupby(['type']).count()\n",
    "gene_df_by_type.sort_values('ID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7005"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(gene_df_by_type.ID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "#transcripts_df_by_type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "#gff_by_type.index.difference(gene_df_by_type.index.union(transcripts_df_by_type.index)).tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Store **Gene Table Annotation's** (gdf)\n",
    "\n",
    "These are individual `.csv` **files** that contain information about associated **features**:\n",
    "- **genes**\n",
    "- **exons** (missing, for now use both `cds` and `utr`)\n",
    "- **CDS**\n",
    "- **UTRs**\n",
    "- **Introns** (missing, but might be necesary for further analysis)\n",
    "\n",
    "Besides the `gene_df`, I'm not sure if we need the rest ... At least for the `gene_counts.py` **script** we can directly parse the `gff` **file**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Genes**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Store `gene_df` as `.csv` **file**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_df_file = os.path.join(annot_dir, 'Schizosaccharomyces_pombe_all_chromosomes.extended.gene.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_df.to_csv(gene_df_file, sep='\\t', index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Exons**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Add **gene information** and `biotype` column, and store `exons_df` as `.csv` **file** "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "#exons_df_file = os.path.join(annot_dir, 'Schizosaccharomyces_pombe_all_chromosomes.extended.exon.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "#exons_df.to_csv(exons_df_file, sep='\\t', index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**CDS**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Add **gene information** and `biotype` column, and store `cds_df` as `.csv` **file** "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "#cds_df_file = os.path.join(annot_dir, 'Schizosaccharomyces_pombe_all_chromosomes.extended.cds.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "#cds_df.to_csv(cds_df_file, sep='\\t', index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**UTRs**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Add **gene information** and `biotype` column, and store `cds_df` as `.csv` **file** "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "#utr_df_file = os.path.join(annot_dir, 'Schizosaccharomyces_pombe_all_chromosomes.extended.utr.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "#utr_df.to_csv(utr_df_file, sep='\\t', index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compare to old GFF: `gene lengths` table from .csv file "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Load `gene lengths` **file** as DataFrame (computed for old GFF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "## don't know how this was generated!\n",
    "#in_lengths = os.path.join(annot_dir, 'schizosaccharomyces_pombe.gene_lengths.V2.csv')\n",
    "in_lengths = os.path.join(project_data_dir, 'annotation/gff/schizosaccharomyces_pombe.chr.extended.gene.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "l_df = pd.read_csv(in_lengths, sep='\\t')\n",
    "#l_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "#l_df = l_df[['gene-id', 'length']]\n",
    "l_df = l_df[['gene_id', 'length']]\n",
    "\n",
    "## rename some columns to fit parastous definitions...\n",
    "#l_df = l_df.rename(columns={\"gene-id\": \"gene_id\", \"length\": \"length_parastou\"})\n",
    "l_df = l_df.rename(columns={\"length\": \"length_old\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>length_old</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SPAC212.11</td>\n",
       "      <td>5662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SPAC212.10</td>\n",
       "      <td>606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SPAC212.09c</td>\n",
       "      <td>1656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SPNCRNA.70</td>\n",
       "      <td>530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SPAC212.08c</td>\n",
       "      <td>1211</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_id  length_old\n",
       "0   SPAC212.11        5662\n",
       "1   SPAC212.10         606\n",
       "2  SPAC212.09c        1656\n",
       "3   SPNCRNA.70         530\n",
       "4  SPAC212.08c        1211"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(6992, 2)"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **Merge** with lengths we just computed:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7024, 5)"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l_df = l_df.merge(gene_df[['gene_id', 'cds_length', 'utr_length', 'gene_length']], on='gene_id', how='outer')\n",
    "#l_df = l_df.merge(gene_df[['gene_id', 'length']], on='gene_id', how='outer')\n",
    "l_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Add `difference` **Column**: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "#l_df['difference'] = abs(l_df['length_parastou'] - l_df['length'])\n",
    "l_df['difference'] = abs(l_df['length_old'] - l_df['cds_length'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>length_old</th>\n",
       "      <th>cds_length</th>\n",
       "      <th>utr_length</th>\n",
       "      <th>gene_length</th>\n",
       "      <th>difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SPAC212.11</td>\n",
       "      <td>5662.0</td>\n",
       "      <td>5662.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5662.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SPAC212.10</td>\n",
       "      <td>606.0</td>\n",
       "      <td>606.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>606.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SPAC212.09c</td>\n",
       "      <td>1656.0</td>\n",
       "      <td>1656.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1656.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SPNCRNA.70</td>\n",
       "      <td>530.0</td>\n",
       "      <td>530.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>530.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SPAC212.08c</td>\n",
       "      <td>1211.0</td>\n",
       "      <td>837.0</td>\n",
       "      <td>374.0</td>\n",
       "      <td>1211.0</td>\n",
       "      <td>374.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_id  length_old  cds_length  utr_length  gene_length  difference\n",
       "0   SPAC212.11      5662.0      5662.0         NaN       5662.0         0.0\n",
       "1   SPAC212.10       606.0       606.0         NaN        606.0         0.0\n",
       "2  SPAC212.09c      1656.0      1656.0         NaN       1656.0         0.0\n",
       "3   SPNCRNA.70       530.0       530.0         NaN        530.0         0.0\n",
       "4  SPAC212.08c      1211.0       837.0       374.0       1211.0       374.0"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7024, 6)"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Add `total_difference` **Column**: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "#l_df['difference'] = abs(l_df['length_parastou'] - l_df['length'])\n",
    "l_df['total_difference'] = abs(l_df['length_old'] - (l_df['gene_length'] ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>length_old</th>\n",
       "      <th>cds_length</th>\n",
       "      <th>utr_length</th>\n",
       "      <th>gene_length</th>\n",
       "      <th>difference</th>\n",
       "      <th>total_difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SPAC212.11</td>\n",
       "      <td>5662.0</td>\n",
       "      <td>5662.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5662.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SPAC212.10</td>\n",
       "      <td>606.0</td>\n",
       "      <td>606.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>606.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SPAC212.09c</td>\n",
       "      <td>1656.0</td>\n",
       "      <td>1656.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1656.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SPNCRNA.70</td>\n",
       "      <td>530.0</td>\n",
       "      <td>530.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>530.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SPAC212.08c</td>\n",
       "      <td>1211.0</td>\n",
       "      <td>837.0</td>\n",
       "      <td>374.0</td>\n",
       "      <td>1211.0</td>\n",
       "      <td>374.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_id  length_old  cds_length  utr_length  gene_length  difference  \\\n",
       "0   SPAC212.11      5662.0      5662.0         NaN       5662.0         0.0   \n",
       "1   SPAC212.10       606.0       606.0         NaN        606.0         0.0   \n",
       "2  SPAC212.09c      1656.0      1656.0         NaN       1656.0         0.0   \n",
       "3   SPNCRNA.70       530.0       530.0         NaN        530.0         0.0   \n",
       "4  SPAC212.08c      1211.0       837.0       374.0       1211.0       374.0   \n",
       "\n",
       "   total_difference  \n",
       "0               0.0  \n",
       "1               0.0  \n",
       "2               0.0  \n",
       "3               0.0  \n",
       "4               0.0  "
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7024, 7)"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Investigate differences: they come from the fact that we are not including either, `five_prime_UTR` or `three_prime_UTR`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "big_deviate_genes = ['SPAC16E8.02', 'SPAC24C9.14', 'SPAC2F3.13c', 'SPBC21B10.13c',\n",
    "       'SPBC713.14c', 'SPBP4G3.02', 'SPCC13B11.02c', 'SPCC70.12c',\n",
    "       'SPNCRNA.1306', 'SPNCRNA.390']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>length_old</th>\n",
       "      <th>cds_length</th>\n",
       "      <th>utr_length</th>\n",
       "      <th>gene_length</th>\n",
       "      <th>difference</th>\n",
       "      <th>total_difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6471</th>\n",
       "      <td>SPCC13B11.02c</td>\n",
       "      <td>3823.0</td>\n",
       "      <td>417.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>417.0</td>\n",
       "      <td>3406.0</td>\n",
       "      <td>3406.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1734</th>\n",
       "      <td>SPAC24C9.14</td>\n",
       "      <td>4377.0</td>\n",
       "      <td>990.0</td>\n",
       "      <td>331.0</td>\n",
       "      <td>1321.0</td>\n",
       "      <td>3387.0</td>\n",
       "      <td>3056.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6897</th>\n",
       "      <td>SPCC70.12c</td>\n",
       "      <td>3597.0</td>\n",
       "      <td>243.0</td>\n",
       "      <td>1107.0</td>\n",
       "      <td>1350.0</td>\n",
       "      <td>3354.0</td>\n",
       "      <td>2247.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3603</th>\n",
       "      <td>SPBC713.14c</td>\n",
       "      <td>1596.0</td>\n",
       "      <td>228.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>228.0</td>\n",
       "      <td>1368.0</td>\n",
       "      <td>1368.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1963</th>\n",
       "      <td>SPAC16E8.02</td>\n",
       "      <td>2212.0</td>\n",
       "      <td>669.0</td>\n",
       "      <td>227.0</td>\n",
       "      <td>896.0</td>\n",
       "      <td>1543.0</td>\n",
       "      <td>1316.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4042</th>\n",
       "      <td>SPBC21B10.13c</td>\n",
       "      <td>1995.0</td>\n",
       "      <td>606.0</td>\n",
       "      <td>82.0</td>\n",
       "      <td>688.0</td>\n",
       "      <td>1389.0</td>\n",
       "      <td>1307.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5628</th>\n",
       "      <td>SPBP4G3.02</td>\n",
       "      <td>2681.0</td>\n",
       "      <td>1362.0</td>\n",
       "      <td>186.0</td>\n",
       "      <td>1548.0</td>\n",
       "      <td>1319.0</td>\n",
       "      <td>1133.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3301</th>\n",
       "      <td>SPNCRNA.103</td>\n",
       "      <td>667.0</td>\n",
       "      <td>1562.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1562.0</td>\n",
       "      <td>895.0</td>\n",
       "      <td>895.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3307</th>\n",
       "      <td>SPBC1271.09</td>\n",
       "      <td>1632.0</td>\n",
       "      <td>1587.0</td>\n",
       "      <td>515.0</td>\n",
       "      <td>2102.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>470.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2745</th>\n",
       "      <td>SPAPB8E5.05</td>\n",
       "      <td>769.0</td>\n",
       "      <td>129.0</td>\n",
       "      <td>224.0</td>\n",
       "      <td>353.0</td>\n",
       "      <td>640.0</td>\n",
       "      <td>416.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6635</th>\n",
       "      <td>SPCC1223.10c</td>\n",
       "      <td>1557.0</td>\n",
       "      <td>756.0</td>\n",
       "      <td>428.0</td>\n",
       "      <td>1184.0</td>\n",
       "      <td>801.0</td>\n",
       "      <td>373.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4044</th>\n",
       "      <td>SPBC21B10.11</td>\n",
       "      <td>626.0</td>\n",
       "      <td>219.0</td>\n",
       "      <td>108.0</td>\n",
       "      <td>327.0</td>\n",
       "      <td>407.0</td>\n",
       "      <td>299.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1966</th>\n",
       "      <td>SPAC16E8.05c</td>\n",
       "      <td>1297.0</td>\n",
       "      <td>630.0</td>\n",
       "      <td>407.0</td>\n",
       "      <td>1037.0</td>\n",
       "      <td>667.0</td>\n",
       "      <td>260.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2562</th>\n",
       "      <td>SPAC12B10.09</td>\n",
       "      <td>1257.0</td>\n",
       "      <td>801.0</td>\n",
       "      <td>219.0</td>\n",
       "      <td>1020.0</td>\n",
       "      <td>456.0</td>\n",
       "      <td>237.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4043</th>\n",
       "      <td>SPBC21B10.12</td>\n",
       "      <td>639.0</td>\n",
       "      <td>783.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>873.0</td>\n",
       "      <td>144.0</td>\n",
       "      <td>234.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>752</th>\n",
       "      <td>SPAC6F12.13c</td>\n",
       "      <td>1205.0</td>\n",
       "      <td>1044.0</td>\n",
       "      <td>329.0</td>\n",
       "      <td>1373.0</td>\n",
       "      <td>161.0</td>\n",
       "      <td>168.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5368</th>\n",
       "      <td>SPBC32C12.02</td>\n",
       "      <td>4089.0</td>\n",
       "      <td>1407.0</td>\n",
       "      <td>2590.0</td>\n",
       "      <td>3997.0</td>\n",
       "      <td>2682.0</td>\n",
       "      <td>92.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2459</th>\n",
       "      <td>SPAC29E6.03c</td>\n",
       "      <td>3204.0</td>\n",
       "      <td>3276.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3276.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>72.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>388</th>\n",
       "      <td>SPAC22F3.11c</td>\n",
       "      <td>709.0</td>\n",
       "      <td>522.0</td>\n",
       "      <td>253.0</td>\n",
       "      <td>775.0</td>\n",
       "      <td>187.0</td>\n",
       "      <td>66.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2869</th>\n",
       "      <td>SPAC29A4.23</td>\n",
       "      <td>489.0</td>\n",
       "      <td>549.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>549.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>60.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3815</th>\n",
       "      <td>SPBC3D6.03c</td>\n",
       "      <td>2352.0</td>\n",
       "      <td>2037.0</td>\n",
       "      <td>262.0</td>\n",
       "      <td>2299.0</td>\n",
       "      <td>315.0</td>\n",
       "      <td>53.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4704</th>\n",
       "      <td>SPBC32F12.08c</td>\n",
       "      <td>1104.0</td>\n",
       "      <td>855.0</td>\n",
       "      <td>301.0</td>\n",
       "      <td>1156.0</td>\n",
       "      <td>249.0</td>\n",
       "      <td>52.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1307</th>\n",
       "      <td>SPAC17G8.01c</td>\n",
       "      <td>2650.0</td>\n",
       "      <td>2412.0</td>\n",
       "      <td>286.0</td>\n",
       "      <td>2698.0</td>\n",
       "      <td>238.0</td>\n",
       "      <td>48.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4602</th>\n",
       "      <td>SPBC25B2.07c</td>\n",
       "      <td>1975.0</td>\n",
       "      <td>1506.0</td>\n",
       "      <td>517.0</td>\n",
       "      <td>2023.0</td>\n",
       "      <td>469.0</td>\n",
       "      <td>48.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2222</th>\n",
       "      <td>SPAC2F3.13c</td>\n",
       "      <td>1138.0</td>\n",
       "      <td>1047.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>1093.0</td>\n",
       "      <td>91.0</td>\n",
       "      <td>45.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1769</th>\n",
       "      <td>SPAC688.08</td>\n",
       "      <td>3405.0</td>\n",
       "      <td>3447.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3447.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>42.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1455</th>\n",
       "      <td>SPAC823.04</td>\n",
       "      <td>831.0</td>\n",
       "      <td>840.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>873.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>42.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>793</th>\n",
       "      <td>SPAC3A12.04c</td>\n",
       "      <td>708.0</td>\n",
       "      <td>744.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>744.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4270</th>\n",
       "      <td>SPBC29A3.08</td>\n",
       "      <td>1582.0</td>\n",
       "      <td>792.0</td>\n",
       "      <td>824.0</td>\n",
       "      <td>1616.0</td>\n",
       "      <td>790.0</td>\n",
       "      <td>34.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6288</th>\n",
       "      <td>SPCC4E9.01c</td>\n",
       "      <td>3150.0</td>\n",
       "      <td>2772.0</td>\n",
       "      <td>408.0</td>\n",
       "      <td>3180.0</td>\n",
       "      <td>378.0</td>\n",
       "      <td>30.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3537</th>\n",
       "      <td>SPBC577.05c</td>\n",
       "      <td>572.0</td>\n",
       "      <td>378.0</td>\n",
       "      <td>167.0</td>\n",
       "      <td>545.0</td>\n",
       "      <td>194.0</td>\n",
       "      <td>27.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2223</th>\n",
       "      <td>SPAC2F3.12c</td>\n",
       "      <td>931.0</td>\n",
       "      <td>840.0</td>\n",
       "      <td>118.0</td>\n",
       "      <td>958.0</td>\n",
       "      <td>91.0</td>\n",
       "      <td>27.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1448</th>\n",
       "      <td>SPATRNAILE.02</td>\n",
       "      <td>99.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>74.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>25.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>740</th>\n",
       "      <td>SPATRNAPRO.02</td>\n",
       "      <td>97.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>73.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>24.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5627</th>\n",
       "      <td>SPBC8E4.01c</td>\n",
       "      <td>2076.0</td>\n",
       "      <td>1719.0</td>\n",
       "      <td>381.0</td>\n",
       "      <td>2100.0</td>\n",
       "      <td>357.0</td>\n",
       "      <td>24.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1660</th>\n",
       "      <td>SPAPB24D3.05c</td>\n",
       "      <td>366.0</td>\n",
       "      <td>387.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>387.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>21.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1932</th>\n",
       "      <td>SPATRNAARG.01</td>\n",
       "      <td>90.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>72.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>18.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2013</th>\n",
       "      <td>SPATRNALEU.03</td>\n",
       "      <td>97.0</td>\n",
       "      <td>79.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>79.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>18.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>609</th>\n",
       "      <td>SPATRNASER.01</td>\n",
       "      <td>98.0</td>\n",
       "      <td>82.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>82.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1580</th>\n",
       "      <td>SPATRNALEU.02</td>\n",
       "      <td>98.0</td>\n",
       "      <td>82.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>82.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6526</th>\n",
       "      <td>SPCTRNASER.11</td>\n",
       "      <td>97.0</td>\n",
       "      <td>82.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>82.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3568</th>\n",
       "      <td>SPBC530.13</td>\n",
       "      <td>1442.0</td>\n",
       "      <td>996.0</td>\n",
       "      <td>434.0</td>\n",
       "      <td>1430.0</td>\n",
       "      <td>446.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2679</th>\n",
       "      <td>SPATRNAALA.06</td>\n",
       "      <td>83.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>72.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>971</th>\n",
       "      <td>SPATRNATYR.01</td>\n",
       "      <td>84.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>74.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>846</th>\n",
       "      <td>SPATRNALEU.01</td>\n",
       "      <td>94.0</td>\n",
       "      <td>84.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>84.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>824</th>\n",
       "      <td>SPATRNAMET.02</td>\n",
       "      <td>83.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>74.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2080</th>\n",
       "      <td>SPATRNAVAL.04</td>\n",
       "      <td>83.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>74.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5165</th>\n",
       "      <td>SPBC16D10.10</td>\n",
       "      <td>1170.0</td>\n",
       "      <td>1161.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1161.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4021</th>\n",
       "      <td>SPBTRNAVAL.06</td>\n",
       "      <td>83.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>74.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1942</th>\n",
       "      <td>SPAC3A11.06</td>\n",
       "      <td>2234.0</td>\n",
       "      <td>2004.0</td>\n",
       "      <td>239.0</td>\n",
       "      <td>2243.0</td>\n",
       "      <td>230.0</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1742</th>\n",
       "      <td>SPATRNALYS.05</td>\n",
       "      <td>83.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>75.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>948</th>\n",
       "      <td>SPATRNALYS.01</td>\n",
       "      <td>83.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>75.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1232</th>\n",
       "      <td>SPATRNAVAL.03</td>\n",
       "      <td>82.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>74.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6223</th>\n",
       "      <td>SPCTRNALYS.10</td>\n",
       "      <td>83.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>75.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3490</th>\n",
       "      <td>SPBTRNAMET.04</td>\n",
       "      <td>81.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>74.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2872</th>\n",
       "      <td>SPAC29A4.03c</td>\n",
       "      <td>1402.0</td>\n",
       "      <td>819.0</td>\n",
       "      <td>586.0</td>\n",
       "      <td>1405.0</td>\n",
       "      <td>583.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4305</th>\n",
       "      <td>SPBC23G7.06c</td>\n",
       "      <td>2639.0</td>\n",
       "      <td>2340.0</td>\n",
       "      <td>302.0</td>\n",
       "      <td>2642.0</td>\n",
       "      <td>299.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4651</th>\n",
       "      <td>SPBC4F6.10</td>\n",
       "      <td>4089.0</td>\n",
       "      <td>1719.0</td>\n",
       "      <td>2373.0</td>\n",
       "      <td>4092.0</td>\n",
       "      <td>2370.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4213</th>\n",
       "      <td>SPBC16E9.16c</td>\n",
       "      <td>2268.0</td>\n",
       "      <td>2271.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2271.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4376</th>\n",
       "      <td>SPBC14C8.09c</td>\n",
       "      <td>891.0</td>\n",
       "      <td>894.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>894.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene_id  length_old  cds_length  utr_length  gene_length  \\\n",
       "6471  SPCC13B11.02c      3823.0       417.0         NaN        417.0   \n",
       "1734    SPAC24C9.14      4377.0       990.0       331.0       1321.0   \n",
       "6897     SPCC70.12c      3597.0       243.0      1107.0       1350.0   \n",
       "3603    SPBC713.14c      1596.0       228.0         NaN        228.0   \n",
       "1963    SPAC16E8.02      2212.0       669.0       227.0        896.0   \n",
       "4042  SPBC21B10.13c      1995.0       606.0        82.0        688.0   \n",
       "5628     SPBP4G3.02      2681.0      1362.0       186.0       1548.0   \n",
       "3301    SPNCRNA.103       667.0      1562.0         NaN       1562.0   \n",
       "3307    SPBC1271.09      1632.0      1587.0       515.0       2102.0   \n",
       "2745    SPAPB8E5.05       769.0       129.0       224.0        353.0   \n",
       "6635   SPCC1223.10c      1557.0       756.0       428.0       1184.0   \n",
       "4044   SPBC21B10.11       626.0       219.0       108.0        327.0   \n",
       "1966   SPAC16E8.05c      1297.0       630.0       407.0       1037.0   \n",
       "2562   SPAC12B10.09      1257.0       801.0       219.0       1020.0   \n",
       "4043   SPBC21B10.12       639.0       783.0        90.0        873.0   \n",
       "752    SPAC6F12.13c      1205.0      1044.0       329.0       1373.0   \n",
       "5368   SPBC32C12.02      4089.0      1407.0      2590.0       3997.0   \n",
       "2459   SPAC29E6.03c      3204.0      3276.0         NaN       3276.0   \n",
       "388    SPAC22F3.11c       709.0       522.0       253.0        775.0   \n",
       "2869    SPAC29A4.23       489.0       549.0         NaN        549.0   \n",
       "3815    SPBC3D6.03c      2352.0      2037.0       262.0       2299.0   \n",
       "4704  SPBC32F12.08c      1104.0       855.0       301.0       1156.0   \n",
       "1307   SPAC17G8.01c      2650.0      2412.0       286.0       2698.0   \n",
       "4602   SPBC25B2.07c      1975.0      1506.0       517.0       2023.0   \n",
       "2222    SPAC2F3.13c      1138.0      1047.0        46.0       1093.0   \n",
       "1769     SPAC688.08      3405.0      3447.0         NaN       3447.0   \n",
       "1455     SPAC823.04       831.0       840.0        33.0        873.0   \n",
       "793    SPAC3A12.04c       708.0       744.0         NaN        744.0   \n",
       "4270    SPBC29A3.08      1582.0       792.0       824.0       1616.0   \n",
       "6288    SPCC4E9.01c      3150.0      2772.0       408.0       3180.0   \n",
       "3537    SPBC577.05c       572.0       378.0       167.0        545.0   \n",
       "2223    SPAC2F3.12c       931.0       840.0       118.0        958.0   \n",
       "1448  SPATRNAILE.02        99.0        74.0         NaN         74.0   \n",
       "740   SPATRNAPRO.02        97.0        73.0         NaN         73.0   \n",
       "5627    SPBC8E4.01c      2076.0      1719.0       381.0       2100.0   \n",
       "1660  SPAPB24D3.05c       366.0       387.0         NaN        387.0   \n",
       "1932  SPATRNAARG.01        90.0        72.0         NaN         72.0   \n",
       "2013  SPATRNALEU.03        97.0        79.0         NaN         79.0   \n",
       "609   SPATRNASER.01        98.0        82.0         NaN         82.0   \n",
       "1580  SPATRNALEU.02        98.0        82.0         NaN         82.0   \n",
       "6526  SPCTRNASER.11        97.0        82.0         NaN         82.0   \n",
       "3568     SPBC530.13      1442.0       996.0       434.0       1430.0   \n",
       "2679  SPATRNAALA.06        83.0        72.0         NaN         72.0   \n",
       "971   SPATRNATYR.01        84.0        74.0         NaN         74.0   \n",
       "846   SPATRNALEU.01        94.0        84.0         NaN         84.0   \n",
       "824   SPATRNAMET.02        83.0        74.0         NaN         74.0   \n",
       "2080  SPATRNAVAL.04        83.0        74.0         NaN         74.0   \n",
       "5165   SPBC16D10.10      1170.0      1161.0         NaN       1161.0   \n",
       "4021  SPBTRNAVAL.06        83.0        74.0         NaN         74.0   \n",
       "1942    SPAC3A11.06      2234.0      2004.0       239.0       2243.0   \n",
       "1742  SPATRNALYS.05        83.0        75.0         NaN         75.0   \n",
       "948   SPATRNALYS.01        83.0        75.0         NaN         75.0   \n",
       "1232  SPATRNAVAL.03        82.0        74.0         NaN         74.0   \n",
       "6223  SPCTRNALYS.10        83.0        75.0         NaN         75.0   \n",
       "3490  SPBTRNAMET.04        81.0        74.0         NaN         74.0   \n",
       "2872   SPAC29A4.03c      1402.0       819.0       586.0       1405.0   \n",
       "4305   SPBC23G7.06c      2639.0      2340.0       302.0       2642.0   \n",
       "4651     SPBC4F6.10      4089.0      1719.0      2373.0       4092.0   \n",
       "4213   SPBC16E9.16c      2268.0      2271.0         NaN       2271.0   \n",
       "4376   SPBC14C8.09c       891.0       894.0         NaN        894.0   \n",
       "\n",
       "      difference  total_difference  \n",
       "6471      3406.0            3406.0  \n",
       "1734      3387.0            3056.0  \n",
       "6897      3354.0            2247.0  \n",
       "3603      1368.0            1368.0  \n",
       "1963      1543.0            1316.0  \n",
       "4042      1389.0            1307.0  \n",
       "5628      1319.0            1133.0  \n",
       "3301       895.0             895.0  \n",
       "3307        45.0             470.0  \n",
       "2745       640.0             416.0  \n",
       "6635       801.0             373.0  \n",
       "4044       407.0             299.0  \n",
       "1966       667.0             260.0  \n",
       "2562       456.0             237.0  \n",
       "4043       144.0             234.0  \n",
       "752        161.0             168.0  \n",
       "5368      2682.0              92.0  \n",
       "2459        72.0              72.0  \n",
       "388        187.0              66.0  \n",
       "2869        60.0              60.0  \n",
       "3815       315.0              53.0  \n",
       "4704       249.0              52.0  \n",
       "1307       238.0              48.0  \n",
       "4602       469.0              48.0  \n",
       "2222        91.0              45.0  \n",
       "1769        42.0              42.0  \n",
       "1455         9.0              42.0  \n",
       "793         36.0              36.0  \n",
       "4270       790.0              34.0  \n",
       "6288       378.0              30.0  \n",
       "3537       194.0              27.0  \n",
       "2223        91.0              27.0  \n",
       "1448        25.0              25.0  \n",
       "740         24.0              24.0  \n",
       "5627       357.0              24.0  \n",
       "1660        21.0              21.0  \n",
       "1932        18.0              18.0  \n",
       "2013        18.0              18.0  \n",
       "609         16.0              16.0  \n",
       "1580        16.0              16.0  \n",
       "6526        15.0              15.0  \n",
       "3568       446.0              12.0  \n",
       "2679        11.0              11.0  \n",
       "971         10.0              10.0  \n",
       "846         10.0              10.0  \n",
       "824          9.0               9.0  \n",
       "2080         9.0               9.0  \n",
       "5165         9.0               9.0  \n",
       "4021         9.0               9.0  \n",
       "1942       230.0               9.0  \n",
       "1742         8.0               8.0  \n",
       "948          8.0               8.0  \n",
       "1232         8.0               8.0  \n",
       "6223         8.0               8.0  \n",
       "3490         7.0               7.0  \n",
       "2872       583.0               3.0  \n",
       "4305       299.0               3.0  \n",
       "4651      2370.0               3.0  \n",
       "4213         3.0               3.0  \n",
       "4376         3.0               3.0  "
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#l_df.sort_values('difference', ascending=False)\n",
    "l_df.sort_values('total_difference', ascending=False).head(60)\n",
    "#l_df[l_df['gene_id'].isin(big_deviate_genes)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_genes = ['SPNCRNA.114']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>length_old</th>\n",
       "      <th>cds_length</th>\n",
       "      <th>utr_length</th>\n",
       "      <th>gene_length</th>\n",
       "      <th>difference</th>\n",
       "      <th>total_difference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5535</th>\n",
       "      <td>SPNCRNA.114</td>\n",
       "      <td>76.0</td>\n",
       "      <td>76.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>76.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene_id  length_old  cds_length  utr_length  gene_length  \\\n",
       "5535  SPNCRNA.114        76.0        76.0         NaN         76.0   \n",
       "\n",
       "      difference  total_difference  \n",
       "5535         0.0               0.0  "
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l_df[l_df['gene_id'].isin(selected_genes)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "#min(l_df['difference'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "#max(l_df['difference'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "#l_df['difference'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Differences arise due to `+ 1` factor for each `exon` present in the `transcript` (also Parastou is missing a couple of genes, `nans` account for the small differences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "#n_exons_per_transcript = exon_df[['exon_id', 'transcript_id']].groupby('transcript_id').count()['exon_id'].value_counts()\n",
    "#n_exons_per_transcript"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "#sum(n_exons_per_transcript)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Visualize `difference`: `scatter_plot`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7fa10b807b38>"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZIAAAEHCAYAAACEKcAKAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8li6FKAAAgAElEQVR4nO3deZhU5Zn38e9dSzcgi4gba0BBHTC4MYrBZIwmaoxRZ9yIGs3EhExeE81rFHHymmiczLgkmsVlYtCJJkZlMAkoMW5ojBvamgYBt1ZUukFFRASB3up+/zinmurq6u5au6q7f5/rKrvqOefUeRqh7nq2+zF3R0REJF+RcldARER6NwUSEREpiAKJiIgURIFEREQKokAiIiIFiZW7AuWw8847+/jx48tdDRGRXuX5559/3913SS/vl4Fk/Pjx1NTUlLsaIiK9ipm9lalcXVsiIlIQBRIRESmIAomIiBREgURERAqiQCIiIgVRIBER6SfWb25k6eoPWb+5sajv2y+n/4qI9DcLahu4+J5lxCMRmhMJrj5pKsfvP7oo760WiYhIH7d+cyMX37OMbc0JNjW2sK05wex7lhWtZaJAIiLSx9Vv2Eo80v7jPh6JUL9ha1HeX4FERKSPGzN8IM2JRLuy5kSCMcMHFuX9FUhERPq4EYOrufqkqQyIRxhSHWNAPMLVJ01lxODqory/BttFRPqB4/cfzYyJO1O/YStjhg8sWhABBRIRkX5jxODqogaQJHVtiYhIQRRIRER6mVItLMxXSQOJmd1qZu+Z2fIMxy40MzeznVPKLjGzOjN7xcyOTik/yMxeDI/9wswsLK82s7vD8iVmNr6Uv4+ISLktqG1gxlWLOXPuEmZctZiFtQ3lrlLJWyS/AY5JLzSzscDngbdTyiYDM4Ep4TU3mlk0PHwTMAuYFD6S73kOsMHdJwLXAVeV5LcQEakApV5YmK+SBhJ3fxz4IMOh64DZgKeUnQDc5e6N7r4KqAMONrORwFB3f9rdHbgdODHlmtvC5/OBI5OtFRGRvqbUCwvz1eNjJGZ2PNDg7kvTDo0GVqe8rg/LRofP08vbXePuLcBGYEQJqi0iUnalXliYrx4NJGY2CPg+8INMhzOUeRflXV2T6d6zzKzGzGrWrVuXTXVFRCpKqRcW5qun15HsCUwAloY9UGOAF8zsYIKWxtiUc8cAa8LyMRnKSbmm3sxiwDAyd6Xh7jcDNwNMmzYtY7AREal0pVxYmK8ebZG4+4vuvqu7j3f38QSB4EB3fwdYCMwMZ2JNIBhUf9bd1wKbzGx6OP5xFrAgfMuFwNnh85OBxeE4iohInzVicDX7jd2xIoIIlH76753A08DeZlZvZud0dq67rwDmASuBvwDnuntrePhbwFyCAfjXgfvD8luAEWZWB1wAzCnJLyIiIp2y/vgFftq0aV5TU1PuaoiI9Cpm9ry7T0sv18p2EREpiAKJiIgURIFEREQKokAiIiIFUSAREZGCKJCIiEhBFEhERKQgCiQiIlIQBRIRESmIAomIiBREgURERAqiQCIiUoD1mxtZuvrDsm93W049vR+JiEifsaC2gYvvWUY8EqE5keDqk6Zy/P6ju7+wj1GLREQkD+s3N3LxPcvY1pxgU2ML25oTzL5nWb9smSiQiIjkoX7DVuKR9h+h8UiE+g1by1Sj8lEgERHJw5jhA2lOJNqVNScSjBk+sEw1Kh8FEhGRPIwYXM3VJ01lQDzCkOoYA+IRrj5pasVsf9uTNNguIpKn4/cfzYyJO1O/YStjhg/sl0EESr9n+61m9p6ZLU8pu8bMXjazZWb2RzPbMeXYJWZWZ2avmNnRKeUHmdmL4bFfmJmF5dVmdndYvsTMxpfy9xERSTdicDX7jd2x3wYRKH3X1m+AY9LKHgL2dfepwKvAJQBmNhmYCUwJr7nRzKLhNTcBs4BJ4SP5nucAG9x9InAdcFXJfhMREcmopIHE3R8HPkgre9DdW8KXzwBjwucnAHe5e6O7rwLqgIPNbCQw1N2fdncHbgdOTLnmtvD5fODIZGtFRCQXWliYv3KPkXwNuDt8PpogsCTVh2XN4fP08uQ1qwHcvcXMNgIjgPdLWGcR6WO0sLAwZZu1ZWbfB1qAO5JFGU7zLsq7uibT/WaZWY2Z1axbty7X6opIH6WFhYUrSyAxs7OB44Azwu4qCFoaY1NOGwOsCcvHZChvd42ZxYBhpHWlJbn7ze4+zd2n7bLLLsX6VUSkl9PCwsL1eCAxs2OAi4Hj3X1LyqGFwMxwJtYEgkH1Z919LbDJzKaH4x9nAQtSrjk7fH4ysDglMImIdEsLCwtX6um/dwJPA3ubWb2ZnQNcDwwBHjKzWjP7bwB3XwHMA1YCfwHOdffW8K2+BcwlGIB/Hbg/LL8FGGFmdcAFwJxS/j4i0vdoYWHhrD9+gZ82bZrX1NSUuxoiUkHWb27s9wsLu2Nmz7v7tPTycs/aEhGpCCMGVyuA5Em5tkREpCAKJCIiUhAFEhERKYgCiYj0OkpnUlk02C4ivUqmdCZK5V5eCiQi0mukpjPZRrCI8Hv/u5SIQVU0qjxZZaKuLRHpNTKlM2ludRpbXHmyykiBRER6jUzpTNIpT1bPUyARkV4jPZ1JdSxCLO1TTHmyep7GSESkV0nfJ/3JuveZnTb4rgH3nqVAIiK9Tmo6k/TAoiDS8xRIRKTXU56s8tIYiYiIFESBRERECqJAIiIiBVEgERGRgiiQiIhIQUq9Z/utZvaemS1PKdvJzB4ys9fCn8NTjl1iZnVm9oqZHZ1SfpCZvRge+4WZWVhebWZ3h+VLzGx8KX8fERHpqNQtkt8Ax6SVzQEecfdJwCPha8xsMjATmBJec6OZRcNrbgJmAZPCR/I9zwE2uPtE4DrgqpL9JiIiklFJA4m7Pw58kFZ8AnBb+Pw24MSU8rvcvdHdVwF1wMFmNhIY6u5Pu7sDt6ddk3yv+cCRydaKiIj0jHKMkezm7msBwp+7huWjgdUp59WHZaPD5+nl7a5x9xZgIzAi003NbJaZ1ZhZzbp164r0q4iISCUNtmdqSXgX5V1d07HQ/WZ3n+bu03bZZZc8qygiIunKEUjeDburCH++F5bXA2NTzhsDrAnLx2Qob3eNmcWAYXTsShORMtB2uP1HOQLJQuDs8PnZwIKU8pnhTKwJBIPqz4bdX5vMbHo4/nFW2jXJ9zoZWByOo4hIGS2obWDGVYs5c+4SZly1mIW1DeWukpRQSZM2mtmdwOHAzmZWD/wQuBKYZ2bnAG8DpwC4+wozmwesBFqAc929NXyrbxHMABsI3B8+AG4BfmtmdQQtkZml/H1EpHuZtsOdfc8yZkzcWYkV+6iSBhJ3/3Inh47s5PwfAz/OUF4D7JuhfBthIBKRypDcDjcZRGD7roUKJH1TJQ22i0gfkGk7XO1a2LcpkIhIUaVvhzsgHtGuhX2cNrYSkYKs39zYYXdC7VrYvyiQiEjeFtQ2cHHafunH7x+sF9auhf2HurZEJC+ps7M2NbawrTnB7HuWad1IP5RTi8TMRgOfSL0uzKclIr1Upq6pbGh2liRlHUjM7CrgNIJ1Hsn1HQ4okIj0Ul11TXVHs7MkKZcWyYnA3u6udqtIH1DowsHk7KzZaYFIrZH+J5dA8gYQBxRIRPqAYnRNaXaWQBaBxMx+SdCFtQWoNbNHSAkm7n5e6aonIqVSrK4pzc6SbFokNeHP5wmSJKZSgkSRXkpdU1Is3QYSd78NwMzOd/efpx4zs/NLVTERKT11TUkx5LKO5OwMZV8tUj1ERKSXymaM5MvA6cAEM0vt2hoCrC9VxUSk9AqZ/iuSlM0YyVPAWmBn4Kcp5ZuAZaWolIiUnvYNkWLJZozkLeAt4NDSV0dEeopWpkux5LKyfRMdZ2ltJJjV9T13f6OYFROR0tLKdCmWXAbbrwUuAkYDY4ALgV8DdwG3Fr9qIlKI9ZsbWbr6w06TKGrfECkWc89uKYiZLXH3Q9LKnnH36Wa21N33y+nGZv8X+DpBK+dF4F+BQcDdwHjgTeBUd98Qnn8JcA5Bnq/z3P2BsPwgtu/n/mfgfO/ml5o2bZrX1NR0dYpIr9BZwsVcBtHzTdoo/Y+ZPe/u09LLc2mRJMzsVDOLhI9TU47ltDAxzCJ8HjDN3fcFosBMYA7wiLtPAh4JX2Nmk8PjU4BjgBvNLBq+3U3ALGBS+Dgml7qI9FYLahuYcdVizpy7hBlXLWZhbQOQe3r3EYOr2W/sjgoikrdcAskZwFeA94B3w+dnmtlA4Nt53DsGDDSzGEFLZA1wAnBbePw2gkSRhOV3uXuju68C6oCDzWwkMNTdnw5bIbenXCPSZ3UVLJKD6KmSg+gipZD1YHs4mP6lTg4/kctN3b3BzH4CvA1sBR509wfNbDd3Xxues9bMdg0vGQ08k/IW9WFZc/g8vbwDM5tF0HJh3LhxuVRXpOJ0NeNKg+jS07JukZjZLmb272Z2s5ndmnzkc1MzG07QypgAjAJ2MLMzu7okQ5l3Ud6x0P1md5/m7tN22WWXXKssUlHGDB/I1uaWdmVbm1vaxjk0iC49KZc08guAvwEPs31jq3x9Dljl7usAzOwPwKeAd81sZNgaGUnQjQZBS2NsyvVjCLrC6sPn6eUifZ6Zkfq9KXgdUA4t6Um5BJJB7n5xke77NjDdzAYRdG0dSbAe5WOCnF5Xhj8XhOcvBH5vZtcStGAmAc+6e6uZbTKz6cAS4Czgl0Wqo0jFqt+wlQGxKM2t21slA2LRdosJld5dekougeQ+MzvW3f9c6E3dfYmZzQdeAFqAvwM3A4OBeWZ2DkGwOSU8f4WZzSPY5rcFONfdk62ib7F9+u/94UOkV8p2Kq7GQaSS5LKOZBOwA9AUPgxwdx9auuqVhtaRSCXKNYHiwtqGDnuJKOGilFJn60hymbU1pLhVEpGkfBIoahxEKkUus7bMzM40s0vD12PN7ODSVU2k/8h37YcWE0olyGVB4o0EGYBPD19vBm4oeo1E+qEdqqI0tpZuzKO7vFsihchlsP0Qdz/QzP4O4O4bzKyqRPUS6fOSA+vLGzZyxaKVWDheOSAefL8r1tqPStu8Srm9+p5cAklzmN/KIVigCCS6vkREMrnjmbe4/N4VxCLGlub2/4wSCefP532aibt1PSyZzQdypW1eVWlBTYojl0DyC+CPwK5m9mPgZOD/laRWIn3YHc+8xff/tByAptaOsyarY1E+bup6zW+2H8iVtHlVpQU1KZ5cZm3dYWbPEyweNOBEd3+pZDUT6YPWb27k8vtWdnnOlqaWLsdGcvlArqT1JpUU1KS4uh1sN7Odkg+ClCV3Ar8nSGeyU6krKNKXBB+mmVLEpejmcC4zvCop71YlBTUprmxaJM/TPkFisi2eTPSzRwnqJdInpI9jLG/Y2G23VXqqk3S5fiBXynqTZFBLX0Sp1kjv120gcfcJ2byRmU1x9xWFV0mkb0gfx7j0i5O5YlHX3VoAre5dfkvP5wO5UvJuVUpQk+LKOkVKt29k9oK7H1iUNysxpUiRUlu/uZEZVy1mW3PqeABUx6NsbszcIqmKGpGIZT2TSdNopacVnCIlm3sU8b1EerUVaz4ikvZPojkBLRmCSDxqXHTU3hyyx4icgkKltDJEihlIitO0EenlFtQ2MHv+MhpbOi6zSv9HEosY92exZkSkkuWSIkVEupGcmpspiGQyMN79mhGRSlfMQNJUxPcSqVjpeatSX2eamtuV1NlWyoclvVXWXVtmNgOodfePw/3VDwR+7u5vAbj79BLVUaRipM/EOnXaGObV1LebmZU+NRdgh+oorQnvcH5ytpVSh0hvlsvGVsuA/YCpwG+BW4B/cfd/Kl31SkOztiQfmWZipRsQj3DpcZP54YLlJHu3YhE4/8i9OP2QcYwYXN1htlWm9x0Qj/DkxUdoMF0qSmeztnLp2mrxIOqcQNAS+TmgEULpN7LptvKEM3b4IKIp57Uk4IbH6tpep+8hku9eJCKVIpdAssnMLgHOBBaFmYDj+d7YzHY0s/lm9rKZvWRmh4apWB4ys9fCn8NTzr/EzOrM7BUzOzql/CAzezE89gsz0zRkKYlMK8rTNbY6H21tpiqafWBQ6hDp7XIJJKcBjcA57v4OMBq4poB7/xz4i7vvQ9Bl9hIwB3jE3ScBj4SvMbPJwExgCnAMcGMYyABuAmYBk8LHMQXUSaRTqXmrqqOZv68MiEcYOjCeU2CopHxYIvko2sr2nG5qNhRYCuzhKRUws1eAw919rZmNBB5z973DlhDu/l/heQ8AlwFvAo+GwQgz+3J4/Te7ur/GSKQrmcYw6jdsZYeqKGs2bqVhw1Z+sHAFzRlSwCfHNp6se79DCpPuBs+7WqmuVexSCfJe2W5mm+hisaG7D82jPnsA64D/MbP9CBJDng/s5u5rw/dda2a7huePBp5Jub4+LGsOn6eXZ/o9ZhG0XBg3blweVZb+4I5n3uKye5cTswgJnNP+cSzzaurxhNOYIXCkqopaW0sin5xSna1UTy5wjEaM1oRzzclBUFJwkUqRTdLGIQBm9iPgHYIZWwacQf6D7TGC6cPfcfclZvZzwm6sTmTqR/AuyjsWut8M3AxBiyS36kp/kLrhVHO4Z8btT7+d1bVVsQh//s5h7VaoFyOFyfrNjVz4v0vbtX6+979L2bSthSsWrdR0YakIuYyRHO3uN7r7Jnf/yN1vAk7K8771QL27LwlfzycILO+GXVqEP99LOX9syvVjgDVh+ZgM5SKdyrTwb/3mRi67N/fk1YPiUQbEI/zk5KklSXOyYs3GDl1oza3OZfcuZ1tzgk2NLWxrTjD7nmVayChlk0uurVYzOwO4i+Bb/5eBvHI7uPs7ZrbazPZ291cIdl1cGT7OBq4Mfy4IL1kI/N7MrgVGEQyqP+vurWa2ycymA0uAs4Bf5lMn6R9SF/41tbby7c9O4vRDxjH3b29kHPPoSlXU+O+vHMSUUUNL2LWUeVA/ZpG2VhNop0Epr1wCyekEM61+ThBIniQIJvn6DnCHmVUBbwD/StBCmmdm5wBvA6cAuPsKM5tHEGhagHPdPRnEvgX8BhgI3B8+RDrItEXtTx96lZ89/CrdxZC4QXPaOQl3PtzSlPWHdz5jGlNGDSUWgdTUXVGDRFoPrqYLSznlEkguB77q7hsg2IIX+AnwtXxu7O61QIfRf4LWSabzfwz8OEN5DbBvPnWQ/iH5Ab5xazPRDMuMsmmIHH/AaL78j2P58txnaA6/wrQkOt8rPV2+KVBGDK7m2lP356L5S4lahFZPcM3J+wFop0GpGLkEkqnJIALg7h+Y2QElqJNIwZLBY3nDxrZB6S1NLVkFjUzuW7aWE/YfzYBYjObWlrbybLqUMrWEsg1A0PmugtppUCpFLoEkYmbD01okxdzPRKQokt/+o2ZtKdq3kV1a984EK9U9rxXoyRQo2woY08g0A0wbW0mlyGXW1k+Bp8zsinAq8FPA1aWplkh+Ur/957rPx6B4hKqo8e9f2IfqWPt/Gs2JBFNGDWtbgb5DVZSqqHHpFyd3+2GuFCjS12XdonD3282sBjiCYCrJv7j7ypLVTCQHde9uonb1hwyIR4lFcku3FosYl58whX1HDWvrJtp92ICMYxDH7z+aTdtauPy+lVTFIlyxaCVDBsS6HO9IpkDRmIb0VWVJkVJuSpHSt/zgTy9y+zPbFw5GDBJZ/LWOGHzzn/bg64ftkfFDPdMsq0JSvhdrJbpWtEu55J0iRaSS1axa3y6IQHZBJArEosY/7N75GpBMYxCFjHcUY0xDG2BJJdKe7dJrLaht4LRfL+n+xAxagcYWZ/Y9y6h7d1PWW9yWc7wjdfxHK9qlkqhFIr1Kaibe2fOX0ZpN86MLnnCO/eUTVEez+4ZfzvGOYsz+EikFBRLpFere3cStT77J/OffJh6NsqWptfOU1DkIMvo6TS3Zr+/IJ7NvMWj2l1QqBRKpeOmD6U2teaV46yAegWg00m7gvBjjHaUaDNfsL6lUCiRSsdZvbuTBFe90GEwvhqjBnd+Yzpm3PtuuPPUbfj4BodSD4eVqDYl0RYFEKtKC2ga+N6+2XbLCYhpUFSMei3b6DT+fgFBoKpRsaUW7VBoFEqko6zc3smLNRi4sMIgcNG4YK9ZuIh6J0NjSQsKNlpSB+WTLY7+xO3b4hp9rQNieFLJJg+HSLymQSEVYv7mRuX97g1ueWAVAcwFBxICzPzWBySOH8kTd++w8uJqNW5s77CiY/HBP/4afy+yo9P1N0ieRaTBc+gMFEim71C1ui8GBC+bV4r49RXw8alx2fPs0KJ3JdnZUppZLLALVsQhVUQ2GS/+hQCJl9au/vs5/3f9y0d83vVusudX50b0reWpO96lMsp0dlanlMjAe44YzDmTYwLgGw6XfUCCRsrnjmbdKEkQ6E40Y9Ru2AmTMoZVals3sqM5aLqXdelek8iiQSI+qe3cTT9StozoW5dIFxevOykZrwlnesJHTbn66XUvDocMMrWym2Gpdh0igrNl/zSwK1AAN7n5cuFnW3cB44E3g1JSNtC4BziFIk3Seuz8Qlh/E9j3b/wyc7938Usr+Wx7n3/kCC5auLcu941HjoqP25icPvdq2ih2gOmaA0ZhSFo8aEYOqaDSrqb+FLkBUNl/pLSo1++/5wEvA0PD1HOARd7/SzOaEry82s8nATGAKMAp42Mz2cvdW4CZgFvAMQSA5Bri/Z38N6c43f/scD6x4r0fvWR2L8NNT9mPowBirP9jKZfeuoDltr92oRYJpXimS5zS2BFvqdrcWpJB1HcrmK31B2bL/mtkY4IvA3JTiE4Dbwue3ASemlN/l7o3uvgqoAw42s5HAUHd/OmyF3J5yjVSAmlXr+ZcbnujxIBKPGtecPJVD9xwBwA8XLu8QRCAY0+gu8WNy6m+xKZuv9BXlbJH8DJgNDEkp283d1wK4+1oz2zUsH03Q4kiqD8uaw+fp5R2Y2SyClgvjxo0rRv2lG8f+7K+sfGdzj983FoH7z/s0K9Z+xIyrFgMdZ3ElRYDWRIJ41BgQi9LUmqA1kWh3fqnWgiibr/QVZWmRmNlxwHvu/ny2l2Qo8y7KOxa63+zu09x92i677JLlbSUfde9u4oDLHyhpEKmKGtWxCF/cd/cOxwbGY6zZuI3Z88M1Hl2sbmxsdVoSwW6JN5xxAE/NOYJrT92fAfEIQ6pjDIhH2gbQ129uzHrfkmwom6/0FeVqkcwAjjezY4EBwFAz+x3wrpmNDFsjI4Fkf0g9MDbl+jHAmrB8TIZyKZN/++1z/KXE3VjxCPznP3+Sz+4TNFgffvldGlvapz95+vX32w2gd6cqGmXYwKpOp/6WYixDs76kryhLIHH3S4BLAMzscOBCdz/TzK4BzgauDH8uCC9ZCPzezK4lGGyfBDzr7q1mtsnMpgNLgLOAX/boLyNA0Ar5yq1LWLux9P37kUiEz+6za9sH7g++NIXL711JPGq0JpxLj5vMj+5d0en1sbAd3lX3VeoAeimTMSqbr/QF5Z61le5KYJ6ZnQO8DZwC4O4rzGwesBJoAc4NZ2wBfIvt03/vRzO2esz6zY08/fp6rn/kVV5+7+Meu++3Pzux7QN3QW0DV9y3knjEaG5J8MMvBWlQqqLRtllXqSIG3/zMHgwbWMW1D7/aliPr3MMndnq/Uo9lKJuv9HZlXUdSLlpHUrgFtQ2cf1dtj983FoG/nP8ZJu42hPWbG5lx1eJ2YyAD4hHu+/ZhHHf9E12OjQCcOm0MY4YP4oZH69rlxkrvsursPk9e3H26FZG+pLN1JGWb/iu91yMr3ylLEAEwM467/gkW1ja0tRRSxSMRPm5q5eqTprYNmFdFM/9Fn1dTz/WLX6Wxpevpt8mxjOpYhEFVUapjEY1liKSotK4tqWDrNzdy6q+e4vV1W8pWh+ZWp7nVuWj+Mn56ytQO2+5m2mdk49YmvnFbTbg/ezojdaJfZ11Wnvyvtz9fRBRIJEtfvWUJj732fo/eM2oQMaM5w4LBxpYEs+e/SMKD7q6B8RjNiQSXfnFy2+LB5NjDHc+81UkQgaa08q7SxQczw4LAVYqdD0V6KwUS6db4OYvKct94LMLXZoznxsfeyHh8S3PwoV4di3DDGQey+oMtHTavmjFxZ65YtDLj9ZnaFhd8fq92gQi0cFCkOxojkU49svKdsgURgFMPGsPcv2UOIqmqohHAuWLRyg7pRlas2dhhHKU6alx01F4Mrm7/Pao6FuEnD7zCmXOXMOOqxSysbQC0cFCkOwokktHh1yzmnNuzTTxQGnc9t5qm1u7Pa2pNAIandYEFr61DELCIcfSU3TuUN7YkaGr1DgPvycH2TKvdRURdW5KmZtV6vvnb51i/JYtP8FLLcmp6ayLBy2s/6jAO0tjqDIpHOPfwiVz/6Gvt0sJP3G0IV580lYvmLyVqQVdYJLwmKbX7qtQLB5VKXnozBRJpc/jVj/DmB9vKXY02TVlmOGlJwDUPvkxV1NoNnscicPotz1IdjQDGrM/swemHjGv7oA7ONAgH9Vu76b4q1cJBpZKX3k5dW0LNqvWMn7OoooJIrppbO87AaklAU7hGpLElwQ2P1bUd2z4TK8GWplYaWxKYGdUx69HuK6WSl75ALZJ+7lP/+RBrPmoqdzWKLtjlMG3nw5SuqkwzsQbEotxwxgEMG1jVY11MmhEmfYFaJP3Un15Yzfg5i3p9EIlm2EhgYCzCUf+wa4fsv6ldVZ3NxJoyahj7jd2xxz7ENSNM+gIFkn7okz+4n+/OW1buahRFJGJUpUWTrS0JFi1/t8O5lx43uS1AVMpMrEqph0gh1LXVj/zphdV9JoAkDYhFmfWZPbjhsTqiZnzcyXzhHaqi7DtqWLuySknhXin1EMmXAkk/UPfuJj5/3eO9JkPUsZ/cjcUvr+syMCQ1tSb4wr67M2HnQSyt38jvnn6LrRk2tGpuzdxdVCkp3CulHiL5UCDp4477xeMsX7Op3NXIyYw9d+GCz+3NwqVrmPu3VW2pUDJpaklw1M8eJ0M6rna+NHVUh9QnIlIc2o+kj3pk5TtlX5mer3jUMJyoRTK2LvJRHYt0ueeIiHSvs/1I1CLpgw740QNs2NJxd8DeokVCCiEAABKOSURBVLltPcj2IFIdi2S9B3s0Emy5m/q6sSXRdn1PZe5NrlbfoSrKx02tGv+QPkuBpA+pe3cTn7vu8XJXI6OqcD/1TrK5d+u8IyZy7UOvZnV9LGLc/Y1DeHP9FgbEo1zyhxfZ1Lg9sPbEOo3kanVPOI2tzoB4MEFSrSHpi8oy/dfMxprZo2b2kpmtMLPzw/KdzOwhM3st/Dk85ZpLzKzOzF4xs6NTyg8ysxfDY78wswwrC/q+Y659rGKDCEBLAUEEYMQOVVx32v5Z/YVtTSRYs3EbJ08by6F7jujxdRqpq9WTubu2NSe0al36rHKtI2kBvufu/wBMB841s8nAHOARd58EPBK+Jjw2E5gCHAPcaGbR8L1uAmYBk8LHMT35i5Rb3bubGD9nES+/93G5q1JSc/64nCfq1hHNtAIxTUuCdpl7L/3iZKqixg5V0R5Zp5FpC+CkZGtIpC8pS9eWu68F1obPN5nZS8Bo4ATg8PC024DHgIvD8rvcvRFYZWZ1wMFm9iYw1N2fBjCz24ETgft77JcpowMv/wsfbK2ALL1Z6G5WVTbm1TQwKB5JGUPpXPID+4m694PNrqIRmloTXHTU3iXvWsq0Wj1Jq9alLyr7ynYzGw8cACwBdguDTDLY7BqeNhpYnXJZfVg2OnyeXp7pPrPMrMbMatatW1fMX6HHJdOb9JYgUkzpiRk705xIsENVtK2L6eOmVppbnf+8/2XueOatvO+/fnMjS1d/2GX3VHK1enXMCIdGGBCPaNW69FllHWw3s8HAPcB33f2jLoY3Mh3wLso7FrrfDNwMwfTf3GtbGfacs4j+Fz62S7gTjxrxiLGlOfO3/upY8IH9cVMrsUjHvyKX37uCY/bdPecP9FzSvSdT1MdjESIJ59zDJ7ZLYV8s2sdEKkHZWiRmFicIIne4+x/C4nfNbGR4fCTwXlheD4xNuXwMsCYsH5OhvM+57oGXGN/PgwgEXWTuzinTxnZI2Bg1+N7n9+KpOUdw/P6jGTN8YMYWTCxqOY9T5JLuPVOK+tQU9sWyoLaBGVct7rA1sEhPK9esLQNuAV5y92tTDi0Ezg6fnw0sSCmfaWbVZjaBYFD92bD7a5OZTQ/f86yUaypCNl0h3Rk/ZxE/f7T7vcv7i5YE3Pb0Wx1mgUUj8IkRg9olZvzhlyZ3uH5LU4LlDRtzumemAfTOBs5zOTdf2sdEKkm5WiQzgK8AR5hZbfg4FrgS+LyZvQZ8PnyNu68A5gErgb8A57p78sv5t4C5QB3wOhU00F7oN8bfPbWK8XMWlah2vUMWk7TaNLXS4cP0jEM+wflHTuxw7hWLVub0oZtLuveeSA3fE8FKJFvlmrX1BJnHNwCO7OSaHwM/zlBeA+xbvNoVR+o3xuSmRbmsqO7vASQp17Un6YsNF9Q2cNNfO7bmcl2UmBxAn502RpLp+lzOzZf2MZFKopXtJZLvznezfrOEB19+vyeq2Celfpgmg3lTpmzAeXzo5pLuvdSp4XsiWIlkS4GkRPL5xqhWSPYM+I8T96V+wxbm/u0NqmJRWt3bfZjWb9iacdYWtN/kKhe5pHsvdWp47WMilaLs60j6qlx2vvvO755TEMlRVSz4q/s/T71JdTxKc8K59LjJ7abjdjZra1A80mGTq95qxODqHt0aWCQTpZEvse7m+SuA5GdQPEKL067bakA8wpMXH9Huz/mOJW/x/T8ub3dtdcx4as6RXX74dvb/Tes2pD9TGvky6ax74/RfPclTqz4sQ40qU1U0guFtSQ6709SaYEA8SlNKtvz0Maj1mxvZd9Qwzj9iIr9cXNc2WpVweLLu/U4XE3a28DCXBYki/YkCSRn0p1bIrE9PwHF2GzKAax9+jS1pW+cOqoqScOf4/UYxr6a+k3fpaPbR+3Dtw6+2K0uOQa3f3MgdS97mhkdfA6dDcGpu9U5n0HU2227yyKEFzcIT6csUSHrQV+c+zWN1H5S7Gj1q1I4DOGDcTuxQFeUnD7X/4K+OGVefNBVwLphXm/V7/vsX9mHWP+3J7sMGdJi19ETd+8yev6zbTbA6m0HX2Wy72tUfdjkLT11e0p8pkJRQ6ofLQf/xcLmrUxaX3fsSA2IRHOcfx+/EE3Xr244dPH4nLpy/lIgZTZ3kfolHoDkRBB13+OGXpnDG9E8AHWctAcy4anFWOynmuphw/7E7djoLT11e0t8pkJRI8sNlWyeJBfuTbeEHe2oQAfhb2utMrjvtAPbZfUinW9WmTvXduLWpQ6shk2RLKJfFhBN3G5KxHFCXl/R7CiQlkNrPLvmrjhpjdxrExN2GdDiWbO0tb9gY7DcSidDU2trpvicD4hHc4duf7T4Lb2frMzKVL+2my0ukP1AgKbL1mxs5+D8e7vdZeovBItah+2n7QHod8aixuTH4k05+kMciQRr5qnAjq29/diJf2Hf3thYN0JaPqqsP+s5m26WXK1WJiAJJUS2obeD8u7IfNO7vPrXHTjz1RubJB1Wxjgs4F9Q2MHv+UhpbgmZHY0vH6wbGY9xwxoEMGxjv0BVWirEMpSoRUSApmqk/XMRHyuCdtVgELvj8XtTMXdJh9Xk8Cn/+zmHturS27/HR9TqT5kSCKaOGZj2ttxhjGUpVIv2dUqQUwfg5CiK5aknAzF8/w8yDxxJPyRUfi8BlX9qXj5ta26V5z5Q2PdWgbrayLXXadaUqkf5MLZICHHPto7z83pZyV6PXaknA3c+t5v7zPs2ajVsBY/UHW9oGz1O7nzKNRSRVxyL86IR9+ew+u3b6Qa6xDJHSUYskT+PnLFIQKQY31mzcxmf22pUpo4ZyxaKVGXf9S45FVMc6ZvM1o8sgArkl0RSR3KhFkqMJcxbR/9Jcdi25aDAfja0JvnF7DdecPJVPjNihy6m0ybGI3y95m+sfraMqmtvgtsYyREqjTwQSMzsG+DkQBea6+5WluE9/ypGVi+9+bi9u+uvrbVNxc9XYErQ87vv2Yd12P40YXM13jpzE6YeMyysglHqPEJH+qNd3bZlZFLgB+AIwGfiymU0u9n36axBJ7gs1IBzM/vTEEe2On3XoOGYePI6WzlYCpqiKGvEoHPfJ3RgY7zjw/XFTa9bdTxrcFqkcfaFFcjBQ5+5vAJjZXcAJwMpi3aBmVfepPCpNxMi4yjtqcMb0cdy55G1aE5AgGKxuaU3gQDxiJBxOmTaGfzlgNPFYlB2qou1SlNS9u4na1R+y/9gd26bopq+lOHXaGObV1BM1o7k1wYVH7c0he4xoa108/PLidvVKtjz2G7ujup9Eepm+EEhGA6tTXtcDhxTzBo+/Vpl7qBswsCpKa8L53lF7URWNsPPgKg7dc2eAcCzhNWJh+pCvH7YHX//0HowYXM35R+5F/Yat7YIEkNUH+MTdhnRIW5Jp/CF5j0zv19UiPnU/ifQuvX6HRDM7BTja3b8evv4KcLC7fyftvFnALIBx48Yd9NZbb2V9j5pV6zn5V88Urc7TJwznhbc/JJFwWjwY2LGIce7he3LQ+J0AZ9SwgazZuJWGDVtpbEmw76ihbGlO0LBhC40tCQ6buDPDd6jq9oO/ktObV3LdRKSjvrxDYj0wNuX1GGBN+knufjNwMwRb7eZyg2kTRvDpiSMyZqsdUh1jc2MLRrAiu9XhuKm7s//Yndh31NC2rqHkOonkquvkh2h6t1GqTMkK03X3AVzJ3+4ruW4ikr2+EEieAyaZ2QSgAZgJnF7sm/z269M7tEye/3+fa7cXRlffrtODgj5ERaSv6PWBxN1bzOzbwAMEvUS3uvuKUtxr2oQRvHnlF9uVpQYDBQYR6Y96fSABcPc/A38udz1ERPqjXr+OREREykuBRERECqJAIiIiBVEgERGRgvT6BYn5MLN1QPYrErfbGajMZe7bqY6Fq/T6QeXXsdLrB6pjPj7h7rukF/bLQJIvM6vJtKqzkqiOhav0+kHl17HS6weqYzGpa0tERAqiQCIiIgVRIMnNzeWuQBZUx8JVev2g8utY6fUD1bFoNEYiIiIFUYtEREQKokAiIiIFUSDJkpkdY2avmFmdmc3pwfuONbNHzewlM1thZueH5TuZ2UNm9lr4c3jKNZeE9XzFzI5OKT/IzF4Mj/3CzKyI9Yya2d/N7L4Krd+OZjbfzF4O/ywPrcA6/t/w//FyM7vTzAaUu45mdquZvWdmy1PKilYnM6s2s7vD8iVmNr4I9bsm/P+8zMz+aGY7lqt+ndUx5diFZuZmtnM561gwd9ejmwdBevrXgT2AKmApMLmH7j0SODB8PgR4FZgMXA3MCcvnAFeFzyeH9asGJoT1jobHngUOJdil937gC0Ws5wXA74H7wteVVr/bgK+Hz6uAHSupjgRbRq8CBoav5wFfLXcdgc8ABwLLU8qKVifg/wD/HT6fCdxdhPodBcTC51eVs36d1TEsH0uw/cVbwM7lrGPBf397+oa98RH+z3sg5fUlwCVlqssC4PPAK8DIsGwk8EqmuoV/UQ8Nz3k5pfzLwK+KVKcxwCPAEWwPJJVUv6EEH9KWVl5JdRwNrAZ2Itje4b7wA7HsdQTG0/6Dumh1Sp4TPo8RrOK2QuqXduyfgTvKWb/O6gjMB/YD3mR7IClbHQt5qGsrO8l/5En1YVmPCpusBwBLgN3cfS1A+HPX8LTO6jo6fJ5eXgw/A2YDiZSySqrfHsA64H/C7re5ZrZDJdXR3RuAnwBvA2uBje7+YCXVMUUx69R2jbu3ABuBEUWs69cIvr1XVP3M7Higwd2Xph2qmDrmQoEkO5n6mHt03rSZDQbuAb7r7h91dWqGMu+ivNB6HQe85+7PZ3tJJ/Uo5Z9xjKBr4SZ3PwD4mKBLpjM9XsdwnOEEgu6MUcAOZnZmV5d0Updy/l3Np06l/DP9PtAC3NHNvXq0fmY2CPg+8INMhzu5X1n+DLOlQJKdeoL+zKQxwJqeurmZxQmCyB3u/oew+F0zGxkeHwm8101d68Pn6eWFmgEcb2ZvAncBR5jZ7yqofsl71rv7kvD1fILAUkl1/Bywyt3XuXsz8AfgUxVWx6Ri1qntGjOLAcOADwqtoJmdDRwHnOFhn08F1W9Pgi8MS8N/N2OAF8xs9wqqY04USLLzHDDJzCaYWRXBgNbCnrhxODPjFuAld7825dBC4Ozw+dkEYyfJ8pnhTI4JwCTg2bALYpOZTQ/f86yUa/Lm7pe4+xh3H0/w57LY3c+slPqFdXwHWG1me4dFRwIrK6mOBF1a081sUPjeRwIvVVgdk4pZp9T3Opng70+h3/iPAS4Gjnf3LWn1Lnv93P1Fd9/V3ceH/27qCSbUvFMpdcxZTw7I9OYHcCzBjKnXge/34H0PI2imLgNqw8exBH2gjwCvhT93Srnm+2E9XyFlxg4wDVgeHrueIg/IAYezfbC9ouoH7A/UhH+OfwKGV2AdLwdeDt//twQzd8paR+BOgjGbZoIPvHOKWSdgAPC/QB3BrKQ9ilC/OoIxg+S/l/8uV/06q2Pa8TcJB9vLVcdCH0qRIiIiBVHXloiIFESBRERECqJAIiIiBVEgERGRgiiQiIhIQRRIRESkIAokIlkys80leM/9zezYlNeXmdmFRXjfjHU1s9+Y2cmFvr9IKgUSkfLan2CBqUivpUAikgczu8jMngs3T7o8LBtvwaZZv7Zgg6oHzWxgeOwfw3OftmDjpeVhup0fAaeZWa2ZnRa+/WQze8zM3jCz87qpxwXhey03s+9mOG5mdr2ZrTSzRWzP1CtSNAokIjkys6MIciAdTNCiOMjMPhMengTc4O5TgA+Bk8Ly/wH+zd0PBVoB3L2JIAPs3e6+v7vfHZ67D3B0+P4/DJN2ZqrHQcC/AocA04FvmNkBaaf9M7A38EngGwSJIEWKSoFEJHdHhY+/Ay8QfPBPCo+tcvfa8PnzwHgLtnod4u5PheW/7+b9F7l7o7u/T5BZd7dOzjsM+KO7f+zumwkyBn867ZzPAHe6e6u7rwEWZ/crimQvVu4KiPRCBvyXu/+qXWGw8VhjSlErMJDM+0V0Jf09Ovt3mu37KqGelJRaJCK5ewD4WrjZGGY22sw6HXtw9w2EKcDDopkphzcBQ/Ksx+PAiWHq+R0IurH+luGcmWYWDfcO+Wye9xLplFokIjly9wfN7B+Ap4OtIdgMnEk49tGJc4Bfm9nHwGME26ECPArMMbNa4L9yrMcLZvYbgtThAHPd/e9pp/0ROAJ4kWAbhL/mcg+RbCiNvEgPMLPB4TgGZjYHGOnu55e5WiJFoRaJSM/4opldQvBv7i3gq+WtjkjxqEUiUuHMLLkjYboj3X19T9dHJJ0CiYiIFESztkREpCAKJCIiUhAFEhERKYgCiYiIFOT/A4VDi1PEZEnOAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#l_df.plot.scatter(x = 'count', y = sample_name, xlim=(0, 1000), ylim=(0, 1000))\n",
    "#l_df.plot.scatter(x = 'length_parastou', y = 'length')\n",
    "#l_df.plot.scatter(x = 'length_old', y = 'length')\n",
    "l_df.plot.scatter(x = 'length_old', y = 'cds_length')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Visualize `total_difference`: `scatter_plot`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7fa10b76b7f0>"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZIAAAEHCAYAAACEKcAKAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8li6FKAAAgAElEQVR4nO3df5yVdZ338dd7+CWGKAIiMtBgoAUmKrOK0bquWlKW2KrFpqsVRbneZXWXQnbb7t62KbW2uZmJZkr5A9QK7tC08FfrIjoooqLmKCiDiDqSMaYjPz73H9c1cBjOzJwz55w5Z2bez8fjPDjne32v6/oMyvnM98f1/SoiMDMz66yqcgdgZmbdmxOJmZkVxInEzMwK4kRiZmYFcSIxM7OC9C13AOUwbNiwqKmpKXcYZmbdyooVK16LiOGty3tlIqmpqaGurq7cYZiZdSuSXshW7q4tMzMriBOJmZkVxInEzMwK4kRiZmYFcSIxM7OCOJGYmfUSjU3NPLbuzzQ2NRf1ur1y+q+ZWW+zaOV6LrhtFf2qqtiyfTtzTz2Ukw8bVZRru0ViZtbDNTY1c8Ftq3h7y3Y2N2/l7S3bOf+2VUVrmTiRmJn1cA2b3qJf1a5f9/2qqmjY9FZRru9EYmbWw1UPGciW7dt3KduyfTvVQwYW5fpOJGZmPdzQQQOYe+qh7NGvir0G9GWPflXMPfVQhg4aUJTre7DdzKwXOPmwUUwdN4yGTW9RPWRg0ZIIOJGYmfUaQwcNKGoCaeGuLTMzK4gTiZlZN1OqBws7y11bZmbdSCkfLOwst0jMzLqJUj9Y2FlOJGZm3USpHyzsLCcSM7NuotQPFnZWSROJpGslvSLpiSzHviEpJA3LKJsjqV7SM5JOzCifLOnx9NjlkpSWD5C0IC1fLqmmlD+PmVk5lfrBws4q9WD7dcCPgfmZhZJGAx8CXswomwDMACYCBwB/kHRQRGwDrgRmAQ8CtwPTgDuAmcCmiBgnaQZwKfCpEv9MZmZlU8oHCzurpC2SiLgfeD3LoR8C5wORUTYduDkimiNiDVAPHClpJDA4IpZFRJAkpVMyzrk+fX8rcHxLa8XMrKcaOmgAk0bvUxFJBMowRiLpZGB9RDzW6tAoYF3G54a0bFT6vnX5LudExFbgDWBoG/edJalOUt2rr75a8M9hZmaJLk0kkvYELgQuynY4S1m0U97eObsXRsyLiNqIqB0+fHgu4ZqZWQ66ukXyHmAs8JiktUA18Iik/UlaGqMz6lYDL6Xl1VnKyTxHUl9gb7J3pZmZWYl0aSKJiMcjYr+IqImIGpJEcEREvAwsBmakM7HGAuOBhyJiA7BZ0pR0/OMsYFF6ycXA2en704C703EUMzPrIqWe/nsTsAw4WFKDpJlt1Y2IJ4GFwGrgd8C56YwtgHOAa0gG4J8jmbEF8DNgqKR64OvA7JL8IGZm1ib1xl/ga2tro66urtxhmJl1K5JWRERt63I/2W5mZgVxIjEzs4I4kZiZWUGcSMzMrCBOJGZmVhAnEjOzAlTatrfl4K12zcw6qRK3vS0Ht0jMzDqhUre9LQcnEjOzTqjUbW/LwYnEzKwTKnXb23JwIjEz64RK3fa2HDzYbmbWSZW47W05OJGYmRVg6KABvTaBtHDXlpmZFcSJxMwMP1hYCHdtmVmv5wcLC+MWiZn1an6wsHCl3mr3WkmvSHoio+z7kp6WtErSryXtk3FsjqR6Sc9IOjGjfLKkx9Njl6d7t5Pu774gLV8uqaaUP4+Z9Tx+sLBwpW6RXAdMa1X2e+CQiDgU+BMwB0DSBGAGMDE95yeS+qTnXAnMAsanr5ZrzgQ2RcQ44IfApSX7ScysR/KDhYUraSKJiPuB11uV3RURW9OPDwLV6fvpwM0R0RwRa4B64EhJI4HBEbEskg3m5wOnZJxzffr+VuD4ltaKmVku/GBh4co92P45YEH6fhRJYmnRkJZtSd+3Lm85Zx1ARGyV9AYwFHit9Y0kzSJp1TBmzJji/QRm1u35wcLClC2RSLoQ2Arc0FKUpVq0U97eObsXRswD5gHU1tZmrWNmvZcfLOy8sszaknQ28DHgjLS7CpKWxuiMatXAS2l5dZbyXc6R1BfYm1ZdaWZmVlpdnkgkTQMuAE6OiL9mHFoMzEhnYo0lGVR/KCI2AJslTUnHP84CFmWcc3b6/jTg7ozEZGZmXaCkXVuSbgKOBYZJagC+QzJLawDw+3Rc/MGI+FJEPClpIbCapMvr3IjYll7qHJIZYAOBO9IXwM+AX0iqJ2mJzCjlz2NmZrtTb/wFvra2Nurq6sodhpkVSWNTswfKu4CkFRFR27q83LO2zMwK4uVNys9LpJhZt+XlTSqDWyRm1u20dGW98dYW+lVV8TY7n0xvWd7EXVxdx4nEzLqVzK6sd7ZtZ5uXNyk7JxIz6zYyu7JaWiH9+ogBfaF/nz47xkjcGulaTiRm1m20rNSb2ZW1R98+XHHG4ew9sL9nbZWJE4mZdRttrdQ78YC9nUDKyLO2zKzb8Eq9lcktEjPrVrxSb+VxIjGzbscr9VYWd22ZmVlBnEjMzKwgTiRmZlYQJxIz61KNTc08tu7PXg+rB/Fgu5l1Ga/U2zO5RWJmXcIr9fZcTiRmVhKtu7BaljfJ1LJSr3VvJU0kkq6V9IqkJzLK9pX0e0nPpn8OyTg2R1K9pGcknZhRPlnS4+mxy9O920n3d1+Qli+XVFPKn8fMcrNo5XqmXno3Z16znKmX3s3ilevbXN7EK/V2f6VukVwHTGtVNhtYGhHjgaXpZyRNINlzfWJ6zk8k9UnPuRKYBYxPXy3XnAlsiohxwA+BS0v2k5hZTtrqwgK8vEkPVdLB9oi4P0srYTpwbPr+euBe4IK0/OaIaAbWSKoHjpS0FhgcEcsAJM0HTgHuSM/5l/RatwI/lqTojRvRm1WIbCv0tnRheXmTnqkcs7ZGRMQGgIjYIGm/tHwU8GBGvYa0bEv6vnV5yznr0mttlfQGMBR4rXThm1mmlt0KWxJDR11YXt6k56mk6b/KUhbtlLd3zu4Xl2aRdI8xZsyYzsRnZq20NZ137qmHcn6rciePnqsciWSjpJFpa2Qk8Epa3gCMzqhXDbyUlldnKc88p0FSX2Bv4PVsN42IecA8gNraWnd9mRUo226F59+2iqnjhrkLq5cpx/TfxcDZ6fuzgUUZ5TPSmVhjSQbVH0q7wTZLmpLO1jqr1Tkt1zoNuNvjI2Zdo6PpvEMHDWDS6H2cRHqBkrZIJN1EMrA+TFID8B3gEmChpJnAi8DpABHxpKSFwGpgK3BuRGxLL3UOyQywgSSD7Hek5T8DfpEOzL9OMuvLzLqAp/NaC+XzC3w6HXcEGQkoIl4sQVwlVVtbG3V1deUOw6zbW7xy/W5jIV7ypOeStCIialuX59wikfRlkhbFRtgxry+AQ4sSoZl1Ox4LMciva+s84OCIaCxVMGbW/Xg6r+Uz2L4OeKNUgZiZWffUYYtE0tfTt88D90paAuxYrjMiLitRbGZm1g3k0rW1V/rni+mrf/qCNh7+MzOz3qPDRBIR/wog6fSIuCXzmKTTSxWYmXWN+o2bWbnuzxw2eh/Gjdir4xPMWslnsH0OcEsOZWbWTVz0m8eZ/+DOGfxnHT2Gf5v+/jJGZN1RLmMkHwE+CoySdHnGocEkDw6aWTdUv3HzLkkEYP6yFzlrSo1bJpaXXGZtvQTUAW8DKzJei4ET2znPzCrYynV/zqvcrC25jJE8Bjwm6caI2NIFMZlZkbVe6h3gsNH7ZK3bVrlZW/IZI3lEUutZWm+QtFYu9oOKZpWpraXex43Yi7OOHsP8ZbuOkbhby/KV81pbkuYC24Ab06IZJPuBvAF8MCI+XpIIS8BrbVlv0NjUzJMv/YUvzK+jeevOxRX36FfFAxcct6Nl4llblquC19oCpkbE1IzPj0t6ICKmSjqz8BDNrFhaWiFVaJckAjuXem9JJONG7OUEYgXJZ4mUQZKOavkg6UhgUPrRs7fMKkTmhlN/3bJtt+Ne6t2KLZ8WyeeBayUNIunS+gvweUnvAr5XiuDMLH8tG069za4tkT3792F7hLe9taLLOZFExMPA+yXtTTK2kjlHcGHRIzOzTsm24dSAvuKnZx7BxAP2dhKxostnP5IBwKlADdA32fUWIuLfShKZmXWobk0j9z/7GseMH0bt2KFAsqz73FMP3W3DqWMO2q/M0VpPlU/X1iKSGVoryFj918zK49Sf/DcrXkx2drj87nr+dtxQfvH5KYA3nLKulU8iqY6IacW6saSvkYy7BPA48FlgT2ABSatnLfDJiNiU1p8DzCSZgvyViLgzLZ/Mzv3cbwfOi3z2DzbrZhqbmjn3xhU7kkiLP9Y3UremcZeWiROIdYV8Zm39j6SirOYmaRTwFaA2Ig4B+pA8lzIbWBoR44Gl6WckTUiPTwSmAT9J948HuBKYBYxPX0VLdmaVZtHK9Xzgkrt58PlNWY/f/+xrXRyRWX6J5IPACknPSFol6XFJqwq4d19goKS+JC2Rl4DpwPXp8euBU9L304GbI6I5ItYA9cCRkkYCgyNiWdoKmZ9xjlmPUb9xM9c9sIZv3vLYbs+FZDpm/LAujMoskU/X1keKddOIWC/pByQbZb0F3BURd0kaEREb0jobJLWMDo4CHsy4RENatiV937p8N5JmkbRcGDNmTLF+FLOSa73Ue1smj9l7R7eWWVfKuUUSES8Ao4Hj0vd/zef8TJKGkLQyxgIHAO/q4Ol4ZQupnfLdCyPmRURtRNQOHz4835DNyiLbUu/ZnDhxP2775w92QURmu8tn+u93gFrgYODnQD/gl8DU9s5rwwnAmoh4Nb32r4APABsljUxbIyOBV9L6DSRJrEU1SVdYQ/q+dblZj9DWku4D+ia/w516xCg+N3Wslzixssqna+sTwOHAIwAR8ZKkzv7f+yIwRdKeJF1bx5OsIvwmcDZwSfrnorT+YuBGSZeRtGDGAw9FxDZJmyVNAZYDZwH/1cmYzMqu9XLvbS3p/h+nH8rR7xnmWVlWEfJJJO9ERLQsJZ8ujdIpEbFc0q0kSWkr8Cgwj2TtroWSZpIkm9PT+k9KWgisTuufGxEtiwidw87pv3ekL7Nuo7GpmbuefJnfrtrAQ2tfZ0DfKrZujx3LvWdb6v1jk7IOBZqVRT7LyH+DpCXwIZK1tT4H3BgR3a4F4GXkrVLc8OALXPibJ7Iey1zu3Uu9WyUoeBn5iPiBpA+RLNZ4MHBRRPy+iDGa9RqNTc3M/d3TLKhraLNOnyrtWO7dS71bJcuna4s0cTh5mBXgqvue45LfPU1HnQFbtoWXe7duocNEImkz2afUCoiIGFz0qMx6qPNvWcnCFetzqvudj0/wYLp1Cx0mkojIqT0taUjLulhmtru6NY05J5HvnnIIZxz17hJHZFYceXVtdWApcEQRr2fWI9Rv3My1D6xlwcPtP1gowScnV3P+tPe6JWLdSjETSbanzM16rbo1jVy06AlWv9zUYd0T3jecS0+d5ARi3VIxE4mXbjdLfeqq/2H5mo57eo989xD+/R/e7xlZ1q0VM5GYGfDd367uMIn07SNu/vxRXmTReoROLbrYBndtWa93w4MvcPV/r2m3Tv8+4rLTJzmJWI+RV4tE0geB8RHxc0nDgUHp/iCQrJdl1uvUrWnkd09uZP/BA5h755/areuxEOuJirb6b0S8XooAzSpVY1MzM697iJUNf8mp/rc+8l5m/d17ShyVWdcr1+q/Zt1WY1MzV//xea6+/3m25TDF5JRJI/k/H5/oVoj1WGVZ/desu5p333N8746nO5yiOLBfFdsieTrdDxZaT5dPIlko6SpgH0lfIFn99+rShGVWeT778+Xc88xrHdbrVyWu+qfJTDxgb7dCrFfw6r9mHWhsaubTVy/jmY1vdli3SvAfn5zEMQft1wWRmVUGr/5r1o5FK9fzjYUr2bK9/XpTD9yXj086gA9N3N+tEOt18pm19Q/ApcB+JM+MePVf67HqN27m14+u46f3relwQN2zsay3y6dFMhf4eEQ8VYwbS9oHuAY4hGR5lc8BzwALgBpgLfDJlhWFJc0BZgLbgK9ExJ1p+WR2brV7O3Be5Lrto1krjU3NfHXBo/zx2cac6juJmOX3ZPvGYiWR1I+A30XEe4FJwFPAbGBpRIwnWU14NoCkCcAMYCIwDfiJpD7pda4EZpFsAzw+PW6WtxsefIG/+e4fck4i3/3EIU4iZuTXIqmTtAD4DdDcUhgRv8r3ppIGA8cAn0mv8Q7wjqTpwLFpteuBe4ELgOnAzRHRDKyRVA8cKWktMDgilqXXnQ+cAtyRb0zWezU2NfP93z3Nze1sewvJb11/d9BwPjxxBB/2WIjZDvkkksHAX4EPZ5QFkHciAQ4EXgV+LmkSsAI4DxgRERsAImKDpJapL6OABzPOb0jLtqTvW5fvRtIskpYLY8aM6UTI1hPNu+85/v2Op3Oqu/CLU7w+llkW+Uz//WyR73sE8OWIWC7pR6TdWG3ItiBktFO+e2HEPGAeQG1trcdQjM9e+xD3/OnVnOpOnzTSScSsDTmPkUg6SNJSSU+knw+V9O1O3rcBaIiI5ennW0kSy0ZJI9PrjwReyag/OuP8auCltLw6S7lZVo1NzVxz/3NM+s7vck4iJ0/anx/9ozf/NGtLPoPtVwNzSLqTiIhVJAPgeYuIl4F1kg5Oi44HVgOLgbPTsrOBRen7xcAMSQMkjSUZVH8o7QbbLGmKJAFnZZxjtour7nuOyRf/gYtvf5o3mrd1WP9jh+zPH752DJf/4+QuiM6s+8pnjGTPiHgo+b7eYWsB9/4ycIOk/sDzwGdJEttCSTOBF4HTASLiSUkLSZLNVuDciGj5JjiHndN/78AD7dZK/cbNfO665by4qbnjysAx44bywxmHezDdLEf5JJLXJL2HdAxC0mnAhs7eOCJWkixL31rWfU0i4rvAd7OU15E8i2K2my/Or+PO1Rtzrv+tj76XWcd4Sq9ZPvJJJOeSDFa/V9J6YA1wRkmiMitQ/cbNnHLFH2l6J7d5FeOG78mCL37ArRCzTsgnkZxC8uT4PSRdUG8CJ0hakbYuzCrC+bc+xsIOngnJdN5x4/jahw/uuKKZZZVPIqlNX4tJpt2eATwMfEnSLRExtwTxmeWssamZb97yGHc/k9tsrCF79uUPXz/WrRCzAuWTSIYCR0REE+zYevdWkifUV5CsxWVWFt9YuJJbH1mfc/139ROPXnRiCSMy6z3ySSRjgHcyPm8B3h0Rb0nKbTqMWZH95pF1fHXhqrzOufjkCZz5gbElisis98knkdwIPCip5TmNjwM3pVvuri56ZGYdeP9Fd7D5nQ42Cslw8qEjuPzT2SYKmlkh8lki5f9Kuh34IMkYyZfSqbfg2VvWherWNHLaVQ92XDH1vhGD+OUXpngsxKxE8t0hcQXJeIhZl2tsauaE/7ibTW/l3gr5ZG01c0+bVMKozCyvRGJWLt9cuJJb8hhM36MP/PYrxzBuxF4ljMrMwInEuoEDZy8h9zYInHnUGC7+xPtLFo+Z7cqJxCpWvmMhBw1/Fzd98WiPhZh1MScSqziNTc1M++F9vPrmlpzPWfHtE5xAzMrEicQqysd+dC9PbHgz5/pD9uzDoxdNK2FEZtYRJxKrCJ15sPBnZ03m+An7lygiM8uVE4mV3bjZS/La2KavoP57J5UsHjPLjxOJlc1V9zzL9+78U17neHkTs8rjRGJlUTN7SV71xw4dyD3fPK5E0ZhZIZxIrEt9Y8Ej3Ppofhtr/uFrfrDQrJJVlfPmkvpIelTSb9PP+0r6vaRn0z+HZNSdI6le0jOSTswonyzp8fTY5Wq1qbxVjprZS/JKIqP32YO1l5zkJGJW4cqaSIDzgKcyPs8GlkbEeGBp+hlJE4AZwERgGvATSX3Sc64EZgHj05fnglaYz1yzLO+urFu/OIU/zj6+RBGZWTGVLZFIqgZOAq7JKJ4OXJ++v55ke9+W8psjojki1gD1wJGSRgKDI2JZRAQwP+McqwA1s5dwb/3rOdcfObg/ay85idqxQ0sYlZkVUznHSP4TOB/I7LcYEREbACJig6T90vJRQOZaGQ1p2Zb0fevy3UiaRdJyYcyYMcWI39rxmWuW5ZVAwGMhZt1VWRKJpI8Br0TECknH5nJKlrJop3z3woh5wDyA2trarHWscPmujwXwhanv5sKPH1KiiMys1MrVIpkKnCzpo8AewGBJvwQ2ShqZtkZGAq+k9RuA0RnnVwMvpeXVWcqtDCb9y+288XZ+OfryGYdx8mFZG5Fm1k2UZYwkIuZERHVE1JAMot8dEWcCi4Gz02pnAy3b+i4GZkgaIGksyaD6Q2k32GZJU9LZWmdlnGNdpG5NIzWzl+SVRMYP25MV3z7BScSsB6i050guARZKmgm8CJwOEBFPSlpIsjf8VuDciNiWnnMOcB0wELgjfVkXaGxqZvLFf8j7vLWXeHkTs55EyWSn3qW2tjbq6uo6rmht+uzPHuSeZxvzOue8vz+Qr534vhJFZGalJmlFRNS2Lq+0FolVsMamZho2vcX0Kx7I+1y3Qsx6LicSy8milev56s0rs0+Ja8eZR1Zz8T9MKklMZlYZnEisQ41NzZx388q8z3MrxKx3cCKxdk277B6efuWveZ3jDafMehcnEmtTvutjgVshZr2RE4nt5pCLltD0Tn7neEaWWe/lRGI7LF39MjPnr8j7PLdCzHo3JxIDOteNNefEg/ji348vQTRm1p04kfRyX/7lw/y/J17puGKGvkC9WyFmlnIi6cU60wrxUu9m1poTSS90yn/dx8r1TXmdM2pwfx741odKFJGZdWdOJL2Mp/SaWbE5kfQSh35nCX9pzu+cMfsM4P7ZJ5QmIDPrMZxIegG3QsyslJxIerD3XbiEt7Z1XC/T34wZzC3//LelCcjMeiQnkh7KrRAz6ypOJN1Iy34g1UMGMnTQgKx1OpNADhs1iN98+e8KDc/MeqmyJBJJo4H5wP7AdmBeRPxI0r7AAqAGWAt8MiI2pefMAWYC24CvRMSdaflkdm61eztwXvTAbR8XrVzPBbetol9VFVu2b2fuqYfust/5bx5Zx1cXrsr7um6FmFmhqsp0363A/46I9wFTgHMlTQBmA0sjYjywNP1MemwGMBGYBvxEUp/0WlcCs4Dx6WtaV/4gXaGxqZkLblvF21u2s7l5K29v2c75t62isSmZhlUze0neSeS9++3pJGJmRVGWFklEbAA2pO83S3oKGAVMB45Nq10P3AtckJbfHBHNwBpJ9cCRktYCgyNiGYCk+cApwB1d9sN0gYZNb9Gvqoq32b6jrF9VFWfMW8bTr7yZ9/WcQMysmMrVItlBUg1wOLAcGJEmmZZks19abRSwLuO0hrRsVPq+dXm2+8ySVCep7tVXXy3mj1By1UMGsmX79l3KNjdvzTuJnHb4SCcRMyu6siYSSYOA24CvRsRf2quapSzaKd+9MGJeRNRGRO3w4cPzD7aMhg4awNxTD2WPfp3/z7X2kpP4waeOKGJUZmaJsiUSSf1IksgNEfGrtHijpJHp8ZFAy7K0DcDojNOrgZfS8uos5T3O1HHDeHvL9o4rtjLnxIPcCjGzkirXrC0BPwOeiojLMg4tBs4GLkn/XJRRfqOky4ADSAbVH4qIbZI2S5pC0jV2FvBfXfRj5CSXKbsd6cyUXsh9LKQYMZpZ71Wu50imAv8EPC5pZVr2LZIEslDSTOBF4HSAiHhS0kJgNcmMr3MjouWZ7XPYOf33DipooL2jKbu5KPWDhcWI0cx6N/XARy46VFtbG3V1dSW9R2NTM1MvvXuX7qg9+lXxwAXH5fRbf6lbIcWI0cx6F0krIqK2dbmfbC+RtqbsNmx6q80v6ZYupulXPJD3/TozDtKZGNtTaBeZu9jMuicnkhLJNmV3y/btVA8ZmLX+opXrOe/mlVmPdaSzg+n5xtieQrvI3MVm1n2V/TmSnipzyu5eA/qyR78q5p56aNbftOs3bu5UEll7yUkFzcjKJ8b2dPTkfanPN7PycoukhE4+bBRTxw1rt7umM2MhVcDzRZrSm0uMHSm0i6zYXWxm1rWcSEps6KABWb8Ml65+mZnzV3Tqmv37VdHY1Fy0L9m2YsxVoV1kxexiM7Ou566tMqiZvaTTSQR2/rZeKQrtIitWF5uZlYdbJCXUehZSZ1shA/pW0bx152/suf623pWzoArtIitGF5uZlYcTSYlkzkJ6Z9v2XRJBPvbs14dzjn0PV9xbv8uMpo6+aMsxC6rQLrJCzzez8nAiKYHMWUiZA8idsS228+mjxvDpo8bk/Nt6tvuff9sqpo4b5i9qMys6J5ISyDYLKVdVsOOsvlXw/dMm7fjyzyWBNGx6izfe2uJZUGbWZZxIiqyxqZlPz3uAN7fkf+7aS06isamZJ196AxATDxic8xd/6660bZ4FZWZdxImkiDr7dHrmQ4VDBw3gmIP2a6f27rJ1ZfXrIwb0hf59+uQ8rmJm1hlOJEUy5eK7eLkp/2ZIMfYKydaVtkffPlxxxuHsPbB/WWdBef0ss57PiaQISr3Ue0faeqBv4gF7l/XL2+tnmfUOfiCxANMuu6dTSeTyGYcVNY5KfKDP62eZ9R5ukXRSZ/cLgc5PxW2vm6jSHujz+llmvYcTSZ4OnL2kwCdDOveFmks3USU90Of1s8x6jx7RtSVpmqRnJNVLml2q+9QUIYlA/l+o3bGbqBK728ysNLp9i0RSH+AK4ENAA/CwpMURsbqY98m3K2vsvgPZsLk56d7Zuo2IYGC/vp2aittdu4kqrbvNzEqj2ycS4EigPiKeB5B0MzAdKFoiqVvTmFf9lhlZmWMaQKe/ULtzN1EldbeZWWn0hEQyCliX8bkBOKqYN7j/2ddyqnfxyRM48wNjd3xu/SXa2S/Ulm6i81uNkfgL2swqQU9IJMpSFrtVkmYBswDGjBmT1w2OGT+My++ub/N4P+DZIj4Xko27icysUvWERNIAjM74XA281LpSRMwD5gHU1tbulmjaUzt2KH87bih/rN+9i2vFt0/osi91dxOZWSXqCbO2HgbGSxorqT8wA1hc7Jv84vNTuPWLU3YpW3vJSf5iN9Sp648AAAdySURBVLNer9u3SCJiq6T/BdwJ9AGujYgnS3Gv2rFDi7q0iZlZT9DtEwlARNwO3F7uOMzMeqOe0LVlZmZl5ERiZmYFcSIxM7OCOJGYmVlBFJHXIxU9gqRXgRc6ceowILfH3MvHMRau0uODyo+x0uMDx9gZ746I4a0Le2Ui6SxJdRFRW+442uMYC1fp8UHlx1jp8YFjLCZ3bZmZWUGcSMzMrCBOJPmZV+4AcuAYC1fp8UHlx1jp8YFjLBqPkZiZWUHcIjEzs4I4kZiZWUGcSHIkaZqkZyTVS5rdhfcdLekeSU9JelLSeWn5vpJ+L+nZ9M8hGefMSeN8RtKJGeWTJT2eHrtcUrZNwTobZx9Jj0r6bYXGt4+kWyU9nf5dHl2BMX4t/W/8hKSbJO1R7hglXSvpFUlPZJQVLSZJAyQtSMuXS6opQnzfT/87r5L0a0n7lCu+tmLMOPYNSSFpWDljLFhE+NXBi2R5+ueAA4H+wGPAhC6690jgiPT9XsCfgAnAXGB2Wj4buDR9PyGNbwAwNo27T3rsIeBokl0l7wA+UsQ4vw7cCPw2/Vxp8V0PfD593x/Yp5JiJNkyeg0wMP28EPhMuWMEjgGOAJ7IKCtaTMA/Az9N388AFhQhvg8DfdP3l5YzvrZiTMtHk2x/8QIwrJwxFvz/b1ffsDu+0v94d2Z8ngPMKVMsi4APAc8AI9OykcAz2WJL/0c9Oq3zdEb5PwJXFSmmamApcBw7E0klxTeY5EtarcorKcZRwDpgX5LtHX6bfiGWPUaghl2/qIsWU0ud9H1fkqe4VUh8rY59ArihnPG1FSNwKzAJWMvORFK2GAt5uWsrNy3/yFs0pGVdKm2yHg4sB0ZExAaA9M/90mptxToqfd+6vBj+Ezgf2J5RVknxHQi8Cvw87X67RtK7KinGiFgP/AB4EdgAvBERd1VSjBmKGdOOcyJiK/AGMLSIsX6O5Lf3iopP0snA+oh4rNWhiokxH04kucnWx9yl86YlDQJuA74aEX9pr2qWsminvNC4Pga8EhErcj2ljThK+Xfcl6Rr4cqIOBx4k6RLpi1dHmM6zjCdpDvjAOBdks5s75Q2Yinn/6udiamUf6cXAluBGzq4V5fGJ2lP4ELgomyH27hfWf4Oc+VEkpsGkv7MFtXAS111c0n9SJLIDRHxq7R4o6SR6fGRwCsdxNqQvm9dXqipwMmS1gI3A8dJ+mUFxddyz4aIWJ5+vpUksVRSjCcAayLi1YjYAvwK+ECFxdiimDHtOEdSX2Bv4PVCA5R0NvAx4IxI+3wqKL73kPzC8Fj676YaeETS/hUUY16cSHLzMDBe0lhJ/UkGtBZ3xY3TmRk/A56KiMsyDi0Gzk7fn00ydtJSPiOdyTEWGA88lHZBbJY0Jb3mWRnndFpEzImI6oioIfl7uTsizqyU+NIYXwbWSTo4LToeWF1JMZJ0aU2RtGd67eOBpyosxhbFjCnzWqeR/P9T6G/804ALgJMj4q+t4i57fBHxeETsFxE16b+bBpIJNS9XSox568oBme78Aj5KMmPqOeDCLrzvB0maqauAlenroyR9oEuBZ9M/980458I0zmfImLED1AJPpMd+TJEH5IBj2TnYXlHxAYcBdenf42+AIRUY478CT6fX/wXJzJ2yxgjcRDJms4XkC29mMWMC9gBuAepJZiUdWIT46knGDFr+vfy0XPG1FWOr42tJB9vLFWOhLy+RYmZmBXHXlpmZFcSJxMzMCuJEYmZmBXEiMTOzgjiRmJlZQZxIzMysIE4kZjmS1FSCax4m6aMZn/9F0jeKcN2ssUq6TtJphV7fLJMTiVl5HUbygKlZt+VEYtYJkr4p6eF086R/TctqlGyadbWSDarukjQwPfY3ad1lSjZeeiJdbuffgE9JWinpU+nlJ0i6V9Lzkr7SQRxfT6/1hKSvZjkuST+WtFrSEnau1GtWNE4kZnmS9GGSNZCOJGlRTJZ0THp4PHBFREwE/gycmpb/HPhSRBwNbAOIiHdIVoBdEBGHRcSCtO57gRPT638nXbQzWxyTgc8CRwFTgC9IOrxVtU8ABwPvB75AshCkWVE5kZjl78Pp61HgEZIv/vHpsTURsTJ9vwKoUbLV614R8T9p+Y0dXH9JRDRHxGskK+uOaKPeB4FfR8SbEdFEsmLw37aqcwxwU0Rsi4iXgLtz+xHNcte33AGYdUMCvhcRV+1SmGw81pxRtA0YSPb9ItrT+hpt/TvN9bpeUM9Kyi0Ss/zdCXwu3WwMSaMktTn2EBGbSJcAT4tmZBzeDOzVyTjuB05Jl55/F0k31h+z1JkhqU+6d8jfd/JeZm1yi8QsTxFxl6T3AcuSrSFoAs4kHftow0zgaklvAveSbIcKcA8wW9JK4Ht5xvGIpOtIlg4HuCYiHm1V7dfAccDjJNsg3JfPPcxy4WXkzbqApEHpOAaSZgMjI+K8ModlVhRukZh1jZMkzSH5N/cC8JnyhmNWPG6RmFU4SS07ErZ2fEQ0dnU8Zq05kZiZWUE8a8vMzAriRGJmZgVxIjEzs4I4kZiZWUH+PxjZYxo9WXnJAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#l_df.plot.scatter(x = 'count', y = sample_name, xlim=(0, 1000), ylim=(0, 1000))\n",
    "#l_df.plot.scatter(x = 'length_parastou', y = 'length')\n",
    "#l_df.plot.scatter(x = 'length_old', y = 'length')\n",
    "l_df.plot.scatter(x = 'length_old', y = 'gene_length')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Alternative: compute lengths using `htseq`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import HTSeq\n",
    "\n",
    "# # A GFF_Reader parses a GFF file and presents its content as\n",
    "# # a sequence of 'GenomicFeature' objects.\n",
    "# gff_file = HTSeq.GFF_Reader(os.path.join(annot_dir, 'schizosaccharomyces_pombe.chr.extended.gff3'))\n",
    "\n",
    "# # The following dict will store, for each `transcript`, a list of it's `exons`:\n",
    "# transcript_dict = {}\n",
    "\n",
    "# # Go through the GFF file and look for all exons:\n",
    "# for feature in gff_file:\n",
    "    \n",
    "#     # 'feature' is an HTSeq.GenomicFeature object, which describes a\n",
    "#     # line in the GFF file. It has several slots, for example, 'iv'\n",
    "#     # for the interval (where it is on the genome), 'type' for its\n",
    "#     # type (3rd column in the GFF file) and 'attr' for the\n",
    "#     # attributes (last column)\n",
    "    \n",
    "#     # We are only interested in lines describing exons:\n",
    "#     if feature.type == \"exon\":\n",
    "        \n",
    "#         # In this GFF file, the ID of the transcript to which the exons\n",
    "#         # belongs is stored in an attribute (i.e., something in the 9th\n",
    "#         # column) called \"Parent\"\n",
    "#         transcript_id = feature.attr['Parent']\n",
    "        \n",
    "#         # Is this the first time we see this transcript?\n",
    "#         if transcript_id not in transcript_dict:\n",
    "            \n",
    "#             # If so, add to the 'transcripts' dict an empty list \n",
    "#             # store store the exons in\n",
    "#             transcript_dict[ transcript_id ] = list()\n",
    "\n",
    "#         # Add the exon to the list\n",
    "#         transcript_dict[ transcript_id ].append( feature )\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Now, go through the transcripts and calculate the length of each\n",
    "# # by adding up the lengths of its exons\n",
    "# for transcript_id in sorted( transcript_dict ):\n",
    "    \n",
    "#     transcript_length = 0\n",
    "#     # Go through the GenomicFeature\n",
    "#     for exon in transcript_dict[ transcript_id ]:\n",
    "        \n",
    "#         # Each GenomicFeature object has a slot called 'iv', which is a\n",
    "#         # GenomicInterval object, which in turn contains slots such as\n",
    "#         # 'chrom', 'start', 'end', 'strand', 'length', etc., to describe\n",
    "#         # where the feature sits in the genome. We use 'length' here.\n",
    "#         transcript_length += exon.iv.length + 1\n",
    "#         # We add the 1 because this GFF file does no\n",
    "#         # Now, 'transcript_length' is the sum of the lengths of all the\n",
    "#         # transcripts exon.\n",
    "        \n",
    "#     # Print this result, and also add the number of exons.\n",
    "#     # This is just the length of the list of exons for the transcript):\n",
    "#     #print(transcript_id, transcript_length, len( transcripts[ transcript_id ] ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "#transcript_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (heterochromatin)",
   "language": "python",
   "name": "heterochromatin"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
