{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from functools import reduce"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.options.mode.chained_assignment = None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---------------------------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "source_dir = '/data/parastou/RNAdeg/data/AllRNA/'\n",
    "#source_dir = '/data/RNAdeg/fastq/RNA/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "#out_dir = '/data/parastou/RNAdeg/results/RipRna/'\n",
    "out_dir = '/data/pablo/RNAdeg/results/RipRna/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "--------------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Workflow"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Unzip fastq files.\n",
    "- Truncate 0-length reads from fastq files.\n",
    "- Run STAR aligner.\n",
    "- Index .bam files.\n",
    "- Remove rRNA reads from .bam and add 'GE' tag to reads which map to genes (RemrRNA).\n",
    "- Calculate and save gene counts table for all given samples from .bam files.\n",
    "- TPM-normalize gene counts table and save."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_sample_names(names, col_name=None):\n",
    "    \n",
    "    ## select col_name\n",
    "    if col_name is None:\n",
    "        if isinstance(names, pd.DataFrame) and names.shape[0] == 1:\n",
    "            col_name = names.columns[0]\n",
    "        else:\n",
    "            col_name = \"sample_name\"\n",
    "    \n",
    "    ## get dataframe\n",
    "    if not isinstance(names, pd.DataFrame):\n",
    "\n",
    "        ## create dataframe from list\n",
    "        df = pd.DataFrame(data = names, columns=[col_name])\n",
    "        \n",
    "    else:\n",
    "        \n",
    "        ## already a dataframe\n",
    "        df = names\n",
    "        \n",
    "    ## sort dataframe by col_name\n",
    "    df = df.sort_values(by=[col_name]).reset_index(drop=True)\n",
    "        \n",
    "    ## get sample's prefix\n",
    "    df[\"prefix\"] = df[col_name].map(lambda x: x.split(\".\")[0])\n",
    "    \n",
    "    ## check for \"duplicated\" files (share same prefix) within the same directory.\n",
    "    ## one should be careful and wonder why is that?\n",
    "    n_samples = df.shape[0]\n",
    "    df = df[~df.duplicated(subset=\"prefix\", keep=False)] ## remove entries with duplicated prefix\n",
    "    \n",
    "    #assert(n_samples == df.shape[0])\n",
    "    if n_samples != df.shape[0]:\n",
    "        print(\"\\n{}\".format(\"-\"*99))\n",
    "        print(\" Warning! Duplicated files (share same prefix) within the same directory. \\n Will be ignored for now!\")        \n",
    "        print(\"{}\\n\".format(\"-\"*99))\n",
    "\n",
    "    #print(\"Number of files in `{}`: {}\".format(col_name, df.shape[0]) )\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 0 - Investigate Samples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A. __Load `valid_samples.txt` file__ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of valid samples (as given by Parastou): 54\n"
     ]
    }
   ],
   "source": [
    "valid_samples = pd.read_csv(source_dir + 'valid_samples.txt', header=None, names = [\"valid_sample\"])\n",
    "valid_samples = process_sample_names(valid_samples, col_name=\"valid_sample\")\n",
    "print(\"Number of valid samples (as given by Parastou): {}\".format(valid_samples.shape[0]) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "#valid_samples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "B. __Load `remote_RNA_file_names.txt` file__ "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of remote samples (as present in /data/cryohalic01/home/ag_halic/share/Conny/fastq_for_Stefan_RNAdeg): 32\n"
     ]
    }
   ],
   "source": [
    "remote_samples = pd.read_csv(source_dir + 'remote_RNA_file_names.txt', header=None, names = [\"remote_sample\"])\n",
    "remote_samples = process_sample_names(remote_samples, col_name=\"remote_sample\")\n",
    "remote_dir = \"/data/cryohalic01/home/ag_halic/share/Conny/fastq_for_Stefan_RNAdeg\"\n",
    "print(\"Number of remote samples (as present in {}): {}\".format(remote_dir, remote_samples.shape[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "#remote_samples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "C. __Check in `source_dir` for sample files ('bz2', '.fastq', '.fastqsanger')__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of sample files in `source_dir` (/data/parastou/RNAdeg/data/AllRNA/): 66\n"
     ]
    }
   ],
   "source": [
    "file_formats = ('bz2', '.fastq', '.fastqsanger')\n",
    "source_samples_names = [ff for ff in os.listdir(source_dir) if ff.endswith(file_formats)]\n",
    "#source_samples_names = pd.read_csv(source_dir + 'sample_names.txt', header=None, names = [\"source_sample\"])\n",
    "source_samples = process_sample_names(source_samples_names, col_name=\"source_sample\")\n",
    "print(\"Number of sample files in `source_dir` ({}): {}\".format(source_dir, source_samples.shape[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "#source_samples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- This means there are __\"duplicated\" files__ (share same prefix) within the same directory. One should be careful and wonder __why is that__?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert(len(source_samples_names) == source_samples.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "duplicated_files = sorted(list(set(source_samples_names).difference(source_samples.source_sample)))\n",
    "duplicated_files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "D. __Merge (Samples) DataFrames: [`source_dir`, `valid_samples.txt`, `remote_RNA_file_names.txt`]__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "# compile the list of dataframes you want to merge\n",
    "data_frames = [valid_samples, remote_samples, source_samples]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_merged = reduce(lambda left, right: pd.merge(left,right,on=['prefix'], how='outer'), data_frames)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sort\n",
    "df_merged = df_merged.sort_values(\"prefix\").reset_index(drop=True)\n",
    "# shuffle columns\n",
    "df_merged = df_merged[[\"prefix\", \"valid_sample\", \"remote_sample\", \"source_sample\"]]\n",
    "# set \"prefix\" as index\n",
    "df_merged = df_merged.set_index(\"prefix\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "67"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_merged.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_merged"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### __Compare  `source_dir` vs `valid_samples.txt`__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "select_cols = [\"valid_sample\", \"source_sample\"]\n",
    "source_vs_valid_samples = df_merged[select_cols]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This are samples that are either __not valid__ or __missing__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "14"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "source_vs_valid_samples = source_vs_valid_samples[source_vs_valid_samples[\"valid_sample\"].isna() != source_vs_valid_samples[\"source_sample\"].isna()]\n",
    "len(source_vs_valid_samples)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- __Not valid__ samples (present in `source_dir` but not in `valid_samples.txt`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of NOT valid samples: 13\n"
     ]
    }
   ],
   "source": [
    "not_valid_samples = source_vs_valid_samples[source_vs_valid_samples.valid_sample.isna()]\n",
    "print(\"Number of NOT valid samples:\", len(not_valid_samples))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#not_valid_samples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- __Missing__ samples (present in `valid_samples.txt` but not in `source_dir`): Ideally this should be 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of missing samples (e.g. valid but are not in the `source_dir`: 1\n"
     ]
    }
   ],
   "source": [
    "missing_samples = source_vs_valid_samples[source_vs_valid_samples.source_sample.isna()]\n",
    "print(\"Number of missing samples (e.g. valid but are not in the `source_dir`:\", len(missing_samples))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- We see that some of the `valid_sample`s (7) seem to be missing an \"A\" at the end.\n",
    "- (1) `valid_sample` \"302_S2RIP_2\" is also missing but this the sample that we ignore because is raising an error."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>valid_sample</th>\n",
       "      <th>source_sample</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prefix</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>302_S2RIP_2</th>\n",
       "      <td>302_S2RIP_2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            valid_sample source_sample\n",
       "prefix                                \n",
       "302_S2RIP_2  302_S2RIP_2           NaN"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "missing_samples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### __Compare  `source_dir` vs `remote_RNA_file_names.txt`__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "select_cols = [\"source_sample\", \"remote_sample\"]\n",
    "source_vs_remote_samples = df_merged[select_cols]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This are samples that are either present in __missing from source_dir__ or __missing from remote_dir__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "36"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "source_vs_remote_samples = source_vs_remote_samples[source_vs_remote_samples[\"source_sample\"].isna() != source_vs_remote_samples[\"remote_sample\"].isna()]\n",
    "len(source_vs_remote_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "#source_vs_remote_samples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- __Missing from remote_dir__ samples (present in `source_dir` but not in `remote_RNA_file_names.txt`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of samples missing in the source_dir: 35\n"
     ]
    }
   ],
   "source": [
    "missing_from_remote = source_vs_remote_samples[source_vs_remote_samples.remote_sample.isna()]\n",
    "print(\"Number of samples missing in the source_dir:\", len(missing_from_remote))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['1167_S5RIP_2',\n",
       " '283_RNA_pA_4',\n",
       " '301_RNA_pA_3',\n",
       " '301_S2RIP_3',\n",
       " '302_S2RIP_3',\n",
       " '324S2RIP_1',\n",
       " '324_RNA_pA_3',\n",
       " '324_S2RIP_3',\n",
       " '491S2RIP_1',\n",
       " '491_S2RIP_3',\n",
       " '504S2RIP_1',\n",
       " '504S2RIP_2',\n",
       " '504_RNA_pA_1',\n",
       " '504_RNA_pA_2',\n",
       " '530S2RIP_1',\n",
       " '530S2RIP_2',\n",
       " '530_RNA_pA_1',\n",
       " '530_RNA_pA_2',\n",
       " '591_S5RIP_1',\n",
       " '63',\n",
       " '638S2RIP_1',\n",
       " '638S2RIP_2',\n",
       " '638_RNA_pA_1',\n",
       " '638_RNA_pA_2',\n",
       " '63_RIPS5P',\n",
       " '63_RNA_pA_3',\n",
       " '63_RNA_pA_4',\n",
       " '63_S2PRIP',\n",
       " '63_S2Ph_RIP',\n",
       " '63_S2RIP_2',\n",
       " '63_S5Ph_RIP',\n",
       " '65',\n",
       " '80S2RIP_1',\n",
       " '80S2RIP_2',\n",
       " '80pARNA_2']"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "missing_from_remote.index.tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- __Missing from source_dir__ samples (present in `remote_RNA_file_names.txt` but not in `source_dir`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of missing samples (e.g. valid but are not in the `source_dir`: 1\n"
     ]
    }
   ],
   "source": [
    "missing_from_source = source_vs_remote_samples[source_vs_remote_samples.source_sample.isna()]\n",
    "print(\"Number of missing samples (e.g. valid but are not in the `source_dir`:\", len(missing_from_source))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- (1) `remote_sample` \"302_S2RIP_2\" is missing but this the sample that we ignore because is raising an error."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source_sample</th>\n",
       "      <th>remote_sample</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>prefix</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>302_S2RIP_2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>302_S2RIP_2.fastq.bz2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            source_sample          remote_sample\n",
       "prefix                                          \n",
       "302_S2RIP_2           NaN  302_S2RIP_2.fastq.bz2"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "missing_from_source"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Samples used in Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Errors__:\n",
    "-  302_S2RIP_2.fastq\n",
    "-  63_RNA_pA_3.fastq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Ignore samples that are not valid, also ignore samples that give errors__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ignore_files = set([\"302_S2RIP_2\", \"63_RNA_pA_3\"])\n",
    "ignore_files = set([\"302_S2RIP_2\"]) ## we already removed the file from directory\n",
    "## union\n",
    "ignore_files = set(not_valid_samples.index | ignore_files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of samples that will be ignored: 22\n"
     ]
    }
   ],
   "source": [
    "ignore_files = sorted([ff + \".\"  for ff in ignore_files])\n",
    "print(\"Number of samples that will be ignored:\", len(ignore_files))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "#ignore_files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Samples used in Analysis are present in `source_dir` and `valid_samples.txt`__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of samples that will be analyzed: 46\n"
     ]
    }
   ],
   "source": [
    "prefix_files = df_merged[(~df_merged.source_sample.isna()) & (~df_merged.valid_sample.isna())].index.tolist()\n",
    "print(\"Total number of samples that will be analyzed:\", len(prefix_files))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "#prefix_files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1 - Prepare output folders"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "out_bam = os.path.join(out_dir, 'bams')\n",
    "out_tagged = os.path.join(out_dir, 'tagged_bams')\n",
    "xp_data = os.path.join(out_dir, 'xp_data')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.isdir(out_bam):    \n",
    "    #!mkdir $out_bam\n",
    "    !mkdir -p $out_bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.isdir(out_tagged):    \n",
    "    #!mkdir $out_tagged\n",
    "    !mkdir -p $out_tagged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.isdir(xp_data):    \n",
    "    #!mkdir $xp_data\n",
    "    !mkdir -p $xp_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-----------------------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2 - Unzip fastq files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "for filename in os.listdir(source_dir):\n",
    "    if filename.endswith('bz2'):\n",
    "        cat('Decompressing file ...\\n', filename)\n",
    "        filepath = os.path.join(source_dir, filename)\n",
    "        !bzip2 -d $filepath"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Optional step: remove truncated reads from fastq files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for filename in os.listdir(source_dir):\n",
    "#     if filename.endswith('.fastq'):\n",
    "        \n",
    "#         filepath = os.path.join(source_dir, filename)\n",
    "#         outfilepath = os.path.join(source_dir, filename.split('.fastq')[0] + '.ztr.fastq')\n",
    "#         !echo $outfilepath\n",
    "#         !bioawk -cfastx 'length($seq) > 0 {print \"@\"$name\"\\n\"$seq\"\\n+\\n\"$qual} $filepath  >> $outfilepath'\n",
    "#         #!rm $filepath"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3 - Align fastq files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set STAR runtime parameters.\n",
    "def star_command(in_file):\n",
    "    \n",
    "    star = '/home/parastou/star_2.5.1b'\n",
    "    n_threads = 20\n",
    "    genome_dir = '/data/parastou/RNAdeg/genomes/spombe/staridx/'\n",
    "    \n",
    "    command = star + ' --runThreadN ' + str(n_threads) + ' --genomeDir ' + genome_dir + ' --readFilesIn ' \\\n",
    "    + source_dir + in_file + ' --outFileNamePrefix ' + os.path.join(out_bam , in_file.split('fastq')[0]) \\\n",
    "    + ' --outSAMtype BAM SortedByCoordinate'\n",
    "    \n",
    "    return command"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Errors__:\n",
    "-  302_S2RIP_2.fastq:\n",
    "    - EXITING because of fatal ERROR: not enough memory for BAM sorting: \n",
    "        - SOLUTION: re-run STAR with at least --limitBAMsortRAM 2821915860\n",
    "-  63_RNA_pA_3.fastq\n",
    "    - EXITING because of FATAL ERROR in reads input: short read sequence line: 1 \n",
    "        - Read Name=@L183:414:CCGY1ANXX:4:1311:17580:12058:1:N:0:0:CACGATAT\n",
    "        - Read Sequence====\n",
    "        - DEF_readNameLengthMax=50000\n",
    "        - DEF_readSeqLengthMax=500\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1167_pA_2.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/1167_pA_2.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/1167_pA_2. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 11:31:14 ..... Started STAR run\n",
      "Nov 20 11:31:14 ..... Loading genome\n",
      "Nov 20 11:31:25 ..... Started mapping\n",
      "Nov 20 11:32:27 ..... Started sorting BAM\n",
      "Nov 20 11:32:41 ..... Finished successfully\n",
      "63.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/63.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/63. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 11:32:42 ..... Started STAR run\n",
      "Nov 20 11:32:42 ..... Loading genome\n",
      "Nov 20 11:32:43 ..... Started mapping\n",
      "Nov 20 11:33:33 ..... Started sorting BAM\n",
      "Nov 20 11:33:49 ..... Finished successfully\n",
      "65.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/65.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/65. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 11:33:50 ..... Started STAR run\n",
      "Nov 20 11:33:50 ..... Loading genome\n",
      "Nov 20 11:33:51 ..... Started mapping\n",
      "Nov 20 11:38:13 ..... Started sorting BAM\n",
      "Nov 20 11:38:26 ..... Finished successfully\n",
      "63_RIPS5P.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/63_RIPS5P.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/63_RIPS5P. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 11:38:27 ..... Started STAR run\n",
      "Nov 20 11:38:27 ..... Loading genome\n",
      "Nov 20 11:38:27 ..... Started mapping\n",
      "Nov 20 11:42:21 ..... Started sorting BAM\n",
      "Nov 20 11:46:37 ..... Finished successfully\n",
      "63_S2Ph_RIP.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/63_S2Ph_RIP.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/63_S2Ph_RIP. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 11:46:38 ..... Started STAR run\n",
      "Nov 20 11:46:38 ..... Loading genome\n",
      "Nov 20 11:46:39 ..... Started mapping\n",
      "Nov 20 11:47:56 ..... Started sorting BAM\n",
      "Nov 20 11:48:51 ..... Finished successfully\n",
      "63_S2PRIP.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/63_S2PRIP.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/63_S2PRIP. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 11:48:53 ..... Started STAR run\n",
      "Nov 20 11:48:53 ..... Loading genome\n",
      "Nov 20 11:48:53 ..... Started mapping\n",
      "Nov 20 11:51:58 ..... Started sorting BAM\n",
      "Nov 20 11:54:18 ..... Finished successfully\n",
      "63_S2RIP_2.ztr.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/63_S2RIP_2.ztr.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/63_S2RIP_2.ztr. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 11:54:20 ..... Started STAR run\n",
      "Nov 20 11:54:20 ..... Loading genome\n",
      "Nov 20 11:54:20 ..... Started mapping\n",
      "Nov 20 11:55:48 ..... Started sorting BAM\n",
      "Nov 20 11:56:47 ..... Finished successfully\n",
      "63_S5Ph_RIP.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/63_S5Ph_RIP.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/63_S5Ph_RIP. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 11:56:48 ..... Started STAR run\n",
      "Nov 20 11:56:48 ..... Loading genome\n",
      "Nov 20 11:56:49 ..... Started mapping\n",
      "Nov 20 11:58:50 ..... Started sorting BAM\n",
      "Nov 20 11:59:04 ..... Finished successfully\n",
      "80_S2RIP.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/80_S2RIP.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/80_S2RIP. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 11:59:05 ..... Started STAR run\n",
      "Nov 20 11:59:05 ..... Loading genome\n",
      "Nov 20 11:59:06 ..... Started mapping\n",
      "Nov 20 12:00:31 ..... Started sorting BAM\n",
      "Nov 20 12:01:40 ..... Finished successfully\n",
      "80S2RIP_1.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/80S2RIP_1.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/80S2RIP_1. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 12:01:41 ..... Started STAR run\n",
      "Nov 20 12:01:41 ..... Loading genome\n",
      "Nov 20 12:01:42 ..... Started mapping\n",
      "Nov 20 12:03:43 ..... Started sorting BAM\n",
      "Nov 20 12:04:24 ..... Finished successfully\n",
      "80S2RIP_2.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/80S2RIP_2.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/80S2RIP_2. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 12:04:25 ..... Started STAR run\n",
      "Nov 20 12:04:25 ..... Loading genome\n",
      "Nov 20 12:04:25 ..... Started mapping\n",
      "Nov 20 12:06:44 ..... Started sorting BAM\n",
      "Nov 20 12:07:43 ..... Finished successfully\n",
      "301_S2RIP_2.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/301_S2RIP_2.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/301_S2RIP_2. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 12:07:44 ..... Started STAR run\n",
      "Nov 20 12:07:44 ..... Loading genome\n",
      "Nov 20 12:07:44 ..... Started mapping\n",
      "Nov 20 12:09:35 ..... Started sorting BAM\n",
      "Nov 20 12:10:40 ..... Finished successfully\n",
      "301_S2RIP_3.ztr.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/301_S2RIP_3.ztr.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/301_S2RIP_3.ztr. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 12:10:42 ..... Started STAR run\n",
      "Nov 20 12:10:42 ..... Loading genome\n",
      "Nov 20 12:10:42 ..... Started mapping\n",
      "Nov 20 12:13:09 ..... Started sorting BAM\n",
      "Nov 20 12:14:31 ..... Finished successfully\n",
      "301_S2RIP.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/301_S2RIP.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/301_S2RIP. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 12:14:32 ..... Started STAR run\n",
      "Nov 20 12:14:32 ..... Loading genome\n",
      "Nov 20 12:14:33 ..... Started mapping\n",
      "Nov 20 12:16:27 ..... Started sorting BAM\n",
      "Nov 20 12:17:33 ..... Finished successfully\n",
      "302_S2RIP_2.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/302_S2RIP_2.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/302_S2RIP_2. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 12:17:35 ..... Started STAR run\n",
      "Nov 20 12:17:35 ..... Loading genome\n",
      "Nov 20 12:17:35 ..... Started mapping\n",
      "Nov 20 12:23:51 ..... Started sorting BAM\n",
      "\n",
      "EXITING because of fatal ERROR: not enough memory for BAM sorting: \n",
      "SOLUTION: re-run STAR with at least --limitBAMsortRAM 2821915860\n",
      "Nov 20 12:23:51 ...... FATAL ERROR, exiting\n",
      "302_S2RIP_3.ztr.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/302_S2RIP_3.ztr.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/302_S2RIP_3.ztr. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 12:23:51 ..... Started STAR run\n",
      "Nov 20 12:23:51 ..... Loading genome\n",
      "Nov 20 12:23:54 ..... Started mapping\n",
      "Nov 20 12:26:18 ..... Started sorting BAM\n",
      "Nov 20 12:27:24 ..... Finished successfully\n",
      "302_S2RIP.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/302_S2RIP.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/302_S2RIP. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 12:27:26 ..... Started STAR run\n",
      "Nov 20 12:27:26 ..... Loading genome\n",
      "Nov 20 12:27:29 ..... Started mapping\n",
      "Nov 20 12:29:40 ..... Started sorting BAM\n",
      "Nov 20 12:30:37 ..... Finished successfully\n",
      "324_S2RIP_2.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/324_S2RIP_2.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/324_S2RIP_2. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 12:30:38 ..... Started STAR run\n",
      "Nov 20 12:30:38 ..... Loading genome\n",
      "Nov 20 12:30:39 ..... Started mapping\n",
      "Nov 20 12:30:56 ..... Started sorting BAM\n",
      "Nov 20 12:31:02 ..... Finished successfully\n",
      "324_S2RIP_3.ztr.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/324_S2RIP_3.ztr.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/324_S2RIP_3.ztr. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 12:31:02 ..... Started STAR run\n",
      "Nov 20 12:31:02 ..... Loading genome\n",
      "Nov 20 12:31:03 ..... Started mapping\n",
      "Nov 20 12:33:26 ..... Started sorting BAM\n",
      "Nov 20 12:34:49 ..... Finished successfully\n",
      "324_S2RIP.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/324_S2RIP.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/324_S2RIP. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 12:34:50 ..... Started STAR run\n",
      "Nov 20 12:34:50 ..... Loading genome\n",
      "Nov 20 12:34:51 ..... Started mapping\n",
      "Nov 20 12:36:17 ..... Started sorting BAM\n",
      "Nov 20 12:37:20 ..... Finished successfully\n",
      "324S2RIP_1.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/324S2RIP_1.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/324S2RIP_1. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 12:37:21 ..... Started STAR run\n",
      "Nov 20 12:37:21 ..... Loading genome\n",
      "Nov 20 12:37:22 ..... Started mapping\n",
      "Nov 20 12:40:30 ..... Started sorting BAM\n",
      "Nov 20 12:40:55 ..... Finished successfully\n",
      "491_S2RIP_2.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/491_S2RIP_2.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/491_S2RIP_2. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 12:40:56 ..... Started STAR run\n",
      "Nov 20 12:40:56 ..... Loading genome\n",
      "Nov 20 12:40:57 ..... Started mapping\n",
      "Nov 20 12:42:44 ..... Started sorting BAM\n",
      "Nov 20 12:43:44 ..... Finished successfully\n",
      "491_S2RIP_3.ztr.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/491_S2RIP_3.ztr.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/491_S2RIP_3.ztr. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 12:43:45 ..... Started STAR run\n",
      "Nov 20 12:43:45 ..... Loading genome\n",
      "Nov 20 12:43:46 ..... Started mapping\n",
      "Nov 20 12:45:08 ..... Started sorting BAM\n",
      "Nov 20 12:45:58 ..... Finished successfully\n",
      "491_S2RIP.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/491_S2RIP.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/491_S2RIP. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 12:45:59 ..... Started STAR run\n",
      "Nov 20 12:45:59 ..... Loading genome\n",
      "Nov 20 12:46:00 ..... Started mapping\n",
      "Nov 20 12:48:18 ..... Started sorting BAM\n",
      "Nov 20 12:51:14 ..... Finished successfully\n",
      "491S2RIP_1.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/491S2RIP_1.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/491S2RIP_1. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 12:51:16 ..... Started STAR run\n",
      "Nov 20 12:51:16 ..... Loading genome\n",
      "Nov 20 12:51:17 ..... Started mapping\n",
      "Nov 20 12:54:07 ..... Started sorting BAM\n",
      "Nov 20 12:54:24 ..... Finished successfully\n",
      "504S2RIP_1.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/504S2RIP_1.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/504S2RIP_1. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 12:54:26 ..... Started STAR run\n",
      "Nov 20 12:54:26 ..... Loading genome\n",
      "Nov 20 12:54:26 ..... Started mapping\n",
      "Nov 20 12:55:57 ..... Started sorting BAM\n",
      "Nov 20 12:56:51 ..... Finished successfully\n",
      "504S2RIP_2.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/504S2RIP_2.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/504S2RIP_2. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 12:56:52 ..... Started STAR run\n",
      "Nov 20 12:56:52 ..... Loading genome\n",
      "Nov 20 12:56:53 ..... Started mapping\n",
      "Nov 20 12:58:27 ..... Started sorting BAM\n",
      "Nov 20 12:59:20 ..... Finished successfully\n",
      "510_S2RIP.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/510_S2RIP.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/510_S2RIP. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 12:59:21 ..... Started STAR run\n",
      "Nov 20 12:59:21 ..... Loading genome\n",
      "Nov 20 12:59:22 ..... Started mapping\n",
      "Nov 20 13:11:21 ..... Started sorting BAM\n",
      "Nov 20 13:12:31 ..... Finished successfully\n",
      "530S2RIP_1.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/530S2RIP_1.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/530S2RIP_1. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 13:12:33 ..... Started STAR run\n",
      "Nov 20 13:12:33 ..... Loading genome\n",
      "Nov 20 13:12:33 ..... Started mapping\n",
      "Nov 20 13:13:37 ..... Started sorting BAM\n",
      "Nov 20 13:14:14 ..... Finished successfully\n",
      "530S2RIP_2.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/530S2RIP_2.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/530S2RIP_2. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 13:14:15 ..... Started STAR run\n",
      "Nov 20 13:14:15 ..... Loading genome\n",
      "Nov 20 13:14:15 ..... Started mapping\n",
      "Nov 20 13:15:43 ..... Started sorting BAM\n",
      "Nov 20 13:16:25 ..... Finished successfully\n",
      "591_S2RIP_2.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/591_S2RIP_2.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/591_S2RIP_2. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 13:16:26 ..... Started STAR run\n",
      "Nov 20 13:16:26 ..... Loading genome\n",
      "Nov 20 13:16:27 ..... Started mapping\n",
      "Nov 20 13:18:29 ..... Started sorting BAM\n",
      "Nov 20 13:19:34 ..... Finished successfully\n",
      "591_S5RIP_1.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/591_S5RIP_1.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/591_S5RIP_1. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 13:19:35 ..... Started STAR run\n",
      "Nov 20 13:19:35 ..... Loading genome\n",
      "Nov 20 13:19:36 ..... Started mapping\n",
      "Nov 20 13:22:10 ..... Started sorting BAM\n",
      "Nov 20 13:23:10 ..... Finished successfully\n",
      "638S2RIP_1.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/638S2RIP_1.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/638S2RIP_1. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 13:23:11 ..... Started STAR run\n",
      "Nov 20 13:23:11 ..... Loading genome\n",
      "Nov 20 13:23:12 ..... Started mapping\n",
      "Nov 20 13:24:37 ..... Started sorting BAM\n",
      "Nov 20 13:25:13 ..... Finished successfully\n",
      "638S2RIP_2.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/638S2RIP_2.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/638S2RIP_2. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 13:25:16 ..... Started STAR run\n",
      "Nov 20 13:25:16 ..... Loading genome\n",
      "Nov 20 13:25:17 ..... Started mapping\n",
      "Nov 20 13:26:21 ..... Started sorting BAM\n",
      "Nov 20 13:26:58 ..... Finished successfully\n",
      "1113_S2RIP.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/1113_S2RIP.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/1113_S2RIP. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 13:26:59 ..... Started STAR run\n",
      "Nov 20 13:26:59 ..... Loading genome\n",
      "Nov 20 13:27:00 ..... Started mapping\n",
      "Nov 20 13:28:17 ..... Started sorting BAM\n",
      "Nov 20 13:28:48 ..... Finished successfully\n",
      "1167_S2RIP_2.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/1167_S2RIP_2.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/1167_S2RIP_2. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 13:28:48 ..... Started STAR run\n",
      "Nov 20 13:28:48 ..... Loading genome\n",
      "Nov 20 13:28:49 ..... Started mapping\n",
      "Nov 20 13:31:39 ..... Started sorting BAM\n",
      "Nov 20 13:33:59 ..... Finished successfully\n",
      "1167_S2RIP.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/1167_S2RIP.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/1167_S2RIP. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 13:34:00 ..... Started STAR run\n",
      "Nov 20 13:34:00 ..... Loading genome\n",
      "Nov 20 13:34:01 ..... Started mapping\n",
      "Nov 20 13:35:50 ..... Started sorting BAM\n",
      "Nov 20 13:36:42 ..... Finished successfully\n",
      "1167_S5RIP_2.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/1167_S5RIP_2.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/1167_S5RIP_2. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 13:36:43 ..... Started STAR run\n",
      "Nov 20 13:36:43 ..... Loading genome\n",
      "Nov 20 13:36:44 ..... Started mapping\n",
      "Nov 20 13:39:02 ..... Started sorting BAM\n",
      "Nov 20 13:41:20 ..... Finished successfully\n",
      "1168_S2RIP_2.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/1168_S2RIP_2.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/1168_S2RIP_2. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 13:41:21 ..... Started STAR run\n",
      "Nov 20 13:41:21 ..... Loading genome\n",
      "Nov 20 13:41:22 ..... Started mapping\n",
      "Nov 20 13:45:14 ..... Started sorting BAM\n",
      "Nov 20 13:47:59 ..... Finished successfully\n",
      "1168_S2RIP.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/1168_S2RIP.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/1168_S2RIP. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 13:48:01 ..... Started STAR run\n",
      "Nov 20 13:48:01 ..... Loading genome\n",
      "Nov 20 13:48:02 ..... Started mapping\n",
      "Nov 20 13:50:04 ..... Started sorting BAM\n",
      "Nov 20 13:51:08 ..... Finished successfully\n",
      "324_RNA_pA_3.ztr.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/324_RNA_pA_3.ztr.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/324_RNA_pA_3.ztr. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 13:51:09 ..... Started STAR run\n",
      "Nov 20 13:51:09 ..... Loading genome\n",
      "Nov 20 13:51:10 ..... Started mapping\n",
      "Nov 20 13:51:43 ..... Started sorting BAM\n",
      "Nov 20 13:52:05 ..... Finished successfully\n",
      "1168_pA_2.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/1168_pA_2.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/1168_pA_2. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 13:52:05 ..... Started STAR run\n",
      "Nov 20 13:52:05 ..... Loading genome\n",
      "Nov 20 13:52:06 ..... Started mapping\n",
      "Nov 20 13:55:02 ..... Started sorting BAM\n",
      "Nov 20 13:55:16 ..... Finished successfully\n",
      "63_RNA_pA_3.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/63_RNA_pA_3.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/63_RNA_pA_3. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 13:55:17 ..... Started STAR run\n",
      "Nov 20 13:55:17 ..... Loading genome\n",
      "Nov 20 13:55:17 ..... Started mapping\n",
      "\n",
      "EXITING because of FATAL ERROR in reads input: short read sequence line: 1\n",
      "Read Name=@L183:414:CCGY1ANXX:4:1311:17580:12058:1:N:0:0:CACGATAT\n",
      "Read Sequence====\n",
      "DEF_readNameLengthMax=50000\n",
      "DEF_readSeqLengthMax=500\n",
      "\n",
      "Nov 20 13:55:18 ...... FATAL ERROR, exiting\n",
      "1113_pA.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/1113_pA.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/1113_pA. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 13:55:18 ..... Started STAR run\n",
      "Nov 20 13:55:18 ..... Loading genome\n",
      "Nov 20 13:55:19 ..... Started mapping\n",
      "Nov 20 13:56:27 ..... Started sorting BAM\n",
      "Nov 20 13:56:58 ..... Finished successfully\n",
      "1167_pA.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/1167_pA.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/1167_pA. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 13:56:59 ..... Started STAR run\n",
      "Nov 20 13:56:59 ..... Loading genome\n",
      "Nov 20 13:57:00 ..... Started mapping\n",
      "Nov 20 13:58:21 ..... Started sorting BAM\n",
      "Nov 20 13:58:57 ..... Finished successfully\n",
      "1168_pA.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/1168_pA.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/1168_pA. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 13:58:57 ..... Started STAR run\n",
      "Nov 20 13:58:57 ..... Loading genome\n",
      "Nov 20 13:58:58 ..... Started mapping\n",
      "Nov 20 13:59:07 ..... Started sorting BAM\n",
      "Nov 20 13:59:10 ..... Finished successfully\n",
      "510_RNA_pA_2.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/510_RNA_pA_2.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/510_RNA_pA_2. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 13:59:11 ..... Started STAR run\n",
      "Nov 20 13:59:11 ..... Loading genome\n",
      "Nov 20 13:59:11 ..... Started mapping\n",
      "Nov 20 14:00:02 ..... Started sorting BAM\n",
      "Nov 20 14:00:18 ..... Finished successfully\n",
      "301_RNA_pA_2.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/301_RNA_pA_2.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/301_RNA_pA_2. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 14:00:19 ..... Started STAR run\n",
      "Nov 20 14:00:19 ..... Loading genome\n",
      "Nov 20 14:00:19 ..... Started mapping\n",
      "Nov 20 14:01:11 ..... Started sorting BAM\n",
      "Nov 20 14:01:24 ..... Finished successfully\n",
      "302_RNA_pA_2.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/302_RNA_pA_2.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/302_RNA_pA_2. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 14:01:25 ..... Started STAR run\n",
      "Nov 20 14:01:25 ..... Loading genome\n",
      "Nov 20 14:01:25 ..... Started mapping\n",
      "Nov 20 14:02:00 ..... Started sorting BAM\n",
      "Nov 20 14:02:20 ..... Finished successfully\n",
      "324_RNA_pA_2.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/324_RNA_pA_2.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/324_RNA_pA_2. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 14:02:20 ..... Started STAR run\n",
      "Nov 20 14:02:20 ..... Loading genome\n",
      "Nov 20 14:02:21 ..... Started mapping\n",
      "Nov 20 14:03:07 ..... Started sorting BAM\n",
      "Nov 20 14:03:21 ..... Finished successfully\n",
      "491_RNA_pA_2.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/491_RNA_pA_2.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/491_RNA_pA_2. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 14:03:21 ..... Started STAR run\n",
      "Nov 20 14:03:21 ..... Loading genome\n",
      "Nov 20 14:03:22 ..... Started mapping\n",
      "Nov 20 14:04:29 ..... Started sorting BAM\n",
      "Nov 20 14:04:47 ..... Finished successfully\n",
      "80pARNA_2.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/80pARNA_2.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/80pARNA_2. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 14:04:52 ..... Started STAR run\n",
      "Nov 20 14:04:52 ..... Loading genome\n",
      "Nov 20 14:04:52 ..... Started mapping\n",
      "Nov 20 14:05:14 ..... Started sorting BAM\n",
      "Nov 20 14:05:31 ..... Finished successfully\n",
      "63_RNA_pA_3.ztr.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/63_RNA_pA_3.ztr.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/63_RNA_pA_3.ztr. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 14:05:35 ..... Started STAR run\n",
      "Nov 20 14:05:35 ..... Loading genome\n",
      "Nov 20 14:05:36 ..... Started mapping\n",
      "Nov 20 14:06:23 ..... Started sorting BAM\n",
      "Nov 20 14:06:36 ..... Finished successfully\n",
      "63_RNA_pA_4.ztr.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/63_RNA_pA_4.ztr.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/63_RNA_pA_4.ztr. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 14:06:37 ..... Started STAR run\n",
      "Nov 20 14:06:37 ..... Loading genome\n",
      "Nov 20 14:06:38 ..... Started mapping\n",
      "Nov 20 14:07:05 ..... Started sorting BAM\n",
      "Nov 20 14:07:27 ..... Finished successfully\n",
      "283_RNA_pA_4.ztr.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/283_RNA_pA_4.ztr.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/283_RNA_pA_4.ztr. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 14:07:27 ..... Started STAR run\n",
      "Nov 20 14:07:27 ..... Loading genome\n",
      "Nov 20 14:07:28 ..... Started mapping\n",
      "Nov 20 14:07:58 ..... Started sorting BAM\n",
      "Nov 20 14:08:22 ..... Finished successfully\n",
      "301_RNA_pA_3.ztr.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/301_RNA_pA_3.ztr.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/301_RNA_pA_3.ztr. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 14:08:22 ..... Started STAR run\n",
      "Nov 20 14:08:22 ..... Loading genome\n",
      "Nov 20 14:08:23 ..... Started mapping\n",
      "Nov 20 14:08:49 ..... Started sorting BAM\n",
      "Nov 20 14:09:04 ..... Finished successfully\n",
      "504_RNA_pA_1.ztr.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/504_RNA_pA_1.ztr.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/504_RNA_pA_1.ztr. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 14:09:08 ..... Started STAR run\n",
      "Nov 20 14:09:08 ..... Loading genome\n",
      "Nov 20 14:09:08 ..... Started mapping\n",
      "Nov 20 14:09:57 ..... Started sorting BAM\n",
      "Nov 20 14:10:10 ..... Finished successfully\n",
      "504_RNA_pA_2.ztr.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/504_RNA_pA_2.ztr.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/504_RNA_pA_2.ztr. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 14:10:11 ..... Started STAR run\n",
      "Nov 20 14:10:11 ..... Loading genome\n",
      "Nov 20 14:10:11 ..... Started mapping\n",
      "Nov 20 14:10:39 ..... Started sorting BAM\n",
      "Nov 20 14:11:00 ..... Finished successfully\n",
      "530_RNA_pA_1.ztr.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/530_RNA_pA_1.ztr.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/530_RNA_pA_1.ztr. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 14:11:00 ..... Started STAR run\n",
      "Nov 20 14:11:00 ..... Loading genome\n",
      "Nov 20 14:11:01 ..... Started mapping\n",
      "Nov 20 14:11:51 ..... Started sorting BAM\n",
      "Nov 20 14:12:06 ..... Finished successfully\n",
      "530_RNA_pA_2.ztr.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/530_RNA_pA_2.ztr.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/530_RNA_pA_2.ztr. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 14:12:06 ..... Started STAR run\n",
      "Nov 20 14:12:06 ..... Loading genome\n",
      "Nov 20 14:12:07 ..... Started mapping\n",
      "Nov 20 14:12:38 ..... Started sorting BAM\n",
      "Nov 20 14:12:55 ..... Finished successfully\n",
      "638_RNA_pA_1.ztr.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/638_RNA_pA_1.ztr.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/638_RNA_pA_1.ztr. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 14:12:58 ..... Started STAR run\n",
      "Nov 20 14:12:58 ..... Loading genome\n",
      "Nov 20 14:12:59 ..... Started mapping\n",
      "Nov 20 14:13:41 ..... Started sorting BAM\n",
      "Nov 20 14:13:59 ..... Finished successfully\n",
      "638_RNA_pA_2.ztr.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/638_RNA_pA_2.ztr.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/638_RNA_pA_2.ztr. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 14:14:00 ..... Started STAR run\n",
      "Nov 20 14:14:00 ..... Loading genome\n",
      "Nov 20 14:14:01 ..... Started mapping\n",
      "Nov 20 14:14:26 ..... Started sorting BAM\n",
      "Nov 20 14:14:39 ..... Finished successfully\n",
      "80_RNA_pA.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/80_RNA_pA.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/80_RNA_pA. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 14:14:43 ..... Started STAR run\n",
      "Nov 20 14:14:43 ..... Loading genome\n",
      "Nov 20 14:14:44 ..... Started mapping\n",
      "Nov 20 14:14:52 ..... Started sorting BAM\n",
      "Nov 20 14:14:56 ..... Finished successfully\n",
      "301_RNA_pA.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/301_RNA_pA.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/301_RNA_pA. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 14:14:56 ..... Started STAR run\n",
      "Nov 20 14:14:56 ..... Loading genome\n",
      "Nov 20 14:14:57 ..... Started mapping\n",
      "Nov 20 14:15:11 ..... Started sorting BAM\n",
      "Nov 20 14:15:17 ..... Finished successfully\n",
      "302_RNA_pA.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/302_RNA_pA.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/302_RNA_pA. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 14:15:17 ..... Started STAR run\n",
      "Nov 20 14:15:17 ..... Loading genome\n",
      "Nov 20 14:15:18 ..... Started mapping\n",
      "Nov 20 14:15:50 ..... Started sorting BAM\n",
      "Nov 20 14:16:09 ..... Finished successfully\n",
      "324_RNA_pA.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/324_RNA_pA.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/324_RNA_pA. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 14:16:11 ..... Started STAR run\n",
      "Nov 20 14:16:11 ..... Loading genome\n",
      "Nov 20 14:16:11 ..... Started mapping\n",
      "Nov 20 14:16:25 ..... Started sorting BAM\n",
      "Nov 20 14:16:30 ..... Finished successfully\n",
      "491_RNA_pA.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/491_RNA_pA.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/491_RNA_pA. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 14:16:30 ..... Started STAR run\n",
      "Nov 20 14:16:30 ..... Loading genome\n",
      "Nov 20 14:16:31 ..... Started mapping\n",
      "Nov 20 14:17:25 ..... Started sorting BAM\n",
      "Nov 20 14:17:41 ..... Finished successfully\n",
      "510_RNA_pA.fastq\n",
      "/home/parastou/star_2.5.1b --runThreadN 20 --genomeDir /data/parastou/RNAdeg/genomes/spombe/staridx/ --readFilesIn /data/parastou/RNAdeg/data/AllRNA/510_RNA_pA.fastq --outFileNamePrefix /data/pablo/RNAdeg/results/RipRna/bams/510_RNA_pA. --outSAMtype BAM SortedByCoordinate\n",
      "Nov 20 14:17:41 ..... Started STAR run\n",
      "Nov 20 14:17:41 ..... Loading genome\n",
      "Nov 20 14:17:42 ..... Started mapping\n",
      "Nov 20 14:18:28 ..... Started sorting BAM\n",
      "Nov 20 14:18:44 ..... Finished successfully\n"
     ]
    }
   ],
   "source": [
    "# Run STAR\n",
    "for filename in os.listdir(source_dir):\n",
    "    \n",
    "    #if filename.endswith(('.fastq', '.fastqsanger')):\n",
    "    if filename.endswith(('.fastq', '.fastqsanger')) and not filename.startswith(ignore_files):\n",
    "   \n",
    "        print(filename)\n",
    "        command = star_command(filename)\n",
    "        print(command)\n",
    "        !$command"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Errors__:\n",
    "-  302_S2RIP_2.fastq:\n",
    "    - samtools index: \"/data/pablo/RNAdeg/results/RipRna/bams/302_S2RIP_2.Aligned.sortedByCoord.out.bam\" is in a format that cannot be usefully indexed\n",
    "\n",
    "-  63_RNA_pA_3.fastq\n",
    "    - samtools index: \"/data/pablo/RNAdeg/results/RipRna/bams/63_RNA_pA_3.Aligned.sortedByCoord.out.bam\" is in a format that cannot be usefully indexed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samtools index: \"/data/pablo/RNAdeg/results/RipRna/bams/302_S2RIP_2.Aligned.sortedByCoord.out.bam\" is in a format that cannot be usefully indexed\n",
      "samtools index: \"/data/pablo/RNAdeg/results/RipRna/bams/63_RNA_pA_3.Aligned.sortedByCoord.out.bam\" is in a format that cannot be usefully indexed\n"
     ]
    }
   ],
   "source": [
    "# Index alignment files.\n",
    "for filename in os.listdir(out_bam):\n",
    " \n",
    "    #if filename.endswith('.Aligned.sortedByCoord.out.bam'):\n",
    "    if filename.endswith('.Aligned.sortedByCoord.out.bam') and not filename.startswith(ignore_files):\n",
    "\n",
    "        filepath = os.path.join(out_bam, filename)\n",
    "        !samtools index $filepath"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4 - Tag bam files with gene IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set RemrRNA command runtime parameters.\n",
    "def remrrna_command():\n",
    "    \n",
    "    remrrna = '/data/parastou/RNAdeg/pyRNAdeg/RemrRNA.py'\n",
    "    annotation = '/data/parastou/RNAdeg/annotation/schizosaccharomyces_pombe.chr.extended.csv'\n",
    "    \n",
    "    command = remrrna + ' -d ' + out_bam + ' -g ' + annotation + ' -o ' + out_tagged\n",
    "    \n",
    "    return command"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO:__main__:Call to RemrRNA module.\n",
      "Call to RemrRNA module.\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:This module removes ribosomal-RNA related reads from given .bam files\n",
      "This module removes ribosomal-RNA related reads from given .bam files\n",
      "INFO:__main__:It produces a gene-tagged .bam file as output.\n",
      "It produces a gene-tagged .bam file as output.\n",
      "INFO:__main__:----------------------------------------\n",
      "----------------------------------------\n",
      "INFO:__main__:Filtering ribosomal-RNA reads and tagging started.\n",
      "Filtering ribosomal-RNA reads and tagging started.\n",
      "INFO:__main__:Source directory:\t/data/pablo/RNAdeg/results/RipRna/bams\n",
      "Source directory:\t/data/pablo/RNAdeg/results/RipRna/bams\n",
      "INFO:__main__:--------------------------------------------------\n",
      "--------------------------------------------------\n",
      "INFO:__main__:Sample name:\t1167_pA_2.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t1167_pA_2.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t19,153,961\n",
      "Total reads:\t\t\t19,153,961\n",
      "INFO:__main__:Spliced reads ignored:\t\t601,077\n",
      "Spliced reads ignored:\t\t601,077\n",
      "INFO:__main__:Total reads mapped to genes:\t17,602,604\n",
      "Total reads mapped to genes:\t17,602,604\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/1167_pA_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/1167_pA_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t63.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t63.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t19,410,169\n",
      "Total reads:\t\t\t19,410,169\n",
      "INFO:__main__:Spliced reads ignored:\t\t546,362\n",
      "Spliced reads ignored:\t\t546,362\n",
      "INFO:__main__:Total reads mapped to genes:\t17,854,944\n",
      "Total reads mapped to genes:\t17,854,944\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/63.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/63.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t65.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t65.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t18,009,267\n",
      "Total reads:\t\t\t18,009,267\n",
      "INFO:__main__:Spliced reads ignored:\t\t497,230\n",
      "Spliced reads ignored:\t\t497,230\n",
      "INFO:__main__:Total reads mapped to genes:\t16,606,288\n",
      "Total reads mapped to genes:\t16,606,288\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/65.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/65.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t63_RIPS5P.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t63_RIPS5P.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t33,202,799\n",
      "Total reads:\t\t\t33,202,799\n",
      "INFO:__main__:Spliced reads ignored:\t\t88,951\n",
      "Spliced reads ignored:\t\t88,951\n",
      "INFO:__main__:Total reads mapped to genes:\t7,883,013\n",
      "Total reads mapped to genes:\t7,883,013\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/63_RIPS5P.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/63_RIPS5P.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t63_S2Ph_RIP.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t63_S2Ph_RIP.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t16,991,670\n",
      "Total reads:\t\t\t16,991,670\n",
      "INFO:__main__:Spliced reads ignored:\t\t37,112\n",
      "Spliced reads ignored:\t\t37,112\n",
      "INFO:__main__:Total reads mapped to genes:\t5,657,208\n",
      "Total reads mapped to genes:\t5,657,208\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/63_S2Ph_RIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/63_S2Ph_RIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t63_S2PRIP.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t63_S2PRIP.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t25,942,299\n",
      "Total reads:\t\t\t25,942,299\n",
      "INFO:__main__:Spliced reads ignored:\t\t85,336\n",
      "Spliced reads ignored:\t\t85,336\n",
      "INFO:__main__:Total reads mapped to genes:\t6,449,064\n",
      "Total reads mapped to genes:\t6,449,064\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/63_S2PRIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/63_S2PRIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t63_S2RIP_2.ztr.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t63_S2RIP_2.ztr.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t15,733,435\n",
      "Total reads:\t\t\t15,733,435\n",
      "INFO:__main__:Spliced reads ignored:\t\t39,900\n",
      "Spliced reads ignored:\t\t39,900\n",
      "INFO:__main__:Total reads mapped to genes:\t1,553,320\n",
      "Total reads mapped to genes:\t1,553,320\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/63_S2RIP_2.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/63_S2RIP_2.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t63_S5Ph_RIP.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t63_S5Ph_RIP.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t6,010,821\n",
      "Total reads:\t\t\t6,010,821\n",
      "INFO:__main__:Spliced reads ignored:\t\t12,215\n",
      "Spliced reads ignored:\t\t12,215\n",
      "INFO:__main__:Total reads mapped to genes:\t1,510,384\n",
      "Total reads mapped to genes:\t1,510,384\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/63_S5Ph_RIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/63_S5Ph_RIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t80_S2RIP.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t80_S2RIP.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t18,074,957\n",
      "Total reads:\t\t\t18,074,957\n",
      "INFO:__main__:Spliced reads ignored:\t\t45,905\n",
      "Spliced reads ignored:\t\t45,905\n",
      "INFO:__main__:Total reads mapped to genes:\t4,234,231\n",
      "Total reads mapped to genes:\t4,234,231\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/80_S2RIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/80_S2RIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t80S2RIP_1.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t80S2RIP_1.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t13,575,895\n",
      "Total reads:\t\t\t13,575,895\n",
      "INFO:__main__:Spliced reads ignored:\t\t37,328\n",
      "Spliced reads ignored:\t\t37,328\n",
      "INFO:__main__:Total reads mapped to genes:\t1,649,853\n",
      "Total reads mapped to genes:\t1,649,853\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/80S2RIP_1.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/80S2RIP_1.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t80S2RIP_2.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t80S2RIP_2.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t16,194,814\n",
      "Total reads:\t\t\t16,194,814\n",
      "INFO:__main__:Spliced reads ignored:\t\t44,572\n",
      "Spliced reads ignored:\t\t44,572\n",
      "INFO:__main__:Total reads mapped to genes:\t1,722,219\n",
      "Total reads mapped to genes:\t1,722,219\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/80S2RIP_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/80S2RIP_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t301_S2RIP_2.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t301_S2RIP_2.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t17,154,070\n",
      "Total reads:\t\t\t17,154,070\n",
      "INFO:__main__:Spliced reads ignored:\t\t39,546\n",
      "Spliced reads ignored:\t\t39,546\n",
      "INFO:__main__:Total reads mapped to genes:\t2,671,327\n",
      "Total reads mapped to genes:\t2,671,327\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/301_S2RIP_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/301_S2RIP_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t301_S2RIP_3.ztr.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t301_S2RIP_3.ztr.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t16,940,467\n",
      "Total reads:\t\t\t16,940,467\n",
      "INFO:__main__:Spliced reads ignored:\t\t47,297\n",
      "Spliced reads ignored:\t\t47,297\n",
      "INFO:__main__:Total reads mapped to genes:\t1,897,600\n",
      "Total reads mapped to genes:\t1,897,600\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/301_S2RIP_3.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/301_S2RIP_3.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t301_S2RIP.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t301_S2RIP.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t18,159,048\n",
      "Total reads:\t\t\t18,159,048\n",
      "INFO:__main__:Spliced reads ignored:\t\t48,150\n",
      "Spliced reads ignored:\t\t48,150\n",
      "INFO:__main__:Total reads mapped to genes:\t5,086,028\n",
      "Total reads mapped to genes:\t5,086,028\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/301_S2RIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/301_S2RIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t302_S2RIP_3.ztr.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t302_S2RIP_3.ztr.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t16,579,898\n",
      "Total reads:\t\t\t16,579,898\n",
      "INFO:__main__:Spliced reads ignored:\t\t52,027\n",
      "Spliced reads ignored:\t\t52,027\n",
      "INFO:__main__:Total reads mapped to genes:\t2,712,915\n",
      "Total reads mapped to genes:\t2,712,915\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/302_S2RIP_3.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/302_S2RIP_3.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t302_S2RIP.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t302_S2RIP.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t15,751,893\n",
      "Total reads:\t\t\t15,751,893\n",
      "INFO:__main__:Spliced reads ignored:\t\t38,557\n",
      "Spliced reads ignored:\t\t38,557\n",
      "INFO:__main__:Total reads mapped to genes:\t3,809,496\n",
      "Total reads mapped to genes:\t3,809,496\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/302_S2RIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/302_S2RIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t324_S2RIP_2.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t324_S2RIP_2.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t2,464,458\n",
      "Total reads:\t\t\t2,464,458\n",
      "INFO:__main__:Spliced reads ignored:\t\t8,622\n",
      "Spliced reads ignored:\t\t8,622\n",
      "INFO:__main__:Total reads mapped to genes:\t415,637\n",
      "Total reads mapped to genes:\t415,637\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/324_S2RIP_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/324_S2RIP_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t324_S2RIP_3.ztr.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t324_S2RIP_3.ztr.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t15,090,396\n",
      "Total reads:\t\t\t15,090,396\n",
      "INFO:__main__:Spliced reads ignored:\t\t47,499\n",
      "Spliced reads ignored:\t\t47,499\n",
      "INFO:__main__:Total reads mapped to genes:\t2,414,397\n",
      "Total reads mapped to genes:\t2,414,397\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/324_S2RIP_3.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/324_S2RIP_3.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t324_S2RIP.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t324_S2RIP.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t17,284,517\n",
      "Total reads:\t\t\t17,284,517\n",
      "INFO:__main__:Spliced reads ignored:\t\t51,130\n",
      "Spliced reads ignored:\t\t51,130\n",
      "INFO:__main__:Total reads mapped to genes:\t4,814,713\n",
      "Total reads mapped to genes:\t4,814,713\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/324_S2RIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/324_S2RIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t324S2RIP_1.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t324S2RIP_1.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t8,831,510\n",
      "Total reads:\t\t\t8,831,510\n",
      "INFO:__main__:Spliced reads ignored:\t\t112,113\n",
      "Spliced reads ignored:\t\t112,113\n",
      "INFO:__main__:Total reads mapped to genes:\t225,047\n",
      "Total reads mapped to genes:\t225,047\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/324S2RIP_1.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/324S2RIP_1.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t491_S2RIP_2.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t491_S2RIP_2.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t15,389,415\n",
      "Total reads:\t\t\t15,389,415\n",
      "INFO:__main__:Spliced reads ignored:\t\t50,826\n",
      "Spliced reads ignored:\t\t50,826\n",
      "INFO:__main__:Total reads mapped to genes:\t2,497,264\n",
      "Total reads mapped to genes:\t2,497,264\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/491_S2RIP_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/491_S2RIP_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t491_S2RIP_3.ztr.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t491_S2RIP_3.ztr.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t14,894,953\n",
      "Total reads:\t\t\t14,894,953\n",
      "INFO:__main__:Spliced reads ignored:\t\t42,748\n",
      "Spliced reads ignored:\t\t42,748\n",
      "INFO:__main__:Total reads mapped to genes:\t2,548,592\n",
      "Total reads mapped to genes:\t2,548,592\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/491_S2RIP_3.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/491_S2RIP_3.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t491_S2RIP.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t491_S2RIP.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t24,641,944\n",
      "Total reads:\t\t\t24,641,944\n",
      "INFO:__main__:Spliced reads ignored:\t\t64,007\n",
      "Spliced reads ignored:\t\t64,007\n",
      "INFO:__main__:Total reads mapped to genes:\t5,306,415\n",
      "Total reads mapped to genes:\t5,306,415\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/491_S2RIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/491_S2RIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t491S2RIP_1.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t491S2RIP_1.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t7,750,987\n",
      "Total reads:\t\t\t7,750,987\n",
      "INFO:__main__:Spliced reads ignored:\t\t103,640\n",
      "Spliced reads ignored:\t\t103,640\n",
      "INFO:__main__:Total reads mapped to genes:\t104,542\n",
      "Total reads mapped to genes:\t104,542\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/491S2RIP_1.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/491S2RIP_1.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t504S2RIP_1.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t504S2RIP_1.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t14,889,605\n",
      "Total reads:\t\t\t14,889,605\n",
      "INFO:__main__:Spliced reads ignored:\t\t51,407\n",
      "Spliced reads ignored:\t\t51,407\n",
      "INFO:__main__:Total reads mapped to genes:\t2,210,429\n",
      "Total reads mapped to genes:\t2,210,429\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/504S2RIP_1.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/504S2RIP_1.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t504S2RIP_2.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t504S2RIP_2.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t14,594,016\n",
      "Total reads:\t\t\t14,594,016\n",
      "INFO:__main__:Spliced reads ignored:\t\t56,209\n",
      "Spliced reads ignored:\t\t56,209\n",
      "INFO:__main__:Total reads mapped to genes:\t2,039,220\n",
      "Total reads mapped to genes:\t2,039,220\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/504S2RIP_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/504S2RIP_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t510_S2RIP.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t510_S2RIP.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t20,638,391\n",
      "Total reads:\t\t\t20,638,391\n",
      "INFO:__main__:Spliced reads ignored:\t\t55,616\n",
      "Spliced reads ignored:\t\t55,616\n",
      "INFO:__main__:Total reads mapped to genes:\t4,123,161\n",
      "Total reads mapped to genes:\t4,123,161\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/510_S2RIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/510_S2RIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t530S2RIP_1.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t530S2RIP_1.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t11,564,816\n",
      "Total reads:\t\t\t11,564,816\n",
      "INFO:__main__:Spliced reads ignored:\t\t36,704\n",
      "Spliced reads ignored:\t\t36,704\n",
      "INFO:__main__:Total reads mapped to genes:\t1,303,290\n",
      "Total reads mapped to genes:\t1,303,290\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/530S2RIP_1.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/530S2RIP_1.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t530S2RIP_2.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t530S2RIP_2.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t14,268,624\n",
      "Total reads:\t\t\t14,268,624\n",
      "INFO:__main__:Spliced reads ignored:\t\t44,915\n",
      "Spliced reads ignored:\t\t44,915\n",
      "INFO:__main__:Total reads mapped to genes:\t1,775,904\n",
      "Total reads mapped to genes:\t1,775,904\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/530S2RIP_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/530S2RIP_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t591_S2RIP_2.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t591_S2RIP_2.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t18,492,744\n",
      "Total reads:\t\t\t18,492,744\n",
      "INFO:__main__:Spliced reads ignored:\t\t61,284\n",
      "Spliced reads ignored:\t\t61,284\n",
      "INFO:__main__:Total reads mapped to genes:\t2,958,945\n",
      "Total reads mapped to genes:\t2,958,945\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/591_S2RIP_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/591_S2RIP_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t591_S5RIP_1.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t591_S5RIP_1.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t16,185,052\n",
      "Total reads:\t\t\t16,185,052\n",
      "INFO:__main__:Spliced reads ignored:\t\t49,649\n",
      "Spliced reads ignored:\t\t49,649\n",
      "INFO:__main__:Total reads mapped to genes:\t3,946,873\n",
      "Total reads mapped to genes:\t3,946,873\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/591_S5RIP_1.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/591_S5RIP_1.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t638S2RIP_1.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t638S2RIP_1.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t13,450,747\n",
      "Total reads:\t\t\t13,450,747\n",
      "INFO:__main__:Spliced reads ignored:\t\t33,753\n",
      "Spliced reads ignored:\t\t33,753\n",
      "INFO:__main__:Total reads mapped to genes:\t1,490,933\n",
      "Total reads mapped to genes:\t1,490,933\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/638S2RIP_1.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/638S2RIP_1.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t638S2RIP_2.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t638S2RIP_2.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t12,221,421\n",
      "Total reads:\t\t\t12,221,421\n",
      "INFO:__main__:Spliced reads ignored:\t\t33,758\n",
      "Spliced reads ignored:\t\t33,758\n",
      "INFO:__main__:Total reads mapped to genes:\t1,422,894\n",
      "Total reads mapped to genes:\t1,422,894\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/638S2RIP_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/638S2RIP_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t1113_S2RIP.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t1113_S2RIP.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t12,273,254\n",
      "Total reads:\t\t\t12,273,254\n",
      "INFO:__main__:Spliced reads ignored:\t\t35,823\n",
      "Spliced reads ignored:\t\t35,823\n",
      "INFO:__main__:Total reads mapped to genes:\t4,374,829\n",
      "Total reads mapped to genes:\t4,374,829\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/1113_S2RIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/1113_S2RIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t1167_S2RIP_2.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t1167_S2RIP_2.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t22,464,998\n",
      "Total reads:\t\t\t22,464,998\n",
      "INFO:__main__:Spliced reads ignored:\t\t57,795\n",
      "Spliced reads ignored:\t\t57,795\n",
      "INFO:__main__:Total reads mapped to genes:\t2,317,690\n",
      "Total reads mapped to genes:\t2,317,690\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/1167_S2RIP_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/1167_S2RIP_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t1167_S2RIP.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t1167_S2RIP.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t14,333,644\n",
      "Total reads:\t\t\t14,333,644\n",
      "INFO:__main__:Spliced reads ignored:\t\t30,959\n",
      "Spliced reads ignored:\t\t30,959\n",
      "INFO:__main__:Total reads mapped to genes:\t2,104,877\n",
      "Total reads mapped to genes:\t2,104,877\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/1167_S2RIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/1167_S2RIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t1167_S5RIP_2.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t1167_S5RIP_2.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t21,117,349\n",
      "Total reads:\t\t\t21,117,349\n",
      "INFO:__main__:Spliced reads ignored:\t\t29,158\n",
      "Spliced reads ignored:\t\t29,158\n",
      "INFO:__main__:Total reads mapped to genes:\t1,243,561\n",
      "Total reads mapped to genes:\t1,243,561\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/1167_S5RIP_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/1167_S5RIP_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t1168_S2RIP_2.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t1168_S2RIP_2.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t23,188,416\n",
      "Total reads:\t\t\t23,188,416\n",
      "INFO:__main__:Spliced reads ignored:\t\t39,898\n",
      "Spliced reads ignored:\t\t39,898\n",
      "INFO:__main__:Total reads mapped to genes:\t2,047,461\n",
      "Total reads mapped to genes:\t2,047,461\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/1168_S2RIP_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/1168_S2RIP_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t1168_S2RIP.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t1168_S2RIP.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t16,081,973\n",
      "Total reads:\t\t\t16,081,973\n",
      "INFO:__main__:Spliced reads ignored:\t\t36,378\n",
      "Spliced reads ignored:\t\t36,378\n",
      "INFO:__main__:Total reads mapped to genes:\t1,197,698\n",
      "Total reads mapped to genes:\t1,197,698\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/1168_S2RIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/1168_S2RIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t324_RNA_pA_3.ztr.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t324_RNA_pA_3.ztr.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t13,349,405\n",
      "Total reads:\t\t\t13,349,405\n",
      "INFO:__main__:Spliced reads ignored:\t\t552,509\n",
      "Spliced reads ignored:\t\t552,509\n",
      "INFO:__main__:Total reads mapped to genes:\t12,345,759\n",
      "Total reads mapped to genes:\t12,345,759\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/324_RNA_pA_3.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/324_RNA_pA_3.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t1168_pA_2.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t1168_pA_2.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t16,075,322\n",
      "Total reads:\t\t\t16,075,322\n",
      "INFO:__main__:Spliced reads ignored:\t\t465,790\n",
      "Spliced reads ignored:\t\t465,790\n",
      "INFO:__main__:Total reads mapped to genes:\t14,639,351\n",
      "Total reads mapped to genes:\t14,639,351\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/1168_pA_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/1168_pA_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t1113_pA.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t1113_pA.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t26,461,506\n",
      "Total reads:\t\t\t26,461,506\n",
      "INFO:__main__:Spliced reads ignored:\t\t814,160\n",
      "Spliced reads ignored:\t\t814,160\n",
      "INFO:__main__:Total reads mapped to genes:\t24,230,150\n",
      "Total reads mapped to genes:\t24,230,150\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/1113_pA.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/1113_pA.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t1167_pA.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t1167_pA.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t27,868,007\n",
      "Total reads:\t\t\t27,868,007\n",
      "INFO:__main__:Spliced reads ignored:\t\t870,482\n",
      "Spliced reads ignored:\t\t870,482\n",
      "INFO:__main__:Total reads mapped to genes:\t25,490,955\n",
      "Total reads mapped to genes:\t25,490,955\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/1167_pA.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/1167_pA.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t1168_pA.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t1168_pA.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t3,078,252\n",
      "Total reads:\t\t\t3,078,252\n",
      "INFO:__main__:Spliced reads ignored:\t\t93,609\n",
      "Spliced reads ignored:\t\t93,609\n",
      "INFO:__main__:Total reads mapped to genes:\t2,807,656\n",
      "Total reads mapped to genes:\t2,807,656\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/1168_pA.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/1168_pA.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t510_RNA_pA_2.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t510_RNA_pA_2.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t21,225,540\n",
      "Total reads:\t\t\t21,225,540\n",
      "INFO:__main__:Spliced reads ignored:\t\t570,146\n",
      "Spliced reads ignored:\t\t570,146\n",
      "INFO:__main__:Total reads mapped to genes:\t19,635,671\n",
      "Total reads mapped to genes:\t19,635,671\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/510_RNA_pA_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/510_RNA_pA_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t301_RNA_pA_2.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t301_RNA_pA_2.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t16,689,040\n",
      "Total reads:\t\t\t16,689,040\n",
      "INFO:__main__:Spliced reads ignored:\t\t662,797\n",
      "Spliced reads ignored:\t\t662,797\n",
      "INFO:__main__:Total reads mapped to genes:\t15,056,234\n",
      "Total reads mapped to genes:\t15,056,234\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/301_RNA_pA_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/301_RNA_pA_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t302_RNA_pA_2.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t302_RNA_pA_2.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t14,098,264\n",
      "Total reads:\t\t\t14,098,264\n",
      "INFO:__main__:Spliced reads ignored:\t\t498,473\n",
      "Spliced reads ignored:\t\t498,473\n",
      "INFO:__main__:Total reads mapped to genes:\t12,986,168\n",
      "Total reads mapped to genes:\t12,986,168\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/302_RNA_pA_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/302_RNA_pA_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t324_RNA_pA_2.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t324_RNA_pA_2.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t16,890,011\n",
      "Total reads:\t\t\t16,890,011\n",
      "INFO:__main__:Spliced reads ignored:\t\t612,845\n",
      "Spliced reads ignored:\t\t612,845\n",
      "INFO:__main__:Total reads mapped to genes:\t15,041,412\n",
      "Total reads mapped to genes:\t15,041,412\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/324_RNA_pA_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/324_RNA_pA_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t491_RNA_pA_2.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t491_RNA_pA_2.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t22,536,290\n",
      "Total reads:\t\t\t22,536,290\n",
      "INFO:__main__:Spliced reads ignored:\t\t852,497\n",
      "Spliced reads ignored:\t\t852,497\n",
      "INFO:__main__:Total reads mapped to genes:\t20,804,301\n",
      "Total reads mapped to genes:\t20,804,301\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/491_RNA_pA_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/491_RNA_pA_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t80pARNA_2.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t80pARNA_2.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t13,323,511\n",
      "Total reads:\t\t\t13,323,511\n",
      "INFO:__main__:Spliced reads ignored:\t\t470,799\n",
      "Spliced reads ignored:\t\t470,799\n",
      "INFO:__main__:Total reads mapped to genes:\t12,423,716\n",
      "Total reads mapped to genes:\t12,423,716\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/80pARNA_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/80pARNA_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t63_RNA_pA_4.ztr.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t63_RNA_pA_4.ztr.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t14,219,928\n",
      "Total reads:\t\t\t14,219,928\n",
      "INFO:__main__:Spliced reads ignored:\t\t569,997\n",
      "Spliced reads ignored:\t\t569,997\n",
      "INFO:__main__:Total reads mapped to genes:\t13,141,023\n",
      "Total reads mapped to genes:\t13,141,023\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/63_RNA_pA_4.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/63_RNA_pA_4.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t283_RNA_pA_4.ztr.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t283_RNA_pA_4.ztr.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t15,268,985\n",
      "Total reads:\t\t\t15,268,985\n",
      "INFO:__main__:Spliced reads ignored:\t\t581,955\n",
      "Spliced reads ignored:\t\t581,955\n",
      "INFO:__main__:Total reads mapped to genes:\t13,909,010\n",
      "Total reads mapped to genes:\t13,909,010\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/283_RNA_pA_4.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/283_RNA_pA_4.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t301_RNA_pA_3.ztr.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t301_RNA_pA_3.ztr.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t13,287,929\n",
      "Total reads:\t\t\t13,287,929\n",
      "INFO:__main__:Spliced reads ignored:\t\t517,675\n",
      "Spliced reads ignored:\t\t517,675\n",
      "INFO:__main__:Total reads mapped to genes:\t12,285,308\n",
      "Total reads mapped to genes:\t12,285,308\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/301_RNA_pA_3.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/301_RNA_pA_3.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t504_RNA_pA_1.ztr.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t504_RNA_pA_1.ztr.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t16,540,223\n",
      "Total reads:\t\t\t16,540,223\n",
      "INFO:__main__:Spliced reads ignored:\t\t534,753\n",
      "Spliced reads ignored:\t\t534,753\n",
      "INFO:__main__:Total reads mapped to genes:\t14,742,015\n",
      "Total reads mapped to genes:\t14,742,015\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/504_RNA_pA_1.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/504_RNA_pA_1.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t504_RNA_pA_2.ztr.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t504_RNA_pA_2.ztr.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t13,542,881\n",
      "Total reads:\t\t\t13,542,881\n",
      "INFO:__main__:Spliced reads ignored:\t\t431,814\n",
      "Spliced reads ignored:\t\t431,814\n",
      "INFO:__main__:Total reads mapped to genes:\t12,213,377\n",
      "Total reads mapped to genes:\t12,213,377\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/504_RNA_pA_2.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/504_RNA_pA_2.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t530_RNA_pA_1.ztr.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t530_RNA_pA_1.ztr.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t16,959,495\n",
      "Total reads:\t\t\t16,959,495\n",
      "INFO:__main__:Spliced reads ignored:\t\t612,733\n",
      "Spliced reads ignored:\t\t612,733\n",
      "INFO:__main__:Total reads mapped to genes:\t15,331,342\n",
      "Total reads mapped to genes:\t15,331,342\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/530_RNA_pA_1.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/530_RNA_pA_1.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t530_RNA_pA_2.ztr.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t530_RNA_pA_2.ztr.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t14,897,898\n",
      "Total reads:\t\t\t14,897,898\n",
      "INFO:__main__:Spliced reads ignored:\t\t569,255\n",
      "Spliced reads ignored:\t\t569,255\n",
      "INFO:__main__:Total reads mapped to genes:\t13,586,095\n",
      "Total reads mapped to genes:\t13,586,095\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/530_RNA_pA_2.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/530_RNA_pA_2.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t638_RNA_pA_1.ztr.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t638_RNA_pA_1.ztr.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t16,339,340\n",
      "Total reads:\t\t\t16,339,340\n",
      "INFO:__main__:Spliced reads ignored:\t\t644,340\n",
      "Spliced reads ignored:\t\t644,340\n",
      "INFO:__main__:Total reads mapped to genes:\t14,987,053\n",
      "Total reads mapped to genes:\t14,987,053\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/638_RNA_pA_1.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/638_RNA_pA_1.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t638_RNA_pA_2.ztr.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t638_RNA_pA_2.ztr.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t12,322,354\n",
      "Total reads:\t\t\t12,322,354\n",
      "INFO:__main__:Spliced reads ignored:\t\t437,074\n",
      "Spliced reads ignored:\t\t437,074\n",
      "INFO:__main__:Total reads mapped to genes:\t11,464,007\n",
      "Total reads mapped to genes:\t11,464,007\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/638_RNA_pA_2.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/638_RNA_pA_2.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t80_RNA_pA.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t80_RNA_pA.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t3,619,460\n",
      "Total reads:\t\t\t3,619,460\n",
      "INFO:__main__:Spliced reads ignored:\t\t127,855\n",
      "Spliced reads ignored:\t\t127,855\n",
      "INFO:__main__:Total reads mapped to genes:\t3,364,049\n",
      "Total reads mapped to genes:\t3,364,049\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/80_RNA_pA.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/80_RNA_pA.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t301_RNA_pA.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t301_RNA_pA.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t7,019,817\n",
      "Total reads:\t\t\t7,019,817\n",
      "INFO:__main__:Spliced reads ignored:\t\t244,756\n",
      "Spliced reads ignored:\t\t244,756\n",
      "INFO:__main__:Total reads mapped to genes:\t6,464,775\n",
      "Total reads mapped to genes:\t6,464,775\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/301_RNA_pA.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/301_RNA_pA.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t302_RNA_pA.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t302_RNA_pA.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t15,257,000\n",
      "Total reads:\t\t\t15,257,000\n",
      "INFO:__main__:Spliced reads ignored:\t\t478,290\n",
      "Spliced reads ignored:\t\t478,290\n",
      "INFO:__main__:Total reads mapped to genes:\t14,024,513\n",
      "Total reads mapped to genes:\t14,024,513\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/302_RNA_pA.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/302_RNA_pA.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t324_RNA_pA.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t324_RNA_pA.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t5,720,621\n",
      "Total reads:\t\t\t5,720,621\n",
      "INFO:__main__:Spliced reads ignored:\t\t176,624\n",
      "Spliced reads ignored:\t\t176,624\n",
      "INFO:__main__:Total reads mapped to genes:\t5,255,239\n",
      "Total reads mapped to genes:\t5,255,239\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/324_RNA_pA.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/324_RNA_pA.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t491_RNA_pA.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t491_RNA_pA.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t20,375,053\n",
      "Total reads:\t\t\t20,375,053\n",
      "INFO:__main__:Spliced reads ignored:\t\t648,253\n",
      "Spliced reads ignored:\t\t648,253\n",
      "INFO:__main__:Total reads mapped to genes:\t18,665,318\n",
      "Total reads mapped to genes:\t18,665,318\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/491_RNA_pA.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/491_RNA_pA.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Sample name:\t510_RNA_pA.Aligned.sortedByCoord.out.bam\n",
      "Sample name:\t510_RNA_pA.Aligned.sortedByCoord.out.bam\n",
      "INFO:__main__:-------------------------\n",
      "-------------------------\n",
      "INFO:__main__:Total reads:\t\t\t17,730,562\n",
      "Total reads:\t\t\t17,730,562\n",
      "INFO:__main__:Spliced reads ignored:\t\t556,417\n",
      "Spliced reads ignored:\t\t556,417\n",
      "INFO:__main__:Total reads mapped to genes:\t16,375,919\n",
      "Total reads mapped to genes:\t16,375,919\n",
      "INFO:__main__:Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/510_RNA_pA.Aligned.sortedByCoord.out.tagged.bam\n",
      "Output file:\t/data/pablo/RNAdeg/results/RipRna/tagged_bams/510_RNA_pA.Aligned.sortedByCoord.out.tagged.bam\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:\n",
      "\n",
      "INFO:__main__:Finished successfully!\n",
      "Finished successfully!\n"
     ]
    }
   ],
   "source": [
    "# Run RemrRNA\n",
    "remrrna = remrrna_command()\n",
    "!$remrrna"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5 - Compute raw and tpm-normalized gene count tables."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set GeneExpressionTable runtime parameters.\n",
    "def gxt_command():\n",
    "    \n",
    "    #gxt = '/data/parastou/RNAdeg/pyRNAdeg/GeneExpressionTable.py' ## doesn't exist\n",
    "    ## not sure which one to use:\n",
    "    gxt = '/data/parastou/RNAdeg/pyRNAdeg/gene_expression_table.py'\n",
    "    #gxt = '/data/parastou/RNAdeg/pyRNAdeg/gene_expression_table_extended.py' ## strand-specific\n",
    "\n",
    "    annotation = '/data/parastou/RNAdeg/annotation/schizosaccharomyces_pombe.chr.extended.csv'\n",
    "    \n",
    "    command = 'python ' + gxt + ' -d ' + out_tagged + ' -g ' + annotation + ' -o ' + xp_data\n",
    "    \n",
    "    return command"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Call to GeneExpressionTable module....\n",
      "\n",
      "This module calculates gene expression in given alignment file(s).\n",
      "Input: folder containing .bam file(s)\n",
      "Output: raw and tpm gene counts data (.csv) files\n",
      "----------------------------------------\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/1113_pA.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 29738993\n",
      "Spliced alignments : 0\n",
      "Non spliced : 29738993\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/301_S2RIP_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 3236804\n",
      "Spliced alignments : 0\n",
      "Non spliced : 3236804\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/302_S2RIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 4597762\n",
      "Spliced alignments : 0\n",
      "Non spliced : 4597762\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/491_S2RIP_3.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 3071513\n",
      "Spliced alignments : 0\n",
      "Non spliced : 3071513\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/1167_pA_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 21344379\n",
      "Spliced alignments : 0\n",
      "Non spliced : 21344379\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/302_S2RIP_3.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 3269901\n",
      "Spliced alignments : 0\n",
      "Non spliced : 3269901\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/491_S2RIP_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 2997211\n",
      "Spliced alignments : 0\n",
      "Non spliced : 2997211\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/63.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 21787274\n",
      "Spliced alignments : 0\n",
      "Non spliced : 21787274\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/301_S2RIP_3.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 2280485\n",
      "Spliced alignments : 0\n",
      "Non spliced : 2280485\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/301_S2RIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 6126317\n",
      "Spliced alignments : 0\n",
      "Non spliced : 6126317\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/65.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 20323060\n",
      "Spliced alignments : 0\n",
      "Non spliced : 20323060\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/1167_pA.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 31125631\n",
      "Spliced alignments : 0\n",
      "Non spliced : 31125631\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/324_S2RIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 5834619\n",
      "Spliced alignments : 0\n",
      "Non spliced : 5834619\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/63_RIPS5P.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 9514580\n",
      "Spliced alignments : 0\n",
      "Non spliced : 9514580\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/1168_pA.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 3408617\n",
      "Spliced alignments : 0\n",
      "Non spliced : 3408617\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/63_S2Ph_RIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 6823978\n",
      "Spliced alignments : 0\n",
      "Non spliced : 6823978\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/324_S2RIP_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 498497\n",
      "Spliced alignments : 0\n",
      "Non spliced : 498497\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/324S2RIP_1.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 287921\n",
      "Spliced alignments : 0\n",
      "Non spliced : 287921\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/491_S2RIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 6443227\n",
      "Spliced alignments : 0\n",
      "Non spliced : 6443227\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/63_S2PRIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 7866524\n",
      "Spliced alignments : 0\n",
      "Non spliced : 7866524\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/510_S2RIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 4976016\n",
      "Spliced alignments : 0\n",
      "Non spliced : 4976016\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/63_S2RIP_2.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 1867093\n",
      "Spliced alignments : 0\n",
      "Non spliced : 1867093\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/491S2RIP_1.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 132433\n",
      "Spliced alignments : 0\n",
      "Non spliced : 132433\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/63_S5Ph_RIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 1829553\n",
      "Spliced alignments : 0\n",
      "Non spliced : 1829553\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/80_S2RIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 5194321\n",
      "Spliced alignments : 0\n",
      "Non spliced : 5194321\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/80S2RIP_1.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 2003218\n",
      "Spliced alignments : 0\n",
      "Non spliced : 2003218\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/80S2RIP_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 2088678\n",
      "Spliced alignments : 0\n",
      "Non spliced : 2088678\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/324_S2RIP_3.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 2908584\n",
      "Spliced alignments : 0\n",
      "Non spliced : 2908584\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/504S2RIP_1.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 2666658\n",
      "Spliced alignments : 0\n",
      "Non spliced : 2666658\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/504S2RIP_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 2467902\n",
      "Spliced alignments : 0\n",
      "Non spliced : 2467902\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/530S2RIP_1.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 1553833\n",
      "Spliced alignments : 0\n",
      "Non spliced : 1553833\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/530S2RIP_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 2124538\n",
      "Spliced alignments : 0\n",
      "Non spliced : 2124538\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/591_S2RIP_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 3620178\n",
      "Spliced alignments : 0\n",
      "Non spliced : 3620178\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/301_RNA_pA_3.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 15000993\n",
      "Spliced alignments : 0\n",
      "Non spliced : 15000993\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/591_S5RIP_1.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 4793897\n",
      "Spliced alignments : 0\n",
      "Non spliced : 4793897\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/638S2RIP_1.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 1797023\n",
      "Spliced alignments : 0\n",
      "Non spliced : 1797023\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/638S2RIP_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 1717818\n",
      "Spliced alignments : 0\n",
      "Non spliced : 1717818\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/504_RNA_pA_1.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 18161544\n",
      "Spliced alignments : 0\n",
      "Non spliced : 18161544\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/1113_S2RIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 5283626\n",
      "Spliced alignments : 0\n",
      "Non spliced : 5283626\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/1167_S2RIP_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 2791797\n",
      "Spliced alignments : 0\n",
      "Non spliced : 2791797\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/1167_S2RIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 2521778\n",
      "Spliced alignments : 0\n",
      "Non spliced : 2521778\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/1167_S5RIP_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 1498893\n",
      "Spliced alignments : 0\n",
      "Non spliced : 1498893\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/1168_S2RIP_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 2488945\n",
      "Spliced alignments : 0\n",
      "Non spliced : 2488945\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/1168_S2RIP.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 1445836\n",
      "Spliced alignments : 0\n",
      "Non spliced : 1445836\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/301_RNA_pA_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 18491678\n",
      "Spliced alignments : 0\n",
      "Non spliced : 18491678\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/324_RNA_pA_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 18429178\n",
      "Spliced alignments : 0\n",
      "Non spliced : 18429178\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/324_RNA_pA_3.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 15081399\n",
      "Spliced alignments : 0\n",
      "Non spliced : 15081399\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/504_RNA_pA_2.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 15064485\n",
      "Spliced alignments : 0\n",
      "Non spliced : 15064485\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/1168_pA_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 17639735\n",
      "Spliced alignments : 0\n",
      "Non spliced : 17639735\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/302_RNA_pA_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 16010878\n",
      "Spliced alignments : 0\n",
      "Non spliced : 16010878\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/63_RNA_pA_4.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 15979307\n",
      "Spliced alignments : 0\n",
      "Non spliced : 15979307\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/510_RNA_pA_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 23628436\n",
      "Spliced alignments : 0\n",
      "Non spliced : 23628436\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/80pARNA_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 15425726\n",
      "Spliced alignments : 0\n",
      "Non spliced : 15425726\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/283_RNA_pA_4.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 16896935\n",
      "Spliced alignments : 0\n",
      "Non spliced : 16896935\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/491_RNA_pA_2.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 25588457\n",
      "Spliced alignments : 0\n",
      "Non spliced : 25588457\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/530_RNA_pA_1.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 18847124\n",
      "Spliced alignments : 0\n",
      "Non spliced : 18847124\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/530_RNA_pA_2.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 16765174\n",
      "Spliced alignments : 0\n",
      "Non spliced : 16765174\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/638_RNA_pA_1.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 18424044\n",
      "Spliced alignments : 0\n",
      "Non spliced : 18424044\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/638_RNA_pA_2.ztr.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 14065835\n",
      "Spliced alignments : 0\n",
      "Non spliced : 14065835\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/80_RNA_pA.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 4144030\n",
      "Spliced alignments : 0\n",
      "Non spliced : 4144030\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/301_RNA_pA.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 8002623\n",
      "Spliced alignments : 0\n",
      "Non spliced : 8002623\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/302_RNA_pA.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 17211463\n",
      "Spliced alignments : 0\n",
      "Non spliced : 17211463\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/324_RNA_pA.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 6455132\n",
      "Spliced alignments : 0\n",
      "Non spliced : 6455132\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/491_RNA_pA.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 22878678\n",
      "Spliced alignments : 0\n",
      "Non spliced : 22878678\n",
      "Input bam: /data/pablo/RNAdeg/results/RipRna/tagged_bams/510_RNA_pA.Aligned.sortedByCoord.out.tagged.bam\n",
      "Total number of alignments : 19789253\n",
      "Spliced alignments : 0\n",
      "Non spliced : 19789253\n",
      "Removed 49 rRNA genes.\n",
      "Calculated raw counts for 6943 genes.\n",
      "Calculated TMP for 6943 genes.\n",
      "Gene count matrix saved in: /data/pablo/RNAdeg/results/RipRna/xp_data/gene_count_table.csv\n",
      "TPM expression matrix saved in: /data/pablo/RNAdeg/results/RipRna/xp_data/tpm_table.csv\n",
      "Finished successfully!\n"
     ]
    }
   ],
   "source": [
    "# Run GeneExpressionTable\n",
    "gxt = gxt_command()\n",
    "!$gxt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----------------"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
